<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
<meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=2">
<meta name="theme-color" content="#222">
<meta name="generator" content="Hexo 5.4.0">
  <link rel="apple-touch-icon" sizes="180x180" href="/images/apple-touch-icon.png">
  <link rel="icon" type="image/png" sizes="32x32" href="/images/favicon-32x32.png">
  <link rel="icon" type="image/png" sizes="16x16" href="/images/favicon-16x16.png">

<link rel="stylesheet" href="/css/main.css">


<link rel="stylesheet" href="/lib/font-awesome/css/all.min.css">

<script id="hexo-configurations">
    var NexT = window.NexT || {};
    var CONFIG = {"hostname":"thereallda.github.io","root":"/","scheme":"Gemini","version":"7.8.0","exturl":false,"sidebar":{"position":"left","display":"post","padding":18,"offset":12,"onmobile":false},"copycode":{"enable":true,"show_result":false,"style":"mac"},"back2top":{"enable":true,"sidebar":false,"scrollpercent":false},"bookmark":{"enable":false,"color":"#222","save":"auto"},"fancybox":false,"mediumzoom":false,"lazyload":false,"pangu":false,"comments":{"style":"tabs","active":"gitalk","storage":true,"lazyload":false,"nav":null,"activeClass":"gitalk"},"algolia":{"hits":{"per_page":10},"labels":{"input_placeholder":"Search for Posts","hits_empty":"We didn't find any results for the search: ${query}","hits_stats":"${hits} results found in ${time} ms"}},"localsearch":{"enable":false,"trigger":"auto","top_n_per_article":1,"unescape":false,"preload":false},"motion":{"enable":true,"async":false,"transition":{"post_block":"fadeIn","post_header":"slideDownIn","post_body":"slideDownIn","coll_header":"slideLeftIn","sidebar":"slideUpIn"}}};
  </script>

  <meta name="description" content="通常，在计算TPM或RPKM&#x2F;FPKM等基因表达量时，除了基因的counts信息外，我们还需要知道基因的长度。这里所用到的基因长度并不是某个基因在基因组上的完整长度。在基因表达分析中，“基因长度”通常指的是成熟转录本的长度，也就是无内含子的碱基序列。因此，单纯地使用基因的染色体起始和结束坐标相减并不能返回转录本的长度信息。目前，对于基因长度有多种定义，包括： 1. 基因最长转录本； 2. 多个转录">
<meta property="og:type" content="article">
<meta property="og:title" content="R-获取基因长度">
<meta property="og:url" content="https://thereallda.github.io/2022/02/07/R-Retrieve-Gene-Length/index.html">
<meta property="og:site_name" content="Dean&#39;s blog">
<meta property="og:description" content="通常，在计算TPM或RPKM&#x2F;FPKM等基因表达量时，除了基因的counts信息外，我们还需要知道基因的长度。这里所用到的基因长度并不是某个基因在基因组上的完整长度。在基因表达分析中，“基因长度”通常指的是成熟转录本的长度，也就是无内含子的碱基序列。因此，单纯地使用基因的染色体起始和结束坐标相减并不能返回转录本的长度信息。目前，对于基因长度有多种定义，包括： 1. 基因最长转录本； 2. 多个转录">
<meta property="og:locale" content="en_US">
<meta property="og:image" content="https://thereallda.github.io/2022/02/07/R-Retrieve-Gene-Length/paste-CBFC7C07.png">
<meta property="og:image" content="https://thereallda.github.io/2022/02/07/R-Retrieve-Gene-Length/paste-2878321A.png">
<meta property="article:published_time" content="2022-02-07T21:12:25.000Z">
<meta property="article:modified_time" content="2022-02-07T09:48:49.726Z">
<meta property="article:author" content="Dean Li">
<meta property="article:tag" content="Data_analysis">
<meta property="article:tag" content="R">
<meta name="twitter:card" content="summary">
<meta name="twitter:image" content="https://thereallda.github.io/2022/02/07/R-Retrieve-Gene-Length/paste-CBFC7C07.png">

<link rel="canonical" href="https://thereallda.github.io/2022/02/07/R-Retrieve-Gene-Length/">


<script id="page-configurations">
  // https://hexo.io/docs/variables.html
  CONFIG.page = {
    sidebar: "",
    isHome : false,
    isPost : true,
    lang   : 'en'
  };
</script>

  <title>R-获取基因长度 | Dean's blog</title>
  






  <noscript>
  <style>
  .use-motion .brand,
  .use-motion .menu-item,
  .sidebar-inner,
  .use-motion .post-block,
  .use-motion .pagination,
  .use-motion .comments,
  .use-motion .post-header,
  .use-motion .post-body,
  .use-motion .collection-header { opacity: initial; }

  .use-motion .site-title,
  .use-motion .site-subtitle {
    opacity: initial;
    top: initial;
  }

  .use-motion .logo-line-before i { left: initial; }
  .use-motion .logo-line-after i { right: initial; }
  </style>
</noscript>

<style>.darkmode--activated{--body-bg-color:#282828;--content-bg-color:#333;--card-bg-color:#555;--text-color:#ccc;--blockquote-color:#bbb;--link-color:#ccc;--link-hover-color:#eee;--brand-color:#ddd;--brand-hover-color:#ddd;--table-row-odd-bg-color:#282828;--table-row-hover-bg-color:#363636;--menu-item-bg-color:#555;--btn-default-bg:#222;--btn-default-color:#ccc;--btn-default-border-color:#555;--btn-default-hover-bg:#666;--btn-default-hover-color:#ccc;--btn-default-hover-border-color:#666;--highlight-background:#282b2e;--highlight-foreground:#a9b7c6;--highlight-gutter-background:#34393d;--highlight-gutter-foreground:#9ca9b6}.darkmode--activated img{opacity:.75}.darkmode--activated img:hover{opacity:.9}.darkmode--activated code{color:#69dbdc;background:0 0}button.darkmode-toggle{z-index:9999}.darkmode-ignore,img{display:flex!important}</style></head>

<body itemscope itemtype="http://schema.org/WebPage">
  <div class="container use-motion">
    <div class="headband"></div>

    <header class="header" itemscope itemtype="http://schema.org/WPHeader">
      <div class="header-inner"><div class="site-brand-container">
  <div class="site-nav-toggle">
    <div class="toggle" aria-label="Toggle navigation bar">
      <span class="toggle-line toggle-line-first"></span>
      <span class="toggle-line toggle-line-middle"></span>
      <span class="toggle-line toggle-line-last"></span>
    </div>
  </div>

  <div class="site-meta">

    <a href="/" class="brand" rel="start">
      <span class="logo-line-before"><i></i></span>
      <h1 class="site-title">Dean's blog</h1>
      <span class="logo-line-after"><i></i></span>
    </a>
  </div>

  <div class="site-nav-right">
    <div class="toggle popup-trigger">
    </div>
  </div>
</div>




<nav class="site-nav">
  <ul id="menu" class="main-menu menu">
        <li class="menu-item menu-item-home">

    <a href="/" rel="section"><i class="fa fa-home fa-fw"></i>Home</a>

  </li>
        <li class="menu-item menu-item-about">

    <a href="/about/" rel="section"><i class="fa fa-user fa-fw"></i>About</a>

  </li>
        <li class="menu-item menu-item-tags">

    <a href="/tags/" rel="section"><i class="fa fa-tags fa-fw"></i>Tags</a>

  </li>
        <li class="menu-item menu-item-archives">

    <a href="/archives/" rel="section"><i class="fa fa-archive fa-fw"></i>Archives</a>

  </li>
  </ul>
</nav>




</div>
    </header>

    
  <div class="back-to-top">
    <i class="fa fa-arrow-up"></i>
    <span>0%</span>
  </div>
  <div class="reading-progress-bar"></div>


    <main class="main">
      <div class="main-inner">
        <div class="content-wrap">
          

          <div class="content post posts-expand">
            

    
  
  
  <article itemscope itemtype="http://schema.org/Article" class="post-block" lang="en">
    <link itemprop="mainEntityOfPage" href="https://thereallda.github.io/2022/02/07/R-Retrieve-Gene-Length/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="image" content="https://avatars.githubusercontent.com/u/55836943?v=4">
      <meta itemprop="name" content="Dean Li">
      <meta itemprop="description" content="">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="Dean's blog">
    </span>
      <header class="post-header">
        <h1 class="post-title" itemprop="name headline">
          R-获取基因长度
        </h1>

        <div class="post-meta">
          
          
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-calendar"></i>
              </span>
              <span class="post-meta-item-text">Posted on</span>
              

              <time title="Created: 2022-02-07 16:12:25 / Modified: 04:48:49" itemprop="dateCreated datePublished" datetime="2022-02-07T16:12:25-05:00">2022-02-07</time>
            </span>

          
            <span class="post-meta-item" title="Views" id="busuanzi_container_page_pv" style="display: none;">
              <span class="post-meta-item-icon">
                <i class="fa fa-eye"></i>
              </span>
              <span class="post-meta-item-text">Views: </span>
              <span id="busuanzi_value_page_pv"></span>
            </span><br>
            <span class="post-meta-item" title="Symbols count in article">
              <span class="post-meta-item-icon">
                <i class="far fa-file-word"></i>
              </span>
                <span class="post-meta-item-text">Symbols count in article: </span>
              <span>2.5k</span>
            </span>
            <span class="post-meta-item" title="Reading time">
              <span class="post-meta-item-icon">
                <i class="far fa-clock"></i>
              </span>
                <span class="post-meta-item-text">Reading time &asymp;</span>
              <span>2 mins.</span>
            </span>

        </div>
      </header>

    
    
    
    <div class="post-body" itemprop="articleBody">

      
        <p>通常，在计算TPM或RPKM/FPKM等基因表达量时，除了基因的counts信息外，我们还需要知道基因的长度。这里所用到的基因长度并不是某个基因在基因组上的完整长度。在基因表达分析中，“基因长度”通常指的是成熟转录本的长度，也就是无内含子的碱基序列。因此，单纯地使用基因的染色体起始和结束坐标相减并不能返回转录本的长度信息。目前，对于基因长度有多种定义，包括：</p>
<p>1. 基因最长转录本；</p>
<p>2. 多个转录本长度的平均值；</p>
<p>3. 非重叠外显子长度之和</p>
<p>4. 非重叠CDS序列长度之和</p>
<p>本文介绍使用<code>gtf</code>文件在R中获取基因长度（<strong>非重叠外显子长度之和</strong>）的方法</p>
<span id="more"></span>

<h1 id="基因长度的定义"><a href="#基因长度的定义" class="headerlink" title="基因长度的定义"></a>基因长度的定义</h1><p><img src="/2022/02/07/R-Retrieve-Gene-Length/paste-CBFC7C07.png"></p>
<p><em>Figure Source</em>: <a target="_blank" rel="noopener" href="https://en.wikipedia.org/wiki/Gene_structure">Gene structure</a></p>
<p>在真核生物，一个基因的DNA序列主要可以分为调控元件和编码序列，而编码序列又可以进一步分为外显子（Exon）和内含子（Intron）序列。经过剪接后，成熟的mRNA一般不包含内含子序列。所以，在统计基因长度时，也只考虑外显子的长度。由于可变剪接的存在，一个基因可能会有多个转录本，在进行基因水平的表达分析时，我们并不会区分各个转录本剪接变体的表达量，而是以基因为单位进行统计。因此，通过合并重叠外显子的区域来获得每个基因的非重叠外显子长度是一种基因长度估算的方法。</p>
<p><img src="/2022/02/07/R-Retrieve-Gene-Length/paste-2878321A.png"><br><em>Figure Source</em>: <a target="_blank" rel="noopener" href="https://davetang.org/muse/2013/01/18/defining-genomic-regions/">Schematic of non-overlapping exons</a></p>
<h1 id="R实现-–-从GTF文件中获取基因长度"><a href="#R实现-–-从GTF文件中获取基因长度" class="headerlink" title="R实现 – 从GTF文件中获取基因长度"></a>R实现 – 从GTF文件中获取基因长度</h1><ol>
<li> 首先，读取计算基因counts时用的GTF文件，并将其转换为<code>TxDb</code>对象；</li>
<li> 然后，提取每个基因的外显子注释信息；</li>
<li> 最后，合并重叠的外显子，计算非重叠外显子的长度作为基因长度（bp）</li>
</ol>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># Non-Overlapping Exon Length ----</span></span><br><span class="line"><span class="comment"># https://www.biostars.org/p/83901/</span></span><br><span class="line"><span class="comment"># First, import the GTF-file that you have also used as input for counting program</span></span><br><span class="line">library(GenomicFeatures)</span><br><span class="line">txdb &lt;- makeTxDbFromGFF(<span class="string">&#x27;gencode.vM23.annotation.gtf&#x27;</span>, format=<span class="string">&#x27;gtf&#x27;</span>)</span><br><span class="line"><span class="comment"># then collect the exons per gene id</span></span><br><span class="line">exons.list.per.gene &lt;- exonsBy(txdb, by=<span class="string">&quot;gene&quot;</span>)</span><br><span class="line"><span class="comment"># then for each gene, reduce all the exons to a set of non overlapping exons, </span></span><br><span class="line"><span class="comment"># calculate their lengths (widths) and sum then</span></span><br><span class="line">exonic.gene.sizes &lt;- <span class="built_in">sum</span>(width(reduce(exons.list.per.gene)))</span><br><span class="line">head(exonic.gene.sizes)</span><br><span class="line"></span><br><span class="line"><span class="comment">##  ENSMUSG00000000001.4 ENSMUSG00000000003.15 ENSMUSG00000000028.15 </span></span><br><span class="line"><span class="comment">##                  3262                   902                  3506 </span></span><br><span class="line"><span class="comment">## ENSMUSG00000000031.16 ENSMUSG00000000037.17 ENSMUSG00000000049.11 </span></span><br><span class="line"><span class="comment">##                  2460                  6397                  1594</span></span><br></pre></td></tr></table></figure>

<p>在使用<code>featureCounts</code>统计基因counts时，其输出的counts.txt文件中通常会包含一列长度信息<code>Length</code></p>
<p>我们可以比较一下使用相同注释文件时，两个程序计算的基因长度是否一致。</p>
<p>查看基因<code>ENSMUSG00000000001.4</code>：</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">$ awk <span class="string">&#x27;NR==2 || /ENSMUSG00000000001.4/&#x27;</span> counts.txt</span><br><span class="line"></span><br><span class="line">Geneid  Chr     Start   End     Strand  Length  a.bam</span><br><span class="line">ENSMUSG00000000001.4    chr3;chr3;chr3;chr3;chr3;chr3;chr3;chr3;chr3    108107280;108109403;108111935;108112473;108115763;108118301;108123542;108123795;108145888    108109316;108109612;108112088;108112602;108115891;108118458;108123683;108123837;108146146    -;-;-;-;-;-;-;-;-       3262    1112</span><br></pre></td></tr></table></figure>

<p>可以发现<code>featureCoutns</code>输出的基因长度与我们计算的一致，这是由于<code>featureCounts</code>也是采用非重叠外显子作为基因长度。</p>
<blockquote>
<p><code>featureCounts</code> Documentation:</p>
<p>“For each meta-feature, the Length column gives the total length of  genomic regions covered by features included in that meta-feature.  Note that this length will be less than the sum of lengths of features  included in the meta-feature when there are features overlapping with  each other.”</p>
</blockquote>
<p>以上就是在R中获取基因长度的方法。</p>
<blockquote>
<p>Ref:</p>
<p><strong>Tutorial: Extract Total Non-Overlapping Exon Length Per Gene With Bioconductor</strong>:<br><a target="_blank" rel="noopener" href="https://www.biostars.org/p/83901/">https://www.biostars.org/p/83901/</a></p>
<p><strong>How featureCounts define the gene length</strong>: <a target="_blank" rel="noopener" href="https://support.bioconductor.org/p/88133/">https://support.bioconductor.org/p/88133/</a></p>
<p><strong>基因长度之多少</strong>: <a target="_blank" rel="noopener" href="https://mp.weixin.qq.com/s/yL6C66C-cMhu_RKiAobCXg">https://mp.weixin.qq.com/s/yL6C66C-cMhu_RKiAobCXg</a></p>
</blockquote>
<p>完。</p>

    </div>

    
    
    
        

<div>
<ul class="post-copyright">
  <li class="post-copyright-author">
    <strong>Post author:  </strong>Dean Li
  </li>
  <li class="post-copyright-link">
    <strong>Post link: </strong>
    <a href="https://thereallda.github.io/2022/02/07/R-Retrieve-Gene-Length/" title="R-获取基因长度">https://thereallda.github.io/2022/02/07/R-Retrieve-Gene-Length/</a>
  </li>
  <li class="post-copyright-license">
    <strong>Copyright Notice:  </strong>All articles in this blog are licensed under <a href="https://creativecommons.org/licenses/by-nc-sa/4.0/" rel="noopener" target="_blank"><i class="fab fa-fw fa-creative-commons"></i>BY-NC-SA</a> unless stating additionally.
  </li>
</ul>
</div>


      <footer class="post-footer">
          <div class="post-tags">
              <a href="/tags/Data-analysis/" rel="tag"># Data_analysis</a>
              <a href="/tags/R/" rel="tag"># R</a>
          </div>

        


        
    <div class="post-nav">
      <div class="post-nav-item">
    <a href="/2022/01/24/%E5%8D%95%E7%BB%86%E8%83%9E%E6%95%B0%E6%8D%AE%E6%A0%BC%E5%BC%8F%E8%BD%AC%E6%8D%A2-Converting-from-loom-or-h5ad-to-Seurat/" rel="prev" title="单细胞数据格式转换:Converting from loom or h5ad to Seurat">
      <i class="fa fa-chevron-left"></i> 单细胞数据格式转换:Converting from loom or h5ad to Seurat
    </a></div>
      <div class="post-nav-item">
    <a href="/2022/02/10/R-RNA-seq-data-simulation/" rel="next" title="RNA-seq测序数据模拟">
      RNA-seq测序数据模拟 <i class="fa fa-chevron-right"></i>
    </a></div>
    </div>
      </footer>
    
  </article>
  
  
  



          </div>
          
    <div class="comments" id="gitalk-container"></div>

<script>
  window.addEventListener('tabs:register', () => {
    let { activeClass } = CONFIG.comments;
    if (CONFIG.comments.storage) {
      activeClass = localStorage.getItem('comments_active') || activeClass;
    }
    if (activeClass) {
      let activeTab = document.querySelector(`a[href="#comment-${activeClass}"]`);
      if (activeTab) {
        activeTab.click();
      }
    }
  });
  if (CONFIG.comments.storage) {
    window.addEventListener('tabs:click', event => {
      if (!event.target.matches('.tabs-comment .tab-content .tab-pane')) return;
      let commentClass = event.target.classList[1];
      localStorage.setItem('comments_active', commentClass);
    });
  }
</script>

        </div>
          
  
  <div class="toggle sidebar-toggle">
    <span class="toggle-line toggle-line-first"></span>
    <span class="toggle-line toggle-line-middle"></span>
    <span class="toggle-line toggle-line-last"></span>
  </div>

  <aside class="sidebar">
    <div class="sidebar-inner">

      <ul class="sidebar-nav motion-element">
        <li class="sidebar-nav-toc">
          Table of Contents
        </li>
        <li class="sidebar-nav-overview">
          Overview
        </li>
      </ul>

      <!--noindex-->
      <div class="post-toc-wrap sidebar-panel">
          <div class="post-toc motion-element"><ol class="nav"><li class="nav-item nav-level-1"><a class="nav-link" href="#%E5%9F%BA%E5%9B%A0%E9%95%BF%E5%BA%A6%E7%9A%84%E5%AE%9A%E4%B9%89"><span class="nav-number">1.</span> <span class="nav-text">基因长度的定义</span></a></li><li class="nav-item nav-level-1"><a class="nav-link" href="#R%E5%AE%9E%E7%8E%B0-%E2%80%93-%E4%BB%8EGTF%E6%96%87%E4%BB%B6%E4%B8%AD%E8%8E%B7%E5%8F%96%E5%9F%BA%E5%9B%A0%E9%95%BF%E5%BA%A6"><span class="nav-number">2.</span> <span class="nav-text">R实现 – 从GTF文件中获取基因长度</span></a></li></ol></div>
      </div>
      <!--/noindex-->

      <div class="site-overview-wrap sidebar-panel">
        <div class="site-author motion-element" itemprop="author" itemscope itemtype="http://schema.org/Person">
    <img class="site-author-image" itemprop="image" alt="Dean Li"
      src="https://avatars.githubusercontent.com/u/55836943?v=4">
  <p class="site-author-name" itemprop="name">Dean Li</p>
  <div class="site-description" itemprop="description"></div>
</div>
<div class="site-state-wrap motion-element">
  <nav class="site-state">
      <div class="site-state-item site-state-posts">
          <a href="/archives/">
        
          <span class="site-state-item-count">66</span>
          <span class="site-state-item-name">posts</span>
        </a>
      </div>
      <div class="site-state-item site-state-tags">
            <a href="/tags/">
          
        <span class="site-state-item-count">21</span>
        <span class="site-state-item-name">tags</span></a>
      </div>
  </nav>
</div>
  <div class="links-of-author motion-element">
      <span class="links-of-author-item">
        <a href="https://github.com/thereallda" title="GitHub → https:&#x2F;&#x2F;github.com&#x2F;thereallda" rel="noopener" target="_blank"><i class="fab fa-github fa-fw"></i>GitHub</a>
      </span>
      <span class="links-of-author-item">
        <a href="mailto:lideanlda@foxmail.com" title="E-Mail → mailto:lideanlda@foxmail.com" rel="noopener" target="_blank"><i class="fa fa-envelope fa-fw"></i>E-Mail</a>
      </span>
      <span class="links-of-author-item">
        <a href="https://www.jianshu.com/u/ab8a1c7d0510" title="简书 → https:&#x2F;&#x2F;www.jianshu.com&#x2F;u&#x2F;ab8a1c7d0510" rel="noopener" target="_blank"><i class="fa fa-book fa-fw"></i>简书</a>
      </span>
  </div>



      </div>

    </div>
  </aside>
  <div id="sidebar-dimmer"></div>


      </div>
    </main>

    <footer class="footer">
      <div class="footer-inner">
        

        

<div class="copyright">
  
  &copy; 
  <span itemprop="copyrightYear">2025</span>
  <span class="with-love">
    <i class="fa fa-heart"></i>
  </span>
  <span class="author" itemprop="copyrightHolder">Dean Li</span>
</div>
  <div class="powered-by">Powered by <a href="https://hexo.io/" class="theme-link" rel="noopener" target="_blank">Hexo</a> & <a href="https://theme-next.org/" class="theme-link" rel="noopener" target="_blank">NexT.Gemini</a>
  </div>



        
<div class="busuanzi-count">
  <script async src="https://busuanzi.ibruce.info/busuanzi/2.3/busuanzi.pure.mini.js"></script>
    <span class="post-meta-item" id="busuanzi_container_site_uv" style="display: none;">
      <span class="post-meta-item-icon">
        <i class="fa fa-user"></i>
      </span>
      <span class="site-uv" title="Total Visitors">
        <span id="busuanzi_value_site_uv"></span>
      </span>
    </span>
    <span class="post-meta-divider">|</span>
    <span class="post-meta-item" id="busuanzi_container_site_pv" style="display: none;">
      <span class="post-meta-item-icon">
        <i class="fa fa-eye"></i>
      </span>
      <span class="site-pv" title="Total Views">
        <span id="busuanzi_value_site_pv"></span>
      </span>
    </span>
</div>








      </div>
    </footer>
  </div>

  
  <script src="/lib/anime.min.js"></script>
  <script src="/lib/velocity/velocity.min.js"></script>
  <script src="/lib/velocity/velocity.ui.min.js"></script>

<script src="/js/utils.js"></script>

<script src="/js/motion.js"></script>


<script src="/js/schemes/pisces.js"></script>


<script src="/js/next-boot.js"></script>




  











<script>
document.querySelectorAll('.pdfobject-container').forEach(element => {
  let url = element.dataset.target;
  let pdfOpenParams = {
    navpanes : 0,
    toolbar  : 0,
    statusbar: 0,
    pagemode : 'thumbs',
    view     : 'FitH'
  };
  let pdfOpenFragment = '#' + Object.entries(pdfOpenParams).map(([key, value]) => `${key}=${encodeURIComponent(value)}`).join('&');
  let fullURL = `/lib/pdf/web/viewer.html?file=${encodeURIComponent(url)}${pdfOpenFragment}`;

  if (NexT.utils.supportsPDFs()) {
    element.innerHTML = `<embed class="pdfobject" src="${url + pdfOpenFragment}" type="application/pdf" style="height: ${element.dataset.height};">`;
  } else {
    element.innerHTML = `<iframe src="${fullURL}" style="height: ${element.dataset.height};" frameborder="0"></iframe>`;
  }
});
</script>




  

  

  
<script src="https://cdn.jsdelivr.net/npm/darkmode-js@1.5.7/lib/darkmode-js.min.js"></script>

<script>
var options = {
  bottom: '64px',
  right: '32px',
  left: 'unset',
  time: '0.5s',
  mixColor: 'transparent',
  backgroundColor: 'transparent',
  buttonColorDark: '#100f2c',
  buttonColorLight: '#fff',
  saveInCookies: true,
  label: '🌓',
  autoMatchOsTheme: true
}
const darkmode = new Darkmode(options);
window.darkmode = darkmode;
darkmode.showWidget();
</script>

<link rel="stylesheet" href="//cdn.jsdelivr.net/npm/gitalk@1/dist/gitalk.min.css">

<script>
NexT.utils.loadComments(document.querySelector('#gitalk-container'), () => {
  NexT.utils.getScript('//cdn.jsdelivr.net/npm/gitalk@1/dist/gitalk.min.js', () => {
    var gitalk = new Gitalk({
      clientID    : 'aea502b6092de143578a',
      clientSecret: '7bb9f5afb99d67ab37d8b296f39782af05dc8614',
      repo        : 'thereallda.GitHub.io',
      owner       : 'thereallda',
      admin       : ['thereallda'],
      id          : '46d3c31ce13be24bf7e7f8f8a63e5914',
        language: '',
      distractionFreeMode: true
    });
    gitalk.render('gitalk-container');
  }, window.Gitalk);
});
</script>

</body>
</html>
