<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
<meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=2">
<meta name="theme-color" content="#222">
<meta name="generator" content="Hexo 5.4.0">
  <link rel="apple-touch-icon" sizes="180x180" href="/images/apple-touch-icon.png">
  <link rel="icon" type="image/png" sizes="32x32" href="/images/favicon-32x32.png">
  <link rel="icon" type="image/png" sizes="16x16" href="/images/favicon-16x16.png">

<link rel="stylesheet" href="/css/main.css">


<link rel="stylesheet" href="/lib/font-awesome/css/all.min.css">

<script id="hexo-configurations">
    var NexT = window.NexT || {};
    var CONFIG = {"hostname":"thereallda.github.io","root":"/","scheme":"Gemini","version":"7.8.0","exturl":false,"sidebar":{"position":"left","display":"post","padding":18,"offset":12,"onmobile":false},"copycode":{"enable":true,"show_result":false,"style":"mac"},"back2top":{"enable":true,"sidebar":false,"scrollpercent":false},"bookmark":{"enable":false,"color":"#222","save":"auto"},"fancybox":false,"mediumzoom":false,"lazyload":false,"pangu":false,"comments":{"style":"tabs","active":"gitalk","storage":true,"lazyload":false,"nav":null,"activeClass":"gitalk"},"algolia":{"hits":{"per_page":10},"labels":{"input_placeholder":"Search for Posts","hits_empty":"We didn't find any results for the search: ${query}","hits_stats":"${hits} results found in ${time} ms"}},"localsearch":{"enable":false,"trigger":"auto","top_n_per_article":1,"unescape":false,"preload":false},"motion":{"enable":true,"async":false,"transition":{"post_block":"fadeIn","post_header":"slideDownIn","post_body":"slideDownIn","coll_header":"slideLeftIn","sidebar":"slideUpIn"}}};
  </script>

  <meta name="description" content="最近在看差异分析当中原始read counts是如何被校正的，自然就不会放过差异分析的经典之一 —— edgeR.">
<meta property="og:type" content="article">
<meta property="og:title" content="理论-edgeR-TMM-normalization">
<meta property="og:url" content="https://thereallda.github.io/2022/01/24/%E7%90%86%E8%AE%BA-edgeR-TMM-normalization/index.html">
<meta property="og:site_name" content="Dean&#39;s blog">
<meta property="og:description" content="最近在看差异分析当中原始read counts是如何被校正的，自然就不会放过差异分析的经典之一 —— edgeR.">
<meta property="og:locale" content="en_US">
<meta property="article:published_time" content="2022-01-24T17:16:36.000Z">
<meta property="article:modified_time" content="2022-01-24T04:17:44.768Z">
<meta property="article:author" content="Dean Li">
<meta property="article:tag" content="RNA-seq">
<meta property="article:tag" content="Statistics">
<meta name="twitter:card" content="summary">

<link rel="canonical" href="https://thereallda.github.io/2022/01/24/%E7%90%86%E8%AE%BA-edgeR-TMM-normalization/">


<script id="page-configurations">
  // https://hexo.io/docs/variables.html
  CONFIG.page = {
    sidebar: "",
    isHome : false,
    isPost : true,
    lang   : 'en'
  };
</script>

  <title>理论-edgeR-TMM-normalization | Dean's blog</title>
  






  <noscript>
  <style>
  .use-motion .brand,
  .use-motion .menu-item,
  .sidebar-inner,
  .use-motion .post-block,
  .use-motion .pagination,
  .use-motion .comments,
  .use-motion .post-header,
  .use-motion .post-body,
  .use-motion .collection-header { opacity: initial; }

  .use-motion .site-title,
  .use-motion .site-subtitle {
    opacity: initial;
    top: initial;
  }

  .use-motion .logo-line-before i { left: initial; }
  .use-motion .logo-line-after i { right: initial; }
  </style>
</noscript>

<style>.darkmode--activated{--body-bg-color:#282828;--content-bg-color:#333;--card-bg-color:#555;--text-color:#ccc;--blockquote-color:#bbb;--link-color:#ccc;--link-hover-color:#eee;--brand-color:#ddd;--brand-hover-color:#ddd;--table-row-odd-bg-color:#282828;--table-row-hover-bg-color:#363636;--menu-item-bg-color:#555;--btn-default-bg:#222;--btn-default-color:#ccc;--btn-default-border-color:#555;--btn-default-hover-bg:#666;--btn-default-hover-color:#ccc;--btn-default-hover-border-color:#666;--highlight-background:#282b2e;--highlight-foreground:#a9b7c6;--highlight-gutter-background:#34393d;--highlight-gutter-foreground:#9ca9b6}.darkmode--activated img{opacity:.75}.darkmode--activated img:hover{opacity:.9}.darkmode--activated code{color:#69dbdc;background:0 0}button.darkmode-toggle{z-index:9999}.darkmode-ignore,img{display:flex!important}</style></head>

<body itemscope itemtype="http://schema.org/WebPage">
  <div class="container use-motion">
    <div class="headband"></div>

    <header class="header" itemscope itemtype="http://schema.org/WPHeader">
      <div class="header-inner"><div class="site-brand-container">
  <div class="site-nav-toggle">
    <div class="toggle" aria-label="Toggle navigation bar">
      <span class="toggle-line toggle-line-first"></span>
      <span class="toggle-line toggle-line-middle"></span>
      <span class="toggle-line toggle-line-last"></span>
    </div>
  </div>

  <div class="site-meta">

    <a href="/" class="brand" rel="start">
      <span class="logo-line-before"><i></i></span>
      <h1 class="site-title">Dean's blog</h1>
      <span class="logo-line-after"><i></i></span>
    </a>
  </div>

  <div class="site-nav-right">
    <div class="toggle popup-trigger">
    </div>
  </div>
</div>




<nav class="site-nav">
  <ul id="menu" class="main-menu menu">
        <li class="menu-item menu-item-home">

    <a href="/" rel="section"><i class="fa fa-home fa-fw"></i>Home</a>

  </li>
        <li class="menu-item menu-item-about">

    <a href="/about/" rel="section"><i class="fa fa-user fa-fw"></i>About</a>

  </li>
        <li class="menu-item menu-item-tags">

    <a href="/tags/" rel="section"><i class="fa fa-tags fa-fw"></i>Tags</a>

  </li>
        <li class="menu-item menu-item-archives">

    <a href="/archives/" rel="section"><i class="fa fa-archive fa-fw"></i>Archives</a>

  </li>
  </ul>
</nav>




</div>
    </header>

    
  <div class="back-to-top">
    <i class="fa fa-arrow-up"></i>
    <span>0%</span>
  </div>
  <div class="reading-progress-bar"></div>


    <main class="main">
      <div class="main-inner">
        <div class="content-wrap">
          

          <div class="content post posts-expand">
            

    
  
  
  <article itemscope itemtype="http://schema.org/Article" class="post-block" lang="en">
    <link itemprop="mainEntityOfPage" href="https://thereallda.github.io/2022/01/24/%E7%90%86%E8%AE%BA-edgeR-TMM-normalization/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="image" content="https://avatars.githubusercontent.com/u/55836943?v=4">
      <meta itemprop="name" content="Dean Li">
      <meta itemprop="description" content="">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="Dean's blog">
    </span>
      <header class="post-header">
        <h1 class="post-title" itemprop="name headline">
          理论-edgeR-TMM-normalization
        </h1>

        <div class="post-meta">
          
          
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-calendar"></i>
              </span>
              <span class="post-meta-item-text">Posted on</span>

              <time title="Created: 2022-01-24 12:16:36" itemprop="dateCreated datePublished" datetime="2022-01-24T12:16:36-05:00">2022-01-24</time>
            </span>
              <span class="post-meta-item">
                <span class="post-meta-item-icon">
                  <i class="far fa-calendar-check"></i>
                </span>
                <span class="post-meta-item-text">Edited on</span>
                <time title="Modified: 2022-01-23 23:17:44" itemprop="dateModified" datetime="2022-01-23T23:17:44-05:00">2022-01-23</time>
              </span>

          
            <span class="post-meta-item" title="Views" id="busuanzi_container_page_pv" style="display: none;">
              <span class="post-meta-item-icon">
                <i class="fa fa-eye"></i>
              </span>
              <span class="post-meta-item-text">Views: </span>
              <span id="busuanzi_value_page_pv"></span>
            </span><br>
            <span class="post-meta-item" title="Symbols count in article">
              <span class="post-meta-item-icon">
                <i class="far fa-file-word"></i>
              </span>
                <span class="post-meta-item-text">Symbols count in article: </span>
              <span>5.5k</span>
            </span>
            <span class="post-meta-item" title="Reading time">
              <span class="post-meta-item-icon">
                <i class="far fa-clock"></i>
              </span>
                <span class="post-meta-item-text">Reading time &asymp;</span>
              <span>5 mins.</span>
            </span>

        </div>
      </header>

    
    
    
    <div class="post-body" itemprop="articleBody">

      
        <p>最近在看差异分析当中原始read counts是如何被校正的，自然就不会放过差异分析的经典之一 —— <code>edgeR</code>.</p>
<span id="more"></span>

<p><code>edgeR</code>使用的校正方法称为trimmed mean of M values (TMM)，其前提假设为样本对照组和处理组间绝大多数基因表达不发生差异。</p>
<blockquote>
<p>如何界定绝大多数基因这一点我个人还没有看到一个量化的指标，是50%还是80%才算绝大多数。</p>
</blockquote>
<p><code>edgeR</code>的TMM校正方法其实在其分析流程中就是一句命令而已</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">y &lt;- calcNormFactors(y)</span><br></pre></td></tr></table></figure>

<p>但由于我十分好奇其背后计算的原理和计算过程，我便搜索一番，才发现无论是中文还是英文的帖子对于TMM的具体运算步骤及代码都没有很好的整理。因此，本文记录我通过阅读<a target="_blank" rel="noopener" href="https://genomebiology.biomedcentral.com/articles/10.1186/gb-2010-11-3-r25">TMM提出的原文</a>和<code>edgeR</code><a target="_blank" rel="noopener" href="https://rdrr.io/bioc/edgeR/src/R/calcNormFactors.R">源码</a>所了解到的TMM校正。</p>
<h2 id="TMM校正示例"><a href="#TMM校正示例" class="headerlink" title="TMM校正示例"></a>TMM校正示例</h2><p>我们通过<code>airway</code>数据展示实际TMM校正的过程</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># TMM normalization</span></span><br><span class="line">data(<span class="string">&quot;airway&quot;</span>, package=<span class="string">&#x27;airway&#x27;</span>)</span><br><span class="line"></span><br><span class="line">set.seed(<span class="number">123</span>)</span><br><span class="line">counts1 &lt;- as.data.frame(assay(airway)[sample(<span class="number">1</span>:nrow(airway), <span class="number">5000</span>),<span class="number">1</span>:<span class="number">4</span>])</span><br></pre></td></tr></table></figure>



<h3 id="Step-1-选择参考样本"><a href="#Step-1-选择参考样本" class="headerlink" title="Step 1: 选择参考样本"></a>Step 1: 选择参考样本</h3><p>TMM normalization首先需要选择一个参考样本，以它为基准进行校正。</p>
<p>默认下，参考样本的选择是通过比较每个样本的<strong>CPM (counts per million)<strong>的上四分位数与所有样本</strong>CPM</strong>的平均上四分位数之间的差值，找出差值最小的样本作为参考样本。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># library size of each samples</span></span><br><span class="line">lib.size &lt;- colSums(counts1)</span><br><span class="line"></span><br><span class="line"><span class="comment"># library size normalized read count for each gene in each sample</span></span><br><span class="line">counts1.cpm &lt;- t(t(counts1)/lib.size) * <span class="number">1e6</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># reference library for TMM normalization</span></span><br><span class="line">f75 &lt;- apply(counts1.cpm, <span class="number">2</span>, <span class="keyword">function</span>(x) quantile(x, probs = <span class="number">0.75</span>))</span><br><span class="line"><span class="comment"># select the one with upper quartile closest to mean upper quartile</span></span><br><span class="line">refColumn &lt;- which.min(<span class="built_in">abs</span>(f75 - mean(f75)))</span><br><span class="line">&gt; refColumn</span><br><span class="line">SRR1039512 </span><br><span class="line">         <span class="number">3</span></span><br></pre></td></tr></table></figure>

<p>在这个例子中<code>refColumn</code>是<code>SRR1039512</code></p>
<blockquote>
<p>在<code>edgeR::calcNormFactors()</code>中，我们也可以通过<code>refColumn=</code>参数指定参考样本</p>
</blockquote>
<h3 id="Step-2-calculation-of-sample-reference-pairwise-M-and-A"><a href="#Step-2-calculation-of-sample-reference-pairwise-M-and-A" class="headerlink" title="Step 2: calculation of sample-reference pairwise M and A"></a>Step 2: calculation of sample-reference pairwise M and A</h3><p>接着，在参考样本和非参考样本间两两计算校正因子（normalization factors）。</p>
<p>我们首先需要计算参考样本和非参考样本间的Fold change (M)和平均表达量 (A)</p>
<p>$M = log2\frac{non-reference\ sample\ count}{reference\ sample\ count}$</p>
<p>$A=\frac{log2(non-reference\ sample\ count)\ +\ log2(reference\ sample\ count)}{2}$</p>
<p><strong>注意这里使用的count都是校正过文库大小的</strong></p>
<p>以下代码用到的变量名含义：</p>
<p><code>obs</code>: 非参考样本的原始count</p>
<p><code>ref</code>: 参考样本的原始count</p>
<p><code>libsize.obs</code>: 非参考样本的原始文库大小</p>
<p><code>libsize.ref</code>: 参考样本的原始文库大小</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># 第一列是非参考样本</span></span><br><span class="line">obs &lt;- <span class="built_in">as.numeric</span>(counts1[, <span class="number">1</span>])</span><br><span class="line">ref &lt;- <span class="built_in">as.numeric</span>(counts1[, refColumn])</span><br><span class="line">libsize.obs &lt;- lib.size[<span class="number">1</span>]</span><br><span class="line">libsize.ref &lt;- lib.size[refColumn]</span><br></pre></td></tr></table></figure>

<p>分别计算M value和A value，为了对M value加权，我们还需要通过delta method估计渐近方差</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># M value: log ratio of expression, accounting for library size</span></span><br><span class="line">M &lt;- log2((obs/libsize.obs)/(ref/libsize.ref))</span><br><span class="line"><span class="comment"># A value:absolute expression</span></span><br><span class="line">A &lt;- (log2(obs/libsize.obs) + log2(ref/libsize.ref))/<span class="number">2</span></span><br><span class="line"><span class="comment"># estimated asymptotic variance</span></span><br><span class="line">v &lt;- (libsize.obs - obs)/libsize.obs/obs + (libsize.ref - ref)/libsize.ref/ref</span><br></pre></td></tr></table></figure>

<p>保留M和A值均为有限值的基因，并过滤极低表达量的基因</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">Acutoff = -<span class="number">1e10</span></span><br><span class="line"><span class="comment">#	remove infinite values, cutoff based on A</span></span><br><span class="line">fin &lt;- <span class="built_in">is.finite</span>(M) &amp; <span class="built_in">is.finite</span>(A) &amp; (A &gt; Acutoff)</span><br><span class="line">M &lt;- M[fin]</span><br><span class="line">A &lt;- A[fin]</span><br><span class="line">v &lt;- v[fin]</span><br></pre></td></tr></table></figure>



<h3 id="Step-3-trimmed-mean-of-M-values"><a href="#Step-3-trimmed-mean-of-M-values" class="headerlink" title="Step 3: trimmed mean of M values"></a>Step 3: trimmed mean of M values</h3><p>接着，我们对M和A值进行双重截值，截掉M值排在前30%和后30%，A值排在前5%和后5%的基因，计算中间这部分基因M值的加权平均值</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br></pre></td><td class="code"><pre><span class="line">logratioTrim &lt;- 0.3</span><br><span class="line">sumTrim &lt;- 0.05</span><br><span class="line"><span class="comment"># Double trim the upper and lower percentages of the data</span></span><br><span class="line"><span class="comment"># trim M values by 30% and A values by 5%</span></span><br><span class="line">  </span><br><span class="line">  n &lt;- <span class="built_in">length</span>(M)</span><br><span class="line">  loM &lt;- <span class="built_in">floor</span>(n * logratioTrim) + <span class="number">1</span></span><br><span class="line">  hiM &lt;- n + <span class="number">1</span> - loM</span><br><span class="line">  loA &lt;- <span class="built_in">floor</span>(n * sumTrim) + <span class="number">1</span></span><br><span class="line">  hiA &lt;- n + <span class="number">1</span> - loA</span><br><span class="line">  </span><br><span class="line">  keep &lt;- (rank(M)&gt;=loM &amp; rank(M)&lt;=hiM) &amp; (rank(A)&gt;=loA &amp; rank(A)&lt;=hiA)</span><br><span class="line">  </span><br><span class="line">  <span class="comment"># Weighted mean of M after trimming </span></span><br><span class="line">  f &lt;- <span class="built_in">sum</span>(M[keep]/v[keep], na.rm=<span class="literal">TRUE</span>) / <span class="built_in">sum</span>(<span class="number">1</span>/v[keep], na.rm=<span class="literal">TRUE</span>)</span><br><span class="line">  f &lt;- 2^f</span><br><span class="line">  <span class="comment">#	Factors should multiple to one</span></span><br><span class="line">  f &lt;- f/<span class="built_in">exp</span>(mean(<span class="built_in">log</span>(f)))</span><br><span class="line"></span><br><span class="line">  <span class="comment">#	Output</span></span><br><span class="line">  <span class="built_in">names</span>(f) &lt;- colnames(counts1)</span><br></pre></td></tr></table></figure>



<p>上述步骤是计算第一个样本与参考样本的校正因子，下面我们通过一个循环计算所有样本的校正因子.</p>
<p><em>以下循环在选取<code>refColumn</code>后开始</em></p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br></pre></td><td class="code"><pre><span class="line">nsamples &lt;- ncol(counts1)</span><br><span class="line"></span><br><span class="line">logratioTrim &lt;- 0.3</span><br><span class="line">sumTrim &lt;- 0.05</span><br><span class="line">Acutoff = -<span class="number">1e10</span></span><br><span class="line"></span><br><span class="line">f &lt;- rep_len(<span class="literal">NA_real_</span>, nsamples)</span><br><span class="line"><span class="keyword">for</span> (i <span class="keyword">in</span> <span class="number">1</span>:nsamples) &#123;</span><br><span class="line">  obs &lt;- <span class="built_in">as.numeric</span>(counts1[, i])</span><br><span class="line">  ref &lt;- <span class="built_in">as.numeric</span>(counts1[, refColumn])</span><br><span class="line">  libsize.obs &lt;- lib.size[i]</span><br><span class="line">  libsize.ref &lt;- lib.size[refColumn]</span><br><span class="line">  </span><br><span class="line">  <span class="comment"># M value: log ratio of expression, accounting for library size</span></span><br><span class="line">  M &lt;- log2((obs/libsize.obs)/(ref/libsize.ref))</span><br><span class="line">  <span class="comment"># A value:absolute expression</span></span><br><span class="line">  A &lt;- (log2(obs/libsize.obs) + log2(ref/libsize.ref))/<span class="number">2</span></span><br><span class="line">  <span class="comment"># estimated asymptotic variance</span></span><br><span class="line">  v &lt;- (libsize.obs - obs)/libsize.obs/obs + (libsize.ref - ref)/libsize.ref/ref</span><br><span class="line">  <span class="comment">#	remove infinite values, cutoff based on A</span></span><br><span class="line">  fin &lt;- <span class="built_in">is.finite</span>(M) &amp; <span class="built_in">is.finite</span>(A) &amp; (A &gt; Acutoff)</span><br><span class="line">  </span><br><span class="line">  M &lt;- M[fin]</span><br><span class="line">  A &lt;- A[fin]</span><br><span class="line">  v &lt;- v[fin]</span><br><span class="line">  </span><br><span class="line">  <span class="comment"># Double trim the upper and lower percentages of the data</span></span><br><span class="line">  <span class="comment"># trim M values by 30% and A values by 5%</span></span><br><span class="line">  </span><br><span class="line">  n &lt;- <span class="built_in">length</span>(M)</span><br><span class="line">  loM &lt;- <span class="built_in">floor</span>(n * logratioTrim) + <span class="number">1</span></span><br><span class="line">  hiM &lt;- n + <span class="number">1</span> - loM</span><br><span class="line">  loA &lt;- <span class="built_in">floor</span>(n * sumTrim) + <span class="number">1</span></span><br><span class="line">  hiA &lt;- n + <span class="number">1</span> - loA</span><br><span class="line">  </span><br><span class="line">  keep &lt;- (rank(M)&gt;=loM &amp; rank(M)&lt;=hiM) &amp; (rank(A)&gt;=loA &amp; rank(A)&lt;=hiA)</span><br><span class="line">  </span><br><span class="line">  <span class="comment"># Weighted mean of M after trimming </span></span><br><span class="line">  f[i] &lt;- <span class="built_in">sum</span>(M[keep]/v[keep], na.rm=<span class="literal">TRUE</span>) / <span class="built_in">sum</span>(<span class="number">1</span>/v[keep], na.rm=<span class="literal">TRUE</span>)</span><br><span class="line">  f[i] &lt;- <span class="number">2</span>^f[i]</span><br><span class="line">&#125;</span><br><span class="line"></span><br><span class="line"><span class="comment">#	Factors should multiple to one</span></span><br><span class="line">f &lt;- f/<span class="built_in">exp</span>(mean(<span class="built_in">log</span>(f)))</span><br><span class="line"></span><br><span class="line"><span class="comment">#	Output</span></span><br><span class="line"><span class="built_in">names</span>(f) &lt;- colnames(counts1)</span><br><span class="line"></span><br><span class="line">f</span><br><span class="line"><span class="comment">#SRR1039508 SRR1039509 SRR1039512 SRR1039513 </span></span><br><span class="line"><span class="comment"># 1.0225375  1.0111616  0.9920515  0.9749133 </span></span><br></pre></td></tr></table></figure>



<p>我们与<code>edgeR::calcNormFactors()</code>比较一下</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">library(edgeR)</span><br><span class="line">y &lt;- DGEList(counts = counts1)</span><br><span class="line">y &lt;- calcNormFactors(y)</span><br><span class="line">nf &lt;- y$samples$norm.factors</span><br><span class="line"></span><br><span class="line">f == nf</span><br><span class="line"><span class="comment">#SRR1039508 SRR1039509 SRR1039512 SRR1039513 </span></span><br><span class="line"><span class="comment">#      TRUE       TRUE       TRUE       TRUE</span></span><br></pre></td></tr></table></figure>



<h3 id="Step-4-TMM-normalized-counts"><a href="#Step-4-TMM-normalized-counts" class="headerlink" title="Step 4: TMM normalized counts"></a>Step 4: TMM normalized counts</h3><p>最后，我们获取TMM校正后的read counts。实际上，上述计算的校正因子是对文库大小的校正，<code>edgeR</code>再利用校正后的文库大小对read counts进行校正。</p>
<p>${TMM\ Normalized\ Counts = \frac{Raw \ Count\ *\ 10^6}{Library\ size\ *\ NormFactor}}$</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># TMM normalized counts</span></span><br><span class="line">counts1.tmm &lt;- t(t(counts1) / (lib.size * f)) * <span class="number">1e6</span></span><br><span class="line">counts1.tmm &lt;- <span class="built_in">round</span>(counts1.tmm, <span class="number">4</span>)</span><br><span class="line"></span><br><span class="line">head(counts1.tmm)</span><br></pre></td></tr></table></figure>

<table>
<thead>
<tr>
<th align="left"></th>
<th align="right">SRR1039508</th>
<th align="right">SRR1039509</th>
<th align="right">SRR1039512</th>
<th align="right">SRR1039513</th>
</tr>
</thead>
<tbody><tr>
<td align="left">ENSG00000260166</td>
<td align="right">0.0000</td>
<td align="right">0.0000</td>
<td align="right">0.000</td>
<td align="right">0.000</td>
</tr>
<tr>
<td align="left">ENSG00000266931</td>
<td align="right">0.0000</td>
<td align="right">0.0000</td>
<td align="right">0.000</td>
<td align="right">0.000</td>
</tr>
<tr>
<td align="left">ENSG00000104774</td>
<td align="right">1375.7505</td>
<td align="right">1510.2195</td>
<td align="right">1431.252</td>
<td align="right">1301.323</td>
</tr>
<tr>
<td align="left">ENSG00000267583</td>
<td align="right">0.0000</td>
<td align="right">0.7924</td>
<td align="right">0.000</td>
<td align="right">0.000</td>
</tr>
<tr>
<td align="left">ENSG00000227581</td>
<td align="right">0.7095</td>
<td align="right">0.0000</td>
<td align="right">0.000</td>
<td align="right">0.000</td>
</tr>
<tr>
<td align="left">ENSG00000227317</td>
<td align="right">0.0000</td>
<td align="right">0.0000</td>
<td align="right">0.000</td>
<td align="right">0.000</td>
</tr>
</tbody></table>
<p>使用<code>edgeR::calcNormFactors()</code>计算校正因子后，我们可以使用<code>edgeR::cpm()</code>提取校正后的counts矩阵</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># by edgeR</span></span><br><span class="line">counts.tmm.edger &lt;- <span class="built_in">round</span>(cpm(y), <span class="number">4</span>)</span><br><span class="line"></span><br><span class="line">identical(counts1.tmm, counts.tmm.edger)</span><br><span class="line"><span class="comment"># [1] TRUE</span></span><br></pre></td></tr></table></figure>

<p>以上就是TMM校正的计算过程。</p>
<p>完。</p>
<blockquote>
<p>Ref:</p>
<p><strong>A scaling normalization method for differential expression analysis of RNA-seq data</strong>: <a target="_blank" rel="noopener" href="https://doi.org/10.1186/gb-2010-11-3-r25">https://doi.org/10.1186/gb-2010-11-3-r25</a></p>
<p><strong>calcNormFactors.R</strong>: <a target="_blank" rel="noopener" href="https://rdrr.io/bioc/edgeR/src/R/calcNormFactors.R">https://rdrr.io/bioc/edgeR/src/R/calcNormFactors.R</a></p>
<p><strong>Gene expression units explained: RPM, RPKM, FPKM, TPM, DESeq, TMM, SCnorm, GeTMM, and ComBat-Seq</strong>: <a target="_blank" rel="noopener" href="https://www.reneshbedre.com/blog/expression_units.html">https://www.reneshbedre.com/blog/expression_units.html</a></p>
<p><strong>edgeR提供的TMM归一化算法详解</strong>: <a target="_blank" rel="noopener" href="https://cloud.tencent.com/developer/article/1625225">https://cloud.tencent.com/developer/article/1625225</a></p>
</blockquote>

    </div>

    
    
    
        

<div>
<ul class="post-copyright">
  <li class="post-copyright-author">
    <strong>Post author:  </strong>Dean Li
  </li>
  <li class="post-copyright-link">
    <strong>Post link: </strong>
    <a href="https://thereallda.github.io/2022/01/24/%E7%90%86%E8%AE%BA-edgeR-TMM-normalization/" title="理论-edgeR-TMM-normalization">https://thereallda.github.io/2022/01/24/理论-edgeR-TMM-normalization/</a>
  </li>
  <li class="post-copyright-license">
    <strong>Copyright Notice:  </strong>All articles in this blog are licensed under <a href="https://creativecommons.org/licenses/by-nc-sa/4.0/" rel="noopener" target="_blank"><i class="fab fa-fw fa-creative-commons"></i>BY-NC-SA</a> unless stating additionally.
  </li>
</ul>
</div>


      <footer class="post-footer">
          <div class="post-tags">
              <a href="/tags/RNA-seq/" rel="tag"># RNA-seq</a>
              <a href="/tags/Statistics/" rel="tag"># Statistics</a>
          </div>

        


        
    <div class="post-nav">
      <div class="post-nav-item">
    <a href="/2022/01/24/%E7%90%86%E8%AE%BA-%E7%AE%80%E8%BF%B0RPKM-FPKM-AND-TPM/" rel="prev" title="理论-简述RPKM-FPKM-AND-TPM">
      <i class="fa fa-chevron-left"></i> 理论-简述RPKM-FPKM-AND-TPM
    </a></div>
      <div class="post-nav-item">
    <a href="/2022/01/24/%E7%90%86%E8%AE%BA-DESeq2-normalization/" rel="next" title="理论-DESeq2-normalization">
      理论-DESeq2-normalization <i class="fa fa-chevron-right"></i>
    </a></div>
    </div>
      </footer>
    
  </article>
  
  
  



          </div>
          
    <div class="comments" id="gitalk-container"></div>

<script>
  window.addEventListener('tabs:register', () => {
    let { activeClass } = CONFIG.comments;
    if (CONFIG.comments.storage) {
      activeClass = localStorage.getItem('comments_active') || activeClass;
    }
    if (activeClass) {
      let activeTab = document.querySelector(`a[href="#comment-${activeClass}"]`);
      if (activeTab) {
        activeTab.click();
      }
    }
  });
  if (CONFIG.comments.storage) {
    window.addEventListener('tabs:click', event => {
      if (!event.target.matches('.tabs-comment .tab-content .tab-pane')) return;
      let commentClass = event.target.classList[1];
      localStorage.setItem('comments_active', commentClass);
    });
  }
</script>

        </div>
          
  
  <div class="toggle sidebar-toggle">
    <span class="toggle-line toggle-line-first"></span>
    <span class="toggle-line toggle-line-middle"></span>
    <span class="toggle-line toggle-line-last"></span>
  </div>

  <aside class="sidebar">
    <div class="sidebar-inner">

      <ul class="sidebar-nav motion-element">
        <li class="sidebar-nav-toc">
          Table of Contents
        </li>
        <li class="sidebar-nav-overview">
          Overview
        </li>
      </ul>

      <!--noindex-->
      <div class="post-toc-wrap sidebar-panel">
          <div class="post-toc motion-element"><ol class="nav"><li class="nav-item nav-level-2"><a class="nav-link" href="#TMM%E6%A0%A1%E6%AD%A3%E7%A4%BA%E4%BE%8B"><span class="nav-number">1.</span> <span class="nav-text">TMM校正示例</span></a><ol class="nav-child"><li class="nav-item nav-level-3"><a class="nav-link" href="#Step-1-%E9%80%89%E6%8B%A9%E5%8F%82%E8%80%83%E6%A0%B7%E6%9C%AC"><span class="nav-number">1.1.</span> <span class="nav-text">Step 1: 选择参考样本</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#Step-2-calculation-of-sample-reference-pairwise-M-and-A"><span class="nav-number">1.2.</span> <span class="nav-text">Step 2: calculation of sample-reference pairwise M and A</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#Step-3-trimmed-mean-of-M-values"><span class="nav-number">1.3.</span> <span class="nav-text">Step 3: trimmed mean of M values</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#Step-4-TMM-normalized-counts"><span class="nav-number">1.4.</span> <span class="nav-text">Step 4: TMM normalized counts</span></a></li></ol></li></ol></div>
      </div>
      <!--/noindex-->

      <div class="site-overview-wrap sidebar-panel">
        <div class="site-author motion-element" itemprop="author" itemscope itemtype="http://schema.org/Person">
    <img class="site-author-image" itemprop="image" alt="Dean Li"
      src="https://avatars.githubusercontent.com/u/55836943?v=4">
  <p class="site-author-name" itemprop="name">Dean Li</p>
  <div class="site-description" itemprop="description"></div>
</div>
<div class="site-state-wrap motion-element">
  <nav class="site-state">
      <div class="site-state-item site-state-posts">
          <a href="/archives/">
        
          <span class="site-state-item-count">66</span>
          <span class="site-state-item-name">posts</span>
        </a>
      </div>
      <div class="site-state-item site-state-tags">
            <a href="/tags/">
          
        <span class="site-state-item-count">21</span>
        <span class="site-state-item-name">tags</span></a>
      </div>
  </nav>
</div>
  <div class="links-of-author motion-element">
      <span class="links-of-author-item">
        <a href="https://github.com/thereallda" title="GitHub → https:&#x2F;&#x2F;github.com&#x2F;thereallda" rel="noopener" target="_blank"><i class="fab fa-github fa-fw"></i>GitHub</a>
      </span>
      <span class="links-of-author-item">
        <a href="mailto:lideanlda@foxmail.com" title="E-Mail → mailto:lideanlda@foxmail.com" rel="noopener" target="_blank"><i class="fa fa-envelope fa-fw"></i>E-Mail</a>
      </span>
      <span class="links-of-author-item">
        <a href="https://scholar.google.com/citations?user=zMno6_8AAAAJ" title="Google → https:&#x2F;&#x2F;scholar.google.com&#x2F;citations?user&#x3D;zMno6_8AAAAJ" rel="noopener" target="_blank"><i class="fab fa-google fa-fw"></i>Google</a>
      </span>
      <span class="links-of-author-item">
        <a href="https://www.jianshu.com/u/ab8a1c7d0510" title="简书 → https:&#x2F;&#x2F;www.jianshu.com&#x2F;u&#x2F;ab8a1c7d0510" rel="noopener" target="_blank"><i class="fa fa-book fa-fw"></i>简书</a>
      </span>
  </div>



      </div>

    </div>
  </aside>
  <div id="sidebar-dimmer"></div>


      </div>
    </main>

    <footer class="footer">
      <div class="footer-inner">
        

        

<div class="copyright">
  
  &copy; 
  <span itemprop="copyrightYear">2025</span>
  <span class="with-love">
    <i class="fa fa-heart"></i>
  </span>
  <span class="author" itemprop="copyrightHolder">Dean Li</span>
</div>
  <div class="powered-by">Powered by <a href="https://hexo.io/" class="theme-link" rel="noopener" target="_blank">Hexo</a> & <a href="https://theme-next.org/" class="theme-link" rel="noopener" target="_blank">NexT.Gemini</a>
  </div>



        
<div class="busuanzi-count">
  <script async src="https://busuanzi.ibruce.info/busuanzi/2.3/busuanzi.pure.mini.js"></script>
    <span class="post-meta-item" id="busuanzi_container_site_uv" style="display: none;">
      <span class="post-meta-item-icon">
        <i class="fa fa-user"></i>
      </span>
      <span class="site-uv" title="Total Visitors">
        <span id="busuanzi_value_site_uv"></span>
      </span>
    </span>
    <span class="post-meta-divider">|</span>
    <span class="post-meta-item" id="busuanzi_container_site_pv" style="display: none;">
      <span class="post-meta-item-icon">
        <i class="fa fa-eye"></i>
      </span>
      <span class="site-pv" title="Total Views">
        <span id="busuanzi_value_site_pv"></span>
      </span>
    </span>
</div>








      </div>
    </footer>
  </div>

  
  <script src="/lib/anime.min.js"></script>
  <script src="/lib/velocity/velocity.min.js"></script>
  <script src="/lib/velocity/velocity.ui.min.js"></script>

<script src="/js/utils.js"></script>

<script src="/js/motion.js"></script>


<script src="/js/schemes/pisces.js"></script>


<script src="/js/next-boot.js"></script>




  











<script>
document.querySelectorAll('.pdfobject-container').forEach(element => {
  let url = element.dataset.target;
  let pdfOpenParams = {
    navpanes : 0,
    toolbar  : 0,
    statusbar: 0,
    pagemode : 'thumbs',
    view     : 'FitH'
  };
  let pdfOpenFragment = '#' + Object.entries(pdfOpenParams).map(([key, value]) => `${key}=${encodeURIComponent(value)}`).join('&');
  let fullURL = `/lib/pdf/web/viewer.html?file=${encodeURIComponent(url)}${pdfOpenFragment}`;

  if (NexT.utils.supportsPDFs()) {
    element.innerHTML = `<embed class="pdfobject" src="${url + pdfOpenFragment}" type="application/pdf" style="height: ${element.dataset.height};">`;
  } else {
    element.innerHTML = `<iframe src="${fullURL}" style="height: ${element.dataset.height};" frameborder="0"></iframe>`;
  }
});
</script>




  

  

  
<script src="https://cdn.jsdelivr.net/npm/darkmode-js@1.5.7/lib/darkmode-js.min.js"></script>

<script>
var options = {
  bottom: '64px',
  right: '32px',
  left: 'unset',
  time: '0.5s',
  mixColor: 'transparent',
  backgroundColor: 'transparent',
  buttonColorDark: '#100f2c',
  buttonColorLight: '#fff',
  saveInCookies: true,
  label: '🌓',
  autoMatchOsTheme: true
}
const darkmode = new Darkmode(options);
window.darkmode = darkmode;
darkmode.showWidget();
</script>

<link rel="stylesheet" href="//cdn.jsdelivr.net/npm/gitalk@1/dist/gitalk.min.css">

<script>
NexT.utils.loadComments(document.querySelector('#gitalk-container'), () => {
  NexT.utils.getScript('//cdn.jsdelivr.net/npm/gitalk@1/dist/gitalk.min.js', () => {
    var gitalk = new Gitalk({
      clientID    : 'aea502b6092de143578a',
      clientSecret: '7bb9f5afb99d67ab37d8b296f39782af05dc8614',
      repo        : 'thereallda.GitHub.io',
      owner       : 'thereallda',
      admin       : ['thereallda'],
      id          : 'f33608fa53b5b1c2fd2513e8e3da5e79',
        language: '',
      distractionFreeMode: true
    });
    gitalk.render('gitalk-container');
  }, window.Gitalk);
});
</script>

</body>
</html>
