<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
<meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=2">
<meta name="theme-color" content="#222">
<meta name="generator" content="Hexo 5.4.0">
  <link rel="apple-touch-icon" sizes="180x180" href="/images/apple-touch-icon.png">
  <link rel="icon" type="image/png" sizes="32x32" href="/images/favicon-32x32.png">
  <link rel="icon" type="image/png" sizes="16x16" href="/images/favicon-16x16.png">

<link rel="stylesheet" href="/css/main.css">


<link rel="stylesheet" href="/lib/font-awesome/css/all.min.css">

<script id="hexo-configurations">
    var NexT = window.NexT || {};
    var CONFIG = {"hostname":"thereallda.github.io","root":"/","scheme":"Gemini","version":"7.8.0","exturl":false,"sidebar":{"position":"left","display":"post","padding":18,"offset":12,"onmobile":false},"copycode":{"enable":true,"show_result":false,"style":"mac"},"back2top":{"enable":true,"sidebar":false,"scrollpercent":false},"bookmark":{"enable":false,"color":"#222","save":"auto"},"fancybox":false,"mediumzoom":false,"lazyload":false,"pangu":false,"comments":{"style":"tabs","active":"gitalk","storage":true,"lazyload":false,"nav":null,"activeClass":"gitalk"},"algolia":{"hits":{"per_page":10},"labels":{"input_placeholder":"Search for Posts","hits_empty":"We didn't find any results for the search: ${query}","hits_stats":"${hits} results found in ${time} ms"}},"localsearch":{"enable":false,"trigger":"auto","top_n_per_article":1,"unescape":false,"preload":false},"motion":{"enable":true,"async":false,"transition":{"post_block":"fadeIn","post_header":"slideDownIn","post_body":"slideDownIn","coll_header":"slideLeftIn","sidebar":"slideUpIn"}}};
  </script>

  <meta name="description" content="在评估不同软件性能的时候，我们会需要模拟一些数据。由于模拟数据当中的情况是已知的，例如差异表达基因的数目。因此，通过比较不同软件在模拟数据上的效果，我们可以获得软件的量化性能指标，例如灵敏度、特异性和准确度等。 本文根据 DESeq2 文章中的方法记录如何进行简单的基于负二项分布（Negative Binomial distribution）模拟RNA-seq基因表达数据。">
<meta property="og:type" content="article">
<meta property="og:title" content="RNA-seq测序数据模拟">
<meta property="og:url" content="https://thereallda.github.io/2022/02/10/R-RNA-seq-data-simulation/index.html">
<meta property="og:site_name" content="Dean&#39;s blog">
<meta property="og:description" content="在评估不同软件性能的时候，我们会需要模拟一些数据。由于模拟数据当中的情况是已知的，例如差异表达基因的数目。因此，通过比较不同软件在模拟数据上的效果，我们可以获得软件的量化性能指标，例如灵敏度、特异性和准确度等。 本文根据 DESeq2 文章中的方法记录如何进行简单的基于负二项分布（Negative Binomial distribution）模拟RNA-seq基因表达数据。">
<meta property="og:locale" content="en_US">
<meta property="og:image" content="https://thereallda.github.io/2022/02/10/R-RNA-seq-data-simulation/paste-E5B09090.png">
<meta property="og:image" content="https://thereallda.github.io/2022/02/10/R-RNA-seq-data-simulation/unnamed-chunk-6-1.png">
<meta property="og:image" content="https://thereallda.github.io/2022/02/10/R-RNA-seq-data-simulation/unnamed-chunk-6-2.png">
<meta property="og:image" content="https://thereallda.github.io/2022/02/10/R-RNA-seq-data-simulation/unnamed-chunk-7-1.png">
<meta property="og:image" content="https://thereallda.github.io/2022/02/10/R-RNA-seq-data-simulation/unnamed-chunk-7-2.png">
<meta property="og:image" content="https://thereallda.github.io/2022/02/10/R-RNA-seq-data-simulation/image-20220210215514028.png">
<meta property="article:published_time" content="2022-02-10T13:49:29.000Z">
<meta property="article:modified_time" content="2022-02-11T01:27:51.497Z">
<meta property="article:author" content="Dean Li">
<meta property="article:tag" content="R">
<meta property="article:tag" content="RNA-seq">
<meta property="article:tag" content="Data_analysis">
<meta name="twitter:card" content="summary">
<meta name="twitter:image" content="https://thereallda.github.io/2022/02/10/R-RNA-seq-data-simulation/paste-E5B09090.png">

<link rel="canonical" href="https://thereallda.github.io/2022/02/10/R-RNA-seq-data-simulation/">


<script id="page-configurations">
  // https://hexo.io/docs/variables.html
  CONFIG.page = {
    sidebar: "",
    isHome : false,
    isPost : true,
    lang   : 'en'
  };
</script>

  <title>RNA-seq测序数据模拟 | Dean's blog</title>
  






  <noscript>
  <style>
  .use-motion .brand,
  .use-motion .menu-item,
  .sidebar-inner,
  .use-motion .post-block,
  .use-motion .pagination,
  .use-motion .comments,
  .use-motion .post-header,
  .use-motion .post-body,
  .use-motion .collection-header { opacity: initial; }

  .use-motion .site-title,
  .use-motion .site-subtitle {
    opacity: initial;
    top: initial;
  }

  .use-motion .logo-line-before i { left: initial; }
  .use-motion .logo-line-after i { right: initial; }
  </style>
</noscript>

<style>.darkmode--activated{--body-bg-color:#282828;--content-bg-color:#333;--card-bg-color:#555;--text-color:#ccc;--blockquote-color:#bbb;--link-color:#ccc;--link-hover-color:#eee;--brand-color:#ddd;--brand-hover-color:#ddd;--table-row-odd-bg-color:#282828;--table-row-hover-bg-color:#363636;--menu-item-bg-color:#555;--btn-default-bg:#222;--btn-default-color:#ccc;--btn-default-border-color:#555;--btn-default-hover-bg:#666;--btn-default-hover-color:#ccc;--btn-default-hover-border-color:#666;--highlight-background:#282b2e;--highlight-foreground:#a9b7c6;--highlight-gutter-background:#34393d;--highlight-gutter-foreground:#9ca9b6}.darkmode--activated img{opacity:.75}.darkmode--activated img:hover{opacity:.9}.darkmode--activated code{color:#69dbdc;background:0 0}button.darkmode-toggle{z-index:9999}.darkmode-ignore,img{display:flex!important}</style></head>

<body itemscope itemtype="http://schema.org/WebPage">
  <div class="container use-motion">
    <div class="headband"></div>

    <header class="header" itemscope itemtype="http://schema.org/WPHeader">
      <div class="header-inner"><div class="site-brand-container">
  <div class="site-nav-toggle">
    <div class="toggle" aria-label="Toggle navigation bar">
      <span class="toggle-line toggle-line-first"></span>
      <span class="toggle-line toggle-line-middle"></span>
      <span class="toggle-line toggle-line-last"></span>
    </div>
  </div>

  <div class="site-meta">

    <a href="/" class="brand" rel="start">
      <span class="logo-line-before"><i></i></span>
      <h1 class="site-title">Dean's blog</h1>
      <span class="logo-line-after"><i></i></span>
    </a>
  </div>

  <div class="site-nav-right">
    <div class="toggle popup-trigger">
    </div>
  </div>
</div>




<nav class="site-nav">
  <ul id="menu" class="main-menu menu">
        <li class="menu-item menu-item-home">

    <a href="/" rel="section"><i class="fa fa-home fa-fw"></i>Home</a>

  </li>
        <li class="menu-item menu-item-about">

    <a href="/about/" rel="section"><i class="fa fa-user fa-fw"></i>About</a>

  </li>
        <li class="menu-item menu-item-tags">

    <a href="/tags/" rel="section"><i class="fa fa-tags fa-fw"></i>Tags</a>

  </li>
        <li class="menu-item menu-item-archives">

    <a href="/archives/" rel="section"><i class="fa fa-archive fa-fw"></i>Archives</a>

  </li>
  </ul>
</nav>




</div>
    </header>

    
  <div class="back-to-top">
    <i class="fa fa-arrow-up"></i>
    <span>0%</span>
  </div>
  <div class="reading-progress-bar"></div>


    <main class="main">
      <div class="main-inner">
        <div class="content-wrap">
          

          <div class="content post posts-expand">
            

    
  
  
  <article itemscope itemtype="http://schema.org/Article" class="post-block" lang="en">
    <link itemprop="mainEntityOfPage" href="https://thereallda.github.io/2022/02/10/R-RNA-seq-data-simulation/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="image" content="https://avatars.githubusercontent.com/u/55836943?v=4">
      <meta itemprop="name" content="Dean Li">
      <meta itemprop="description" content="">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="Dean's blog">
    </span>
      <header class="post-header">
        <h1 class="post-title" itemprop="name headline">
          RNA-seq测序数据模拟
        </h1>

        <div class="post-meta">
          
          
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-calendar"></i>
              </span>
              <span class="post-meta-item-text">Posted on</span>

              <time title="Created: 2022-02-10 21:49:29" itemprop="dateCreated datePublished" datetime="2022-02-10T21:49:29+08:00">2022-02-10</time>
            </span>
              <span class="post-meta-item">
                <span class="post-meta-item-icon">
                  <i class="far fa-calendar-check"></i>
                </span>
                <span class="post-meta-item-text">Edited on</span>
                <time title="Modified: 2022-02-11 09:27:51" itemprop="dateModified" datetime="2022-02-11T09:27:51+08:00">2022-02-11</time>
              </span>

          
            <span class="post-meta-item" title="Views" id="busuanzi_container_page_pv" style="display: none;">
              <span class="post-meta-item-icon">
                <i class="fa fa-eye"></i>
              </span>
              <span class="post-meta-item-text">Views: </span>
              <span id="busuanzi_value_page_pv"></span>
            </span><br>
            <span class="post-meta-item" title="Symbols count in article">
              <span class="post-meta-item-icon">
                <i class="far fa-file-word"></i>
              </span>
                <span class="post-meta-item-text">Symbols count in article: </span>
              <span>7.4k</span>
            </span>
            <span class="post-meta-item" title="Reading time">
              <span class="post-meta-item-icon">
                <i class="far fa-clock"></i>
              </span>
                <span class="post-meta-item-text">Reading time &asymp;</span>
              <span>7 mins.</span>
            </span>

        </div>
      </header>

    
    
    
    <div class="post-body" itemprop="articleBody">

      
        <p>在评估不同软件性能的时候，我们会需要模拟一些数据。由于模拟数据当中的情况是已知的，例如差异表达基因的数目。因此，通过比较不同软件在模拟数据上的效果，我们可以获得软件的量化性能指标，例如灵敏度、特异性和准确度等。</p>
<p>本文根据 <code>DESeq2</code> 文章中的方法记录如何进行简单的基于负二项分布（Negative Binomial distribution）模拟RNA-seq基因表达数据。</p>
<span id="more"></span>
<p>为了模拟基因表达数据，我们需要：</p>
<ol>
<li>从已知的测序数据中获取基因表达的均值与离散程度作为负二项分布的参数</li>
<li>构建样本信息，包括基因数目、样本数目、差异表达程度和样本分组等</li>
<li>基于负二项分布模拟基因表达数据</li>
</ol>
<h1>Mean-Dispersion Estimation</h1>
<p>首先，我们从<a target="_blank" rel="noopener" href="https://www.nature.com/articles/nature08872">2010 Nature</a>发布的RNA-seq数据中，获取真实基因表达的均值(mean)与离散程度(dispersion)，以此作为我们模拟表达的基因的均值与离散程度。</p>
<p><code>DESeq2paper</code> 提供了该数据的 <code>RData</code> (<a target="_blank" rel="noopener" href="https://www.huber.embl.de/DESeq2paper/data/pickrell_sumexp.RData">https://www.huber.embl.de/DESeq2paper/data/pickrell_sumexp.RData</a>)</p>
<p>但是，在实际操作过程中，发现从 <code>DESeq2paper</code> 主页上下载的数据在读入后会出错</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">load(<span class="string">&quot;data/pickrell_sumexp.RData&quot;</span>)</span><br><span class="line">pickrell</span><br><span class="line"></span><br><span class="line"><span class="comment"># class: SummarizedExperiment </span></span><br><span class="line"><span class="comment"># Error in h(simpleError(msg, call)) : </span></span><br><span class="line"><span class="comment">#   error in evaluating the argument &#x27;x&#x27; in selecting a method for function &#x27;nrow&#x27;: no slot of name &quot;elementMetadata&quot; for this object of class &quot;SummarizedExperiment&quot;</span></span><br><span class="line"><span class="comment"># Error during wrapup: no slot of name &quot;NAMES&quot; for this object of class &quot;SummarizedExperiment&quot;</span></span><br><span class="line"><span class="comment"># Error: no more error handlers available (recursive errors?); invoking &#x27;abort&#x27; restart</span></span><br></pre></td></tr></table></figure>
<p>搜索发现 <code>DESeq2</code> 作者 Michael Love 本人也遇到这个问题</p>
<blockquote>
<p><a target="_blank" rel="noopener" href="https://support.bioconductor.org/p/131099/">https://support.bioconductor.org/p/131099/</a></p>
<p><img src="paste-E5B09090.png" alt=""></p>
</blockquote>
<p>但他也没有修复这个官网上的数据，而是把修复的数据放到：<a target="_blank" rel="noopener" href="https://github.com/mikelove/apeglmPaper/blob/master/sim/rdata/pickrell_sumexp.RData">https://github.com/mikelove/apeglmPaper/blob/master/sim/rdata/pickrell_sumexp.RData</a></p>
<p>下载该数据，放到项目目录的 <code>data/</code> 文件夹下。</p>
<p>接下来，使用 <code>DESeq2</code> 中的函数从数据中获取基因表达的均值与离散程度。</p>
<p><code>pickrell</code> 数据中包含69个样本共56299个基因的表达数据</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># mean-dispersion estimation from pickrell</span></span><br><span class="line">library(DESeq2)</span><br><span class="line"><span class="comment"># https://github.com/mikelove/apeglmPaper/blob/master/sim/rdata/pickrell_sumexp.RData</span></span><br><span class="line">load(<span class="string">&quot;data/pickrell_sumexp.RData&quot;</span>)</span><br><span class="line">pickrell</span><br><span class="line"></span><br><span class="line"><span class="comment">## class: RangedSummarizedExperiment </span></span><br><span class="line"><span class="comment">## dim: 56299 69 </span></span><br><span class="line"><span class="comment">## metadata(0):</span></span><br><span class="line"><span class="comment">## assays(1): counts</span></span><br><span class="line"><span class="comment">## rownames(56299): ENSG00000000003 ENSG00000000005 ... ENSG00000261841</span></span><br><span class="line"><span class="comment">##   ENSG00000261842</span></span><br><span class="line"><span class="comment">## rowData names(0):</span></span><br><span class="line"><span class="comment">## colnames: NULL</span></span><br><span class="line"><span class="comment">## colData names(1): rownames</span></span><br></pre></td></tr></table></figure>
<p>数据的均值和离散程度分别存储在 <code>mcols(ddspickrell)</code> 的 <code>baseMean</code> 和 <code>dispGeneEst</code> 列。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br></pre></td><td class="code"><pre><span class="line">ddspickrell &lt;- DESeqDataSet(pickrell, ~<span class="number">1</span>)</span><br><span class="line">ddspickrell &lt;- estimateSizeFactors(ddspickrell)</span><br><span class="line"><span class="comment"># estimate gene-wise dispersion</span></span><br><span class="line">ddspickrell &lt;- estimateDispersionsGeneEst(ddspickrell)</span><br><span class="line"><span class="comment"># estimate fitted dispersion</span></span><br><span class="line">ddspickrell &lt;- estimateDispersionsFit(ddspickrell)</span><br><span class="line">head(mcols(ddspickrell))</span><br><span class="line"></span><br><span class="line"><span class="comment">## DataFrame with 6 rows and 6 columns</span></span><br><span class="line"><span class="comment">##                   baseMean     baseVar   allZero dispGeneEst dispGeneIter</span></span><br><span class="line"><span class="comment">##                  &lt;numeric&gt;   &lt;numeric&gt; &lt;logical&gt;   &lt;numeric&gt;    &lt;numeric&gt;</span></span><br><span class="line"><span class="comment">## ENSG00000000003   0.992996 4.91552e+00     FALSE  1.92047047            8</span></span><br><span class="line"><span class="comment">## ENSG00000000005   0.018963 1.22328e-02     FALSE  0.00000001           28</span></span><br><span class="line"><span class="comment">## ENSG00000000419 128.895411 1.11440e+03     FALSE  0.05723383           10</span></span><br><span class="line"><span class="comment">## ENSG00000000457  79.020253 5.77887e+02     FALSE  0.04260266           11</span></span><br><span class="line"><span class="comment">## ENSG00000000460  57.365964 6.06033e+02     FALSE  0.11135132           14</span></span><br><span class="line"><span class="comment">## ENSG00000000938 713.350827 1.31691e+05     FALSE  0.25701274            8</span></span><br><span class="line"><span class="comment">##                   dispFit</span></span><br><span class="line"><span class="comment">##                 &lt;numeric&gt;</span></span><br><span class="line"><span class="comment">## ENSG00000000003  2.029100</span></span><br><span class="line"><span class="comment">## ENSG00000000005 98.513543</span></span><br><span class="line"><span class="comment">## ENSG00000000419  0.165160</span></span><br><span class="line"><span class="comment">## ENSG00000000457  0.174293</span></span><br><span class="line"><span class="comment">## ENSG00000000460  0.183204</span></span><br><span class="line"><span class="comment">## ENSG00000000938  0.153303</span></span><br></pre></td></tr></table></figure>
<p>提取离散程度大于 <code>1e-06</code> 的基因的均值与离散程度，并保存供后续分析所用</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># keep gene with dispersion greater than 1e-06</span></span><br><span class="line"><span class="comment"># extract the gene-wise mean and dispersion</span></span><br><span class="line">meanDispPairs &lt;- mcols(ddspickrell)[which(mcols(ddspickrell)$dispGeneEst &gt; <span class="number">1e-06</span>), <span class="built_in">c</span>(<span class="string">&quot;baseMean&quot;</span>, <span class="string">&quot;dispGeneEst&quot;</span>)]</span><br><span class="line"><span class="built_in">names</span>(meanDispPairs) &lt;- <span class="built_in">c</span>(<span class="string">&quot;mean&quot;</span>, <span class="string">&quot;disp&quot;</span>)</span><br><span class="line"><span class="comment"># save the mean-dispersion pairs for further usage</span></span><br><span class="line">save(meanDispPairs, file=<span class="string">&quot;data/meanDispPairs.RData&quot;</span>)</span><br></pre></td></tr></table></figure>
<h1>Simple Simulation</h1>
<p>假设我们需要模拟一组数据，包含两组样本，每组4个重复，20000个基因，其中有20%的基因发生两倍差异表达。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># gene expression data simulation following negative binomial ditribution</span></span><br><span class="line">load(<span class="string">&quot;data/meanDispPairs.RData&quot;</span>)</span><br><span class="line"><span class="comment"># total number of genes</span></span><br><span class="line">n &lt;- 20000</span><br><span class="line"><span class="comment"># total sample size</span></span><br><span class="line">m &lt;- 8</span><br><span class="line"><span class="comment"># log fold changes used</span></span><br><span class="line">es &lt;- log2(<span class="number">2</span>)</span><br><span class="line"><span class="comment"># formula</span></span><br><span class="line">condition &lt;- factor(<span class="built_in">rep</span>(<span class="built_in">c</span>(<span class="string">&quot;A&quot;</span>,<span class="string">&quot;B&quot;</span>), each = m/<span class="number">2</span>))</span><br><span class="line">x &lt;- model.matrix(~ condition)</span><br><span class="line"><span class="comment"># beta indicates the fold-change of genes</span></span><br><span class="line"><span class="comment"># 80% non-DE genes and 20% DE genes with up- or down-regulated</span></span><br><span class="line">beta &lt;- <span class="built_in">c</span>(<span class="built_in">rep</span>(<span class="number">0</span>, n * <span class="number">8</span>/<span class="number">10</span>), sample(<span class="built_in">c</span>(-es,es), n * <span class="number">2</span>/<span class="number">10</span>, <span class="literal">TRUE</span>))  </span><br><span class="line"></span><br><span class="line">set.seed(<span class="number">123</span>)</span><br><span class="line">sf=<span class="built_in">rep</span>(<span class="number">1</span>,m)</span><br><span class="line"><span class="comment"># randomly sample mean-dispersion from real gene expression dataset</span></span><br><span class="line">idx &lt;- sample(nrow(meanDispPairs), n, replace=<span class="literal">TRUE</span>)</span><br><span class="line">mu0 &lt;- meanDispPairs[idx,<span class="number">1</span>]</span><br><span class="line">disp &lt;- meanDispPairs[idx,<span class="number">2</span>]</span><br><span class="line">betafull &lt;- cbind(log2(mu0), beta)</span><br><span class="line"><span class="comment"># betafull: [n,2]; t(x): [2,m]</span></span><br><span class="line"><span class="comment"># mu: [n,m]; mu matrix contains the mean of each genes for each sample</span></span><br><span class="line">mu &lt;- 2^(betafull %*% t(x))</span><br><span class="line">head(mu)</span><br><span class="line"></span><br><span class="line"><span class="comment">##           1      2      3      4      5      6      7      8</span></span><br><span class="line"><span class="comment">## [1,]   2.55   2.55   2.55   2.55   2.55   2.55   2.55   2.55</span></span><br><span class="line"><span class="comment">## [2,]   0.90   0.90   0.90   0.90   0.90   0.90   0.90   0.90</span></span><br><span class="line"><span class="comment">## [3,]  79.34  79.34  79.34  79.34  79.34  79.34  79.34  79.34</span></span><br><span class="line"><span class="comment">## [4,]   4.81   4.81   4.81   4.81   4.81   4.81   4.81   4.81</span></span><br><span class="line"><span class="comment">## [5,] 368.22 368.22 368.22 368.22 368.22 368.22 368.22 368.22</span></span><br><span class="line"><span class="comment">## [6,] 719.99 719.99 719.99 719.99 719.99 719.99 719.99 719.99</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># replicate mu matrix m times</span></span><br><span class="line">muMat &lt;- matrix(<span class="built_in">rep</span>(mu, times=m) * <span class="built_in">rep</span>(sf, each=n), ncol=m)</span><br><span class="line"><span class="comment"># n*m values to be simulated</span></span><br><span class="line">mat &lt;- matrix(rnbinom(n*m, mu=muMat, size=<span class="number">1</span>/disp), ncol=m)</span><br><span class="line">head(mat)</span><br><span class="line"></span><br><span class="line"><span class="comment">##      [,1] [,2] [,3] [,4] [,5] [,6] [,7] [,8]</span></span><br><span class="line"><span class="comment">## [1,]    3    2    2    4    1    1    3    3</span></span><br><span class="line"><span class="comment">## [2,]    0    1    0    0    0    0    1    2</span></span><br><span class="line"><span class="comment">## [3,]   89   82  103  137   50   82   62   79</span></span><br><span class="line"><span class="comment">## [4,]    6    2    3    4    3    5    3    1</span></span><br><span class="line"><span class="comment">## [5,]  519  490  213  321  306  256  460  375</span></span><br><span class="line"><span class="comment">## [6,]  580 1459  829  855  508  340  647  410</span></span><br></pre></td></tr></table></figure>
<p>简单可视化展示模拟数据。这里把不表达的基因去除后进行可视化，基因表达的直方图表明我们的数据整体还是符合负二项分布的</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">par(mfrow=<span class="built_in">c</span>(<span class="number">2</span>,<span class="number">4</span>))</span><br><span class="line">sapply(<span class="built_in">seq_len</span>(ncol(mat)), <span class="keyword">function</span>(i) hist(<span class="built_in">log</span>(mat[rowSums(mat)&gt;<span class="number">0</span>,] + <span class="number">1</span>)[,i], main=paste(<span class="string">&#x27;Col&#x27;</span>,i)))</span><br></pre></td></tr></table></figure>
<p><img src="unnamed-chunk-6-1.png" alt=""></p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">par(mfrow=<span class="built_in">c</span>(<span class="number">1</span>,<span class="number">1</span>))</span><br><span class="line">boxplot(<span class="built_in">log</span>(mat[rowSums(mat)&gt;<span class="number">0</span>,] + <span class="number">1</span>), main=<span class="string">&#x27;Simple simulation&#x27;</span>)</span><br></pre></td></tr></table></figure>
<p><img src="unnamed-chunk-6-2.png" alt=""></p>
<p>与Pickrell数据进行比较。这里使用稍微严格一点的过滤条件</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># pickrell dataset</span></span><br><span class="line">counts.pick &lt;- assay(pickrell)</span><br><span class="line">counts.pick &lt;- counts.pick[rowSums(counts.pick&gt;<span class="number">0</span>) &gt; <span class="number">10</span>, ]</span><br><span class="line"><span class="comment"># demonstrate first 8 columns</span></span><br><span class="line">par(mfrow=<span class="built_in">c</span>(<span class="number">2</span>,<span class="number">4</span>))</span><br><span class="line">sapply(<span class="number">1</span>:<span class="number">8</span>, <span class="keyword">function</span>(i) hist(<span class="built_in">log</span>(counts.pick[,i]+<span class="number">1</span>), main=paste(<span class="string">&#x27;Col&#x27;</span>,i)))</span><br></pre></td></tr></table></figure>
<p><img src="unnamed-chunk-7-1.png" alt=""></p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">par(mfrow=<span class="built_in">c</span>(<span class="number">1</span>,<span class="number">1</span>))</span><br><span class="line">boxplot(<span class="built_in">log</span>(counts.pick+<span class="number">1</span>))</span><br></pre></td></tr></table></figure>
<p><img src="unnamed-chunk-7-2.png" alt=""></p>
<p>在Pickrell数据当中样本之间的整体表达情况会有一点波动，而我们模拟的数据没有。这可能是由于实验操作或者生物体之间的误差引起的，而模拟数据并不存在这种这种误差。我们也可以人为引入这种误差来更好地模拟真实数据。</p>
<h1>Assessment</h1>
<p>最后，我们使用 <code>DESeq2</code> 对模拟数据进行差异分析，测试模拟数据</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># DE with DESeq2</span></span><br><span class="line">e &lt;- DESeqDataSetFromMatrix(mat,</span><br><span class="line">                            colData = as.data.frame(condition),</span><br><span class="line">                            design = ~condition)</span><br><span class="line"></span><br><span class="line"><span class="comment">## converting counts to integer mode</span></span><br><span class="line"></span><br><span class="line">de &lt;- DESeq(e)</span><br><span class="line"></span><br><span class="line"><span class="comment">## estimating size factors</span></span><br><span class="line"></span><br><span class="line"><span class="comment">## estimating dispersions</span></span><br><span class="line"></span><br><span class="line"><span class="comment">## gene-wise dispersion estimates</span></span><br><span class="line"></span><br><span class="line"><span class="comment">## mean-dispersion relationship</span></span><br><span class="line"></span><br><span class="line"><span class="comment">## final dispersion estimates</span></span><br><span class="line"></span><br><span class="line"><span class="comment">## fitting model and testing</span></span><br><span class="line"></span><br><span class="line">res &lt;- results(de)</span><br><span class="line">table(<span class="built_in">abs</span>(res$log2FoldChange) &gt;= <span class="number">1</span>)/nrow(res)</span><br><span class="line"></span><br><span class="line"><span class="comment">## </span></span><br><span class="line"><span class="comment">## FALSE  TRUE </span></span><br><span class="line"><span class="comment">## 0.710 0.232</span></span><br><span class="line"></span><br><span class="line">table(<span class="built_in">abs</span>(res$log2FoldChange) &gt;= <span class="number">1</span> &amp; res$padj &lt; <span class="number">0.1</span>)/nrow(res)</span><br><span class="line"></span><br><span class="line"><span class="comment">## </span></span><br><span class="line"><span class="comment">## FALSE  TRUE </span></span><br><span class="line"><span class="comment">## 0.753 0.048</span></span><br><span class="line"></span><br><span class="line">summary(res)</span><br><span class="line"></span><br><span class="line"><span class="comment">## </span></span><br><span class="line"><span class="comment">## out of 18828 with nonzero total read count</span></span><br><span class="line"><span class="comment">## adjusted p-value &lt; 0.1</span></span><br><span class="line"><span class="comment">## LFC &gt; 0 (up)       : 663, 3.5%</span></span><br><span class="line"><span class="comment">## LFC &lt; 0 (down)     : 640, 3.4%</span></span><br><span class="line"><span class="comment">## outliers [1]       : 62, 0.33%</span></span><br><span class="line"><span class="comment">## low counts [2]     : 6545, 35%</span></span><br><span class="line"><span class="comment">## (mean count &lt; 5)</span></span><br><span class="line"><span class="comment">## [1] see &#x27;cooksCutoff&#x27; argument of ?results</span></span><br><span class="line"><span class="comment">## [2] see &#x27;independentFiltering&#x27; argument of ?results</span></span><br></pre></td></tr></table></figure>
<p>使用模拟数据进行差异分析发现约23%的基因具有2倍以上的差异表达，但如果考虑 <code>adjusted p-value &lt; 0.1</code> 的话，则只有5%左右的基因具有显著差异表达。同时，我们可以发现有37%的基因被认为是低表达（基因的平均counts &lt; 5）的情况。</p>
<p>进一步，我们通过灵敏度（sensitivity）和特异性（specificity）评估 <code>DESeq2</code> 在模拟数据上的表现</p>
<p><img src="image-20220210215514028.png" alt="Sensitivity-and-Specificity"></p>
<p><em>Figure</em>: <a target="_blank" rel="noopener" href="https://en.wikipedia.org/wiki/Sensitivity_and_specificity">Wiki</a></p>
<p><strong>sensitivity, recall, hit rate, or true positive rate (TPR)</strong><br>
<span class="katex"><span class="katex-mathml"><math><semantics><mrow><mrow><mi>T</mi><mi>P</mi><mi>R</mi><mo>=</mo><mfrac><mrow><mi>T</mi><mi>P</mi></mrow><mrow><mi>P</mi></mrow></mfrac><mo>=</mo><mfrac><mrow><mi>T</mi><mi>P</mi></mrow><mrow><mi>T</mi><mi>P</mi><mo>+</mo><mi>F</mi><mi>N</mi></mrow></mfrac><mo>=</mo><mn>1</mn><mo>−</mo><mi>F</mi><mi>N</mi><mi>R</mi></mrow></mrow><annotation encoding="application/x-tex">{TPR=\frac{TP}{P}=\frac{TP}{TP+FN}=1-FNR}</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="strut" style="height:0.872331em;"></span><span class="strut bottom" style="height:1.275662em;vertical-align:-0.403331em;"></span><span class="base textstyle uncramped"><span class="mord textstyle uncramped"><span class="mord mathit" style="margin-right:0.13889em;">T</span><span class="mord mathit" style="margin-right:0.13889em;">P</span><span class="mord mathit" style="margin-right:0.00773em;">R</span><span class="mrel">=</span><span class="mord reset-textstyle textstyle uncramped"><span class="sizing reset-size5 size5 reset-textstyle textstyle uncramped nulldelimiter"></span><span class="mfrac"><span class="vlist"><span style="top:0.345em;"><span class="fontsize-ensurer reset-size5 size5"><span style="font-size:0em;">​</span></span><span class="reset-textstyle scriptstyle cramped"><span class="mord scriptstyle cramped"><span class="mord mathit" style="margin-right:0.13889em;">P</span></span></span></span><span style="top:-0.22999999999999998em;"><span class="fontsize-ensurer reset-size5 size5"><span style="font-size:0em;">​</span></span><span class="reset-textstyle textstyle uncramped frac-line"></span></span><span style="top:-0.394em;"><span class="fontsize-ensurer reset-size5 size5"><span style="font-size:0em;">​</span></span><span class="reset-textstyle scriptstyle uncramped"><span class="mord scriptstyle uncramped"><span class="mord mathit" style="margin-right:0.13889em;">T</span><span class="mord mathit" style="margin-right:0.13889em;">P</span></span></span></span><span class="baseline-fix"><span class="fontsize-ensurer reset-size5 size5"><span style="font-size:0em;">​</span></span>​</span></span></span><span class="sizing reset-size5 size5 reset-textstyle textstyle uncramped nulldelimiter"></span></span><span class="mrel">=</span><span class="mord reset-textstyle textstyle uncramped"><span class="sizing reset-size5 size5 reset-textstyle textstyle uncramped nulldelimiter"></span><span class="mfrac"><span class="vlist"><span style="top:0.345em;"><span class="fontsize-ensurer reset-size5 size5"><span style="font-size:0em;">​</span></span><span class="reset-textstyle scriptstyle cramped"><span class="mord scriptstyle cramped"><span class="mord mathit" style="margin-right:0.13889em;">T</span><span class="mord mathit" style="margin-right:0.13889em;">P</span><span class="mbin">+</span><span class="mord mathit" style="margin-right:0.13889em;">F</span><span class="mord mathit" style="margin-right:0.10903em;">N</span></span></span></span><span style="top:-0.22999999999999998em;"><span class="fontsize-ensurer reset-size5 size5"><span style="font-size:0em;">​</span></span><span class="reset-textstyle textstyle uncramped frac-line"></span></span><span style="top:-0.394em;"><span class="fontsize-ensurer reset-size5 size5"><span style="font-size:0em;">​</span></span><span class="reset-textstyle scriptstyle uncramped"><span class="mord scriptstyle uncramped"><span class="mord mathit" style="margin-right:0.13889em;">T</span><span class="mord mathit" style="margin-right:0.13889em;">P</span></span></span></span><span class="baseline-fix"><span class="fontsize-ensurer reset-size5 size5"><span style="font-size:0em;">​</span></span>​</span></span></span><span class="sizing reset-size5 size5 reset-textstyle textstyle uncramped nulldelimiter"></span></span><span class="mrel">=</span><span class="mord mathrm">1</span><span class="mbin">−</span><span class="mord mathit" style="margin-right:0.13889em;">F</span><span class="mord mathit" style="margin-right:0.10903em;">N</span><span class="mord mathit" style="margin-right:0.00773em;">R</span></span></span></span></span></p>
<p><strong>specificity, selectivity or true negative rate (TNR)</strong><br>
<span class="katex"><span class="katex-mathml"><math><semantics><mrow><mrow><mi>T</mi><mi>N</mi><mi>R</mi><mo>=</mo><mfrac><mrow><mi>T</mi><mi>N</mi></mrow><mrow><mi>N</mi></mrow></mfrac><mo>=</mo><mfrac><mrow><mi>T</mi><mi>N</mi></mrow><mrow><mi>T</mi><mi>N</mi><mo>+</mo><mi>F</mi><mi>P</mi></mrow></mfrac><mo>=</mo><mn>1</mn><mo>−</mo><mi>F</mi><mi>P</mi><mi>R</mi></mrow></mrow><annotation encoding="application/x-tex">{TNR=\frac{TN}{N}=\frac{TN}{TN+FP}=1-FPR}</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="strut" style="height:0.872331em;"></span><span class="strut bottom" style="height:1.275662em;vertical-align:-0.403331em;"></span><span class="base textstyle uncramped"><span class="mord textstyle uncramped"><span class="mord mathit" style="margin-right:0.13889em;">T</span><span class="mord mathit" style="margin-right:0.10903em;">N</span><span class="mord mathit" style="margin-right:0.00773em;">R</span><span class="mrel">=</span><span class="mord reset-textstyle textstyle uncramped"><span class="sizing reset-size5 size5 reset-textstyle textstyle uncramped nulldelimiter"></span><span class="mfrac"><span class="vlist"><span style="top:0.345em;"><span class="fontsize-ensurer reset-size5 size5"><span style="font-size:0em;">​</span></span><span class="reset-textstyle scriptstyle cramped"><span class="mord scriptstyle cramped"><span class="mord mathit" style="margin-right:0.10903em;">N</span></span></span></span><span style="top:-0.22999999999999998em;"><span class="fontsize-ensurer reset-size5 size5"><span style="font-size:0em;">​</span></span><span class="reset-textstyle textstyle uncramped frac-line"></span></span><span style="top:-0.394em;"><span class="fontsize-ensurer reset-size5 size5"><span style="font-size:0em;">​</span></span><span class="reset-textstyle scriptstyle uncramped"><span class="mord scriptstyle uncramped"><span class="mord mathit" style="margin-right:0.13889em;">T</span><span class="mord mathit" style="margin-right:0.10903em;">N</span></span></span></span><span class="baseline-fix"><span class="fontsize-ensurer reset-size5 size5"><span style="font-size:0em;">​</span></span>​</span></span></span><span class="sizing reset-size5 size5 reset-textstyle textstyle uncramped nulldelimiter"></span></span><span class="mrel">=</span><span class="mord reset-textstyle textstyle uncramped"><span class="sizing reset-size5 size5 reset-textstyle textstyle uncramped nulldelimiter"></span><span class="mfrac"><span class="vlist"><span style="top:0.345em;"><span class="fontsize-ensurer reset-size5 size5"><span style="font-size:0em;">​</span></span><span class="reset-textstyle scriptstyle cramped"><span class="mord scriptstyle cramped"><span class="mord mathit" style="margin-right:0.13889em;">T</span><span class="mord mathit" style="margin-right:0.10903em;">N</span><span class="mbin">+</span><span class="mord mathit" style="margin-right:0.13889em;">F</span><span class="mord mathit" style="margin-right:0.13889em;">P</span></span></span></span><span style="top:-0.22999999999999998em;"><span class="fontsize-ensurer reset-size5 size5"><span style="font-size:0em;">​</span></span><span class="reset-textstyle textstyle uncramped frac-line"></span></span><span style="top:-0.394em;"><span class="fontsize-ensurer reset-size5 size5"><span style="font-size:0em;">​</span></span><span class="reset-textstyle scriptstyle uncramped"><span class="mord scriptstyle uncramped"><span class="mord mathit" style="margin-right:0.13889em;">T</span><span class="mord mathit" style="margin-right:0.10903em;">N</span></span></span></span><span class="baseline-fix"><span class="fontsize-ensurer reset-size5 size5"><span style="font-size:0em;">​</span></span>​</span></span></span><span class="sizing reset-size5 size5 reset-textstyle textstyle uncramped nulldelimiter"></span></span><span class="mrel">=</span><span class="mord mathrm">1</span><span class="mbin">−</span><span class="mord mathit" style="margin-right:0.13889em;">F</span><span class="mord mathit" style="margin-right:0.13889em;">P</span><span class="mord mathit" style="margin-right:0.00773em;">R</span></span></span></span></span></p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br></pre></td><td class="code"><pre><span class="line">nonzero &lt;- rowSums(counts(e)) &gt; <span class="number">0</span></span><br><span class="line"><span class="comment"># TPR = Sensitivity</span></span><br><span class="line">sensidx &lt;- <span class="built_in">abs</span>(beta) &gt; <span class="number">0</span> &amp; nonzero</span><br><span class="line">sensitivity &lt;- <span class="built_in">sum</span>((res$padj&lt;<span class="number">0.1</span>)[sensidx], na.rm=<span class="literal">TRUE</span>)/nrow(res)</span><br><span class="line">sensitivity</span><br><span class="line"></span><br><span class="line"><span class="comment">## [1] 0.0565</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># FPR = 1 - Specificity</span></span><br><span class="line">nonDEidx &lt;- <span class="built_in">abs</span>(beta) == <span class="number">0</span> &amp; nonzero</span><br><span class="line">oneMinusSpecificity &lt;- <span class="built_in">sum</span>((res$padj&lt;<span class="number">0.1</span>)[nonDEidx], na.rm=<span class="literal">TRUE</span>)/nrow(res)</span><br><span class="line">oneMinusSpecificity</span><br><span class="line"></span><br><span class="line"><span class="comment">## [1] 0.00865</span></span><br></pre></td></tr></table></figure>
<p>以上只是单次模拟的结果，如果需要对软件的性能进行评估，我们应当进行多次模拟，然后再进行性能评估。</p>
<blockquote>
<p>Ref:</p>
<p><strong>DESeq2</strong>: <a target="_blank" rel="noopener" href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-014-0550-8">https://genomebiology.biomedcentral.com/articles/10.1186/s13059-014-0550-8</a></p>
<p><strong>DESeq2paper:</strong> <a target="_blank" rel="noopener" href="https://www.huber.embl.de/DESeq2paper/">https://www.huber.embl.de/DESeq2paper/</a></p>
<p><strong>biostar handbook|How to simulate sequencing results:</strong> <a target="_blank" rel="noopener" href="https://blog.actorsfit.com/a?ID=01000-a374125f-99d7-452a-99f9-61466d0a3ee6">https://blog.actorsfit.com/a?ID=01000-a374125f-99d7-452a-99f9-61466d0a3ee6</a></p>
</blockquote>
<p>完</p>

    </div>

    
    
    
        

<div>
<ul class="post-copyright">
  <li class="post-copyright-author">
    <strong>Post author:  </strong>Dean Li
  </li>
  <li class="post-copyright-link">
    <strong>Post link: </strong>
    <a href="https://thereallda.github.io/2022/02/10/R-RNA-seq-data-simulation/" title="RNA-seq测序数据模拟">https://thereallda.github.io/2022/02/10/R-RNA-seq-data-simulation/</a>
  </li>
  <li class="post-copyright-license">
    <strong>Copyright Notice:  </strong>All articles in this blog are licensed under <a href="https://creativecommons.org/licenses/by-nc-sa/4.0/" rel="noopener" target="_blank"><i class="fab fa-fw fa-creative-commons"></i>BY-NC-SA</a> unless stating additionally.
  </li>
</ul>
</div>


      <footer class="post-footer">
          <div class="post-tags">
              <a href="/tags/R/" rel="tag"># R</a>
              <a href="/tags/RNA-seq/" rel="tag"># RNA-seq</a>
              <a href="/tags/Data-analysis/" rel="tag"># Data_analysis</a>
          </div>

        


        
    <div class="post-nav">
      <div class="post-nav-item">
    <a href="/2022/02/07/R-Retrieve-Gene-Length/" rel="prev" title="R-获取基因长度">
      <i class="fa fa-chevron-left"></i> R-获取基因长度
    </a></div>
      <div class="post-nav-item">
    <a href="/2022/02/28/PathwayEnrichment/" rel="next" title="通路富集分析">
      通路富集分析 <i class="fa fa-chevron-right"></i>
    </a></div>
    </div>
      </footer>
    
  </article>
  
  
  



          </div>
          
    <div class="comments" id="gitalk-container"></div>

<script>
  window.addEventListener('tabs:register', () => {
    let { activeClass } = CONFIG.comments;
    if (CONFIG.comments.storage) {
      activeClass = localStorage.getItem('comments_active') || activeClass;
    }
    if (activeClass) {
      let activeTab = document.querySelector(`a[href="#comment-${activeClass}"]`);
      if (activeTab) {
        activeTab.click();
      }
    }
  });
  if (CONFIG.comments.storage) {
    window.addEventListener('tabs:click', event => {
      if (!event.target.matches('.tabs-comment .tab-content .tab-pane')) return;
      let commentClass = event.target.classList[1];
      localStorage.setItem('comments_active', commentClass);
    });
  }
</script>

        </div>
          
  
  <div class="toggle sidebar-toggle">
    <span class="toggle-line toggle-line-first"></span>
    <span class="toggle-line toggle-line-middle"></span>
    <span class="toggle-line toggle-line-last"></span>
  </div>

  <aside class="sidebar">
    <div class="sidebar-inner">

      <ul class="sidebar-nav motion-element">
        <li class="sidebar-nav-toc">
          Table of Contents
        </li>
        <li class="sidebar-nav-overview">
          Overview
        </li>
      </ul>

      <!--noindex-->
      <div class="post-toc-wrap sidebar-panel">
          <div class="post-toc motion-element"><ol class="nav"><li class="nav-item nav-level-1"><a class="nav-link"><span class="nav-number">1.</span> <span class="nav-text">Mean-Dispersion Estimation</span></a></li><li class="nav-item nav-level-1"><a class="nav-link"><span class="nav-number">2.</span> <span class="nav-text">Simple Simulation</span></a></li><li class="nav-item nav-level-1"><a class="nav-link"><span class="nav-number">3.</span> <span class="nav-text">Assessment</span></a></li></ol></div>
      </div>
      <!--/noindex-->

      <div class="site-overview-wrap sidebar-panel">
        <div class="site-author motion-element" itemprop="author" itemscope itemtype="http://schema.org/Person">
    <img class="site-author-image" itemprop="image" alt="Dean Li"
      src="https://avatars.githubusercontent.com/u/55836943?v=4">
  <p class="site-author-name" itemprop="name">Dean Li</p>
  <div class="site-description" itemprop="description"></div>
</div>
<div class="site-state-wrap motion-element">
  <nav class="site-state">
      <div class="site-state-item site-state-posts">
          <a href="/archives/">
        
          <span class="site-state-item-count">55</span>
          <span class="site-state-item-name">posts</span>
        </a>
      </div>
      <div class="site-state-item site-state-tags">
            <a href="/tags/">
          
        <span class="site-state-item-count">18</span>
        <span class="site-state-item-name">tags</span></a>
      </div>
  </nav>
</div>
  <div class="links-of-author motion-element">
      <span class="links-of-author-item">
        <a href="https://github.com/thereallda" title="GitHub → https:&#x2F;&#x2F;github.com&#x2F;thereallda" rel="noopener" target="_blank"><i class="fab fa-github fa-fw"></i>GitHub</a>
      </span>
      <span class="links-of-author-item">
        <a href="mailto:lideanlda@foxmail.com" title="E-Mail → mailto:lideanlda@foxmail.com" rel="noopener" target="_blank"><i class="fa fa-envelope fa-fw"></i>E-Mail</a>
      </span>
      <span class="links-of-author-item">
        <a href="https://www.jianshu.com/u/ab8a1c7d0510" title="简书 → https:&#x2F;&#x2F;www.jianshu.com&#x2F;u&#x2F;ab8a1c7d0510" rel="noopener" target="_blank"><i class="fa fa-book fa-fw"></i>简书</a>
      </span>
  </div>



      </div>

    </div>
  </aside>
  <div id="sidebar-dimmer"></div>


      </div>
    </main>

    <footer class="footer">
      <div class="footer-inner">
        

        

<div class="copyright">
  
  &copy; 
  <span itemprop="copyrightYear">2022</span>
  <span class="with-love">
    <i class="fa fa-heart"></i>
  </span>
  <span class="author" itemprop="copyrightHolder">Dean Li</span>
</div>
  <div class="powered-by">Powered by <a href="https://hexo.io/" class="theme-link" rel="noopener" target="_blank">Hexo</a> & <a href="https://theme-next.org/" class="theme-link" rel="noopener" target="_blank">NexT.Gemini</a>
  </div>



        
<div class="busuanzi-count">
  <script async src="https://busuanzi.ibruce.info/busuanzi/2.3/busuanzi.pure.mini.js"></script>
    <span class="post-meta-item" id="busuanzi_container_site_uv" style="display: none;">
      <span class="post-meta-item-icon">
        <i class="fa fa-user"></i>
      </span>
      <span class="site-uv" title="Total Visitors">
        <span id="busuanzi_value_site_uv"></span>
      </span>
    </span>
    <span class="post-meta-divider">|</span>
    <span class="post-meta-item" id="busuanzi_container_site_pv" style="display: none;">
      <span class="post-meta-item-icon">
        <i class="fa fa-eye"></i>
      </span>
      <span class="site-pv" title="Total Views">
        <span id="busuanzi_value_site_pv"></span>
      </span>
    </span>
</div>








      </div>
    </footer>
  </div>

  
  <script src="/lib/anime.min.js"></script>
  <script src="/lib/velocity/velocity.min.js"></script>
  <script src="/lib/velocity/velocity.ui.min.js"></script>

<script src="/js/utils.js"></script>

<script src="/js/motion.js"></script>


<script src="/js/schemes/pisces.js"></script>


<script src="/js/next-boot.js"></script>




  















  

  

  
<script src="https://cdn.jsdelivr.net/npm/darkmode-js@1.5.7/lib/darkmode-js.min.js"></script>

<script>
var options = {
  bottom: '64px',
  right: '32px',
  left: 'unset',
  time: '0.5s',
  mixColor: 'transparent',
  backgroundColor: 'transparent',
  buttonColorDark: '#100f2c',
  buttonColorLight: '#fff',
  saveInCookies: true,
  label: '🌓',
  autoMatchOsTheme: true
}
const darkmode = new Darkmode(options);
window.darkmode = darkmode;
darkmode.showWidget();
</script>

<link rel="stylesheet" href="//cdn.jsdelivr.net/npm/gitalk@1/dist/gitalk.min.css">

<script>
NexT.utils.loadComments(document.querySelector('#gitalk-container'), () => {
  NexT.utils.getScript('//cdn.jsdelivr.net/npm/gitalk@1/dist/gitalk.min.js', () => {
    var gitalk = new Gitalk({
      clientID    : 'aea502b6092de143578a',
      clientSecret: '7bb9f5afb99d67ab37d8b296f39782af05dc8614',
      repo        : 'thereallda.GitHub.io',
      owner       : 'thereallda',
      admin       : ['thereallda'],
      id          : 'b099ee30c08971e181e5f895fab5e6c2',
        language: '',
      distractionFreeMode: true
    });
    gitalk.render('gitalk-container');
  }, window.Gitalk);
});
</script>

</body>
</html>
