<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
<meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=2">
<meta name="theme-color" content="#222">
<meta name="generator" content="Hexo 5.4.0">
  <link rel="apple-touch-icon" sizes="180x180" href="/images/apple-touch-icon.png">
  <link rel="icon" type="image/png" sizes="32x32" href="/images/favicon-32x32.png">
  <link rel="icon" type="image/png" sizes="16x16" href="/images/favicon-16x16.png">

<link rel="stylesheet" href="/css/main.css">


<link rel="stylesheet" href="/lib/font-awesome/css/all.min.css">

<script id="hexo-configurations">
    var NexT = window.NexT || {};
    var CONFIG = {"hostname":"thereallda.github.io","root":"/","scheme":"Gemini","version":"7.8.0","exturl":false,"sidebar":{"position":"left","display":"post","padding":18,"offset":12,"onmobile":false},"copycode":{"enable":true,"show_result":false,"style":"mac"},"back2top":{"enable":true,"sidebar":false,"scrollpercent":false},"bookmark":{"enable":false,"color":"#222","save":"auto"},"fancybox":false,"mediumzoom":false,"lazyload":false,"pangu":false,"comments":{"style":"tabs","active":"gitalk","storage":true,"lazyload":false,"nav":null,"activeClass":"gitalk"},"algolia":{"hits":{"per_page":10},"labels":{"input_placeholder":"Search for Posts","hits_empty":"We didn't find any results for the search: ${query}","hits_stats":"${hits} results found in ${time} ms"}},"localsearch":{"enable":false,"trigger":"auto","top_n_per_article":1,"unescape":false,"preload":false},"motion":{"enable":true,"async":false,"transition":{"post_block":"fadeIn","post_header":"slideDownIn","post_body":"slideDownIn","coll_header":"slideLeftIn","sidebar":"slideUpIn"}}};
  </script>

  <meta name="description" content="Following tutorial from 2019 Nature Protocol (https:&#x2F;&#x2F;www.nature.com&#x2F;articles&#x2F;s41596-018-0103-9) {width&#x3D;â€800â€}">
<meta property="og:type" content="article">
<meta property="og:title" content="é€šè·¯å¯Œé›†åˆ†æ">
<meta property="og:url" content="https://thereallda.github.io/2022/02/28/PathwayEnrichment/index.html">
<meta property="og:site_name" content="Dean&#39;s blog">
<meta property="og:description" content="Following tutorial from 2019 Nature Protocol (https:&#x2F;&#x2F;www.nature.com&#x2F;articles&#x2F;s41596-018-0103-9) {width&#x3D;â€800â€}">
<meta property="og:locale" content="en_US">
<meta property="og:image" content="https://thereallda.github.io/2022/02/28/PathwayEnrichment/image-20211231184636912.png">
<meta property="og:image" content="https://thereallda.github.io/2022/02/28/PathwayEnrichment/image-20220228165156178.png">
<meta property="og:image" content="https://thereallda.github.io/2022/02/28/PathwayEnrichment/Gmt_format_snapshot.gif">
<meta property="og:image" content="https://thereallda.github.io/g_Profiler_a_web_server_for_functional_enrichment_analysis_and_conversions_of_gene_lists_Google_Chrome_2022_1_4%2014_25_27.png">
<meta property="og:image" content="https://thereallda.github.io/2022/02/28/PathwayEnrichment/tmp1.png">
<meta property="og:image" content="https://thereallda.github.io/2022/02/28/PathwayEnrichment/tmp2.png">
<meta property="og:image" content="https://thereallda.github.io/Session_%20New%20Session%202022_1_4%2015_49_00_LI.jpg">
<meta property="og:image" content="https://thereallda.github.io/2022/02/28/PathwayEnrichment/New_enrichmentmap.png">
<meta property="og:image" content="https://thereallda.github.io/Session_%20New%20Session%202022_1_4%2016_44_23.png">
<meta property="og:image" content="https://thereallda.github.io/2022/02/28/PathwayEnrichment/img_20220104195150.png">
<meta property="og:image" content="https://thereallda.github.io/2022/02/28/PathwayEnrichment/autoannotate.png">
<meta property="og:image" content="https://thereallda.github.io/2022/02/28/PathwayEnrichment/nice_formated_network.png">
<meta property="article:published_time" content="2022-02-28T08:48:26.000Z">
<meta property="article:modified_time" content="2022-02-28T08:56:50.595Z">
<meta property="article:author" content="Dean Li">
<meta property="article:tag" content="Data_analysis">
<meta property="article:tag" content="R">
<meta property="article:tag" content="note">
<meta name="twitter:card" content="summary">
<meta name="twitter:image" content="https://thereallda.github.io/2022/02/28/PathwayEnrichment/image-20211231184636912.png">

<link rel="canonical" href="https://thereallda.github.io/2022/02/28/PathwayEnrichment/">


<script id="page-configurations">
  // https://hexo.io/docs/variables.html
  CONFIG.page = {
    sidebar: "",
    isHome : false,
    isPost : true,
    lang   : 'en'
  };
</script>

  <title>é€šè·¯å¯Œé›†åˆ†æ | Dean's blog</title>
  






  <noscript>
  <style>
  .use-motion .brand,
  .use-motion .menu-item,
  .sidebar-inner,
  .use-motion .post-block,
  .use-motion .pagination,
  .use-motion .comments,
  .use-motion .post-header,
  .use-motion .post-body,
  .use-motion .collection-header { opacity: initial; }

  .use-motion .site-title,
  .use-motion .site-subtitle {
    opacity: initial;
    top: initial;
  }

  .use-motion .logo-line-before i { left: initial; }
  .use-motion .logo-line-after i { right: initial; }
  </style>
</noscript>

<style>.darkmode--activated{--body-bg-color:#282828;--content-bg-color:#333;--card-bg-color:#555;--text-color:#ccc;--blockquote-color:#bbb;--link-color:#ccc;--link-hover-color:#eee;--brand-color:#ddd;--brand-hover-color:#ddd;--table-row-odd-bg-color:#282828;--table-row-hover-bg-color:#363636;--menu-item-bg-color:#555;--btn-default-bg:#222;--btn-default-color:#ccc;--btn-default-border-color:#555;--btn-default-hover-bg:#666;--btn-default-hover-color:#ccc;--btn-default-hover-border-color:#666;--highlight-background:#282b2e;--highlight-foreground:#a9b7c6;--highlight-gutter-background:#34393d;--highlight-gutter-foreground:#9ca9b6}.darkmode--activated img{opacity:.75}.darkmode--activated img:hover{opacity:.9}.darkmode--activated code{color:#69dbdc;background:0 0}button.darkmode-toggle{z-index:9999}.darkmode-ignore,img{display:flex!important}</style></head>

<body itemscope itemtype="http://schema.org/WebPage">
  <div class="container use-motion">
    <div class="headband"></div>

    <header class="header" itemscope itemtype="http://schema.org/WPHeader">
      <div class="header-inner"><div class="site-brand-container">
  <div class="site-nav-toggle">
    <div class="toggle" aria-label="Toggle navigation bar">
      <span class="toggle-line toggle-line-first"></span>
      <span class="toggle-line toggle-line-middle"></span>
      <span class="toggle-line toggle-line-last"></span>
    </div>
  </div>

  <div class="site-meta">

    <a href="/" class="brand" rel="start">
      <span class="logo-line-before"><i></i></span>
      <h1 class="site-title">Dean's blog</h1>
      <span class="logo-line-after"><i></i></span>
    </a>
  </div>

  <div class="site-nav-right">
    <div class="toggle popup-trigger">
    </div>
  </div>
</div>




<nav class="site-nav">
  <ul id="menu" class="main-menu menu">
        <li class="menu-item menu-item-home">

    <a href="/" rel="section"><i class="fa fa-home fa-fw"></i>Home</a>

  </li>
        <li class="menu-item menu-item-about">

    <a href="/about/" rel="section"><i class="fa fa-user fa-fw"></i>About</a>

  </li>
        <li class="menu-item menu-item-tags">

    <a href="/tags/" rel="section"><i class="fa fa-tags fa-fw"></i>Tags</a>

  </li>
        <li class="menu-item menu-item-archives">

    <a href="/archives/" rel="section"><i class="fa fa-archive fa-fw"></i>Archives</a>

  </li>
  </ul>
</nav>




</div>
    </header>

    
  <div class="back-to-top">
    <i class="fa fa-arrow-up"></i>
    <span>0%</span>
  </div>
  <div class="reading-progress-bar"></div>


    <main class="main">
      <div class="main-inner">
        <div class="content-wrap">
          

          <div class="content post posts-expand">
            

    
  
  
  <article itemscope itemtype="http://schema.org/Article" class="post-block" lang="en">
    <link itemprop="mainEntityOfPage" href="https://thereallda.github.io/2022/02/28/PathwayEnrichment/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="image" content="https://avatars.githubusercontent.com/u/55836943?v=4">
      <meta itemprop="name" content="Dean Li">
      <meta itemprop="description" content="">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="Dean's blog">
    </span>
      <header class="post-header">
        <h1 class="post-title" itemprop="name headline">
          é€šè·¯å¯Œé›†åˆ†æ
        </h1>

        <div class="post-meta">
          
          
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-calendar"></i>
              </span>
              <span class="post-meta-item-text">Posted on</span>
              

              <time title="Created: 2022-02-28 16:48:26 / Modified: 16:56:50" itemprop="dateCreated datePublished" datetime="2022-02-28T16:48:26+08:00">2022-02-28</time>
            </span>

          
            <span class="post-meta-item" title="Views" id="busuanzi_container_page_pv" style="display: none;">
              <span class="post-meta-item-icon">
                <i class="fa fa-eye"></i>
              </span>
              <span class="post-meta-item-text">Views: </span>
              <span id="busuanzi_value_page_pv"></span>
            </span><br>
            <span class="post-meta-item" title="Symbols count in article">
              <span class="post-meta-item-icon">
                <i class="far fa-file-word"></i>
              </span>
                <span class="post-meta-item-text">Symbols count in article: </span>
              <span>9.9k</span>
            </span>
            <span class="post-meta-item" title="Reading time">
              <span class="post-meta-item-icon">
                <i class="far fa-clock"></i>
              </span>
                <span class="post-meta-item-text">Reading time &asymp;</span>
              <span>9 mins.</span>
            </span>

        </div>
      </header>

    
    
    
    <div class="post-body" itemprop="articleBody">

      
        <p>Following tutorial from 2019 Nature Protocol (<a target="_blank" rel="noopener" href="https://www.nature.com/articles/s41596-018-0103-9">https://www.nature.com/articles/s41596-018-0103-9</a>)</p>
<p><img src="/2022/02/28/PathwayEnrichment/image-20211231184636912.png">{width=â€800â€}</p>
<span id="more"></span>

<h1 id="Prerequisite"><a href="#Prerequisite" class="headerlink" title="Prerequisite"></a>Prerequisite</h1><p><strong>Project directory:</strong> <code>Pathway_Enrichment_Turorial/</code></p>
<p><strong>Data:</strong> <a target="_blank" rel="noopener" href="https://static-content.springer.com/esm/art%3A10.1038%2Fs41596-018-0103-9/MediaObjects/41596_2018_103_MOESM1_ESM.zip">https://static-content.springer.com/esm/art%3A10.1038%2Fs41596-018-0103-9/MediaObjects/41596_2018_103_MOESM1_ESM.zip</a></p>
<p>ä¸‹è½½åè§£å‹äº<code>Pathway_Enrichment_Tutorial/data/</code>ï¼Œå‹ç¼©åŒ…å†…åŒ…æ‹¬ä»¥ä¸‹æ–‡ä»¶ï¼š</p>
<ul>
<li>  Supplementary_Methods_Supplementary_protocols.pdf</li>
<li>  Supplementary_Table1_Cancer_drivers.txt</li>
<li>  Supplementary_Table2_MesenvsImmuno_RNASeq_ranks.rnk</li>
<li>  Supplementary_Table3_Human_GOBP_AllPathways_no_GO_iea_July_01_2017_symbol.gmt</li>
<li>  Supplementary_Table4_gprofiler_results.txt</li>
<li>  Supplementary_Table5_hsapiens.pathways.NAME.gmt</li>
<li>  Supplementary_Table6_TCGA_OV_RNAseq_expression.txt</li>
<li>  Supplementary_Table7_TCGA_OV_RNAseq_classes.cls</li>
<li>  Supplementary_Table8_gsea_report_for_na_pos.xls</li>
<li>  Supplementary_Table9_gsea_report_for_na_neg.xls</li>
<li>  Supplementary_Table10_TCGA_RNASeq_rawcounts.txt</li>
<li>  Supplementary_Table11_RNASeq_classdefinitions.txt</li>
<li>  Supplementary_Table12_TCGA_Microarray_rmanormalized.txt</li>
<li>  Supplementary_Table13_Microarray_classdefinitions.txt</li>
</ul>
<p><strong>Software:</strong></p>
<ul>
<li>  g:Profiler (R client)</li>
<li>  Java Standard Edition (<a target="_blank" rel="noopener" href="http://www.oracle.com/technetwork/java/javase/downloads/index.html">http://www.oracle.com/technetwork/java/javase/downloads/index.html</a>) is required to run GSEA and Cytoscape.</li>
<li>  GSEA (optional)</li>
<li>  The Cytoscape desktop application (<a target="_blank" rel="noopener" href="http://www.cytoscape.org/download.php">http://www.cytoscape.org/download.php</a>)</li>
<li>  Cytoscape pluginsï¼šâ€EnrichmentMap Pipeline Collectionâ€ (<a target="_blank" rel="noopener" href="http://apps.cytoscape.org/apps/enrichmentmappipelinecollection">http://apps.cytoscape.org/apps/enrichmentmappipelinecollection</a> ); å¯ä»¥åœ¨Cytoscape app storeä¸­æœåˆ°</li>
</ul>
<h1 id="R-Pathway-enrichment-analysis"><a href="#R-Pathway-enrichment-analysis" class="headerlink" title="R | Pathway enrichment analysis"></a>R | Pathway enrichment analysis</h1><h2 id="using-g-Profiler"><a href="#using-g-Profiler" class="headerlink" title="using g:Profiler"></a>using g:Profiler</h2><blockquote>
<p>è¿™é‡Œä½¿ç”¨g:Profilerçš„RåŒ…è¿›è¡Œå¯Œé›†åˆ†æï¼Œä¹Ÿå¯ä»¥ä½¿ç”¨å®ƒçš„ç½‘é¡µç‰ˆï¼ˆ<a target="_blank" rel="noopener" href="https://biit.cs.ut.ee/gprofiler/gost">https://biit.cs.ut.ee/gprofiler/gost</a>è¿›è¡Œåˆ†æ</p>
</blockquote>
<ol>
<li> é¦–å…ˆï¼Œä½¿ç”¨Supplementary_Table1_Cancer_drivers.txtä¸­çš„åŸºå› åˆ—è¡¨è¿›è¡Œå¸¸è§„çš„é€šè·¯å¯Œé›†ã€‚</li>
</ol>
<p>In R</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">if</span> (!requireNamespace(<span class="string">&quot;gprofiler2&quot;</span>)) install.packages(<span class="string">&quot;gprofiler2&quot;</span>)</span><br><span class="line">library(gprofiler2)</span><br><span class="line"><span class="comment"># read in query gene list</span></span><br><span class="line">queryG &lt;- read.delim(<span class="string">&#x27;data/supplementary_files/Supplementary_Table1_Cancer_drivers.txt&#x27;</span>, header=<span class="built_in">F</span>)</span><br><span class="line"></span><br><span class="line">gores1 &lt;- gost(queryG$V1, </span><br><span class="line">               organism=<span class="string">&#x27;hsapiens&#x27;</span>,</span><br><span class="line">               exclude_iea=<span class="literal">TRUE</span>, </span><br><span class="line">               ordered_query=<span class="literal">TRUE</span>,</span><br><span class="line">               evcodes=<span class="literal">TRUE</span>,</span><br><span class="line">               sources=<span class="built_in">c</span>(<span class="string">&#x27;GO:BP&#x27;</span>, <span class="string">&#x27;REAC&#x27;</span>)) </span><br></pre></td></tr></table></figure>

<p>è¿™é‡Œé€‰æ‹©ç‰©ç§ä¸ºâ€hsapiensâ€ - Homo sapiens.</p>
<p><code>exclude_iea</code>: æ’é™¤æ²¡æœ‰ç»è¿‡äººå·¥å®¡æ ¸çš„ä¸é è°±GOæ³¨é‡Š (less reliable GO annotations, IEAs)</p>
<p><code>ordered_query</code>: æŠŠè¾“å…¥çš„åŸºå› è®¤ä¸ºæ˜¯æ’åºè¿‡çš„åŸºå› åˆ—è¡¨ï¼Œä»¥ä¸€ç§GSEAçš„æ–¹å¼è¿›è¡Œé€šè·¯å¯Œé›†</p>
<p><code>evcodes</code>: è¿”å›è¾“å…¥åŸºå› ä¸ç›¸åº”é€šè·¯äº§ç”Ÿäº¤é›†çš„åŸºå› ID</p>
<p><code>sources</code>: å¯Œé›†çš„æ•°æ®åº“ï¼Œè¿™é‡Œé€‰æ‹©GOçš„BPé¡¹å’ŒReactomeçš„é€šè·¯ã€‚</p>
<blockquote>
<p>Available data sources and their abbreviations are:</p>
<ul>
<li>  <a target="_blank" rel="noopener" href="http://geneontology.org/">Gene Ontology</a> (GO or by branch <a href="GO:MF">GO:MF</a>, <a href="GO:BP">GO:BP</a>, <a href="GO:CC">GO:CC</a>)</li>
<li>  <a target="_blank" rel="noopener" href="https://www.genome.jp/kegg/">KEGG</a> (KEGG)</li>
<li>  <a target="_blank" rel="noopener" href="https://reactome.org/">Reactome</a> (REAC): an open-source, open access, manually curated and peer-reviewed pathway database.</li>
<li>  <a target="_blank" rel="noopener" href="https://www.wikipathways.org/index.php/WikiPathways">WikiPathways</a> (WP): an open, collaborative platform dedicated to the curation of biological pathways.</li>
<li>  <a target="_blank" rel="noopener" href="https://genexplain.com/transfac/">TRANSFAC</a> (TF): is the database of eukaryotic transcription factors, their genomic binding sites and DNA-binding profiles.</li>
<li>  <a target="_blank" rel="noopener" href="https://mirtarbase.cuhk.edu.cn/~miRTarBase/miRTarBase_2019/php/index.php">miRTarBase</a> (MIRNA): <strong>miRTarBase</strong> has accumulated more than three hundred and sixty thousand miRNA-target interactions (MTIs), which are collected by manually surveying pertinent literature after NLP of the text systematically to filter research articles related to functional studies of miRNAs.</li>
<li>  <a target="_blank" rel="noopener" href="https://www.proteinatlas.org/">Human Protein Atlas</a> (HPA): including all the human proteins in cells, tissues, and organs using an integration of various omics technologies.</li>
<li>  <a target="_blank" rel="noopener" href="https://mips.helmholtz-muenchen.de/corum/">CORUM</a> (CORUM): provides a resource of manually annotated protein complexes from mammalian organisms.</li>
<li>  <a target="_blank" rel="noopener" href="http://hpo.jax.org/app/">Human phenotype ontology</a> (HP): provides a standardized vocabulary of phenotypic abnormalities encountered in human disease.</li>
</ul>
</blockquote>
<p>åŒæ—¶ï¼Œæˆ‘ä»¬å¯¹å¯Œé›†åˆ°çš„é€šè·¯è¿›è¡Œç­›é€‰ï¼Œæ’é™¤åŒ…å«è¿‡å¤šåŸºå› çš„é€šè·¯ï¼ˆå¤§äº350ä¸ªï¼‰ï¼Œè¿‡å°‘åŸºå› çš„é€šè·¯ï¼ˆå°äº5ä¸ªï¼‰ï¼Œä»¥åŠä¸è¾“å…¥åŸºå› äº¤é›†è¿‡å°çš„é€šè·¯ï¼ˆå°äº3ä¸ªï¼‰</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># filtered by term_size and intersection_size</span></span><br><span class="line"><span class="comment"># term_size - number of genes that are annotated to the term</span></span><br><span class="line"><span class="comment"># intersection_size - the number of genes in the input query that are annotated to the term</span></span><br><span class="line">gores1_filtered &lt;- <span class="built_in">list</span>(<span class="string">&quot;result&quot;</span> = subset(gores1$result, </span><br><span class="line">                                          term_size &gt; <span class="number">5</span> &amp; </span><br><span class="line">                                            term_size &lt; <span class="number">350</span> &amp; </span><br><span class="line">                                            intersection_size &gt; <span class="number">3</span>),</span><br><span class="line">                        <span class="string">&quot;meta&quot;</span> = gores1$meta)</span><br></pre></td></tr></table></figure>

<p>ç†è®ºä¸Šï¼Œæˆ‘ä»¬åº”å½“åœ¨è¿›è¡Œé€šè·¯å¯Œé›†ä¹‹å‰è¿›è¡Œæ’é™¤ï¼Œè¿™æ ·æ‰å¯ä»¥é€šè¿‡æ’é™¤è¿™ç±»é€šè·¯æ¥é€šè¿‡ç»Ÿè®¡æ£€éªŒçš„æ•ˆåŠ›ï¼ˆpowerï¼‰ï¼Œä½†æ˜¯<code>gprofiler2::gost()</code>å¹¶æ²¡æœ‰æä¾›è¿™æ ·çš„å‚æ•°ä»¥ä¾›æˆ‘ä»¬è®¾å®šç›¸åº”çš„cutoff</p>
<p>å¯Œé›†çš„ç»“æœä¿å­˜åœ¨<code>$table</code>ä¸­</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># pathway enrichment results table</span></span><br><span class="line">head(gores1_filtered$result, n=<span class="number">3</span>)</span><br></pre></td></tr></table></figure>

<pre><code>##       query significant  p_value term_size query_size intersection_size
## 95  query_1        TRUE 3.82e-13       142         81                15
## 124 query_1        TRUE 6.31e-12       197        120                18
## 125 query_1        TRUE 6.62e-12       230        120                19
##     precision recall    term_id source                        term_name
## 95      0.185 0.1056 GO:0048732  GO:BP                gland development
## 124     0.150 0.0914 GO:0048762  GO:BP mesenchymal cell differentiation
## 125     0.158 0.0826 GO:0060485  GO:BP           mesenchyme development
##     effective_domain_size source_order                parents
## 95                  16592        15077             GO:0048513
## 124                 16592        15103 GO:0030154, GO:0060485
## 125                 16592        17301 GO:0009888, GO:0048513
##                                                                                      evidence_codes
## 95                          ISS,IDA,ISS,ISS,IMP,ISS,TAS,IEP ISS,IGI ISS,IDA,IMP ISS,TAS,ISS,ISS,IDA
## 124         IMP,IMP,IMP ISS,IMP IGI TAS,IMP,ISS,ISS,TAS,ISS,IMP,IMP,ISS,ISS,IMP,IDA ISS,ISS,IMP,IDA
## 125 IMP,IMP,IMP ISS,IMP IGI TAS,IMP,ISS,ISS,ISS,TAS,ISS,IMP,IMP,ISS,ISS,IMP,IDA ISS,IMP ISS,IMP,IDA
##                                                                                                    intersection
## 95                          NF1,GATA3,FBXW7,ARHGAP35,TBX3,CEBPA,AKT1,SOX9,WT1,HGF,ERBB4,ARID5B,ELF3,FGFR2,BRCA2
## 124     PTEN,GATA3,NOTCH1,CTNNB1,TBX3,EPHA3,SOX9,HGF,ERBB4,MTOR,FOXA1,FGFR2,SMAD4,FOXA2,TGFBR2,SMAD2,AXIN2,EZH2
## 125 PTEN,GATA3,NOTCH1,CTNNB1,TBX3,EPHA3,SOX9,WT1,HGF,ERBB4,MTOR,FOXA1,FGFR2,SMAD4,FOXA2,TGFBR2,SMAD2,AXIN2,EZH2
</code></pre>
<blockquote>
<p>The result <code>data.frame</code> contains the following columns:</p>
<ul>
<li>  term_id - unique term identifier (e.g <a href="GO:0005005">GO:0005005</a>)</li>
<li>  <strong>p_values</strong> - hypergeometric p-values after correction for multiple testing in the order of input queries</li>
<li>  <strong>significant</strong> - indicators in the order of input queries for statistically significant results</li>
<li>  term_size - number of genes that are annotated to the term</li>
<li>  <strong>query_sizes</strong> - number of genes that were included in the query in the order of input queries</li>
<li>  <strong>intersection_sizes</strong> - the number of genes in the input query that are annotated to the corresponding term in the order of input queries</li>
<li>  source - the abbreviation of the data source for the term (e.g. <a href="GO:BP">GO:BP</a>)</li>
<li>  term_name - the short name of the function</li>
<li>  effective_domain_size - the total number of genes â€œin the universeâ€ used for the hypergeometric test</li>
<li>  source_order - numeric order for the term within its data source (this is important for drawing the results)</li>
<li>  source_order - numeric order for the term within its data source (this is important for drawing the results)</li>
<li>  parents - list of term IDs that are hierarchically directly above the term. For non-hierarchical data sources this points to an artificial root node.</li>
</ul>
</blockquote>
<p><code>gProfiler2</code>æä¾›Manhattan plotè¿›è¡Œé€šè·¯å¯Œé›†ç»“æœçš„å¯è§†åŒ–</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">gostplot(gores1_filtered, capped = <span class="literal">TRUE</span>, <span class="built_in">interactive</span> = <span class="literal">TRUE</span>)</span><br></pre></td></tr></table></figure>

<p><img src="/2022/02/28/PathwayEnrichment/image-20220228165156178.png"></p>
<h2 id="Preparing-data-for-EnrichmentMap"><a href="#Preparing-data-for-EnrichmentMap" class="headerlink" title="Preparing data for EnrichmentMap"></a>Preparing data for EnrichmentMap</h2><h3 id="Creating-a-Generic-Enrichment-Map-GEM-file"><a href="#Creating-a-Generic-Enrichment-Map-GEM-file" class="headerlink" title="Creating a Generic Enrichment Map (GEM) file"></a>Creating a Generic Enrichment Map (GEM) file</h3><p>GEMæ ¼å¼æ–‡ä»¶æ˜¯Cytoscape EnrichmentMap appä¸­éœ€è¦çš„é€šè·¯å¯Œé›†ç»“æœæ–‡ä»¶ã€‚å…¶éœ€è¦åŒ…æ‹¬GO ID, Description, p-value, FDR, Phenotype and Genes.</p>
<p>ç”±äº<code>gost</code>å‡½æ•°è¾“å‡ºçš„p-valueå·²ç»æ˜¯å¤šé‡æ£€éªŒæ ¡æ­£åçš„å€¼ï¼Œæ‰€ä»¥è¿™é‡Œç›´æ¥å¤åˆ¶ä¸¤åˆ—ä¸€æ ·çš„p-valueï¼Œå…¶å®éƒ½æ˜¯FDR</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">gem &lt;- gores1_filtered$result[,<span class="built_in">c</span>(<span class="string">&quot;term_id&quot;</span>, <span class="string">&quot;term_name&quot;</span>, <span class="string">&quot;p_value&quot;</span>, <span class="string">&quot;intersection&quot;</span>)]</span><br><span class="line">colnames(gem) &lt;- <span class="built_in">c</span>(<span class="string">&quot;GO.ID&quot;</span>, <span class="string">&quot;Description&quot;</span>, <span class="string">&quot;p.Val&quot;</span>, <span class="string">&quot;Genes&quot;</span>)</span><br><span class="line">gem$FDR &lt;- gem$p.Val</span><br><span class="line">gem$Phenotype = <span class="string">&quot;+1&quot;</span></span><br><span class="line">gem &lt;- gem[,<span class="built_in">c</span>(<span class="string">&quot;GO.ID&quot;</span>, <span class="string">&quot;Description&quot;</span>, <span class="string">&quot;p.Val&quot;</span>, <span class="string">&quot;FDR&quot;</span>, <span class="string">&quot;Phenotype&quot;</span>, <span class="string">&quot;Genes&quot;</span>)]</span><br><span class="line">head(gem, n=<span class="number">3</span>)</span><br></pre></td></tr></table></figure>

<pre><code>##          GO.ID                      Description    p.Val      FDR Phenotype
## 95  GO:0048732                gland development 3.82e-13 3.82e-13        +1
## 124 GO:0048762 mesenchymal cell differentiation 6.31e-12 6.31e-12        +1
## 125 GO:0060485           mesenchyme development 6.62e-12 6.62e-12        +1
##                                                                                                           Genes
## 95                          NF1,GATA3,FBXW7,ARHGAP35,TBX3,CEBPA,AKT1,SOX9,WT1,HGF,ERBB4,ARID5B,ELF3,FGFR2,BRCA2
## 124     PTEN,GATA3,NOTCH1,CTNNB1,TBX3,EPHA3,SOX9,HGF,ERBB4,MTOR,FOXA1,FGFR2,SMAD4,FOXA2,TGFBR2,SMAD2,AXIN2,EZH2
## 125 PTEN,GATA3,NOTCH1,CTNNB1,TBX3,EPHA3,SOX9,WT1,HGF,ERBB4,MTOR,FOXA1,FGFR2,SMAD4,FOXA2,TGFBR2,SMAD2,AXIN2,EZH2
</code></pre>
<p>ä¿å­˜è¯¥æ–‡ä»¶ï¼Œåœ¨Cytoscapeä¸­ä½¿ç”¨</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">write.table(gem, file = <span class="string">&quot;extdata/gProfiler_gem.txt&quot;</span>, sep = <span class="string">&quot;\t&quot;</span>, <span class="built_in">quote</span> = <span class="built_in">F</span>, row.names = <span class="built_in">F</span>)</span><br></pre></td></tr></table></figure>

<h3 id="the-Gene-Matrix-Transposed-file-format-GMT"><a href="#the-Gene-Matrix-Transposed-file-format-GMT" class="headerlink" title="the Gene Matrix Transposed file format (GMT)"></a>the Gene Matrix Transposed file format (GMT)</h3><p>GMTæ ¼å¼æ–‡ä»¶æ˜¯ä¸€ç§æè¿°åŸºå› é›†çš„æ–‡ä»¶æ ¼å¼ï¼›å…¶ä¸­æ¯ä¸€è¡Œä»£è¡¨ä¸€ä¸ªåŸºå› é›† ï¼ŒGMTç¤ºä¾‹å¦‚ä¸‹å›¾ï¼š<img src="/2022/02/28/PathwayEnrichment/Gmt_format_snapshot.gif"></p>
<p>å…¶ä¸­ç¬¬ä¸€åˆ—ä¸ºåŸºå› é›†çš„IDï¼Œç¬¬äºŒåˆ—æ˜¯åŸºå› é›†çš„æè¿°ï¼Œåç»­çš„åˆ—ä¸ºè¯¥åŸºå› é›†ä¸­çš„åŸºå› ID</p>
<p>å¦‚æœæ˜¯ä½¿ç”¨<code>gprofiler</code>è¿›è¡Œåˆ†æï¼Œå¯ä»¥åœ¨å…¶ç½‘ç«™ä¸Šä¸‹è½½å¯¹åº”è¾“å…¥åŸºå› IDçš„GMTæ–‡ä»¶ï¼ˆ<a target="_blank" rel="noopener" href="https://biit.cs.ut.ee/gprofiler/gost">https://biit.cs.ut.ee/gprofiler/gost</a>ï¼‰</p>
<p>é¦–å…ˆï¼Œåœ¨Organismçš„ä¸‹æ‹‰èœå•ä¸­é€‰æ‹©ç›¸åº”ç‰©ç§ï¼Œæˆ‘ä»¬è¿™é‡Œé€‰æ‹©Homo sapiens</p>
<p>![](g_Profiler_a_web_server_for_functional_enrichment_analysis_and_conversions_of_gene_lists_Google_Chrome_2022_1_4 14_25_46.png)</p>
<p>ç„¶åï¼Œæ‰“å¼€[Data sources]èœå•ï¼Œä¸‹è½½ç›¸åº”åŸºå› IDçš„GMT</p>
<p><img src="/g_Profiler_a_web_server_for_functional_enrichment_analysis_and_conversions_of_gene_lists_Google_Chrome_2022_1_4%2014_25_27.png"></p>
<p>æˆ‘ä»¬è¿™é‡Œä½¿ç”¨symbolä½œä¸ºåŸºå› IDï¼Œä¸‹è½½<code>name.gmt.zip</code>è¿™ä¸ªæ–‡ä»¶</p>
<p><img src="/2022/02/28/PathwayEnrichment/tmp1.png"></p>
<p>æˆ‘ä»¬è¿™é‡Œåªä½¿ç”¨<code>GO:BP</code>å’Œ<code>REAC</code>çš„æ³¨é‡Šï¼Œæ•…åˆå¹¶è¿™ä¸¤ä¸ªæ–‡ä»¶ä¸º<code>hsapiens.GOBP_REAC.name.gmt</code>ä½œä¸ºåç»­åˆ†æä½¿ç”¨çš„æ–‡ä»¶ ï¼Œæ–‡ä»¶çš„å‰å‡ è¡Œå¦‚ä¸‹æ‰€ç¤ºï¼š</p>
<p><img src="/2022/02/28/PathwayEnrichment/tmp2.png"></p>
<p><strong>P.S.</strong> <code>combined.name.gmt.zip</code>æä¾›çš„æ˜¯å„ä¸ªæ•°æ®åº“åˆå¹¶åçš„GMT</p>
<p>å¦å¤–ï¼Œgprofileræä¾›çš„gmtä¸åŒ…å«KEGGå’ŒTRANSFACçš„æ•°æ®</p>
<blockquote>
<p>â€œNote that this file does not include annotations from KEGG and Transfac as we are restricted by data source licenses that do not allow us to share these two data sources with our users.â€</p>
</blockquote>
<h1 id="Cytoscape-Visualization-of-enrichment-results-with-EnrichmentMap"><a href="#Cytoscape-Visualization-of-enrichment-results-with-EnrichmentMap" class="headerlink" title="Cytoscape | Visualization of enrichment results with EnrichmentMap"></a>Cytoscape | Visualization of enrichment results with EnrichmentMap</h1><p>æ¥ä¸‹æ¥çš„æ“ä½œéœ€è¦åœ¨Cytoscapeä¸­è¿›è¡Œï¼Œç¡®ä¿ä½ çš„ç”µè„‘å®‰è£…äº†Cytoscapeï¼Œä»¥åŠå…¶ä¸­çš„EnrichmentMap app</p>
<ol>
<li> æ‰“å¼€Cytoscapeåï¼Œç‚¹å‡»ä¸Šæ–¹çš„[Apps]é€‰é¡¹å¡</li>
<li> æ‰“å¼€[EnrichmentMap]</li>
</ol>
<p><img src="/Session_%20New%20Session%202022_1_4%2015_49_00_LI.jpg"></p>
<p>åœ¨å¼¹å‡ºçš„Enrichment Mapçª—å£ä¸­ï¼Œ</p>
<p><img src="/2022/02/28/PathwayEnrichment/New_enrichmentmap.png"></p>
<ol>
<li> ç‚¹å‡»â€+â€æ–°å»ºåˆ†æ</li>
<li> ä¿®æ”¹åˆ†æçš„åå­—</li>
<li> é€‰æ‹©å¯Œé›†åˆ†æçš„ç±»å‹ï¼Œè¿™é‡Œæˆ‘ä»¬é€‰æ‹©â€Generic/gProfiler/Enrichrâ€</li>
<li> è¾“å…¥GEMæ ¼å¼çš„å¯Œé›†åˆ†æç»“æœï¼Œè¿™é‡Œè¾“å…¥æˆ‘ä»¬å‰é¢ä¿å­˜çš„<code>extdata/gProfiler_gem.txt</code></li>
<li> è¾“å…¥GMTæ ¼å¼çš„åŸºå› é›†æ³¨é‡Šæ–‡ä»¶ï¼Œè¿™é‡Œè¾“å…¥åˆå¹¶çš„<code>hsapiens.GOBP_REAC.name.gmt</code></li>
<li> (å¯é€‰)è¾“å…¥åŸºå› è¡¨è¾¾çš„æ’åºæ–‡ä»¶ï¼Œè¿™é‡Œæˆ‘ä»¬ç”¨æ–‡çŒ®æä¾›çš„<code>Supplementary_Table2_MesenvsImmuno_RNASeq_ranks.rnk</code></li>
<li> (å¯é€‰)è¾“å…¥åŸºå› è¡¨è¾¾çŸ©é˜µæ–‡ä»¶ï¼Œè¿™é‡Œæˆ‘ä»¬ç”¨æ–‡çŒ®æä¾›çš„<code>Supplementary_Table6_TCGA_OV_RNAseq_expression.txt</code></li>
<li> (å¯é€‰)æ ¹æ®è¡¨è¾¾æ–‡ä»¶è¿›è¡ŒåŸºå› ç­›é€‰ï¼Œå‹¾é€‰è¯¥é€‰é¡¹ä¼šå¯¹GMTæ–‡ä»¶çš„åŸºå› è¿›è¡Œç­›é€‰ï¼Œæ’é™¤ä¸åœ¨è¡¨è¾¾æ–‡ä»¶ä¸­çš„åŸºå› ã€‚</li>
<li> (å¯é€‰)è®¾ç½®æ¯ä¸ªå¯Œé›†æ¡ç›®çš„FDR cutoffï¼Œé«˜äºè¯¥cutoffçš„å¯Œé›†é¡¹å°†ä¼šè¢«æ’é™¤ï¼Œé»˜è®¤æ˜¯0.05ï¼Œè¿™é‡Œè®¾ç½®ä¸º0.01</li>
<li>(å¯é€‰)å¯¹ç½‘ç»œè¿æ¥åº¦è¿›è¡Œè°ƒæ•´ã€‚è¶Šå¾€å·¦ç”Ÿæˆçš„ç½‘ç»œè¶Šç¨€ç–ï¼Œè¶Šå¾€å³ç”Ÿæˆçš„ç½‘ç»œè¶Šå¯†é›†ã€‚æœ¬è´¨ä¸Šæ˜¯å¯¹å¯Œé›†é¡¹ä¹‹é—´çš„ç›¸å…³æ€§é˜ˆå€¼è¿›è¡Œè°ƒæ•´ï¼Œä»è€Œå‡å°‘/å¢åŠ æ˜¾ç¤ºçš„è¾¹ã€‚ç‚¹å‡»[Show advanced options]ä¼šæ˜¾ç¤ºæ›´å¤šå…³äºç½‘ç»œçš„è°ƒæ•´é¡¹ç›®</li>
</ol>
<p>æœ€åï¼Œç‚¹å‡»**[Build]**ç”Ÿæˆç½‘ç»œ</p>
<p><img src="/Session_%20New%20Session%202022_1_4%2016_44_23.png"></p>
<p>åœ¨[Layout]èœå•ä¸­ï¼Œå¯ä»¥å°è¯•æ›´æ¢ä¸ºä¸åŒçš„ç½‘ç»œå¸ƒå±€ï¼Œçœ‹å“ªä¸€ç§å¸ƒå±€å±•ç¤ºçš„ç½‘ç»œæ•ˆæœæœ€ä½³ï¼Œé»˜è®¤ä¸ºâ€the unweighted Prefuse Force Directed layoutâ€.</p>
<h1 id="Cytoscape-Auto-annotate-clustered-enriched-terms"><a href="#Cytoscape-Auto-annotate-clustered-enriched-terms" class="headerlink" title="Cytoscape | Auto annotate clustered enriched terms"></a>Cytoscape | Auto annotate clustered enriched terms</h1><blockquote>
<p>Enrichment maps typically include clusters of similar pathways representing major biological themes.</p>
</blockquote>
<p>å¯Œé›†è°±å›¾(Enrichment Map)å¸¸å¸¸åŒ…æ‹¬äº†ä»£è¡¨ä¸»è¦çš„ç”Ÿç‰©å­¦ä¸»é¢˜çš„</p>
<p>è¿™ä¸€èŠ‚ï¼Œæˆ‘ä»¬å°†ä½¿ç”¨Cytoscapeä¸­çš„â€AutoAnnotateâ€ appè¿›è¡Œé€šè·¯clustersçš„è‡ªåŠ¨æ³¨é‡Šã€‚</p>
<p>åœ¨è¿›è¡ŒAutoAnnotateå‰ï¼Œè¦ç¡®è®¤ç½‘ç»œçš„layoutæ˜¯è¾ƒä¼˜çš„æ–¹æ¡ˆã€‚è¿™æ˜¯ç”±äºå®ƒä¼šæ ¹æ®å½“å‰çš„layoutè¿›è¡Œlabelã€‚</p>
<ol>
<li> ç‚¹å‡»[Apps] â€“&gt; [AutoAnnotate] â€“&gt; [New Annotation Setâ€¦]</li>
</ol>
<p>![](Session_ New Session 2022_1_4 19_51_19.png)</p>
<ol start="2">
<li> åœ¨â€Quick Startâ€é¡µé¢ä¸­ï¼Œç‚¹å‡»[Create Annotations]</li>
</ol>
<p><img src="/2022/02/28/PathwayEnrichment/img_20220104195150.png"></p>
<p>è¿è¡Œåï¼Œç½‘ç»œä¸ŠåŠŸèƒ½ç›¸è¿‘çš„é€šè·¯èŠ‚ç‚¹ä¼šè¢«åœˆä¸ºä¸€ä¸ªclusterï¼Œå¹¶æ ‡æ³¨ä¸Šç›¸åº”çš„biological themesã€‚æ ‡æ³¨çš„å­—ä½“è¶Šå¤§è¯´æ˜è¯¥clusteråŒ…å«çš„èŠ‚ç‚¹è¶Šå¤šã€‚å¦å¤–ï¼Œåœ¨Control panelä¸­ä¼šå¤šäº†â€AutoAnnotationsâ€è¿™ä¸€ä¸ªé¡µé¢ã€‚</p>
<p><img src="/2022/02/28/PathwayEnrichment/autoannotate.png"></p>
<p>è¿™æ ·ç”Ÿæˆçš„æ³¨é‡Šç»“æœä¸€èˆ¬æ˜¯æ²¡æ³•çœ‹çš„ï¼Œæˆ‘ä»¬éœ€è¦å°†é‡å çš„åœˆå’Œæ ‡æ³¨é”™å¼€ï¼Œå°†â€æ— å…³çš„â€clusteræ³¨é‡Šå»é™¤ï¼Œæœ€ç»ˆè¾¾åˆ°å‘è¡¨çº§åˆ«çš„æ ·å­</p>
<p><img src="/2022/02/28/PathwayEnrichment/nice_formated_network.png"></p>
<p>æœ€åï¼Œæˆ‘ä»¬å¯ä»¥å°†è¿è¡Œçš„cytoscape sessionä¿å­˜ä¸º<code>.cys</code>æ–‡ä»¶ï¼Œæ–¹ä¾¿ä¹‹åå†åˆ†æã€‚</p>
<hr>
<p><strong>åœ¨ç”Ÿæˆå¯Œé›†è°±å›¾(Enrichment Map)åæ›´ä¸ºå…³é”®çš„æ˜¯å¯¹ç½‘ç»œçš„è§£è¯»ï¼</strong></p>
<p>é€šè·¯ä¿¡æ¯å¤©ç„¶çš„å…·æœ‰å†—ä½™çš„ç‰¹è´¨ï¼Œå› æ­¤é€šè·¯å¯Œé›†åˆ†æå¾€å¾€ä¼šç»™å‡ºå¯¹åŒä¸€ä¸ªé€šè·¯ä¸åŒæè¿°çš„æ³¨é‡Šç»“æœã€‚å¯Œé›†è°±å›¾ <strong>(Enrichment Map)</strong> ç®€åŒ–äº†å†—ä½™çš„é€šè·¯ä¿¡æ¯ï¼Œå°†å…¶å±•ç¤ºä¸ºæœ‰ä»£è¡¨æ€§çš„ç”Ÿç‰©å­¦ä¸»é¢˜ (biological themes)ï¼Œæ›´æœ‰åˆ©äºæˆ‘ä»¬å¯¹åŸºå› é€šè·¯å¯Œé›†ç»“æœçš„è§£è¯»ã€‚ä½†è¿™åªæ˜¯ç›¸å½“äºå¸®åŠ©æˆ‘ä»¬å‡ç»ƒäº†é€šè·¯å¯Œé›†çš„ä¿¡æ¯ï¼Œè€Œå¯¹äºä¿¡æ¯çš„è§£è¯»è¿˜æ˜¯ä¾èµ–äºæˆ‘ä»¬çš„ç”Ÿç‰©å­¦çŸ¥è¯†ã€‚æˆ‘ä»¬éœ€è¦ä¾æ®å®éªŒçš„å‡è®¾ã€è¯¾é¢˜çš„èƒŒæ™¯åšå‡ºåˆç†çš„è§£é‡Šã€‚ä¾‹å¦‚åœ¨æˆ‘ä»¬è¿™ç¯‡æ–‡ç« çš„ä¾‹å­ä¸­å¯¹ç™Œç—‡å¸¸è§çš„é©±åŠ¨çªå˜åŸºå› è¿›è¡Œåˆ†æï¼Œå°±å‘ç°äº†è®¸å¤šä¸è‚¿ç˜¤ç”Ÿé•¿ç›¸å…³çš„é€šè·¯ã€‚å½“å…¨é¢å±•ç°å‡ºåŸºå› é›†çš„é€šè·¯ä¿¡æ¯åï¼Œæˆ‘ä»¬è¿˜å¯ä»¥æ ¹æ®æ„Ÿå…´è¶£çš„é€šè·¯æ¥å¯»æ‰¾ç›¸å…³çš„åŸºå› è¿›è¡Œåç»­åˆ†æã€‚æ€»è€Œè¨€ä¹‹ï¼Œè¿›è¡Œæ•°æ®åˆ†ææ›´ä¸ºå…³é”®çš„æ˜¯å¯¹æ•°æ®ç»“æœçš„è§£è¯»ã€‚</p>
<blockquote>
<p>Ref</p>
<p><strong>2019 Nature Protocol</strong>: <a target="_blank" rel="noopener" href="https://doi.org/10.1038/s41596-018-0103-9">https://doi.org/10.1038/s41596-018-0103-9</a></p>
<p><strong>gprofiler2 vignettes</strong>: <a target="_blank" rel="noopener" href="https://cran.r-project.org/web/packages/gprofiler2/vignettes/gprofiler2.html">https://cran.r-project.org/web/packages/gprofiler2/vignettes/gprofiler2.html</a></p>
<p><strong>Cytoscape</strong>: <a target="_blank" rel="noopener" href="https://cytoscape.org/">https://cytoscape.org/</a></p>
</blockquote>

    </div>

    
    
    
        

<div>
<ul class="post-copyright">
  <li class="post-copyright-author">
    <strong>Post author:  </strong>Dean Li
  </li>
  <li class="post-copyright-link">
    <strong>Post link: </strong>
    <a href="https://thereallda.github.io/2022/02/28/PathwayEnrichment/" title="é€šè·¯å¯Œé›†åˆ†æ">https://thereallda.github.io/2022/02/28/PathwayEnrichment/</a>
  </li>
  <li class="post-copyright-license">
    <strong>Copyright Notice:  </strong>All articles in this blog are licensed under <a href="https://creativecommons.org/licenses/by-nc-sa/4.0/" rel="noopener" target="_blank"><i class="fab fa-fw fa-creative-commons"></i>BY-NC-SA</a> unless stating additionally.
  </li>
</ul>
</div>


      <footer class="post-footer">
          <div class="post-tags">
              <a href="/tags/Data-analysis/" rel="tag"># Data_analysis</a>
              <a href="/tags/R/" rel="tag"># R</a>
              <a href="/tags/note/" rel="tag"># note</a>
          </div>

        


        
    <div class="post-nav">
      <div class="post-nav-item">
    <a href="/2022/02/10/R-RNA-seq-data-simulation/" rel="prev" title="RNA-seqæµ‹åºæ•°æ®æ¨¡æ‹Ÿ">
      <i class="fa fa-chevron-left"></i> RNA-seqæµ‹åºæ•°æ®æ¨¡æ‹Ÿ
    </a></div>
      <div class="post-nav-item"></div>
    </div>
      </footer>
    
  </article>
  
  
  



          </div>
          
    <div class="comments" id="gitalk-container"></div>

<script>
  window.addEventListener('tabs:register', () => {
    let { activeClass } = CONFIG.comments;
    if (CONFIG.comments.storage) {
      activeClass = localStorage.getItem('comments_active') || activeClass;
    }
    if (activeClass) {
      let activeTab = document.querySelector(`a[href="#comment-${activeClass}"]`);
      if (activeTab) {
        activeTab.click();
      }
    }
  });
  if (CONFIG.comments.storage) {
    window.addEventListener('tabs:click', event => {
      if (!event.target.matches('.tabs-comment .tab-content .tab-pane')) return;
      let commentClass = event.target.classList[1];
      localStorage.setItem('comments_active', commentClass);
    });
  }
</script>

        </div>
          
  
  <div class="toggle sidebar-toggle">
    <span class="toggle-line toggle-line-first"></span>
    <span class="toggle-line toggle-line-middle"></span>
    <span class="toggle-line toggle-line-last"></span>
  </div>

  <aside class="sidebar">
    <div class="sidebar-inner">

      <ul class="sidebar-nav motion-element">
        <li class="sidebar-nav-toc">
          Table of Contents
        </li>
        <li class="sidebar-nav-overview">
          Overview
        </li>
      </ul>

      <!--noindex-->
      <div class="post-toc-wrap sidebar-panel">
          <div class="post-toc motion-element"><ol class="nav"><li class="nav-item nav-level-1"><a class="nav-link" href="#Prerequisite"><span class="nav-number">1.</span> <span class="nav-text">Prerequisite</span></a></li><li class="nav-item nav-level-1"><a class="nav-link" href="#R-Pathway-enrichment-analysis"><span class="nav-number">2.</span> <span class="nav-text">R | Pathway enrichment analysis</span></a><ol class="nav-child"><li class="nav-item nav-level-2"><a class="nav-link" href="#using-g-Profiler"><span class="nav-number">2.1.</span> <span class="nav-text">using g:Profiler</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#Preparing-data-for-EnrichmentMap"><span class="nav-number">2.2.</span> <span class="nav-text">Preparing data for EnrichmentMap</span></a><ol class="nav-child"><li class="nav-item nav-level-3"><a class="nav-link" href="#Creating-a-Generic-Enrichment-Map-GEM-file"><span class="nav-number">2.2.1.</span> <span class="nav-text">Creating a Generic Enrichment Map (GEM) file</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#the-Gene-Matrix-Transposed-file-format-GMT"><span class="nav-number">2.2.2.</span> <span class="nav-text">the Gene Matrix Transposed file format (GMT)</span></a></li></ol></li></ol></li><li class="nav-item nav-level-1"><a class="nav-link" href="#Cytoscape-Visualization-of-enrichment-results-with-EnrichmentMap"><span class="nav-number">3.</span> <span class="nav-text">Cytoscape | Visualization of enrichment results with EnrichmentMap</span></a></li><li class="nav-item nav-level-1"><a class="nav-link" href="#Cytoscape-Auto-annotate-clustered-enriched-terms"><span class="nav-number">4.</span> <span class="nav-text">Cytoscape | Auto annotate clustered enriched terms</span></a></li></ol></div>
      </div>
      <!--/noindex-->

      <div class="site-overview-wrap sidebar-panel">
        <div class="site-author motion-element" itemprop="author" itemscope itemtype="http://schema.org/Person">
    <img class="site-author-image" itemprop="image" alt="Dean Li"
      src="https://avatars.githubusercontent.com/u/55836943?v=4">
  <p class="site-author-name" itemprop="name">Dean Li</p>
  <div class="site-description" itemprop="description"></div>
</div>
<div class="site-state-wrap motion-element">
  <nav class="site-state">
      <div class="site-state-item site-state-posts">
          <a href="/archives/">
        
          <span class="site-state-item-count">50</span>
          <span class="site-state-item-name">posts</span>
        </a>
      </div>
      <div class="site-state-item site-state-tags">
            <a href="/tags/">
          
        <span class="site-state-item-count">16</span>
        <span class="site-state-item-name">tags</span></a>
      </div>
  </nav>
</div>
  <div class="links-of-author motion-element">
      <span class="links-of-author-item">
        <a href="https://github.com/thereallda" title="GitHub â†’ https:&#x2F;&#x2F;github.com&#x2F;thereallda" rel="noopener" target="_blank"><i class="fab fa-github fa-fw"></i>GitHub</a>
      </span>
      <span class="links-of-author-item">
        <a href="mailto:lideanlda@foxmail.com" title="E-Mail â†’ mailto:lideanlda@foxmail.com" rel="noopener" target="_blank"><i class="fa fa-envelope fa-fw"></i>E-Mail</a>
      </span>
      <span class="links-of-author-item">
        <a href="https://www.jianshu.com/u/ab8a1c7d0510" title="ç®€ä¹¦ â†’ https:&#x2F;&#x2F;www.jianshu.com&#x2F;u&#x2F;ab8a1c7d0510" rel="noopener" target="_blank"><i class="fa fa-book fa-fw"></i>ç®€ä¹¦</a>
      </span>
  </div>



      </div>

    </div>
  </aside>
  <div id="sidebar-dimmer"></div>


      </div>
    </main>

    <footer class="footer">
      <div class="footer-inner">
        

        

<div class="copyright">
  
  &copy; 
  <span itemprop="copyrightYear">2022</span>
  <span class="with-love">
    <i class="fa fa-heart"></i>
  </span>
  <span class="author" itemprop="copyrightHolder">Dean Li</span>
</div>
  <div class="powered-by">Powered by <a href="https://hexo.io/" class="theme-link" rel="noopener" target="_blank">Hexo</a> & <a href="https://theme-next.org/" class="theme-link" rel="noopener" target="_blank">NexT.Gemini</a>
  </div>



        
<div class="busuanzi-count">
  <script async src="https://busuanzi.ibruce.info/busuanzi/2.3/busuanzi.pure.mini.js"></script>
    <span class="post-meta-item" id="busuanzi_container_site_uv" style="display: none;">
      <span class="post-meta-item-icon">
        <i class="fa fa-user"></i>
      </span>
      <span class="site-uv" title="Total Visitors">
        <span id="busuanzi_value_site_uv"></span>
      </span>
    </span>
    <span class="post-meta-divider">|</span>
    <span class="post-meta-item" id="busuanzi_container_site_pv" style="display: none;">
      <span class="post-meta-item-icon">
        <i class="fa fa-eye"></i>
      </span>
      <span class="site-pv" title="Total Views">
        <span id="busuanzi_value_site_pv"></span>
      </span>
    </span>
</div>








      </div>
    </footer>
  </div>

  
  <script src="/lib/anime.min.js"></script>
  <script src="/lib/velocity/velocity.min.js"></script>
  <script src="/lib/velocity/velocity.ui.min.js"></script>

<script src="/js/utils.js"></script>

<script src="/js/motion.js"></script>


<script src="/js/schemes/pisces.js"></script>


<script src="/js/next-boot.js"></script>




  















  

  

  
<script src="https://cdn.jsdelivr.net/npm/darkmode-js@1.5.7/lib/darkmode-js.min.js"></script>

<script>
var options = {
  bottom: '64px',
  right: '32px',
  left: 'unset',
  time: '0.5s',
  mixColor: 'transparent',
  backgroundColor: 'transparent',
  buttonColorDark: '#100f2c',
  buttonColorLight: '#fff',
  saveInCookies: true,
  label: 'ğŸŒ“',
  autoMatchOsTheme: true
}
const darkmode = new Darkmode(options);
window.darkmode = darkmode;
darkmode.showWidget();
</script>

<link rel="stylesheet" href="//cdn.jsdelivr.net/npm/gitalk@1/dist/gitalk.min.css">

<script>
NexT.utils.loadComments(document.querySelector('#gitalk-container'), () => {
  NexT.utils.getScript('//cdn.jsdelivr.net/npm/gitalk@1/dist/gitalk.min.js', () => {
    var gitalk = new Gitalk({
      clientID    : 'aea502b6092de143578a',
      clientSecret: '7bb9f5afb99d67ab37d8b296f39782af05dc8614',
      repo        : 'thereallda.GitHub.io',
      owner       : 'thereallda',
      admin       : ['thereallda'],
      id          : '767850ef6006d888c28098fd82ab6df1',
        language: '',
      distractionFreeMode: true
    });
    gitalk.render('gitalk-container');
  }, window.Gitalk);
});
</script>

</body>
</html>
