<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
<meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=2">
<meta name="theme-color" content="#222">
<meta name="generator" content="Hexo 5.4.0">
  <link rel="apple-touch-icon" sizes="180x180" href="/images/apple-touch-icon.png">
  <link rel="icon" type="image/png" sizes="32x32" href="/images/favicon-32x32.png">
  <link rel="icon" type="image/png" sizes="16x16" href="/images/favicon-16x16.png">

<link rel="stylesheet" href="/css/main.css">


<link rel="stylesheet" href="/lib/font-awesome/css/all.min.css">

<script id="hexo-configurations">
    var NexT = window.NexT || {};
    var CONFIG = {"hostname":"thereallda.github.io","root":"/","scheme":"Gemini","version":"7.8.0","exturl":false,"sidebar":{"position":"left","display":"post","padding":18,"offset":12,"onmobile":false},"copycode":{"enable":true,"show_result":false,"style":"mac"},"back2top":{"enable":true,"sidebar":false,"scrollpercent":false},"bookmark":{"enable":false,"color":"#222","save":"auto"},"fancybox":false,"mediumzoom":false,"lazyload":false,"pangu":false,"comments":{"style":"tabs","active":"gitalk","storage":true,"lazyload":false,"nav":null,"activeClass":"gitalk"},"algolia":{"hits":{"per_page":10},"labels":{"input_placeholder":"Search for Posts","hits_empty":"We didn't find any results for the search: ${query}","hits_stats":"${hits} results found in ${time} ms"}},"localsearch":{"enable":false,"trigger":"auto","top_n_per_article":1,"unescape":false,"preload":false},"motion":{"enable":true,"async":false,"transition":{"post_block":"fadeIn","post_header":"slideDownIn","post_body":"slideDownIn","coll_header":"slideLeftIn","sidebar":"slideUpIn"}}};
  </script>

  <meta name="description" content="本文写于观看生信技能树公众号（vx: biotrainee）的七步走纯R代码通过数据挖掘复现一篇实验文章（第1到6步）一文后，感觉生信技能树优秀学徒的工作十分吸引人，就自己动手复现了一次。  Step 00 问题概述  本文的任务是全代码复现一篇paper，标题为 ：Co-expression networks revealed potential core lncRNAs in the tri">
<meta property="og:type" content="article">
<meta property="og:title" content="R-TCGA数据挖掘复现之BRCA">
<meta property="og:url" content="https://thereallda.github.io/2022/01/23/R-TCGA%E6%95%B0%E6%8D%AE%E6%8C%96%E6%8E%98%E5%A4%8D%E7%8E%B0%E4%B9%8BBRCA/index.html">
<meta property="og:site_name" content="Dean&#39;s blog">
<meta property="og:description" content="本文写于观看生信技能树公众号（vx: biotrainee）的七步走纯R代码通过数据挖掘复现一篇实验文章（第1到6步）一文后，感觉生信技能树优秀学徒的工作十分吸引人，就自己动手复现了一次。  Step 00 问题概述  本文的任务是全代码复现一篇paper，标题为 ：Co-expression networks revealed potential core lncRNAs in the tri">
<meta property="og:locale" content="en_US">
<meta property="og:image" content="https://thereallda.github.io/2022/01/23/R-TCGA%E6%95%B0%E6%8D%AE%E6%8C%96%E6%8E%98%E5%A4%8D%E7%8E%B0%E4%B9%8BBRCA/image-20220123161900658.png">
<meta property="og:image" content="https://thereallda.github.io/2022/01/23/R-TCGA%E6%95%B0%E6%8D%AE%E6%8C%96%E6%8E%98%E5%A4%8D%E7%8E%B0%E4%B9%8BBRCA/image-20220123161906036.png">
<meta property="og:image" content="https://thereallda.github.io/2022/01/23/R-TCGA%E6%95%B0%E6%8D%AE%E6%8C%96%E6%8E%98%E5%A4%8D%E7%8E%B0%E4%B9%8BBRCA/image-20220123161912689.png">
<meta property="og:image" content="https://thereallda.github.io/2022/01/23/R-TCGA%E6%95%B0%E6%8D%AE%E6%8C%96%E6%8E%98%E5%A4%8D%E7%8E%B0%E4%B9%8BBRCA/image-20220123161921831.png">
<meta property="og:image" content="https://thereallda.github.io/2022/01/23/R-TCGA%E6%95%B0%E6%8D%AE%E6%8C%96%E6%8E%98%E5%A4%8D%E7%8E%B0%E4%B9%8BBRCA/image-20220123161946993.png">
<meta property="og:image" content="https://thereallda.github.io/2022/01/23/R-TCGA%E6%95%B0%E6%8D%AE%E6%8C%96%E6%8E%98%E5%A4%8D%E7%8E%B0%E4%B9%8BBRCA/image-20220123161958678.png">
<meta property="og:image" content="https://thereallda.github.io/2022/01/23/R-TCGA%E6%95%B0%E6%8D%AE%E6%8C%96%E6%8E%98%E5%A4%8D%E7%8E%B0%E4%B9%8BBRCA/image-20220123162007199.png">
<meta property="og:image" content="https://thereallda.github.io/2022/01/23/R-TCGA%E6%95%B0%E6%8D%AE%E6%8C%96%E6%8E%98%E5%A4%8D%E7%8E%B0%E4%B9%8BBRCA/image-20220123162030904.png">
<meta property="og:image" content="https://thereallda.github.io/2022/01/23/R-TCGA%E6%95%B0%E6%8D%AE%E6%8C%96%E6%8E%98%E5%A4%8D%E7%8E%B0%E4%B9%8BBRCA/image-20220123162037193.png">
<meta property="og:image" content="https://thereallda.github.io/2022/01/23/R-TCGA%E6%95%B0%E6%8D%AE%E6%8C%96%E6%8E%98%E5%A4%8D%E7%8E%B0%E4%B9%8BBRCA/image-20220123162126780.png">
<meta property="og:image" content="https://thereallda.github.io/2022/01/23/R-TCGA%E6%95%B0%E6%8D%AE%E6%8C%96%E6%8E%98%E5%A4%8D%E7%8E%B0%E4%B9%8BBRCA/image-20220123162139971.png">
<meta property="og:image" content="https://thereallda.github.io/2022/01/23/R-TCGA%E6%95%B0%E6%8D%AE%E6%8C%96%E6%8E%98%E5%A4%8D%E7%8E%B0%E4%B9%8BBRCA/image-20220123162206227.png">
<meta property="article:published_time" content="2022-01-23T08:17:38.000Z">
<meta property="article:modified_time" content="2022-01-24T01:17:32.112Z">
<meta property="article:author" content="Dean Li">
<meta property="article:tag" content="Data_analysis">
<meta property="article:tag" content="R">
<meta property="article:tag" content="RNA-seq">
<meta name="twitter:card" content="summary">
<meta name="twitter:image" content="https://thereallda.github.io/2022/01/23/R-TCGA%E6%95%B0%E6%8D%AE%E6%8C%96%E6%8E%98%E5%A4%8D%E7%8E%B0%E4%B9%8BBRCA/image-20220123161900658.png">

<link rel="canonical" href="https://thereallda.github.io/2022/01/23/R-TCGA%E6%95%B0%E6%8D%AE%E6%8C%96%E6%8E%98%E5%A4%8D%E7%8E%B0%E4%B9%8BBRCA/">


<script id="page-configurations">
  // https://hexo.io/docs/variables.html
  CONFIG.page = {
    sidebar: "",
    isHome : false,
    isPost : true,
    lang   : 'en'
  };
</script>

  <title>R-TCGA数据挖掘复现之BRCA | Dean's blog</title>
  






  <noscript>
  <style>
  .use-motion .brand,
  .use-motion .menu-item,
  .sidebar-inner,
  .use-motion .post-block,
  .use-motion .pagination,
  .use-motion .comments,
  .use-motion .post-header,
  .use-motion .post-body,
  .use-motion .collection-header { opacity: initial; }

  .use-motion .site-title,
  .use-motion .site-subtitle {
    opacity: initial;
    top: initial;
  }

  .use-motion .logo-line-before i { left: initial; }
  .use-motion .logo-line-after i { right: initial; }
  </style>
</noscript>

<style>.darkmode--activated{--body-bg-color:#282828;--content-bg-color:#333;--card-bg-color:#555;--text-color:#ccc;--blockquote-color:#bbb;--link-color:#ccc;--link-hover-color:#eee;--brand-color:#ddd;--brand-hover-color:#ddd;--table-row-odd-bg-color:#282828;--table-row-hover-bg-color:#363636;--menu-item-bg-color:#555;--btn-default-bg:#222;--btn-default-color:#ccc;--btn-default-border-color:#555;--btn-default-hover-bg:#666;--btn-default-hover-color:#ccc;--btn-default-hover-border-color:#666;--highlight-background:#282b2e;--highlight-foreground:#a9b7c6;--highlight-gutter-background:#34393d;--highlight-gutter-foreground:#9ca9b6}.darkmode--activated img{opacity:.75}.darkmode--activated img:hover{opacity:.9}.darkmode--activated code{color:#69dbdc;background:0 0}button.darkmode-toggle{z-index:9999}.darkmode-ignore,img{display:flex!important}</style></head>

<body itemscope itemtype="http://schema.org/WebPage">
  <div class="container use-motion">
    <div class="headband"></div>

    <header class="header" itemscope itemtype="http://schema.org/WPHeader">
      <div class="header-inner"><div class="site-brand-container">
  <div class="site-nav-toggle">
    <div class="toggle" aria-label="Toggle navigation bar">
      <span class="toggle-line toggle-line-first"></span>
      <span class="toggle-line toggle-line-middle"></span>
      <span class="toggle-line toggle-line-last"></span>
    </div>
  </div>

  <div class="site-meta">

    <a href="/" class="brand" rel="start">
      <span class="logo-line-before"><i></i></span>
      <h1 class="site-title">Dean's blog</h1>
      <span class="logo-line-after"><i></i></span>
    </a>
  </div>

  <div class="site-nav-right">
    <div class="toggle popup-trigger">
    </div>
  </div>
</div>




<nav class="site-nav">
  <ul id="menu" class="main-menu menu">
        <li class="menu-item menu-item-home">

    <a href="/" rel="section"><i class="fa fa-home fa-fw"></i>Home</a>

  </li>
        <li class="menu-item menu-item-about">

    <a href="/about/" rel="section"><i class="fa fa-user fa-fw"></i>About</a>

  </li>
        <li class="menu-item menu-item-tags">

    <a href="/tags/" rel="section"><i class="fa fa-tags fa-fw"></i>Tags</a>

  </li>
        <li class="menu-item menu-item-archives">

    <a href="/archives/" rel="section"><i class="fa fa-archive fa-fw"></i>Archives</a>

  </li>
  </ul>
</nav>




</div>
    </header>

    
  <div class="back-to-top">
    <i class="fa fa-arrow-up"></i>
    <span>0%</span>
  </div>
  <div class="reading-progress-bar"></div>


    <main class="main">
      <div class="main-inner">
        <div class="content-wrap">
          

          <div class="content post posts-expand">
            

    
  
  
  <article itemscope itemtype="http://schema.org/Article" class="post-block" lang="en">
    <link itemprop="mainEntityOfPage" href="https://thereallda.github.io/2022/01/23/R-TCGA%E6%95%B0%E6%8D%AE%E6%8C%96%E6%8E%98%E5%A4%8D%E7%8E%B0%E4%B9%8BBRCA/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="image" content="https://avatars.githubusercontent.com/u/55836943?v=4">
      <meta itemprop="name" content="Dean Li">
      <meta itemprop="description" content="">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="Dean's blog">
    </span>
      <header class="post-header">
        <h1 class="post-title" itemprop="name headline">
          R-TCGA数据挖掘复现之BRCA
        </h1>

        <div class="post-meta">
          
          
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-calendar"></i>
              </span>
              <span class="post-meta-item-text">Posted on</span>

              <time title="Created: 2022-01-23 16:17:38" itemprop="dateCreated datePublished" datetime="2022-01-23T16:17:38+08:00">2022-01-23</time>
            </span>
              <span class="post-meta-item">
                <span class="post-meta-item-icon">
                  <i class="far fa-calendar-check"></i>
                </span>
                <span class="post-meta-item-text">Edited on</span>
                <time title="Modified: 2022-01-24 09:17:32" itemprop="dateModified" datetime="2022-01-24T09:17:32+08:00">2022-01-24</time>
              </span>

          
            <span class="post-meta-item" title="Views" id="busuanzi_container_page_pv" style="display: none;">
              <span class="post-meta-item-icon">
                <i class="fa fa-eye"></i>
              </span>
              <span class="post-meta-item-text">Views: </span>
              <span id="busuanzi_value_page_pv"></span>
            </span><br>
            <span class="post-meta-item" title="Symbols count in article">
              <span class="post-meta-item-icon">
                <i class="far fa-file-word"></i>
              </span>
                <span class="post-meta-item-text">Symbols count in article: </span>
              <span>13k</span>
            </span>
            <span class="post-meta-item" title="Reading time">
              <span class="post-meta-item-icon">
                <i class="far fa-clock"></i>
              </span>
                <span class="post-meta-item-text">Reading time &asymp;</span>
              <span>12 mins.</span>
            </span>

        </div>
      </header>

    
    
    
    <div class="post-body" itemprop="articleBody">

      
        <blockquote>
<p>本文写于观看生信技能树公众号（vx: biotrainee）的<a target="_blank" rel="noopener" href="https://mp.weixin.qq.com/s/2AW5OnCMte9rkp1hCbX8Rg">七步走纯R代码通过数据挖掘复现一篇实验文章（第1到6步）</a>一文后，感觉生信技能树优秀学徒的工作十分吸引人，就自己动手复现了一次。</p>
</blockquote>
<h1>Step 00 问题概述</h1>
<blockquote>
<p>本文的任务是全代码复现一篇paper，标题为 ：Co-expression networks revealed potential core lncRNAs in the triple-negative breast cancer. PMID：<em>27380926</em></p>
<p><em>ref：</em> 生信技能树–<a target="_blank" rel="noopener" href="https://mp.weixin.qq.com/s/2AW5OnCMte9rkp1hCbX8Rg">七步走纯R代码通过数据挖掘复现一篇实验文章（第1到6步）</a></p>
</blockquote>
<p>我们复现的文章是对8名乳腺癌的患者的转录组测序数据的分析。复现测序的流程恐怕不太现实，但是我们可以通过TCGA数据库中的肿瘤数据复现文章的数据分析流程。</p>
<p>本文的<strong>分析流程</strong>包括：</p>
<ul>
<li>
<p>下载数据</p>
</li>
<li>
<p>数据清洗</p>
</li>
<li>
<p>质量控制</p>
</li>
<li>
<p>差异分析</p>
</li>
<li>
<p>注释mRNA,lncRNA</p>
</li>
<li>
<p>富集分析</p>
</li>
</ul>
<p>至于WGCNA分析在本文就不再复现了，有兴趣的同学也可以查阅生信技能树的文章<a target="_blank" rel="noopener" href="https://mp.weixin.qq.com/s/lReIjUeU4B-cC1Ehkod4Ow">七步走纯R代码通过数据挖掘复现一篇实验文章（第七步WGCNA）</a></p>
<span id="more"></span>
<h1>Step 01 数据下载</h1>
<ul>
<li>
<p>TCGA database</p>
<p>TCGA数据库上的数据下载可以参考生信技能树上有关的文章<a target="_blank" rel="noopener" href="https://mp.weixin.qq.com/s/t48BBhj19LZtLNcuHM3QEQ">送你一篇TCGA数据挖掘文章</a>。在本文中也简要地复述下载流程。</p>
<p>首先，登入<a target="_blank" rel="noopener" href="https://xenabrowser.net/datapages/">UCSC Xena</a><br>
<img src="image-20220123161900658.png" alt=""></p>
</li>
</ul>
<p><strong>选择TCGA的breast cancer data</strong></p>
<p><img src="image-20220123161906036.png" alt=""></p>
<p><img src="image-20220123161912689.png" alt=""></p>
<p><strong>下载RNAseq表达矩阵和临床信息</strong></p>
<p><strong>P.S.</strong> 要注意的是在生信技能树中使用的是GDC的breast cancer dataset，而本文使用的是TCGA 的。两个dataset分析出来的数据差异颇大。</p>
<ul>
<li>
<p>Ensembl GTF file – annotation information</p>
<p>在Ensembl的FTP download页(<a target="_blank" rel="noopener" href="http://asia.ensembl.org/info/data/ftp/index.html">http://asia.ensembl.org/info/data/ftp/index.html</a>)中，选择人的GTF文件：</p>
</li>
</ul>
<p><img src="image-20220123161921831.png" alt=""></p>
<p>随后，下载文件**“Homo_sapiens.GRCh38.98.chr.gtf.gz”**即可。</p>
<h1>Step 02 数据清洗</h1>
<p>该步骤需要从临床信息中提取中三阴性乳腺癌样本的临床信息与表达矩阵，并将肿瘤样本与正常样本进行配对。</p>
<blockquote>
<p><strong>三阴性乳腺癌(Triple-negative breast cancer, TNBC) :</strong> 指的是以下三种受体均不表达的乳腺癌类型：</p>
<ul>
<li>雌激素受体：estrogen receptor (ER) ;</li>
<li>孕激素受体：progesterone receptor(PR) ;</li>
<li>人类表皮生长因子受体2： HER2/neu</li>
</ul>
</blockquote>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br></pre></td><td class="code"><pre><span class="line">rm(<span class="built_in">list</span> = ls())</span><br><span class="line"><span class="comment">#selecting triple-negative breast cancer samples from phenotype data</span></span><br><span class="line"><span class="comment">#extracting clinical information</span></span><br><span class="line"></span><br><span class="line">p &lt;- read.table(<span class="string">&#x27;.../data/BRCA_clinicalMatrix&#x27;</span>,header = <span class="built_in">T</span>,</span><br><span class="line">                sep = <span class="string">&#x27;\t&#x27;</span>,<span class="built_in">quote</span> = <span class="string">&#x27;&#x27;</span>)</span><br><span class="line"></span><br><span class="line">colnames(p)[grep(<span class="string">&quot;receptor_status&quot;</span>, colnames(p))]</span><br><span class="line"><span class="comment">## [1] &quot;breast_carcinoma_estrogen_receptor_status&quot;               </span></span><br><span class="line"><span class="comment">## [2] &quot;breast_carcinoma_progesterone_receptor_status&quot;           </span></span><br><span class="line"><span class="comment">## [3] &quot;lab_proc_her2_neu_immunohistochemistry_receptor_status&quot;  </span></span><br><span class="line"><span class="comment">## [4] &quot;metastatic_breast_carcinoma_estrogen_receptor_status&quot;    </span></span><br><span class="line"><span class="comment">## [5] &quot;metastatic_breast_carcinoma_progesterone_receptor_status&quot;</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># examining how many triple-negative receptors samples </span></span><br><span class="line">table(p$breast_carcinoma_estrogen_receptor_status == <span class="string">&#x27;Negative&#x27;</span> &amp;</span><br><span class="line">        p$breast_carcinoma_progesterone_receptor_status == <span class="string">&#x27;Negative&#x27;</span> &amp;</span><br><span class="line">        p$lab_proc_her2_neu_immunohistochemistry_receptor_status == <span class="string">&#x27;Negative&#x27;</span>)</span><br><span class="line"><span class="comment">## FALSE  TRUE </span></span><br><span class="line"><span class="comment">## 1117   130 </span></span><br><span class="line"></span><br><span class="line"><span class="comment"># extracting tnbc samples </span></span><br><span class="line">tnbc_samples &lt;- p[p$breast_carcinoma_estrogen_receptor_status == <span class="string">&#x27;Negative&#x27;</span> &amp;</span><br><span class="line">                    p$breast_carcinoma_progesterone_receptor_status == <span class="string">&#x27;Negative&#x27;</span> &amp;</span><br><span class="line">                    p$lab_proc_her2_neu_immunohistochemistry_receptor_status == <span class="string">&#x27;Negative&#x27;</span>, ]</span><br></pre></td></tr></table></figure>
<p>在TCGA的命名规则中样本名字的第14,15个字符是以两位数字表示的，其中01-09表示肿瘤样本，10-16表示正常对照样本，具体对应关系可查看其帮助网页：<a target="_blank" rel="noopener" href="https://gdc.cancer.gov/resources-tcga-users/tcga-code-tables/sample-type-codes">https://gdc.cancer.gov/resources-tcga-users/tcga-code-tables/sample-type-codes</a></p>
<p>因此，在后续分析中我们分别将对应位置为01的分到tumor group，11的分为normal group</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">#pairing tumor samples with normal samples</span></span><br><span class="line">library(stringr)</span><br><span class="line"></span><br><span class="line">tab1 &lt;- tnbc_samples[<span class="number">1</span>:<span class="number">2</span>] <span class="comment"># includes &#x27;sampleID&#x27; &amp; &quot;AJCC_Stage_nature2012&quot;  </span></span><br><span class="line">tumor &lt;- tab1[substr(tab1$sampleID,<span class="number">14</span>,<span class="number">15</span>) &lt; <span class="number">10</span>,]         </span><br><span class="line">tumor$TCGAID &lt;- str_sub(tumor$sampleID,<span class="number">1</span>,<span class="number">12</span>)</span><br><span class="line">normal &lt;- tab1[!substr(tab1$sampleID,<span class="number">14</span>,<span class="number">15</span>) &lt;<span class="number">10</span>,]</span><br><span class="line">normal$TCGAID &lt;- str_sub(normal$sampleID,<span class="number">1</span>,<span class="number">12</span>)</span><br><span class="line"><span class="built_in">dim</span>(tumor)</span><br><span class="line"><span class="comment">## [1] 117   3</span></span><br><span class="line"><span class="built_in">dim</span>(normal)</span><br><span class="line"><span class="comment">## [1] 13  3</span></span><br><span class="line">tnbc_samples_paired &lt;- merge(tumor,normal,by = <span class="string">&#x27;TCGAID&#x27;</span>) <span class="comment"># return samples only have N-T pairing</span></span><br><span class="line"></span><br><span class="line"><span class="comment">#because samples &#x27;TCGA-BH-A18V&#x27; have been detected twice, </span></span><br><span class="line"><span class="comment">#so we remove the duplicated one &#x27;TCGA-BH-A18V-06&#x27;</span></span><br><span class="line">tnbc_samples_paired &lt;- tnbc_samples_paired[-<span class="number">6</span>, ]</span><br><span class="line">save(tnbc_samples_paired, file = <span class="string">&quot;data/tnbc_samples_paired.Rdata&quot;</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment">#gene expression matrix</span></span><br><span class="line">rawdata &lt;- read.csv(<span class="string">&quot;data/HiSeqV2&quot;</span>, sep = <span class="string">&#x27;\t&#x27;</span>, header = <span class="built_in">T</span>)</span><br><span class="line">rawdata &lt;- as.data.frame(rawdata)</span><br><span class="line">rawdata[<span class="number">1</span>:<span class="number">3</span>,<span class="number">1</span>:<span class="number">3</span>]</span><br><span class="line"><span class="comment">##     sample TCGA.AR.A5QQ.01 TCGA.D8.A1JA.01</span></span><br><span class="line"><span class="comment">##  1 ARHGEF10L          9.5074          7.4346</span></span><br><span class="line"><span class="comment">##  2     HIF3A          1.5787          3.6607</span></span><br><span class="line"><span class="comment">##  3     RNF17          0.0000          0.6245</span></span><br><span class="line">tnbc_samples_paired[<span class="number">1</span>,<span class="string">&quot;sampleID.x&quot;</span>]</span><br><span class="line"><span class="comment">## [1] TCGA-A7-A4SE-01</span></span><br><span class="line"></span><br><span class="line"><span class="comment">#make sampleid suitable for comparing with the id in rawdata</span></span><br><span class="line">t_idfordata &lt;- tnbc_samples_paired$sampleID.x</span><br><span class="line">t_idfordata &lt;- gsub(<span class="string">&#x27;-&#x27;</span>,<span class="string">&#x27;.&#x27;</span>,t_idfordata)</span><br><span class="line">tnbc_samples_paired$t_dataid &lt;- t_idfordata</span><br><span class="line">n_idfordata &lt;- tnbc_samples_paired$sampleID.y</span><br><span class="line">n_idfordata &lt;- gsub(<span class="string">&#x27;-&#x27;</span>,<span class="string">&#x27;.&#x27;</span>,n_idfordata)</span><br><span class="line">tnbc_samples_paired$n_dataid &lt;- n_idfordata</span><br><span class="line"></span><br><span class="line">table(colnames(rawdata) %in% tnbc_samples_paired$t_dataid)</span><br><span class="line"><span class="comment">##  FALSE  TRUE </span></span><br><span class="line"><span class="comment">##   1205    14 </span></span><br><span class="line">tab2 &lt;- rawdata[ ,colnames(rawdata) %in% tnbc_samples_paired$t_dataid]</span><br><span class="line">tab3 &lt;- rawdata[ ,colnames(rawdata) %in% tnbc_samples_paired$n_dataid]</span><br><span class="line">tab2 &lt;- tab2[, str_sub(colnames(tab2),<span class="number">1</span>,<span class="number">12</span>) %in% str_sub(colnames(tab3),<span class="number">1</span>,<span class="number">12</span>)]</span><br><span class="line">expr &lt;- cbind(tab2, tab3)</span><br><span class="line">rownames(expr) &lt;- rawdata[ ,<span class="number">1</span>]</span><br><span class="line">expr &lt;- t(expr)</span><br><span class="line">expr[<span class="number">1</span>:<span class="number">3</span>,<span class="number">1</span>:<span class="number">3</span>]</span><br><span class="line"><span class="comment">##                  ARHGEF10L  HIF3A  RNF17</span></span><br><span class="line"><span class="comment">##  TCGA.E2.A1L7.01    9.8265 1.7767 0.0000</span></span><br><span class="line"><span class="comment">##  TCGA.BH.A1FC.01    9.6724 2.2705 0.6677</span></span><br><span class="line"><span class="comment">##  TCGA.E2.A1LS.01    9.3743 7.8902 0.0000</span></span><br><span class="line"></span><br><span class="line">save(expr,file = <span class="string">&quot;data/TNBC_pair_expr.Rdata&quot;</span>)</span><br></pre></td></tr></table></figure>
<h1>Step 03 质量控制</h1>
<p>提取表达矩阵后，我们需要对提取到的数据进行质量检测，看看分组是否正确等等。在这里分别使用<strong>PCA</strong>和<strong>聚类</strong>的方法对表达矩阵进行分析。一般而言，两者之一都可以作为表达矩阵质量分析的可视化结果，在此处为了展示方式方法的多样性，我们都将其进行展示。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># using pca to exmain the data quality</span></span><br><span class="line">library(factoextra)</span><br><span class="line">library(FactoMineR)</span><br><span class="line"></span><br><span class="line">group &lt;- <span class="built_in">c</span>(<span class="built_in">rep</span>(<span class="string">&#x27;tumor&#x27;</span>,<span class="number">11</span>), <span class="built_in">rep</span>(<span class="string">&#x27;normal&#x27;</span>,<span class="number">11</span>))</span><br><span class="line">expr.pca &lt;- PCA(expr,graph = <span class="built_in">F</span>)</span><br><span class="line"></span><br><span class="line">fviz_pca_ind(expr.pca,</span><br><span class="line">             geom.ind = <span class="string">&quot;point&quot;</span>,</span><br><span class="line">             col.ind = group, </span><br><span class="line">             addEllipses = <span class="literal">TRUE</span>, </span><br><span class="line">             legend.title = <span class="string">&quot;Groups&quot;</span>)</span><br></pre></td></tr></table></figure>
<p><img src="image-20220123161946993.png" alt=""></p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">#cluster for exmaining the data quality</span></span><br><span class="line">plot(hclust(dist(expr)))</span><br></pre></td></tr></table></figure>
<p><img src="image-20220123161958678.png" alt=""></p>
<p>两种分析都将tumor和normal group清晰地分开，说明表达矩阵质量良好。</p>
<h1>Step 04 差异表达分析</h1>
<p>本次差异分析使用<code>DESeq2</code>进行，由于<code>DESeq2</code><a target="_blank" rel="noopener" href="https://bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html">要求</a>输入的表达矩阵数据是原始的counts数据，而TCGA上表达矩阵的值是进行过<code>log2(count + 1)</code>校正的，因此在差异分析之前，需要进行un-normalized</p>
<p><img src="image-20220123162007199.png" alt=""></p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br></pre></td><td class="code"><pre><span class="line">library(DESeq2)</span><br><span class="line"></span><br><span class="line"><span class="comment"># un-normalization</span></span><br><span class="line">dat &lt;- as.data.frame(t(expr)) </span><br><span class="line"><span class="comment"># un-normalization</span></span><br><span class="line">dat &lt;- 2^dat - <span class="number">1</span></span><br><span class="line">dat &lt;- <span class="built_in">ceiling</span>(dat)</span><br><span class="line">dat[<span class="number">1</span>:<span class="number">3</span>,<span class="number">1</span>:<span class="number">3</span>]</span><br><span class="line"><span class="comment">##            TCGA.E2.A1L7.01 TCGA.BH.A1FC.01 TCGA.E2.A1LS.01</span></span><br><span class="line"><span class="comment">##  ARHGEF10L             907             815             663</span></span><br><span class="line"><span class="comment">##  HIF3A                   3               4             237</span></span><br><span class="line"><span class="comment">##  RNF17                   0               1               0</span></span><br></pre></td></tr></table></figure>
<p>在<code>DESeq2</code>分析过程中，会将表达矩阵存储在<code>dds</code>对象中，以存储中间变量和进行一部分计算。<code>dds</code>对象的构建需要包括以下几方面数据：</p>
<ul>
<li><strong>Raw counts matrix</strong>：行为基因，列为样本的基因表达数据</li>
<li><strong>colData</strong> ：存储样本信息</li>
<li><strong>design formula</strong>：指明在模型中的变量，并用于估计模型的<strong>离散值</strong>和<strong>log2 fold changes</strong></li>
</ul>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># Transforming to dds object</span></span><br><span class="line">group_list &lt;- factor(<span class="built_in">rep</span>(<span class="built_in">c</span>(<span class="string">&#x27;tumor&#x27;</span>,<span class="string">&#x27;normal&#x27;</span>), each = <span class="number">11</span>))</span><br><span class="line">colData &lt;- data.frame(row.names=colnames(dat),</span><br><span class="line">                      group_list=group_list)</span><br><span class="line"></span><br><span class="line">dds &lt;- DESeqDataSetFromMatrix(countData = dat,</span><br><span class="line">                              colData = colData,</span><br><span class="line">                              design = ~group_list,</span><br><span class="line">                              tidy = <span class="built_in">F</span>)</span><br><span class="line"><span class="built_in">dim</span>(dds)</span><br><span class="line"><span class="comment">## [1] 20530    22</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># filtering very low-expression data </span></span><br><span class="line">table(rowSums(counts(dds)==<span class="number">0</span>))</span><br><span class="line"><span class="comment"># keep rows at least have almost 70% samples being detected </span></span><br><span class="line">keep &lt;- rowSums(counts(dds)==<span class="number">0</span>)&lt; <span class="number">16</span></span><br><span class="line">dds &lt;- dds[keep, ]</span><br><span class="line">counts(dds)[<span class="number">1</span>:<span class="number">10</span>,<span class="number">1</span>:<span class="number">3</span>]</span><br><span class="line"><span class="built_in">dim</span>(dds) <span class="comment"># more than 2,000 genes being romoved</span></span><br><span class="line"><span class="comment">## [1] 18423    22</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># Performing differential expression analysis</span></span><br><span class="line">dds &lt;- DESeq(dds)</span><br><span class="line"><span class="comment"># Extracting transformed values</span></span><br><span class="line">vsd &lt;- vst(dds, blind = <span class="built_in">F</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment"># specifying the contrast in model using to estimate the fold change and p-value</span></span><br><span class="line">contrast &lt;- <span class="built_in">c</span>(<span class="string">&quot;group_list&quot;</span>,<span class="string">&quot;tumor&quot;</span>,<span class="string">&quot;normal&quot;</span>)</span><br><span class="line">dd1 &lt;- results(dds, contrast=contrast, alpha = <span class="number">0.05</span>)</span><br><span class="line">plotMA(dd1, ylim=<span class="built_in">c</span>(-<span class="number">2</span>,<span class="number">2</span>))</span><br></pre></td></tr></table></figure>
<p><img src="image-20220123162030904.png" alt="Before lfcShrink"></p>
<p>MA-plot用于可视化fold change与gene counts之间的关系，默认情况下<em>P</em> &lt; 0.1的值会被标红，而超过y轴范围的值则以三角形表示<br>
<code>lfcShrink</code>可对log fold change进行矫正以消除低表达基因带来的误差</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># lfcShrink</span></span><br><span class="line">dd3 &lt;- lfcShrink(dds, coef = <span class="string">&quot;group_list_tumor_vs_normal&quot;</span>, res=dd1, type=<span class="string">&#x27;apeglm&#x27;</span>)</span><br><span class="line">dd3</span><br><span class="line">plotMA(dd3, ylim=<span class="built_in">c</span>(-<span class="number">2</span>,<span class="number">2</span>))</span><br><span class="line">summary(dd3, alpha = <span class="number">0.05</span>)</span><br><span class="line"><span class="comment">##  out of 18423 with nonzero total read count</span></span><br><span class="line"><span class="comment">##  adjusted p-value &lt; 0.05</span></span><br><span class="line"><span class="comment">##  LFC &gt; 0 (up)       : 4054, 22%</span></span><br><span class="line"><span class="comment">##  LFC &lt; 0 (down)     : 2837, 15%</span></span><br><span class="line"><span class="comment">##  outliers [1]       : 0, 0%</span></span><br><span class="line"><span class="comment">##  low counts [2]     : 0, 0%</span></span><br><span class="line"><span class="comment">##  (mean count &lt; 0)</span></span><br><span class="line"><span class="comment">##  [1] see &#x27;cooksCutoff&#x27; argument of ?results</span></span><br><span class="line"><span class="comment">##  [2] see &#x27;independentFiltering&#x27; argument of ?results</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># considering genes which fold change &gt; 2 or &lt;0.5 and adjusted-p &lt;0.05 as significantly differential expressed</span></span><br><span class="line">sig &lt;- <span class="built_in">abs</span>(dd3$log2FoldChange)&gt;<span class="number">1</span> &amp; dd3$padj&lt;<span class="number">0.05</span></span><br><span class="line">res_sig &lt;- dd3[sig,]</span><br><span class="line">summary(res_sig)</span><br><span class="line"><span class="comment">##  out of 4215 with nonzero total read count</span></span><br><span class="line"><span class="comment">##  adjusted p-value &lt; 0.05</span></span><br><span class="line"><span class="comment">##  LFC &gt; 0 (up)       : 2255, 53%</span></span><br><span class="line"><span class="comment">##  LFC &lt; 0 (down)     : 1960, 47%</span></span><br><span class="line"><span class="comment">##  outliers [1]       : 0, 0%</span></span><br><span class="line"><span class="comment">##  low counts [2]     : 0, 0%</span></span><br><span class="line"></span><br><span class="line">save(dd3,res_sig,vsd, file = <span class="string">&#x27;.../data/TCGA_TNBC_DE.Rdata&#x27;</span>)</span><br></pre></td></tr></table></figure>
<p><img src="image-20220123162037193.png" alt="After lfcShrink"></p>
<p>火山图展示fold-change与p-value的关系</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># visualization</span></span><br><span class="line">library(ggplot2)</span><br><span class="line">library(ggthemes)</span><br><span class="line"></span><br><span class="line">res &lt;- as.data.frame(dd3)</span><br><span class="line">res$threshold &lt;- as.factor(ifelse(res$padj &lt; <span class="number">0.05</span> &amp; <span class="built_in">abs</span>(res$log2FoldChange) &gt;=log2(<span class="number">2</span>),ifelse(res$log2FoldChange &gt; log2(<span class="number">2</span>) ,<span class="string">&#x27;Up&#x27;</span>,<span class="string">&#x27;Down&#x27;</span>),<span class="string">&#x27;Not&#x27;</span>))</span><br><span class="line"></span><br><span class="line">plot2 &lt;- ggplot(data=res, aes(x=log2FoldChange, y =-log10(padj), colour=threshold,fill=threshold)) +</span><br><span class="line">  scale_color_manual(values=<span class="built_in">c</span>(<span class="string">&quot;blue&quot;</span>, <span class="string">&quot;grey&quot;</span>,<span class="string">&quot;red&quot;</span>))+</span><br><span class="line">  geom_point(alpha=<span class="number">0.4</span>, size=<span class="number">1.2</span>) +</span><br><span class="line">  theme_bw(base_size = <span class="number">12</span>, base_family = <span class="string">&quot;Times&quot;</span>) +</span><br><span class="line">  geom_vline(xintercept=<span class="built_in">c</span>(-<span class="number">0.5</span>,<span class="number">0.5</span>),lty=<span class="number">4</span>,col=<span class="string">&quot;grey&quot;</span>,lwd=<span class="number">0.6</span>)+</span><br><span class="line">  geom_hline(yintercept = -log10(<span class="number">0.05</span>),lty=<span class="number">4</span>,col=<span class="string">&quot;grey&quot;</span>,lwd=<span class="number">0.6</span>)+</span><br><span class="line">  theme(legend.position=<span class="string">&quot;right&quot;</span>,</span><br><span class="line">        panel.grid=element_blank(),</span><br><span class="line">        legend.title = element_blank(),</span><br><span class="line">        legend.text= element_text(face=<span class="string">&quot;bold&quot;</span>, color=<span class="string">&quot;black&quot;</span>,family = <span class="string">&quot;Times&quot;</span>, size=<span class="number">8</span>),</span><br><span class="line">        plot.title = element_text(hjust = <span class="number">0.5</span>),</span><br><span class="line">        axis.text.x = element_text(face=<span class="string">&quot;bold&quot;</span>, color=<span class="string">&quot;black&quot;</span>, size=<span class="number">12</span>),</span><br><span class="line">        axis.text.y = element_text(face=<span class="string">&quot;bold&quot;</span>,  color=<span class="string">&quot;black&quot;</span>, size=<span class="number">12</span>),</span><br><span class="line">        axis.title.x = element_text(face=<span class="string">&quot;bold&quot;</span>, color=<span class="string">&quot;black&quot;</span>, size=<span class="number">12</span>),</span><br><span class="line">        axis.title.y = element_text(face=<span class="string">&quot;bold&quot;</span>,color=<span class="string">&quot;black&quot;</span>, size=<span class="number">12</span>)) +</span><br><span class="line">  labs( x=<span class="string">&quot;log2 (Fold Change)&quot;</span>,y=<span class="string">&quot;-log10 (p-value)&quot;</span>)</span><br><span class="line">plot2</span><br></pre></td></tr></table></figure>
<p><img src="image-20220123162126780.png" alt=""></p>
<h1>Step 05 注释</h1>
<p>文章中对数据注释后分为了mRNA和lncRNA，并对两者分别进行了分析。接下来我们也将利用<strong>Ensembl</strong>的GTF进行注释。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br></pre></td><td class="code"><pre><span class="line">library(rtracklayer)</span><br><span class="line">library(tidyr)</span><br><span class="line">library(dplyr)</span><br><span class="line">library(pheatmap)</span><br><span class="line">require(org.Hs.eg.db)</span><br><span class="line"></span><br><span class="line">gtf1 &lt;- import(<span class="string">&#x27;data/Homo_sapiens.GRCh38.98.chr.gtf&#x27;</span>)</span><br><span class="line">gtf_df &lt;- as.data.frame(gtf1)</span><br><span class="line">colnames(gtf_df)</span><br><span class="line"><span class="comment"># extracting &quot;gene_id&quot; ,&quot;gene_biotype&quot;</span></span><br><span class="line">gtf &lt;- gtf_df[,<span class="built_in">c</span>(<span class="number">10</span>,<span class="number">14</span>)]</span><br><span class="line">head(gtf)</span><br><span class="line">save(gtf,file = <span class="string">&quot;data/Homo_sapiens.GRCh38.98.chr.Rdata&quot;</span>)</span><br><span class="line"></span><br><span class="line">keytypes(org.Hs.eg.db)</span><br><span class="line">res_sig$gene_names &lt;- rownames(res_sig)</span><br><span class="line"></span><br><span class="line"><span class="comment"># ID transformation</span></span><br><span class="line">res_id &lt;- clusterProfiler::bitr(res_sig$gene_names, </span><br><span class="line">                                fromType = <span class="string">&#x27;SYMBOL&#x27;</span>, </span><br><span class="line">                                toType = <span class="string">&quot;ENSEMBL&quot;</span>, </span><br><span class="line">                                OrgDb = <span class="string">&#x27;org.Hs.eg.db&#x27;</span>)</span><br><span class="line"></span><br><span class="line">k &lt;- res_id[res_id$ENSEMBL %in% gtf$gene_id, <span class="number">2</span>] %&gt;%</span><br><span class="line">  match(gtf$gene_id)</span><br><span class="line">id_keep &lt;- gtf[k,]</span><br><span class="line">colnames(res_id) &lt;- <span class="built_in">c</span>(<span class="string">&quot;gene_names&quot;</span>,<span class="string">&#x27;gene_id&#x27;</span>)</span><br><span class="line">id_keep &lt;- merge(id_keep, res_id, by=<span class="string">&#x27;gene_id&#x27;</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment">## lncRNA                           polymorphic_pseudogene               </span></span><br><span class="line"><span class="comment">## 30                                2                                 </span></span><br><span class="line"><span class="comment">## processed_pseudogene             transcribed_processed_pseudogene </span></span><br><span class="line"><span class="comment">## 1                                 6 </span></span><br><span class="line"><span class="comment">## protein_coding                   snoRNA   </span></span><br><span class="line"><span class="comment">## 3650                              3                               </span></span><br><span class="line"><span class="comment">## transcribed_unitary_pseudogene   transcribed_unprocessed_pseudogene </span></span><br><span class="line"><span class="comment">## 6                                 15 </span></span><br><span class="line"></span><br><span class="line">res_ord &lt;- as.data.frame(res_sig[order(res_sig$padj),])</span><br><span class="line"></span><br><span class="line"><span class="comment"># extracting mRNA and lncRNA results respectively</span></span><br><span class="line">res_mrna &lt;- id_keep[id_keep$gene_biotype==<span class="string">&#x27;protein_coding&#x27;</span>,] %&gt;%</span><br><span class="line">  merge(as.data.frame(res_ord), by = <span class="string">&quot;gene_names&quot;</span>)</span><br><span class="line"></span><br><span class="line">res_lncrna &lt;- id_keep[id_keep$gene_biotype==<span class="string">&#x27;lncRNA&#x27;</span>,] %&gt;%</span><br><span class="line">  merge(as.data.frame(res_ord), by = <span class="string">&quot;gene_names&quot;</span>)</span><br><span class="line">save(res_ord,res_mrna,res_lncrna, file = <span class="string">&#x27;.../data/TCGA_annotation_results.Rdata&#x27;</span>)</span><br></pre></td></tr></table></figure>
<h1>Step 06 富集分析</h1>
<p>富集分析及其可视化采用<code>clusterProfiler</code>进行，由于kegg识别的ID为&quot;ENTREZID&quot;，因此在分析之前也进行了一次转换。同时，在转换的过程中出现了&quot;ENSEMBL&quot;–“ENTREZID” multi-mapping的情况，因此我们移除了冗余的id。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br></pre></td><td class="code"><pre><span class="line">library(clusterProfiler)</span><br><span class="line">library(org.Hs.eg.db)</span><br><span class="line">library(ggplot2)</span><br><span class="line">library(RColorBrewer)</span><br><span class="line">library(gridExtra)</span><br><span class="line">library(enrichplot)</span><br><span class="line"></span><br><span class="line">deid &lt;- bitr(res_mrna$gene_id, </span><br><span class="line">             fromType = <span class="string">&quot;ENSEMBL&quot;</span>, </span><br><span class="line">             toType = <span class="string">&quot;ENTREZID&quot;</span>, </span><br><span class="line">             OrgDb = <span class="string">&#x27;org.Hs.eg.db&#x27;</span>)</span><br><span class="line"><span class="comment">## &#x27;select()&#x27; returned 1:many mapping between keys and columns</span></span><br><span class="line">deid &lt;- a[!duplicated(a$ENSEMBL),]</span><br><span class="line"></span><br><span class="line"><span class="comment"># cc,MF not showed</span></span><br><span class="line">ego_BP &lt;- enrichGO(gene          = deid$ENTREZID,</span><br><span class="line">                   OrgDb         = org.Hs.eg.db,</span><br><span class="line">                   keyType       = <span class="string">&quot;ENTREZID&quot;</span>,</span><br><span class="line">                   ont           = <span class="string">&quot;BP&quot;</span>,</span><br><span class="line">                   pvalueCutoff  = <span class="number">0.05</span>,</span><br><span class="line">                   qvalueCutoff  = <span class="number">0.05</span>,</span><br><span class="line">                   readable      = <span class="literal">TRUE</span>)</span><br><span class="line"></span><br><span class="line">dotplot(ego_BP, showCategory = <span class="number">20</span>,font.size = <span class="number">8</span>)</span><br><span class="line"></span><br><span class="line">ego_KEGG &lt;- enrichKEGG(gene = deid$ENTREZID, organism = <span class="string">&quot;hsa&quot;</span>, </span><br><span class="line">                       keyType = <span class="string">&#x27;kegg&#x27;</span>,</span><br><span class="line">                       pvalueCutoff = <span class="number">0.05</span>,</span><br><span class="line">                       pAdjustMethod = <span class="string">&quot;BH&quot;</span>,</span><br><span class="line">                       minGSSize = <span class="number">10</span>, maxGSSize = <span class="number">500</span>, </span><br><span class="line">                       qvalueCutoff = <span class="number">0.05</span>,</span><br><span class="line">                       use_internal_data = <span class="literal">FALSE</span>)</span><br><span class="line"></span><br><span class="line">dotplot(ego_KEGG, showCategory = <span class="number">20</span>,font.size = <span class="number">8</span>)</span><br></pre></td></tr></table></figure>
<p><img src="image-20220123162139971.png" alt="GO:BP dotplot"></p>
<p><img src="image-20220123162206227.png" alt="KEGG dotplot"></p>
<p>本次代码复现也到此暂告一段落，可能是由于数据集或是分析代码的改动，我们注释到的显著差异表达的lncRNA只有30个，远远小于文章报道的1,211个。原则上，我们应先对差异分析结果注释，再行设定cutoff以找出显著的差异表达基因。但是在先行注释的情况下，仍然只能在本数据集中找到143个lncRNA，让我不禁怀疑不同数据集的差异性真的有这么大？亦或是由样本的差异性所导致的？</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">#dd3是lfcshrink的差异分析结果</span></span><br><span class="line">dd3$gene_names &lt;- rownames(dd3)</span><br><span class="line">res_id2 &lt;- clusterProfiler::bitr(dd3$gene_names, </span><br><span class="line">                                 fromType = <span class="string">&#x27;SYMBOL&#x27;</span>, </span><br><span class="line">                                 toType = <span class="string">&quot;ENSEMBL&quot;</span>, </span><br><span class="line">                                 OrgDb = <span class="string">&#x27;org.Hs.eg.db&#x27;</span>)</span><br><span class="line"><span class="built_in">dim</span>(res_id2)</span><br><span class="line"><span class="comment">## [1] 17830     2</span></span><br><span class="line">k2 &lt;- res_id2[res_id2$ENSEMBL %in% gtf$gene_id, <span class="number">2</span>] %&gt;%</span><br><span class="line">  match(gtf$gene_id)</span><br><span class="line">id_keep2 &lt;- gtf[k2,]</span><br><span class="line">colnames(res_id2) &lt;- <span class="built_in">c</span>(<span class="string">&quot;gene_names&quot;</span>,<span class="string">&#x27;gene_id&#x27;</span>)</span><br><span class="line">id_keep2 &lt;- merge(id_keep2, res_id2, by=<span class="string">&#x27;gene_id&#x27;</span>)</span><br><span class="line">table(id_keep2$gene_biotype)</span><br><span class="line"></span><br><span class="line">                            lncRNA                           misc_RNA </span><br><span class="line">                               <span class="number">143</span>                                  <span class="number">1</span> </span><br><span class="line">            polymorphic_pseudogene               processed_pseudogene </span><br><span class="line">                                <span class="number">12</span>                                 <span class="number">12</span> </span><br><span class="line">                    protein_coding                           ribozyme </span><br><span class="line">                             <span class="number">15534</span>                                  <span class="number">1</span> </span><br><span class="line">                            scaRNA                             snoRNA </span><br><span class="line">                                 <span class="number">5</span>                                 <span class="number">26</span> </span><br><span class="line">                               TEC                          TR_C_gene </span><br><span class="line">                                 <span class="number">3</span>                                  <span class="number">1</span> </span><br><span class="line">  transcribed_processed_pseudogene     transcribed_unitary_pseudogene </span><br><span class="line">                                <span class="number">33</span>                                 <span class="number">14</span> </span><br><span class="line">transcribed_unprocessed_pseudogene                 unitary_pseudogene </span><br><span class="line">                                <span class="number">79</span>                                  <span class="number">1</span> </span><br><span class="line">            unprocessed_pseudogene </span><br><span class="line">                                 <span class="number">3</span> </span><br></pre></td></tr></table></figure>
<p>不论分析结果如何，本次分析流程也是十分值得学习的。至于文中的疑问在解决后会回来填坑的！最后，再次感谢生信技能树（vx: biotrainee）的分享，大家快去关注吧！</p>
<blockquote>
<p><strong>补坑</strong><br>
在咨询生信技能树的jimmy老师后，对本文中的一些问题也得到了解答。关于为何最终注释的lncRNA较少主要是因为选取的表达矩阵是RSEM normalized count matrix，该数据集含有的non-coding genes 的数量本来就较少，故能够注释到的lncRNA也会较少。但在<a target="_blank" rel="noopener" href="https://xenabrowser.net/datapages/?cohort=TCGA%20Breast%20Cancer%20(BRCA)&amp;removeHub=https%3A%2F%2Fxena.treehouse.gi.ucsc.edu%3A443">TCGA Breast Cancer (BRCA)</a>的数据集中我暂时还没发现到转录组的表达矩阵，该数据集的RNA-seq数据基本上是使用polyA+ IlluminaHiSeq，意味着测序的基本上都是mRNA。miRNA的data倒是有，但整个转录组的data还没找到，如果有找到的朋友也可以告知我。</p>
<p><em>补充于2019/10/13</em></p>
</blockquote>
<p>完。</p>

    </div>

    
    
    
        

<div>
<ul class="post-copyright">
  <li class="post-copyright-author">
    <strong>Post author:  </strong>Dean Li
  </li>
  <li class="post-copyright-link">
    <strong>Post link: </strong>
    <a href="https://thereallda.github.io/2022/01/23/R-TCGA%E6%95%B0%E6%8D%AE%E6%8C%96%E6%8E%98%E5%A4%8D%E7%8E%B0%E4%B9%8BBRCA/" title="R-TCGA数据挖掘复现之BRCA">https://thereallda.github.io/2022/01/23/R-TCGA数据挖掘复现之BRCA/</a>
  </li>
  <li class="post-copyright-license">
    <strong>Copyright Notice:  </strong>All articles in this blog are licensed under <a href="https://creativecommons.org/licenses/by-nc-sa/4.0/" rel="noopener" target="_blank"><i class="fab fa-fw fa-creative-commons"></i>BY-NC-SA</a> unless stating additionally.
  </li>
</ul>
</div>


      <footer class="post-footer">
          <div class="post-tags">
              <a href="/tags/Data-analysis/" rel="tag"># Data_analysis</a>
              <a href="/tags/R/" rel="tag"># R</a>
              <a href="/tags/RNA-seq/" rel="tag"># RNA-seq</a>
          </div>

        


        
    <div class="post-nav">
      <div class="post-nav-item">
    <a href="/2022/01/22/R-%E7%94%A8R%E8%BF%9B%E8%A1%8CRNA-seq%E6%95%B0%E6%8D%AE%E5%88%86%E6%9E%90/" rel="prev" title="R-用R进行RNA-seq数据分析">
      <i class="fa fa-chevron-left"></i> R-用R进行RNA-seq数据分析
    </a></div>
      <div class="post-nav-item">
    <a href="/2022/01/24/DESeq2-%E7%AE%80%E5%8D%95%E7%9A%84%E8%BD%AC%E5%BD%95%E7%BB%84%E5%B7%AE%E5%BC%82%E8%A1%A8%E8%BE%BE%E5%88%86%E6%9E%90/" rel="next" title="DESeq2-简单的转录组差异表达分析">
      DESeq2-简单的转录组差异表达分析 <i class="fa fa-chevron-right"></i>
    </a></div>
    </div>
      </footer>
    
  </article>
  
  
  



          </div>
          
    <div class="comments" id="gitalk-container"></div>

<script>
  window.addEventListener('tabs:register', () => {
    let { activeClass } = CONFIG.comments;
    if (CONFIG.comments.storage) {
      activeClass = localStorage.getItem('comments_active') || activeClass;
    }
    if (activeClass) {
      let activeTab = document.querySelector(`a[href="#comment-${activeClass}"]`);
      if (activeTab) {
        activeTab.click();
      }
    }
  });
  if (CONFIG.comments.storage) {
    window.addEventListener('tabs:click', event => {
      if (!event.target.matches('.tabs-comment .tab-content .tab-pane')) return;
      let commentClass = event.target.classList[1];
      localStorage.setItem('comments_active', commentClass);
    });
  }
</script>

        </div>
          
  
  <div class="toggle sidebar-toggle">
    <span class="toggle-line toggle-line-first"></span>
    <span class="toggle-line toggle-line-middle"></span>
    <span class="toggle-line toggle-line-last"></span>
  </div>

  <aside class="sidebar">
    <div class="sidebar-inner">

      <ul class="sidebar-nav motion-element">
        <li class="sidebar-nav-toc">
          Table of Contents
        </li>
        <li class="sidebar-nav-overview">
          Overview
        </li>
      </ul>

      <!--noindex-->
      <div class="post-toc-wrap sidebar-panel">
          <div class="post-toc motion-element"><ol class="nav"><li class="nav-item nav-level-1"><a class="nav-link"><span class="nav-number">1.</span> <span class="nav-text">Step 00 问题概述</span></a></li><li class="nav-item nav-level-1"><a class="nav-link"><span class="nav-number">2.</span> <span class="nav-text">Step 01 数据下载</span></a></li><li class="nav-item nav-level-1"><a class="nav-link"><span class="nav-number">3.</span> <span class="nav-text">Step 02 数据清洗</span></a></li><li class="nav-item nav-level-1"><a class="nav-link"><span class="nav-number">4.</span> <span class="nav-text">Step 03 质量控制</span></a></li><li class="nav-item nav-level-1"><a class="nav-link"><span class="nav-number">5.</span> <span class="nav-text">Step 04 差异表达分析</span></a></li><li class="nav-item nav-level-1"><a class="nav-link"><span class="nav-number">6.</span> <span class="nav-text">Step 05 注释</span></a></li><li class="nav-item nav-level-1"><a class="nav-link"><span class="nav-number">7.</span> <span class="nav-text">Step 06 富集分析</span></a></li></ol></div>
      </div>
      <!--/noindex-->

      <div class="site-overview-wrap sidebar-panel">
        <div class="site-author motion-element" itemprop="author" itemscope itemtype="http://schema.org/Person">
    <img class="site-author-image" itemprop="image" alt="Dean Li"
      src="https://avatars.githubusercontent.com/u/55836943?v=4">
  <p class="site-author-name" itemprop="name">Dean Li</p>
  <div class="site-description" itemprop="description"></div>
</div>
<div class="site-state-wrap motion-element">
  <nav class="site-state">
      <div class="site-state-item site-state-posts">
          <a href="/archives/">
        
          <span class="site-state-item-count">51</span>
          <span class="site-state-item-name">posts</span>
        </a>
      </div>
      <div class="site-state-item site-state-tags">
            <a href="/tags/">
          
        <span class="site-state-item-count">16</span>
        <span class="site-state-item-name">tags</span></a>
      </div>
  </nav>
</div>
  <div class="links-of-author motion-element">
      <span class="links-of-author-item">
        <a href="https://github.com/thereallda" title="GitHub → https:&#x2F;&#x2F;github.com&#x2F;thereallda" rel="noopener" target="_blank"><i class="fab fa-github fa-fw"></i>GitHub</a>
      </span>
      <span class="links-of-author-item">
        <a href="mailto:lideanlda@foxmail.com" title="E-Mail → mailto:lideanlda@foxmail.com" rel="noopener" target="_blank"><i class="fa fa-envelope fa-fw"></i>E-Mail</a>
      </span>
      <span class="links-of-author-item">
        <a href="https://www.jianshu.com/u/ab8a1c7d0510" title="简书 → https:&#x2F;&#x2F;www.jianshu.com&#x2F;u&#x2F;ab8a1c7d0510" rel="noopener" target="_blank"><i class="fa fa-book fa-fw"></i>简书</a>
      </span>
  </div>



      </div>

    </div>
  </aside>
  <div id="sidebar-dimmer"></div>


      </div>
    </main>

    <footer class="footer">
      <div class="footer-inner">
        

        

<div class="copyright">
  
  &copy; 
  <span itemprop="copyrightYear">2022</span>
  <span class="with-love">
    <i class="fa fa-heart"></i>
  </span>
  <span class="author" itemprop="copyrightHolder">Dean Li</span>
</div>
  <div class="powered-by">Powered by <a href="https://hexo.io/" class="theme-link" rel="noopener" target="_blank">Hexo</a> & <a href="https://theme-next.org/" class="theme-link" rel="noopener" target="_blank">NexT.Gemini</a>
  </div>



        
<div class="busuanzi-count">
  <script async src="https://busuanzi.ibruce.info/busuanzi/2.3/busuanzi.pure.mini.js"></script>
    <span class="post-meta-item" id="busuanzi_container_site_uv" style="display: none;">
      <span class="post-meta-item-icon">
        <i class="fa fa-user"></i>
      </span>
      <span class="site-uv" title="Total Visitors">
        <span id="busuanzi_value_site_uv"></span>
      </span>
    </span>
    <span class="post-meta-divider">|</span>
    <span class="post-meta-item" id="busuanzi_container_site_pv" style="display: none;">
      <span class="post-meta-item-icon">
        <i class="fa fa-eye"></i>
      </span>
      <span class="site-pv" title="Total Views">
        <span id="busuanzi_value_site_pv"></span>
      </span>
    </span>
</div>








      </div>
    </footer>
  </div>

  
  <script src="/lib/anime.min.js"></script>
  <script src="/lib/velocity/velocity.min.js"></script>
  <script src="/lib/velocity/velocity.ui.min.js"></script>

<script src="/js/utils.js"></script>

<script src="/js/motion.js"></script>


<script src="/js/schemes/pisces.js"></script>


<script src="/js/next-boot.js"></script>




  















  

  

  
<script src="https://cdn.jsdelivr.net/npm/darkmode-js@1.5.7/lib/darkmode-js.min.js"></script>

<script>
var options = {
  bottom: '64px',
  right: '32px',
  left: 'unset',
  time: '0.5s',
  mixColor: 'transparent',
  backgroundColor: 'transparent',
  buttonColorDark: '#100f2c',
  buttonColorLight: '#fff',
  saveInCookies: true,
  label: '🌓',
  autoMatchOsTheme: true
}
const darkmode = new Darkmode(options);
window.darkmode = darkmode;
darkmode.showWidget();
</script>

<link rel="stylesheet" href="//cdn.jsdelivr.net/npm/gitalk@1/dist/gitalk.min.css">

<script>
NexT.utils.loadComments(document.querySelector('#gitalk-container'), () => {
  NexT.utils.getScript('//cdn.jsdelivr.net/npm/gitalk@1/dist/gitalk.min.js', () => {
    var gitalk = new Gitalk({
      clientID    : 'aea502b6092de143578a',
      clientSecret: '7bb9f5afb99d67ab37d8b296f39782af05dc8614',
      repo        : 'thereallda.GitHub.io',
      owner       : 'thereallda',
      admin       : ['thereallda'],
      id          : '0fd0a78f5bce7f25c69a71d7c049897f',
        language: '',
      distractionFreeMode: true
    });
    gitalk.render('gitalk-container');
  }, window.Gitalk);
});
</script>

</body>
</html>
