<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
<meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=2">
<meta name="theme-color" content="#222">
<meta name="generator" content="Hexo 5.4.0">
  <link rel="apple-touch-icon" sizes="180x180" href="/images/apple-touch-icon.png">
  <link rel="icon" type="image/png" sizes="32x32" href="/images/favicon-32x32.png">
  <link rel="icon" type="image/png" sizes="16x16" href="/images/favicon-16x16.png">

<link rel="stylesheet" href="/css/main.css">


<link rel="stylesheet" href="/lib/font-awesome/css/all.min.css">

<script id="hexo-configurations">
    var NexT = window.NexT || {};
    var CONFIG = {"hostname":"thereallda.github.io","root":"/","scheme":"Gemini","version":"7.8.0","exturl":false,"sidebar":{"position":"left","display":"post","padding":18,"offset":12,"onmobile":false},"copycode":{"enable":true,"show_result":false,"style":"mac"},"back2top":{"enable":true,"sidebar":false,"scrollpercent":false},"bookmark":{"enable":false,"color":"#222","save":"auto"},"fancybox":false,"mediumzoom":false,"lazyload":false,"pangu":false,"comments":{"style":"tabs","active":"gitalk","storage":true,"lazyload":false,"nav":null,"activeClass":"gitalk"},"algolia":{"hits":{"per_page":10},"labels":{"input_placeholder":"Search for Posts","hits_empty":"We didn't find any results for the search: ${query}","hits_stats":"${hits} results found in ${time} ms"}},"localsearch":{"enable":false,"trigger":"auto","top_n_per_article":1,"unescape":false,"preload":false},"motion":{"enable":true,"async":false,"transition":{"post_block":"fadeIn","post_header":"slideDownIn","post_body":"slideDownIn","coll_header":"slideLeftIn","sidebar":"slideUpIn"}}};
  </script>

  <meta name="description" content="在高通量测序数据的分析中，仅仅靠raw read counts描述基因的表达量是远远不够的。受限于测序过程中的技术因素影响，read counts对于基因表达量的反映存在一定偏好（bias）。因此，Mortazavi等人提出了RPKM&#x2F;FPKM的方法对read counts进行normalization以使基因表达量的比较可以在不同文库间进行。随后，Mortazavi等人更是提出了考虑转录本长度分">
<meta property="og:type" content="article">
<meta property="og:title" content="理论-简述RPKM-FPKM-AND-TPM">
<meta property="og:url" content="https://thereallda.github.io/2022/01/24/%E7%90%86%E8%AE%BA-%E7%AE%80%E8%BF%B0RPKM-FPKM-AND-TPM/index.html">
<meta property="og:site_name" content="Dean&#39;s blog">
<meta property="og:description" content="在高通量测序数据的分析中，仅仅靠raw read counts描述基因的表达量是远远不够的。受限于测序过程中的技术因素影响，read counts对于基因表达量的反映存在一定偏好（bias）。因此，Mortazavi等人提出了RPKM&#x2F;FPKM的方法对read counts进行normalization以使基因表达量的比较可以在不同文库间进行。随后，Mortazavi等人更是提出了考虑转录本长度分">
<meta property="og:locale" content="en_US">
<meta property="og:image" content="https://thereallda.github.io/image-20220124102415711.png">
<meta property="og:image" content="https://thereallda.github.io/image-20220124102844133.png">
<meta property="og:image" content="https://thereallda.github.io/2022/01/24/%E7%90%86%E8%AE%BA-%E7%AE%80%E8%BF%B0RPKM-FPKM-AND-TPM/image-20220124111816290.png">
<meta property="og:image" content="https://thereallda.github.io/2022/01/24/%E7%90%86%E8%AE%BA-%E7%AE%80%E8%BF%B0RPKM-FPKM-AND-TPM/image-20220124111831499.png">
<meta property="og:image" content="https://thereallda.github.io/2022/01/24/%E7%90%86%E8%AE%BA-%E7%AE%80%E8%BF%B0RPKM-FPKM-AND-TPM/image-20220124111839423.png">
<meta property="article:published_time" content="2022-01-24T16:15:43.000Z">
<meta property="article:modified_time" content="2022-01-24T03:18:54.076Z">
<meta property="article:author" content="Dean Li">
<meta property="article:tag" content="RNA-seq">
<meta property="article:tag" content="Statistics">
<meta name="twitter:card" content="summary">
<meta name="twitter:image" content="https://thereallda.github.io/image-20220124102415711.png">

<link rel="canonical" href="https://thereallda.github.io/2022/01/24/%E7%90%86%E8%AE%BA-%E7%AE%80%E8%BF%B0RPKM-FPKM-AND-TPM/">


<script id="page-configurations">
  // https://hexo.io/docs/variables.html
  CONFIG.page = {
    sidebar: "",
    isHome : false,
    isPost : true,
    lang   : 'en'
  };
</script>

  <title>理论-简述RPKM-FPKM-AND-TPM | Dean's blog</title>
  






  <noscript>
  <style>
  .use-motion .brand,
  .use-motion .menu-item,
  .sidebar-inner,
  .use-motion .post-block,
  .use-motion .pagination,
  .use-motion .comments,
  .use-motion .post-header,
  .use-motion .post-body,
  .use-motion .collection-header { opacity: initial; }

  .use-motion .site-title,
  .use-motion .site-subtitle {
    opacity: initial;
    top: initial;
  }

  .use-motion .logo-line-before i { left: initial; }
  .use-motion .logo-line-after i { right: initial; }
  </style>
</noscript>

<style>.darkmode--activated{--body-bg-color:#282828;--content-bg-color:#333;--card-bg-color:#555;--text-color:#ccc;--blockquote-color:#bbb;--link-color:#ccc;--link-hover-color:#eee;--brand-color:#ddd;--brand-hover-color:#ddd;--table-row-odd-bg-color:#282828;--table-row-hover-bg-color:#363636;--menu-item-bg-color:#555;--btn-default-bg:#222;--btn-default-color:#ccc;--btn-default-border-color:#555;--btn-default-hover-bg:#666;--btn-default-hover-color:#ccc;--btn-default-hover-border-color:#666;--highlight-background:#282b2e;--highlight-foreground:#a9b7c6;--highlight-gutter-background:#34393d;--highlight-gutter-foreground:#9ca9b6}.darkmode--activated img{opacity:.75}.darkmode--activated img:hover{opacity:.9}.darkmode--activated code{color:#69dbdc;background:0 0}button.darkmode-toggle{z-index:9999}.darkmode-ignore,img{display:flex!important}</style></head>

<body itemscope itemtype="http://schema.org/WebPage">
  <div class="container use-motion">
    <div class="headband"></div>

    <header class="header" itemscope itemtype="http://schema.org/WPHeader">
      <div class="header-inner"><div class="site-brand-container">
  <div class="site-nav-toggle">
    <div class="toggle" aria-label="Toggle navigation bar">
      <span class="toggle-line toggle-line-first"></span>
      <span class="toggle-line toggle-line-middle"></span>
      <span class="toggle-line toggle-line-last"></span>
    </div>
  </div>

  <div class="site-meta">

    <a href="/" class="brand" rel="start">
      <span class="logo-line-before"><i></i></span>
      <h1 class="site-title">Dean's blog</h1>
      <span class="logo-line-after"><i></i></span>
    </a>
  </div>

  <div class="site-nav-right">
    <div class="toggle popup-trigger">
    </div>
  </div>
</div>




<nav class="site-nav">
  <ul id="menu" class="main-menu menu">
        <li class="menu-item menu-item-home">

    <a href="/" rel="section"><i class="fa fa-home fa-fw"></i>Home</a>

  </li>
        <li class="menu-item menu-item-about">

    <a href="/about/" rel="section"><i class="fa fa-user fa-fw"></i>About</a>

  </li>
        <li class="menu-item menu-item-tags">

    <a href="/tags/" rel="section"><i class="fa fa-tags fa-fw"></i>Tags</a>

  </li>
        <li class="menu-item menu-item-archives">

    <a href="/archives/" rel="section"><i class="fa fa-archive fa-fw"></i>Archives</a>

  </li>
  </ul>
</nav>




</div>
    </header>

    
  <div class="back-to-top">
    <i class="fa fa-arrow-up"></i>
    <span>0%</span>
  </div>
  <div class="reading-progress-bar"></div>


    <main class="main">
      <div class="main-inner">
        <div class="content-wrap">
          

          <div class="content post posts-expand">
            

    
  
  
  <article itemscope itemtype="http://schema.org/Article" class="post-block" lang="en">
    <link itemprop="mainEntityOfPage" href="https://thereallda.github.io/2022/01/24/%E7%90%86%E8%AE%BA-%E7%AE%80%E8%BF%B0RPKM-FPKM-AND-TPM/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="image" content="https://avatars.githubusercontent.com/u/55836943?v=4">
      <meta itemprop="name" content="Dean Li">
      <meta itemprop="description" content="">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="Dean's blog">
    </span>
      <header class="post-header">
        <h1 class="post-title" itemprop="name headline">
          理论-简述RPKM-FPKM-AND-TPM
        </h1>

        <div class="post-meta">
          
          
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-calendar"></i>
              </span>
              <span class="post-meta-item-text">Posted on</span>

              <time title="Created: 2022-01-24 11:15:43" itemprop="dateCreated datePublished" datetime="2022-01-24T11:15:43-05:00">2022-01-24</time>
            </span>
              <span class="post-meta-item">
                <span class="post-meta-item-icon">
                  <i class="far fa-calendar-check"></i>
                </span>
                <span class="post-meta-item-text">Edited on</span>
                <time title="Modified: 2022-01-23 22:18:54" itemprop="dateModified" datetime="2022-01-23T22:18:54-05:00">2022-01-23</time>
              </span>

          
            <span class="post-meta-item" title="Views" id="busuanzi_container_page_pv" style="display: none;">
              <span class="post-meta-item-icon">
                <i class="fa fa-eye"></i>
              </span>
              <span class="post-meta-item-text">Views: </span>
              <span id="busuanzi_value_page_pv"></span>
            </span><br>
            <span class="post-meta-item" title="Symbols count in article">
              <span class="post-meta-item-icon">
                <i class="far fa-file-word"></i>
              </span>
                <span class="post-meta-item-text">Symbols count in article: </span>
              <span>4.7k</span>
            </span>
            <span class="post-meta-item" title="Reading time">
              <span class="post-meta-item-icon">
                <i class="far fa-clock"></i>
              </span>
                <span class="post-meta-item-text">Reading time &asymp;</span>
              <span>4 mins.</span>
            </span>

        </div>
      </header>

    
    
    
    <div class="post-body" itemprop="articleBody">

      
        <p>在高通量测序数据的分析中，仅仅靠raw read counts描述基因的表达量是远远不够的。受限于测序过程中的技术因素影响，read counts对于基因表达量的反映存在一定偏好（bias）。因此，<strong>Mortazavi等人</strong>提出了<strong>RPKM/FPKM</strong>的方法对read counts进行normalization以使基因表达量的比较可以在不同文库间进行。随后，<strong>Mortazavi等人</strong>更是提出了考虑<strong>转录本长度分布</strong>情况的TPM方法。本文将会简要说明为什么我们要对read counts进行normalization，以及RPKM，FPKM，TPM是什么，并通过一个简单地例子阐述为什么TPM才是被更多人认同的方法。</p>
<span id="more"></span>

<h2 id="为什么我们要进行Normalization"><a href="#为什么我们要进行Normalization" class="headerlink" title="为什么我们要进行Normalization"></a>为什么我们要进行Normalization</h2><p>  我们之所以说测序得到的read counts并不是其mRNA丰度的忠实反映，是因为read counts的数量会受到多种因素的影响，例如：</p>
<ul>
<li><p><strong>测序深度</strong>：测序深度直接影响样本的read counts。例如，<strong>测序深度高</strong>的样本，同一个基因的read counts相对<strong>测序深度较低</strong>的样本而言也会较高，即使这个基因并没有发生差异表达。另外，某些低表达量的基因只有在较高的测序深度时才能检测到。一般而言，随着测序深度的增加，基因种类以及可变剪接体的数目也会增加。</p>
<p><img src="/image-20220124102415711.png" alt="Effect of sequencing depth"></p>
</li>
</ul>
<p>在上图中，样本A中的基因X表达量是样本B的两倍，但这是由于测序深度引起的结果，而非真实存在的差异。</p>
<ul>
<li><p><strong>基因长度</strong>：由于高通量测序是将cDNA碎片化后再进行测序的，因此越长的基因产生的片段会越多，其read counts也更多。这就导致了较长的基因read counts比较短的基因要更多，从而混淆了不同长度基因间表达量的比较，所以对基因长度的校正也是十分有必要的。</p>
<p><img src="/image-20220124102844133.png" alt="Effect of gene length"></p>
</li>
</ul>
<p>在这个图里，基因X和Y的真实表达量是一致的，但是基因X的read counts会比基因Y要多，这是由于基因X的基因长度较长所致的。</p>
<p>除了上述两个主要因素外，还会有其他因素对read counts的检测有所影响，例如<strong>转录组的组成，GC含量，random hexamers引起的测序偏好</strong>等等。由于上述因素的存在，导致在不同样本间使用read counts 进行比较是不太现实的，人们便提出了许多对read counts进行normalization的方法。本文在此简单地介绍使用最为广泛却最受质疑的RPKM/FPKM，以及受到更多人认可的TPM。</p>
<h2 id="RPKM-FPKM"><a href="#RPKM-FPKM" class="headerlink" title="RPKM/FPKM"></a>RPKM/FPKM</h2><blockquote>
<p>RPKM: <strong>Reads per kilo base per million mapped reads</strong> (single-end sequencing)</p>
<p>FPKM: <strong>Fragments per kilo base per million mapped reads</strong> (paired-end sequencing)</p>
</blockquote>
<p>RPKM与FPKM实际上一样的单位，只不过RPKM是在单端测序文库中使用，而FPKM是双端测序所用的</p>
<p><img src="/2022/01/24/%E7%90%86%E8%AE%BA-%E7%AE%80%E8%BF%B0RPKM-FPKM-AND-TPM/image-20220124111816290.png" alt="RPKM"></p>
<p>从RPKM的公式中我们可以看到，RPKM对<strong>基因长度（gene length）</strong>以<strong>及测序深度（mapped reads from library）</strong>都进行了校正。而FPKM只是RPKM的双端测序（pair-end）版本。</p>
<p>其R语言实现如下:</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">RPKM &lt;- <span class="keyword">function</span>(x, gene.length)&#123; </span><br><span class="line">  lib.size &lt;- <span class="built_in">sum</span>(x)</span><br><span class="line">  rpm &lt;- 1e6 * x/lib.size</span><br><span class="line">  rpkm &lt;- 1e3 * rpm/gene.length</span><br><span class="line">  <span class="built_in">return</span>(rpkm)</span><br><span class="line">&#125;</span><br></pre></td></tr></table></figure>

<p>其中<code>x</code>为一个向量存储了每个基因的counts，而<code>gene.length</code>是相应于<code>x</code>中每个基因的长度的一个向量.</p>
<h2 id="TPM"><a href="#TPM" class="headerlink" title="TPM"></a>TPM</h2><blockquote>
<p>TPM: <strong>Transcript per million</strong></p>
</blockquote>
<p>TPM的计算公式如下所示：</p>
<p><img src="/2022/01/24/%E7%90%86%E8%AE%BA-%E7%AE%80%E8%BF%B0RPKM-FPKM-AND-TPM/image-20220124111831499.png" alt="TPM"></p>
<p>同RPKM一样，TPM对基因的长度进行了校正，计算<strong>比对到基因上的reads</strong>/<strong>基因长度</strong>得到<strong>长度校正的表达量 reads per kilobase (RPK)<strong>。再以文库中</strong>RPK之和</strong>作为Scale Factor求出TPM。</p>
<p>其R语言实现如下:</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">TPM &lt;- <span class="keyword">function</span>(x, gene.length)&#123;</span><br><span class="line">  rpk &lt;- 1e3 * x/gene.length</span><br><span class="line">  scale.factor &lt;- <span class="built_in">sum</span>(rpk)</span><br><span class="line">  tpm &lt;- 1e6 * rpk/scale.factor</span><br><span class="line">  <span class="built_in">return</span>(tpm)</span><br><span class="line">&#125;</span><br></pre></td></tr></table></figure>

<p>其中<code>x</code>为一个向量存储了每个基因的counts，而<code>gene.length</code>是相应于<code>x</code>中每个基因的长度的一个向量.</p>
<p>相比于RPKM使用文库大小（<strong>read counts之和</strong>）来作为文库校正因子，TPM使用<strong>RPK之和</strong>作为文库校正因子的好处是考虑了不同样本间的基因长度的分布。因为RPK是一个对基因长度进行校正后的表达量单位，所以<strong>RPK之和</strong>也不会再带入基因长度的bias。因此，如果需要比较的样本之间转录本分布不一致时（例如不同物种RNA-seq的比较），使用TPM是一个较佳的normalization方案。</p>
<h2 id="例子：RPKM-FPKM-vs-TPM"><a href="#例子：RPKM-FPKM-vs-TPM" class="headerlink" title="例子：RPKM/FPKM vs TPM"></a>例子：RPKM/FPKM vs TPM</h2><p>接下来，我们使用一个例子展示RPKM/FPKM和TPM方法校正的区别。</p>
<p>以下有<strong>ABCD四个基因</strong>，并同时进行了<strong>三次重复的测序（rep1，rep2，rep3）</strong>。首先，可以看出由于<strong>基因长度</strong>的关系，B基因的read counts都是较其余的基因较高的（排除D而言）。其次，可以看出rep3的<strong>测序深度</strong>较高，得到的read counts也较多，同时还检测到其余两个重复中没有检测到的D基因（低丰度基因）。不管是测序深度还是基因长度的严重地干扰了我们对不同样本的基因表达量比较。因此，我们采用RPKM和TPM的方法进行Normalization后，再来比较其中的差异。</p>
<p><img src="/2022/01/24/%E7%90%86%E8%AE%BA-%E7%AE%80%E8%BF%B0RPKM-FPKM-AND-TPM/image-20220124111839423.png"></p>
<p>为了方便起见，以下的函数省略了kilo base，million base的转换</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br></pre></td><td class="code"><pre><span class="line">RPKM &lt;- function(x, gene.length)&#123; </span><br><span class="line">  lib.size &lt;- sum(x)</span><br><span class="line">  rpm &lt;- x/lib.size</span><br><span class="line">  rpkm &lt;- rpm/gene.length</span><br><span class="line">  return(rpkm)</span><br><span class="line">&#125;</span><br><span class="line"></span><br><span class="line">TPM &lt;- function(x, gene.length)&#123;</span><br><span class="line">  rpk &lt;- x/gene.length</span><br><span class="line">  scale.factor &lt;- sum(rpk)</span><br><span class="line">  tpm &lt;- rpk/scale.factor</span><br><span class="line">  return(tpm)</span><br><span class="line">&#125;</span><br><span class="line"></span><br><span class="line">expr &lt;- data.frame(rep1=c(10,20,5,0), rep2=c(12,25,8,0), rep3=c(30,60,15,1), </span><br><span class="line">                   row.names = c(&quot;A&quot;, &quot;B&quot;, &quot;C&quot;, &quot;D&quot;))</span><br><span class="line">gl &lt;- c(2,4,1,10)</span><br><span class="line"></span><br><span class="line">rpkm1 &lt;- apply(expr, 2, FUN = RPKM, gene.length = gl)</span><br><span class="line">tpm1 &lt;- apply(expr, 2, FUN = TPM, gene.length = gl)</span><br><span class="line"># 对结果都乘1000，看起来直观点</span><br><span class="line">rbind(rpkm1*1000, colSums(rpkm1)*1000)</span><br><span class="line">##       rep1     rep2        rep3</span><br><span class="line">## A 142.8571 133.3333 141.5094340</span><br><span class="line">## B 142.8571 138.8889 141.5094340</span><br><span class="line">## C 142.8571 177.7778 141.5094340</span><br><span class="line">## D   0.0000   0.0000   0.9433962</span><br><span class="line">##   428.5714 450.0000 425.4716981</span><br><span class="line"></span><br><span class="line">rbind(tpm1*1000, colSums(tpm1)*1000)</span><br><span class="line">##        rep1      rep2        rep3</span><br><span class="line">## A  333.3333  296.2963  332.594235</span><br><span class="line">## B  333.3333  308.6420  332.594235</span><br><span class="line">## C  333.3333  395.0617  332.594235</span><br><span class="line">## D    0.0000    0.0000    2.217295</span><br><span class="line">##   1000.0000 1000.0000 1000.000000</span><br></pre></td></tr></table></figure>

<p>通过以上的例子，我们可以发现RPKM和TPM最重要且直接的区别就是，TPM校正后的表达量之和在不同样本中是相等的，而在RPKM中并不是。</p>
<p>由于RPKM在各样本之和不一致，导致我们不能<strong>直接</strong>比较同一基因在不同样本间的RPKM。例如，对于基因A而言，<strong>RPKM校正后</strong>，其在rep1和rep2中的占比分别为<code>142.8571/428.5714</code>和<code>133.3333/450</code>。相反，由于<strong>TPM</strong>在各样本之和是一致的，我们可以直接比较不同样本的TPM来比较该基因表达量在样本中的占比。</p>
<p>简而言之，RPKM和TPM这类方法就是为了使不同样本间的总体表达量趋于一致，让不同样本间的基因表达量有可比较性，而TPM能够更好地校正样本间的差异。</p>
<ul>
<li>常用的Normalization 方法总结</li>
</ul>
<table>
<thead>
<tr>
<th>Normalization method</th>
<th>描述</th>
<th>考虑因素</th>
<th>使用场景</th>
</tr>
</thead>
<tbody><tr>
<td><strong>CPM</strong> (counts per million)</td>
<td>使用read counts的总和校正counts</td>
<td>测序深度</td>
<td>同一样本组重复样本之间的基因counts比较；<strong>不适用于样品内的比较或差异分析</strong></td>
</tr>
<tr>
<td><strong>TPM</strong> (transcripts per kilobase million)</td>
<td>每百万mapped reads中每kb转录本上的reads数</td>
<td>测序深度和基因长度</td>
<td>样本内或同一样本组样本之间的基因counts比较； <strong>不适用于差异分析</strong></td>
</tr>
<tr>
<td><strong>RPKM/FPKM</strong> (reads/fragments per kilobase of exon per million reads/fragments mapped)</td>
<td>如TPM</td>
<td>测序深度和基因长度</td>
<td>样本内基因间的counts比较; <strong>不适用于样本间比较和差异分析</strong></td>
</tr>
<tr>
<td>DESeq2’s <strong>median of ratios</strong> [<a target="_blank" rel="noopener" href="https://genomebiology.biomedcentral.com/articles/10.1186/gb-2010-11-10-r106">1</a>]</td>
<td>counts除以样本特异的校正因子，该因子由基因计数相对于每个基因的几何平均值的中位数比率确定</td>
<td>测序深度及RNA组成</td>
<td>样本之间的基因counts比较以及差异分析; <strong>不适用于样本内比较</strong></td>
</tr>
<tr>
<td>EdgeR’s <strong>trimmed mean of M values (TMM)</strong> [<a target="_blank" rel="noopener" href="https://genomebiology.biomedcentral.com/articles/10.1186/gb-2010-11-3-r25">2</a>]</td>
<td>使用样本之间的加权截尾的对数表达量比值的均值进行TMM校正</td>
<td>测序深度，RNA组成以及基因长度</td>
<td>样品之间和样品内部的基因counts比较，适用于差异分析</td>
</tr>
</tbody></table>
<p>这里要补充的是在进行差异表达分析时，我们实际上并不会用到RPKM/FPKM, TPM，而是使用raw counts给到差异分析的工具。这是由于RPKM/FPKM和TPM都没有考虑到样本间文库组成的差异，因而不适合用于样本间差异分析。例如，直接比较小鼠肝脏与肾脏之间基因表达的TPM是不太合理的。</p>
<blockquote>
<p>Ref:</p>
<ol>
<li><a target="_blank" rel="noopener" href="https://www.rna-seqblog.com/rpkm-fpkm-and-tpm-clearly-explained/">https://www.rna-seqblog.com/rpkm-fpkm-and-tpm-clearly-explained/</a></li>
<li><a target="_blank" rel="noopener" href="https://www.biostars.org/p/273537/">https://www.biostars.org/p/273537/</a></li>
<li><a target="_blank" rel="noopener" href="https://www.jianshu.com/p/1940c5954c81">https://www.jianshu.com/p/1940c5954c81</a></li>
<li>Mortazavi A, Williams BA, McCue K, Schaeffer L, Wold B. Mapping and quantifying mammalian transcriptomes by RNA-Seq. Nature methods. 2008 Jul 1;5(7):621-8.</li>
<li><a target="_blank" rel="noopener" href="https://hbctraining.github.io/DGE_workshop/lessons/02_DGE_count_normalization.html">https://hbctraining.github.io/DGE_workshop/lessons/02_DGE_count_normalization.html</a></li>
</ol>
</blockquote>
<p>完。</p>

    </div>

    
    
    
        

<div>
<ul class="post-copyright">
  <li class="post-copyright-author">
    <strong>Post author:  </strong>Dean Li
  </li>
  <li class="post-copyright-link">
    <strong>Post link: </strong>
    <a href="https://thereallda.github.io/2022/01/24/%E7%90%86%E8%AE%BA-%E7%AE%80%E8%BF%B0RPKM-FPKM-AND-TPM/" title="理论-简述RPKM-FPKM-AND-TPM">https://thereallda.github.io/2022/01/24/理论-简述RPKM-FPKM-AND-TPM/</a>
  </li>
  <li class="post-copyright-license">
    <strong>Copyright Notice:  </strong>All articles in this blog are licensed under <a href="https://creativecommons.org/licenses/by-nc-sa/4.0/" rel="noopener" target="_blank"><i class="fab fa-fw fa-creative-commons"></i>BY-NC-SA</a> unless stating additionally.
  </li>
</ul>
</div>


      <footer class="post-footer">
          <div class="post-tags">
              <a href="/tags/RNA-seq/" rel="tag"># RNA-seq</a>
              <a href="/tags/Statistics/" rel="tag"># Statistics</a>
          </div>

        


        
    <div class="post-nav">
      <div class="post-nav-item">
    <a href="/2022/01/24/edgeR-%E7%AE%80%E5%8D%95%E7%9A%84%E8%BD%AC%E5%BD%95%E7%BB%84%E5%B7%AE%E5%BC%82%E8%A1%A8%E8%BE%BE%E5%88%86%E6%9E%90/" rel="prev" title="edgeR-简单的转录组差异表达分析">
      <i class="fa fa-chevron-left"></i> edgeR-简单的转录组差异表达分析
    </a></div>
      <div class="post-nav-item">
    <a href="/2022/01/24/%E7%90%86%E8%AE%BA-edgeR-TMM-normalization/" rel="next" title="理论-edgeR-TMM-normalization">
      理论-edgeR-TMM-normalization <i class="fa fa-chevron-right"></i>
    </a></div>
    </div>
      </footer>
    
  </article>
  
  
  



          </div>
          
    <div class="comments" id="gitalk-container"></div>

<script>
  window.addEventListener('tabs:register', () => {
    let { activeClass } = CONFIG.comments;
    if (CONFIG.comments.storage) {
      activeClass = localStorage.getItem('comments_active') || activeClass;
    }
    if (activeClass) {
      let activeTab = document.querySelector(`a[href="#comment-${activeClass}"]`);
      if (activeTab) {
        activeTab.click();
      }
    }
  });
  if (CONFIG.comments.storage) {
    window.addEventListener('tabs:click', event => {
      if (!event.target.matches('.tabs-comment .tab-content .tab-pane')) return;
      let commentClass = event.target.classList[1];
      localStorage.setItem('comments_active', commentClass);
    });
  }
</script>

        </div>
          
  
  <div class="toggle sidebar-toggle">
    <span class="toggle-line toggle-line-first"></span>
    <span class="toggle-line toggle-line-middle"></span>
    <span class="toggle-line toggle-line-last"></span>
  </div>

  <aside class="sidebar">
    <div class="sidebar-inner">

      <ul class="sidebar-nav motion-element">
        <li class="sidebar-nav-toc">
          Table of Contents
        </li>
        <li class="sidebar-nav-overview">
          Overview
        </li>
      </ul>

      <!--noindex-->
      <div class="post-toc-wrap sidebar-panel">
          <div class="post-toc motion-element"><ol class="nav"><li class="nav-item nav-level-2"><a class="nav-link" href="#%E4%B8%BA%E4%BB%80%E4%B9%88%E6%88%91%E4%BB%AC%E8%A6%81%E8%BF%9B%E8%A1%8CNormalization"><span class="nav-number">1.</span> <span class="nav-text">为什么我们要进行Normalization</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#RPKM-FPKM"><span class="nav-number">2.</span> <span class="nav-text">RPKM&#x2F;FPKM</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#TPM"><span class="nav-number">3.</span> <span class="nav-text">TPM</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#%E4%BE%8B%E5%AD%90%EF%BC%9ARPKM-FPKM-vs-TPM"><span class="nav-number">4.</span> <span class="nav-text">例子：RPKM&#x2F;FPKM vs TPM</span></a></li></ol></div>
      </div>
      <!--/noindex-->

      <div class="site-overview-wrap sidebar-panel">
        <div class="site-author motion-element" itemprop="author" itemscope itemtype="http://schema.org/Person">
    <img class="site-author-image" itemprop="image" alt="Dean Li"
      src="https://avatars.githubusercontent.com/u/55836943?v=4">
  <p class="site-author-name" itemprop="name">Dean Li</p>
  <div class="site-description" itemprop="description"></div>
</div>
<div class="site-state-wrap motion-element">
  <nav class="site-state">
      <div class="site-state-item site-state-posts">
          <a href="/archives/">
        
          <span class="site-state-item-count">66</span>
          <span class="site-state-item-name">posts</span>
        </a>
      </div>
      <div class="site-state-item site-state-tags">
            <a href="/tags/">
          
        <span class="site-state-item-count">21</span>
        <span class="site-state-item-name">tags</span></a>
      </div>
  </nav>
</div>
  <div class="links-of-author motion-element">
      <span class="links-of-author-item">
        <a href="https://github.com/thereallda" title="GitHub → https:&#x2F;&#x2F;github.com&#x2F;thereallda" rel="noopener" target="_blank"><i class="fab fa-github fa-fw"></i>GitHub</a>
      </span>
      <span class="links-of-author-item">
        <a href="mailto:lideanlda@foxmail.com" title="E-Mail → mailto:lideanlda@foxmail.com" rel="noopener" target="_blank"><i class="fa fa-envelope fa-fw"></i>E-Mail</a>
      </span>
      <span class="links-of-author-item">
        <a href="https://www.jianshu.com/u/ab8a1c7d0510" title="简书 → https:&#x2F;&#x2F;www.jianshu.com&#x2F;u&#x2F;ab8a1c7d0510" rel="noopener" target="_blank"><i class="fa fa-book fa-fw"></i>简书</a>
      </span>
  </div>



      </div>

    </div>
  </aside>
  <div id="sidebar-dimmer"></div>


      </div>
    </main>

    <footer class="footer">
      <div class="footer-inner">
        

        

<div class="copyright">
  
  &copy; 
  <span itemprop="copyrightYear">2025</span>
  <span class="with-love">
    <i class="fa fa-heart"></i>
  </span>
  <span class="author" itemprop="copyrightHolder">Dean Li</span>
</div>
  <div class="powered-by">Powered by <a href="https://hexo.io/" class="theme-link" rel="noopener" target="_blank">Hexo</a> & <a href="https://theme-next.org/" class="theme-link" rel="noopener" target="_blank">NexT.Gemini</a>
  </div>



        
<div class="busuanzi-count">
  <script async src="https://busuanzi.ibruce.info/busuanzi/2.3/busuanzi.pure.mini.js"></script>
    <span class="post-meta-item" id="busuanzi_container_site_uv" style="display: none;">
      <span class="post-meta-item-icon">
        <i class="fa fa-user"></i>
      </span>
      <span class="site-uv" title="Total Visitors">
        <span id="busuanzi_value_site_uv"></span>
      </span>
    </span>
    <span class="post-meta-divider">|</span>
    <span class="post-meta-item" id="busuanzi_container_site_pv" style="display: none;">
      <span class="post-meta-item-icon">
        <i class="fa fa-eye"></i>
      </span>
      <span class="site-pv" title="Total Views">
        <span id="busuanzi_value_site_pv"></span>
      </span>
    </span>
</div>








      </div>
    </footer>
  </div>

  
  <script src="/lib/anime.min.js"></script>
  <script src="/lib/velocity/velocity.min.js"></script>
  <script src="/lib/velocity/velocity.ui.min.js"></script>

<script src="/js/utils.js"></script>

<script src="/js/motion.js"></script>


<script src="/js/schemes/pisces.js"></script>


<script src="/js/next-boot.js"></script>




  











<script>
document.querySelectorAll('.pdfobject-container').forEach(element => {
  let url = element.dataset.target;
  let pdfOpenParams = {
    navpanes : 0,
    toolbar  : 0,
    statusbar: 0,
    pagemode : 'thumbs',
    view     : 'FitH'
  };
  let pdfOpenFragment = '#' + Object.entries(pdfOpenParams).map(([key, value]) => `${key}=${encodeURIComponent(value)}`).join('&');
  let fullURL = `/lib/pdf/web/viewer.html?file=${encodeURIComponent(url)}${pdfOpenFragment}`;

  if (NexT.utils.supportsPDFs()) {
    element.innerHTML = `<embed class="pdfobject" src="${url + pdfOpenFragment}" type="application/pdf" style="height: ${element.dataset.height};">`;
  } else {
    element.innerHTML = `<iframe src="${fullURL}" style="height: ${element.dataset.height};" frameborder="0"></iframe>`;
  }
});
</script>




  

  

  
<script src="https://cdn.jsdelivr.net/npm/darkmode-js@1.5.7/lib/darkmode-js.min.js"></script>

<script>
var options = {
  bottom: '64px',
  right: '32px',
  left: 'unset',
  time: '0.5s',
  mixColor: 'transparent',
  backgroundColor: 'transparent',
  buttonColorDark: '#100f2c',
  buttonColorLight: '#fff',
  saveInCookies: true,
  label: '🌓',
  autoMatchOsTheme: true
}
const darkmode = new Darkmode(options);
window.darkmode = darkmode;
darkmode.showWidget();
</script>

<link rel="stylesheet" href="//cdn.jsdelivr.net/npm/gitalk@1/dist/gitalk.min.css">

<script>
NexT.utils.loadComments(document.querySelector('#gitalk-container'), () => {
  NexT.utils.getScript('//cdn.jsdelivr.net/npm/gitalk@1/dist/gitalk.min.js', () => {
    var gitalk = new Gitalk({
      clientID    : 'aea502b6092de143578a',
      clientSecret: '7bb9f5afb99d67ab37d8b296f39782af05dc8614',
      repo        : 'thereallda.GitHub.io',
      owner       : 'thereallda',
      admin       : ['thereallda'],
      id          : '111af392d9156431d9fe46b12c073330',
        language: '',
      distractionFreeMode: true
    });
    gitalk.render('gitalk-container');
  }, window.Gitalk);
});
</script>

</body>
</html>
