<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
<meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=2">
<meta name="theme-color" content="#222">
<meta name="generator" content="Hexo 5.4.0">
  <link rel="apple-touch-icon" sizes="180x180" href="/images/apple-touch-icon.png">
  <link rel="icon" type="image/png" sizes="32x32" href="/images/favicon-32x32.png">
  <link rel="icon" type="image/png" sizes="16x16" href="/images/favicon-16x16.png">

<link rel="stylesheet" href="/css/main.css">


<link rel="stylesheet" href="/lib/font-awesome/css/all.min.css">

<script id="hexo-configurations">
    var NexT = window.NexT || {};
    var CONFIG = {"hostname":"thereallda.github.io","root":"/","scheme":"Gemini","version":"7.8.0","exturl":false,"sidebar":{"position":"left","display":"post","padding":18,"offset":12,"onmobile":false},"copycode":{"enable":true,"show_result":false,"style":"mac"},"back2top":{"enable":true,"sidebar":false,"scrollpercent":false},"bookmark":{"enable":false,"color":"#222","save":"auto"},"fancybox":false,"mediumzoom":false,"lazyload":false,"pangu":false,"comments":{"style":"tabs","active":"gitalk","storage":true,"lazyload":false,"nav":null,"activeClass":"gitalk"},"algolia":{"hits":{"per_page":10},"labels":{"input_placeholder":"Search for Posts","hits_empty":"We didn't find any results for the search: ${query}","hits_stats":"${hits} results found in ${time} ms"}},"localsearch":{"enable":false,"trigger":"auto","top_n_per_article":1,"unescape":false,"preload":false},"motion":{"enable":true,"async":false,"transition":{"post_block":"fadeIn","post_header":"slideDownIn","post_body":"slideDownIn","coll_header":"slideLeftIn","sidebar":"slideUpIn"}}};
  </script>

  <meta name="description" content="整理ChIP-seq &#x2F; CUT &amp; Tag 分析时用到的工具。本文只对使用的工具用法进行简单介绍。   Bowtie 2是常用的基因组比对软件。其原理在此展开，有兴趣的同学可以参阅其官方文档以及其发表的文章(https:&#x2F;&#x2F;doi.org&#x2F;10.1038&#x2F;nmeth.1923)。下面简单介绍Bowtie 2 Index和比对的命令及个人常用参数。">
<meta property="og:type" content="article">
<meta property="og:title" content="BOWTIE2-进行基因组比对">
<meta property="og:url" content="https://thereallda.github.io/2022/01/24/BOWTIE2-%E8%BF%9B%E8%A1%8C%E5%9F%BA%E5%9B%A0%E7%BB%84%E6%AF%94%E5%AF%B9/index.html">
<meta property="og:site_name" content="Dean&#39;s blog">
<meta property="og:description" content="整理ChIP-seq &#x2F; CUT &amp; Tag 分析时用到的工具。本文只对使用的工具用法进行简单介绍。   Bowtie 2是常用的基因组比对软件。其原理在此展开，有兴趣的同学可以参阅其官方文档以及其发表的文章(https:&#x2F;&#x2F;doi.org&#x2F;10.1038&#x2F;nmeth.1923)。下面简单介绍Bowtie 2 Index和比对的命令及个人常用参数。">
<meta property="og:locale" content="en_US">
<meta property="og:image" content="https://thereallda.github.io/2022/01/24/BOWTIE2-%E8%BF%9B%E8%A1%8C%E5%9F%BA%E5%9B%A0%E7%BB%84%E6%AF%94%E5%AF%B9/image-20220124160935067.png">
<meta property="og:image" content="https://thereallda.github.io/2022/01/24/BOWTIE2-%E8%BF%9B%E8%A1%8C%E5%9F%BA%E5%9B%A0%E7%BB%84%E6%AF%94%E5%AF%B9/image-20220124183540041.png">
<meta property="article:published_time" content="2022-01-24T21:06:37.000Z">
<meta property="article:modified_time" content="2022-01-24T10:35:47.424Z">
<meta property="article:author" content="Dean Li">
<meta property="article:tag" content="genome">
<meta property="article:tag" content="Data_analysis">
<meta name="twitter:card" content="summary">
<meta name="twitter:image" content="https://thereallda.github.io/2022/01/24/BOWTIE2-%E8%BF%9B%E8%A1%8C%E5%9F%BA%E5%9B%A0%E7%BB%84%E6%AF%94%E5%AF%B9/image-20220124160935067.png">

<link rel="canonical" href="https://thereallda.github.io/2022/01/24/BOWTIE2-%E8%BF%9B%E8%A1%8C%E5%9F%BA%E5%9B%A0%E7%BB%84%E6%AF%94%E5%AF%B9/">


<script id="page-configurations">
  // https://hexo.io/docs/variables.html
  CONFIG.page = {
    sidebar: "",
    isHome : false,
    isPost : true,
    lang   : 'en'
  };
</script>

  <title>BOWTIE2-进行基因组比对 | Dean's blog</title>
  






  <noscript>
  <style>
  .use-motion .brand,
  .use-motion .menu-item,
  .sidebar-inner,
  .use-motion .post-block,
  .use-motion .pagination,
  .use-motion .comments,
  .use-motion .post-header,
  .use-motion .post-body,
  .use-motion .collection-header { opacity: initial; }

  .use-motion .site-title,
  .use-motion .site-subtitle {
    opacity: initial;
    top: initial;
  }

  .use-motion .logo-line-before i { left: initial; }
  .use-motion .logo-line-after i { right: initial; }
  </style>
</noscript>

<style>.darkmode--activated{--body-bg-color:#282828;--content-bg-color:#333;--card-bg-color:#555;--text-color:#ccc;--blockquote-color:#bbb;--link-color:#ccc;--link-hover-color:#eee;--brand-color:#ddd;--brand-hover-color:#ddd;--table-row-odd-bg-color:#282828;--table-row-hover-bg-color:#363636;--menu-item-bg-color:#555;--btn-default-bg:#222;--btn-default-color:#ccc;--btn-default-border-color:#555;--btn-default-hover-bg:#666;--btn-default-hover-color:#ccc;--btn-default-hover-border-color:#666;--highlight-background:#282b2e;--highlight-foreground:#a9b7c6;--highlight-gutter-background:#34393d;--highlight-gutter-foreground:#9ca9b6}.darkmode--activated img{opacity:.75}.darkmode--activated img:hover{opacity:.9}.darkmode--activated code{color:#69dbdc;background:0 0}button.darkmode-toggle{z-index:9999}.darkmode-ignore,img{display:flex!important}</style></head>

<body itemscope itemtype="http://schema.org/WebPage">
  <div class="container use-motion">
    <div class="headband"></div>

    <header class="header" itemscope itemtype="http://schema.org/WPHeader">
      <div class="header-inner"><div class="site-brand-container">
  <div class="site-nav-toggle">
    <div class="toggle" aria-label="Toggle navigation bar">
      <span class="toggle-line toggle-line-first"></span>
      <span class="toggle-line toggle-line-middle"></span>
      <span class="toggle-line toggle-line-last"></span>
    </div>
  </div>

  <div class="site-meta">

    <a href="/" class="brand" rel="start">
      <span class="logo-line-before"><i></i></span>
      <h1 class="site-title">Dean's blog</h1>
      <span class="logo-line-after"><i></i></span>
    </a>
  </div>

  <div class="site-nav-right">
    <div class="toggle popup-trigger">
    </div>
  </div>
</div>




<nav class="site-nav">
  <ul id="menu" class="main-menu menu">
        <li class="menu-item menu-item-home">

    <a href="/" rel="section"><i class="fa fa-home fa-fw"></i>Home</a>

  </li>
        <li class="menu-item menu-item-about">

    <a href="/about/" rel="section"><i class="fa fa-user fa-fw"></i>About</a>

  </li>
        <li class="menu-item menu-item-tags">

    <a href="/tags/" rel="section"><i class="fa fa-tags fa-fw"></i>Tags</a>

  </li>
        <li class="menu-item menu-item-archives">

    <a href="/archives/" rel="section"><i class="fa fa-archive fa-fw"></i>Archives</a>

  </li>
  </ul>
</nav>




</div>
    </header>

    
  <div class="back-to-top">
    <i class="fa fa-arrow-up"></i>
    <span>0%</span>
  </div>
  <div class="reading-progress-bar"></div>


    <main class="main">
      <div class="main-inner">
        <div class="content-wrap">
          

          <div class="content post posts-expand">
            

    
  
  
  <article itemscope itemtype="http://schema.org/Article" class="post-block" lang="en">
    <link itemprop="mainEntityOfPage" href="https://thereallda.github.io/2022/01/24/BOWTIE2-%E8%BF%9B%E8%A1%8C%E5%9F%BA%E5%9B%A0%E7%BB%84%E6%AF%94%E5%AF%B9/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="image" content="https://avatars.githubusercontent.com/u/55836943?v=4">
      <meta itemprop="name" content="Dean Li">
      <meta itemprop="description" content="">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="Dean's blog">
    </span>
      <header class="post-header">
        <h1 class="post-title" itemprop="name headline">
          BOWTIE2-进行基因组比对
        </h1>

        <div class="post-meta">
          
          
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-calendar"></i>
              </span>
              <span class="post-meta-item-text">Posted on</span>
              

              <time title="Created: 2022-01-24 16:06:37 / Modified: 05:35:47" itemprop="dateCreated datePublished" datetime="2022-01-24T16:06:37-05:00">2022-01-24</time>
            </span>

          
            <span class="post-meta-item" title="Views" id="busuanzi_container_page_pv" style="display: none;">
              <span class="post-meta-item-icon">
                <i class="fa fa-eye"></i>
              </span>
              <span class="post-meta-item-text">Views: </span>
              <span id="busuanzi_value_page_pv"></span>
            </span><br>
            <span class="post-meta-item" title="Symbols count in article">
              <span class="post-meta-item-icon">
                <i class="far fa-file-word"></i>
              </span>
                <span class="post-meta-item-text">Symbols count in article: </span>
              <span>4.5k</span>
            </span>
            <span class="post-meta-item" title="Reading time">
              <span class="post-meta-item-icon">
                <i class="far fa-clock"></i>
              </span>
                <span class="post-meta-item-text">Reading time &asymp;</span>
              <span>4 mins.</span>
            </span>

        </div>
      </header>

    
    
    
    <div class="post-body" itemprop="articleBody">

      
        <blockquote>
<p>整理ChIP-seq / CUT &amp; Tag 分析时用到的工具。本文只对使用的工具用法进行简单介绍。</p>
</blockquote>
<p><img src="/2022/01/24/BOWTIE2-%E8%BF%9B%E8%A1%8C%E5%9F%BA%E5%9B%A0%E7%BB%84%E6%AF%94%E5%AF%B9/image-20220124160935067.png"></p>
<p>Bowtie 2是常用的基因组比对软件。其原理在此展开，有兴趣的同学可以参阅其<a target="_blank" rel="noopener" href="http://bowtie-bio.sourceforge.net/bowtie2/manual.shtml#how-is-bowtie-2-different-from-bowtie-1">官方文档</a>以及其发表的文章(<a target="_blank" rel="noopener" href="https://doi.org/10.1038/nmeth.1923)%E3%80%82%E4%B8%8B%E9%9D%A2%E7%AE%80%E5%8D%95%E4%BB%8B%E7%BB%8DBowtie">https://doi.org/10.1038/nmeth.1923)。下面简单介绍Bowtie</a> 2 Index和比对的命令及个人常用参数。</p>
<span id="more"></span>

<h1 id="用法"><a href="#用法" class="headerlink" title="用法"></a>用法</h1><h2 id="Index"><a href="#Index" class="headerlink" title="Index"></a>Index</h2><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">bowtie2-build [options]* &lt;reference_in&gt; &lt;bt2_base&gt;</span><br></pre></td></tr></table></figure>
<p><reference_in>：如果此处使用<code>-f</code> 参数，则指明index的参考fasta 文件；如果使用<code>-c</code>参数，则指明index的参考序列，例如，<code>GGTCATCCT</code>,<code>ACGGGTCGT</code>,<code>CCGTTCTATGCGGCTTA</code>.<br><bt2_base>：指的是生成的index文件的前缀，默认情况，<code>bowtie2-build</code>产生NAME.1.bt2, NAME.2.bt2, NAME.3.bt2, NAME.4.bt2, NAME.rev.1.bt2, and NAME.rev.2.bt2, where NAME is <bt2_base>.<br><code>--threads</code> 使用的线程数</p>
<h3 id="Index例子"><a href="#Index例子" class="headerlink" title="Index例子"></a>Index例子</h3><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">bowtie2-build -f /public/Reference/GRCh38.primary_assembly.genome.fa --threads 24 GRCh38</span><br></pre></td></tr></table></figure>
<p>上述命令使用该fasta文件<code>/public/Reference/GRCh38.primary_assembly.genome.fa </code> ，在当前位置产生前缀为<code>GRCh38</code>的index文件。<br><img src="/2022/01/24/BOWTIE2-%E8%BF%9B%E8%A1%8C%E5%9F%BA%E5%9B%A0%E7%BB%84%E6%AF%94%E5%AF%B9/image-20220124183540041.png"></p>
<h2 id="Alignment"><a href="#Alignment" class="headerlink" title="Alignment"></a>Alignment</h2><h3 id="单端测序比对"><a href="#单端测序比对" class="headerlink" title="单端测序比对"></a>单端测序比对</h3><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">bowtie2 [options]* -x &lt;bt2-idx&gt; -U &lt;fq&gt; -S &lt;sam_output&gt; -p &lt;threads&gt; 2&gt;Align.summary</span><br></pre></td></tr></table></figure>

<p><code>-x</code>：参考基因组index文件的前缀（包括路径）<br><code>-U</code>：单端测序的fastq文件<br><code>-S</code>：输出的SAM文件，包含比对结果<br><code>-p</code>：使用的线程数<br><code>2&gt;Align.summary</code>：将输出到屏幕的标准误(standard error)重导向到”Align.summary”文件，其格式通常如下</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br></pre></td><td class="code"><pre><span class="line">## Single-end</span><br><span class="line">20000 reads; of these:</span><br><span class="line">  20000 (100.00%) were unpaired; of these:</span><br><span class="line">    1247 (6.24%) aligned 0 times</span><br><span class="line">    18739 (93.69%) aligned exactly 1 time</span><br><span class="line">    14 (0.07%) aligned &gt;1 times</span><br><span class="line">93.77% overall alignment rate</span><br><span class="line"></span><br><span class="line">## Paired-end</span><br><span class="line">10000 reads; of these:</span><br><span class="line">  10000 (100.00%) were paired; of these:</span><br><span class="line">    650 (6.50%) aligned concordantly 0 times</span><br><span class="line">    8823 (88.23%) aligned concordantly exactly 1 time</span><br><span class="line">    527 (5.27%) aligned concordantly &gt;1 times</span><br><span class="line">    ----</span><br><span class="line">    650 pairs aligned concordantly 0 times; of these:</span><br><span class="line">      34 (5.23%) aligned discordantly 1 time</span><br><span class="line">    ----</span><br><span class="line">    616 pairs aligned 0 times concordantly or discordantly; of these:</span><br><span class="line">      1232 mates make up the pairs; of these:</span><br><span class="line">        660 (53.57%) aligned 0 times</span><br><span class="line">        571 (46.35%) aligned exactly 1 time</span><br><span class="line">        1 (0.08%) aligned &gt;1 times</span><br><span class="line">96.70% overall alignment rate</span><br><span class="line">The indentation indicates how subtotals relate to t</span><br><span class="line"></span><br></pre></td></tr></table></figure>
<h3 id="双端测序比对"><a href="#双端测序比对" class="headerlink" title="双端测序比对"></a>双端测序比对</h3><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">bowtie2 [options]* -x &lt;bt2-idx&gt; -1 &lt;fq1&gt; -2 &lt;fq2&gt; -S &lt;sam_output&gt; -p &lt;threads&gt; 2&gt;Align.summary</span><br></pre></td></tr></table></figure>
<p>双端比对模式基本与单端一致，只需替换fastq文件传入的参数即可<br><code>-1</code>：一链fastq文件<br><code>-2</code>：二链fastq文件</p>
<p>Bowtie2 还有更多详细的比对参数可以调整，这里就不一一介绍了。下面再介绍其输出的SAM文件中各列的含义。</p>
<h2 id="Alignment-OUTPUT"><a href="#Alignment-OUTPUT" class="headerlink" title="Alignment OUTPUT"></a>Alignment OUTPUT</h2><p>比对结果以SAM文件保存。SAM文件的每一行代表一个reads的比对情况，至少包含了12列（tab分割），从左往右，每一列的含义依次为：</p>
<ol>
<li><p>Read的名字</p>
</li>
<li><p>flags之和</p>
<blockquote>
<p>在bowtie2中，flags的含义为<br>1<br>The read is one of a pair<br>2<br>The alignment is one end of a proper paired-end alignment<br>4<br>The read has no reported alignments<br>8<br>The read is one of a pair and has no reported alignments<br>16<br>The alignment is to the reverse reference strand<br>32<br>The other mate in the paired-end alignment is aligned to the reverse reference strand<br>64<br>The read is mate 1 in a pair<br>128<br>The read is mate 2 in a pair<br>注意每个比对软件flags的含义有所区别</p>
</blockquote>
</li>
<li><p>比对到的参考基因组染色体名称</p>
</li>
<li><p>read 5’端比对到的参考基因组正链染色体坐标（1-based）</p>
</li>
<li><p>比对质量</p>
</li>
<li><p>CIGAR字符串，用以表征比对的结果</p>
</li>
<li><p>双端测序中，二链所比对上的染色体名称，如果与一链相同则为<code>=</code>，如果没有二链则为<code>*</code></p>
</li>
<li><p>双端测序中，二链read 5’端比对到的参考基因组正链染色体坐标（1-based），如果没有二链则为<code>0</code></p>
</li>
<li><p>推测的一链与二链之间的片段长度。该值为负表明，二链比对到一链的上游；该值为0表明二链没有比对上；该值为non-0表明二链与一链比对到不同的染色体上（non-0如何理解？）</p>
</li>
<li><p>Read的序列</p>
</li>
<li><p>ASCII 编码的read碱基质量</p>
</li>
<li><p>可选的列，包括以下这些</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><span class="line">AS:i:&lt;N&gt; Alignment score. Can be negative. Can be greater than 0 in --local mode (but not in --end-to-end mode). Only present if SAM record is for an aligned read. </span><br><span class="line">XS:i:&lt;N&gt; Alignment score for the best-scoring alignment found other than the alignment reported. Can be negative. Can be greater than 0 in --local mode (but not in --end-to-end mode). Only present if the SAM record is for an aligned read and more than one alignment was found for the read. Note that, when the read is part of a concordantly-aligned pair, this score could be greater than AS:i. </span><br><span class="line">YS:i:&lt;N&gt; Alignment score for opposite mate in the paired-end alignment. Only present if the SAM record is for a read that aligned as part of a paired-end alignment. </span><br><span class="line">XN:i:&lt;N&gt; The number of ambiguous bases in the reference covering this alignment. Only present if SAM record is for an aligned read. </span><br><span class="line">XM:i:&lt;N&gt; The number of mismatches in the alignment. Only present if SAM record is for an aligned read. </span><br><span class="line">XO:i:&lt;N&gt; The number of gap opens, for both read and reference gaps, in the alignment. Only present if SAM record is for an aligned read. </span><br><span class="line">XG:i:&lt;N&gt; The number of gap extensions, for both read and reference gaps, in the alignment. Only present if SAM record is for an aligned read. </span><br><span class="line">NM:i:&lt;N&gt; The edit distance; that is, the minimal number of one-nucleotide edits (substitutions, insertions and deletions) needed to transform the read string into the reference string. Only present if SAM record is for an aligned read. </span><br><span class="line">YF:Z:&lt;S&gt; String indicating reason why the read was filtered out. See also: Filtering. Only appears for reads that were filtered out. </span><br><span class="line">YT:Z:&lt;S&gt; Value of UU indicates the read was not part of a pair. Value of CP indicates the read was part of a pair and the pair aligned concordantly. Value of DP indicates the read was part of a pair and the pair aligned discordantly. Value of UP indicates the read was part of a pair but the pair failed to aligned either concordantly or discordantly. </span><br><span class="line">MD:Z:&lt;S&gt; A string representation of the mismatched reference bases in the alignm</span><br></pre></td></tr></table></figure></li>
</ol>
<p>以上就是对Bowtie 2进行基因组比对的一些总结，以后有新的心得再做补充。</p>
<blockquote>
<p>Ref:<br><a target="_blank" rel="noopener" href="http://bowtie-bio.sourceforge.net/bowtie2/manual.shtml#how-is-bowtie-2-different-from-bowtie-1">http://bowtie-bio.sourceforge.net/bowtie2/manual.shtml#how-is-bowtie-2-different-from-bowtie-1</a></p>
</blockquote>
<p>完。</p>

    </div>

    
    
    
        

<div>
<ul class="post-copyright">
  <li class="post-copyright-author">
    <strong>Post author:  </strong>Dean Li
  </li>
  <li class="post-copyright-link">
    <strong>Post link: </strong>
    <a href="https://thereallda.github.io/2022/01/24/BOWTIE2-%E8%BF%9B%E8%A1%8C%E5%9F%BA%E5%9B%A0%E7%BB%84%E6%AF%94%E5%AF%B9/" title="BOWTIE2-进行基因组比对">https://thereallda.github.io/2022/01/24/BOWTIE2-进行基因组比对/</a>
  </li>
  <li class="post-copyright-license">
    <strong>Copyright Notice:  </strong>All articles in this blog are licensed under <a href="https://creativecommons.org/licenses/by-nc-sa/4.0/" rel="noopener" target="_blank"><i class="fab fa-fw fa-creative-commons"></i>BY-NC-SA</a> unless stating additionally.
  </li>
</ul>
</div>


      <footer class="post-footer">
          <div class="post-tags">
              <a href="/tags/genome/" rel="tag"># genome</a>
              <a href="/tags/Data-analysis/" rel="tag"># Data_analysis</a>
          </div>

        


        
    <div class="post-nav">
      <div class="post-nav-item">
    <a href="/2022/01/24/%E4%BD%BF%E7%94%A8deeptools%E5%B0%86bam%E6%96%87%E4%BB%B6%E8%BD%AC%E6%8D%A2%E4%B8%BAbw%E6%96%87%E4%BB%B6/" rel="prev" title="使用deeptools将bam文件转换为bw文件">
      <i class="fa fa-chevron-left"></i> 使用deeptools将bam文件转换为bw文件
    </a></div>
      <div class="post-nav-item">
    <a href="/2022/01/24/STAR-%E8%BD%AC%E5%BD%95%E7%BB%84%E6%AF%94%E5%AF%B9/" rel="next" title="STAR-转录组比对">
      STAR-转录组比对 <i class="fa fa-chevron-right"></i>
    </a></div>
    </div>
      </footer>
    
  </article>
  
  
  



          </div>
          
    <div class="comments" id="gitalk-container"></div>

<script>
  window.addEventListener('tabs:register', () => {
    let { activeClass } = CONFIG.comments;
    if (CONFIG.comments.storage) {
      activeClass = localStorage.getItem('comments_active') || activeClass;
    }
    if (activeClass) {
      let activeTab = document.querySelector(`a[href="#comment-${activeClass}"]`);
      if (activeTab) {
        activeTab.click();
      }
    }
  });
  if (CONFIG.comments.storage) {
    window.addEventListener('tabs:click', event => {
      if (!event.target.matches('.tabs-comment .tab-content .tab-pane')) return;
      let commentClass = event.target.classList[1];
      localStorage.setItem('comments_active', commentClass);
    });
  }
</script>

        </div>
          
  
  <div class="toggle sidebar-toggle">
    <span class="toggle-line toggle-line-first"></span>
    <span class="toggle-line toggle-line-middle"></span>
    <span class="toggle-line toggle-line-last"></span>
  </div>

  <aside class="sidebar">
    <div class="sidebar-inner">

      <ul class="sidebar-nav motion-element">
        <li class="sidebar-nav-toc">
          Table of Contents
        </li>
        <li class="sidebar-nav-overview">
          Overview
        </li>
      </ul>

      <!--noindex-->
      <div class="post-toc-wrap sidebar-panel">
          <div class="post-toc motion-element"><ol class="nav"><li class="nav-item nav-level-1"><a class="nav-link" href="#%E7%94%A8%E6%B3%95"><span class="nav-number">1.</span> <span class="nav-text">用法</span></a><ol class="nav-child"><li class="nav-item nav-level-2"><a class="nav-link" href="#Index"><span class="nav-number">1.1.</span> <span class="nav-text">Index</span></a><ol class="nav-child"><li class="nav-item nav-level-3"><a class="nav-link" href="#Index%E4%BE%8B%E5%AD%90"><span class="nav-number">1.1.1.</span> <span class="nav-text">Index例子</span></a></li></ol></li><li class="nav-item nav-level-2"><a class="nav-link" href="#Alignment"><span class="nav-number">1.2.</span> <span class="nav-text">Alignment</span></a><ol class="nav-child"><li class="nav-item nav-level-3"><a class="nav-link" href="#%E5%8D%95%E7%AB%AF%E6%B5%8B%E5%BA%8F%E6%AF%94%E5%AF%B9"><span class="nav-number">1.2.1.</span> <span class="nav-text">单端测序比对</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#%E5%8F%8C%E7%AB%AF%E6%B5%8B%E5%BA%8F%E6%AF%94%E5%AF%B9"><span class="nav-number">1.2.2.</span> <span class="nav-text">双端测序比对</span></a></li></ol></li><li class="nav-item nav-level-2"><a class="nav-link" href="#Alignment-OUTPUT"><span class="nav-number">1.3.</span> <span class="nav-text">Alignment OUTPUT</span></a></li></ol></li></ol></div>
      </div>
      <!--/noindex-->

      <div class="site-overview-wrap sidebar-panel">
        <div class="site-author motion-element" itemprop="author" itemscope itemtype="http://schema.org/Person">
    <img class="site-author-image" itemprop="image" alt="Dean Li"
      src="https://avatars.githubusercontent.com/u/55836943?v=4">
  <p class="site-author-name" itemprop="name">Dean Li</p>
  <div class="site-description" itemprop="description"></div>
</div>
<div class="site-state-wrap motion-element">
  <nav class="site-state">
      <div class="site-state-item site-state-posts">
          <a href="/archives/">
        
          <span class="site-state-item-count">66</span>
          <span class="site-state-item-name">posts</span>
        </a>
      </div>
      <div class="site-state-item site-state-tags">
            <a href="/tags/">
          
        <span class="site-state-item-count">21</span>
        <span class="site-state-item-name">tags</span></a>
      </div>
  </nav>
</div>
  <div class="links-of-author motion-element">
      <span class="links-of-author-item">
        <a href="https://github.com/thereallda" title="GitHub → https:&#x2F;&#x2F;github.com&#x2F;thereallda" rel="noopener" target="_blank"><i class="fab fa-github fa-fw"></i>GitHub</a>
      </span>
      <span class="links-of-author-item">
        <a href="mailto:lideanlda@foxmail.com" title="E-Mail → mailto:lideanlda@foxmail.com" rel="noopener" target="_blank"><i class="fa fa-envelope fa-fw"></i>E-Mail</a>
      </span>
      <span class="links-of-author-item">
        <a href="https://www.jianshu.com/u/ab8a1c7d0510" title="简书 → https:&#x2F;&#x2F;www.jianshu.com&#x2F;u&#x2F;ab8a1c7d0510" rel="noopener" target="_blank"><i class="fa fa-book fa-fw"></i>简书</a>
      </span>
  </div>



      </div>

    </div>
  </aside>
  <div id="sidebar-dimmer"></div>


      </div>
    </main>

    <footer class="footer">
      <div class="footer-inner">
        

        

<div class="copyright">
  
  &copy; 
  <span itemprop="copyrightYear">2025</span>
  <span class="with-love">
    <i class="fa fa-heart"></i>
  </span>
  <span class="author" itemprop="copyrightHolder">Dean Li</span>
</div>
  <div class="powered-by">Powered by <a href="https://hexo.io/" class="theme-link" rel="noopener" target="_blank">Hexo</a> & <a href="https://theme-next.org/" class="theme-link" rel="noopener" target="_blank">NexT.Gemini</a>
  </div>



        
<div class="busuanzi-count">
  <script async src="https://busuanzi.ibruce.info/busuanzi/2.3/busuanzi.pure.mini.js"></script>
    <span class="post-meta-item" id="busuanzi_container_site_uv" style="display: none;">
      <span class="post-meta-item-icon">
        <i class="fa fa-user"></i>
      </span>
      <span class="site-uv" title="Total Visitors">
        <span id="busuanzi_value_site_uv"></span>
      </span>
    </span>
    <span class="post-meta-divider">|</span>
    <span class="post-meta-item" id="busuanzi_container_site_pv" style="display: none;">
      <span class="post-meta-item-icon">
        <i class="fa fa-eye"></i>
      </span>
      <span class="site-pv" title="Total Views">
        <span id="busuanzi_value_site_pv"></span>
      </span>
    </span>
</div>








      </div>
    </footer>
  </div>

  
  <script src="/lib/anime.min.js"></script>
  <script src="/lib/velocity/velocity.min.js"></script>
  <script src="/lib/velocity/velocity.ui.min.js"></script>

<script src="/js/utils.js"></script>

<script src="/js/motion.js"></script>


<script src="/js/schemes/pisces.js"></script>


<script src="/js/next-boot.js"></script>




  











<script>
document.querySelectorAll('.pdfobject-container').forEach(element => {
  let url = element.dataset.target;
  let pdfOpenParams = {
    navpanes : 0,
    toolbar  : 0,
    statusbar: 0,
    pagemode : 'thumbs',
    view     : 'FitH'
  };
  let pdfOpenFragment = '#' + Object.entries(pdfOpenParams).map(([key, value]) => `${key}=${encodeURIComponent(value)}`).join('&');
  let fullURL = `/lib/pdf/web/viewer.html?file=${encodeURIComponent(url)}${pdfOpenFragment}`;

  if (NexT.utils.supportsPDFs()) {
    element.innerHTML = `<embed class="pdfobject" src="${url + pdfOpenFragment}" type="application/pdf" style="height: ${element.dataset.height};">`;
  } else {
    element.innerHTML = `<iframe src="${fullURL}" style="height: ${element.dataset.height};" frameborder="0"></iframe>`;
  }
});
</script>




  

  

  
<script src="https://cdn.jsdelivr.net/npm/darkmode-js@1.5.7/lib/darkmode-js.min.js"></script>

<script>
var options = {
  bottom: '64px',
  right: '32px',
  left: 'unset',
  time: '0.5s',
  mixColor: 'transparent',
  backgroundColor: 'transparent',
  buttonColorDark: '#100f2c',
  buttonColorLight: '#fff',
  saveInCookies: true,
  label: '🌓',
  autoMatchOsTheme: true
}
const darkmode = new Darkmode(options);
window.darkmode = darkmode;
darkmode.showWidget();
</script>

<link rel="stylesheet" href="//cdn.jsdelivr.net/npm/gitalk@1/dist/gitalk.min.css">

<script>
NexT.utils.loadComments(document.querySelector('#gitalk-container'), () => {
  NexT.utils.getScript('//cdn.jsdelivr.net/npm/gitalk@1/dist/gitalk.min.js', () => {
    var gitalk = new Gitalk({
      clientID    : 'aea502b6092de143578a',
      clientSecret: '7bb9f5afb99d67ab37d8b296f39782af05dc8614',
      repo        : 'thereallda.GitHub.io',
      owner       : 'thereallda',
      admin       : ['thereallda'],
      id          : 'f6ff05cf9e3206b7938aa1750c07f4b2',
        language: '',
      distractionFreeMode: true
    });
    gitalk.render('gitalk-container');
  }, window.Gitalk);
});
</script>

</body>
</html>
