<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
<meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=2">
<meta name="theme-color" content="#222">
<meta name="generator" content="Hexo 5.4.0">
  <link rel="apple-touch-icon" sizes="180x180" href="/images/apple-touch-icon.png">
  <link rel="icon" type="image/png" sizes="32x32" href="/images/favicon-32x32.png">
  <link rel="icon" type="image/png" sizes="16x16" href="/images/favicon-16x16.png">

<link rel="stylesheet" href="/css/main.css">


<link rel="stylesheet" href="/lib/font-awesome/css/all.min.css">

<script id="hexo-configurations">
    var NexT = window.NexT || {};
    var CONFIG = {"hostname":"thereallda.github.io","root":"/","scheme":"Gemini","version":"7.8.0","exturl":false,"sidebar":{"position":"left","display":"post","padding":18,"offset":12,"onmobile":false},"copycode":{"enable":true,"show_result":false,"style":"mac"},"back2top":{"enable":true,"sidebar":false,"scrollpercent":false},"bookmark":{"enable":false,"color":"#222","save":"auto"},"fancybox":false,"mediumzoom":false,"lazyload":false,"pangu":false,"comments":{"style":"tabs","active":"gitalk","storage":true,"lazyload":false,"nav":null,"activeClass":"gitalk"},"algolia":{"hits":{"per_page":10},"labels":{"input_placeholder":"Search for Posts","hits_empty":"We didn't find any results for the search: ${query}","hits_stats":"${hits} results found in ${time} ms"}},"localsearch":{"enable":false,"trigger":"auto","top_n_per_article":1,"unescape":false,"preload":false},"motion":{"enable":true,"async":false,"transition":{"post_block":"fadeIn","post_header":"slideDownIn","post_body":"slideDownIn","coll_header":"slideLeftIn","sidebar":"slideUpIn"}}};
  </script>

  <meta name="description" content="学习最好的方式就是分享。  最近看到一个在R上进行的RNA-seq 分析流程，恰好自己也有过RNA-seq分析的经验，所以就想结合以前的经验分享这个流程出来。P.S. RNA-seq  分析有多种流程，本文仅是举出其中一个例子，抛砖引玉。  本文将要介绍的是由Combine Australia所提供的一个针对有参基因组的基因差异表达分析流程。">
<meta property="og:type" content="article">
<meta property="og:title" content="R-用R进行RNA-seq数据分析">
<meta property="og:url" content="https://thereallda.github.io/2022/01/22/R-%E7%94%A8R%E8%BF%9B%E8%A1%8CRNA-seq%E6%95%B0%E6%8D%AE%E5%88%86%E6%9E%90/index.html">
<meta property="og:site_name" content="Dean&#39;s blog">
<meta property="og:description" content="学习最好的方式就是分享。  最近看到一个在R上进行的RNA-seq 分析流程，恰好自己也有过RNA-seq分析的经验，所以就想结合以前的经验分享这个流程出来。P.S. RNA-seq  分析有多种流程，本文仅是举出其中一个例子，抛砖引玉。  本文将要介绍的是由Combine Australia所提供的一个针对有参基因组的基因差异表达分析流程。">
<meta property="og:locale" content="en_US">
<meta property="og:image" content="https://thereallda.github.io/2022/01/22/R-%E7%94%A8R%E8%BF%9B%E8%A1%8CRNA-seq%E6%95%B0%E6%8D%AE%E5%88%86%E6%9E%90/image-20220122200228135.png">
<meta property="og:image" content="https://thereallda.github.io/2022/01/22/R-%E7%94%A8R%E8%BF%9B%E8%A1%8CRNA-seq%E6%95%B0%E6%8D%AE%E5%88%86%E6%9E%90/image-20220122200240029.png">
<meta property="og:image" content="https://thereallda.github.io/2022/01/22/R-%E7%94%A8R%E8%BF%9B%E8%A1%8CRNA-seq%E6%95%B0%E6%8D%AE%E5%88%86%E6%9E%90/image-20220122200248845.png">
<meta property="og:image" content="https://thereallda.github.io/2022/01/22/R-%E7%94%A8R%E8%BF%9B%E8%A1%8CRNA-seq%E6%95%B0%E6%8D%AE%E5%88%86%E6%9E%90/image-20220122200312506.png">
<meta property="og:image" content="https://thereallda.github.io/2022/01/22/R-%E7%94%A8R%E8%BF%9B%E8%A1%8CRNA-seq%E6%95%B0%E6%8D%AE%E5%88%86%E6%9E%90/image-20220122200322356.png">
<meta property="og:image" content="https://thereallda.github.io/2022/01/22/R-%E7%94%A8R%E8%BF%9B%E8%A1%8CRNA-seq%E6%95%B0%E6%8D%AE%E5%88%86%E6%9E%90/image-20220122200348849.png">
<meta property="og:image" content="https://thereallda.github.io/2022/01/22/R-%E7%94%A8R%E8%BF%9B%E8%A1%8CRNA-seq%E6%95%B0%E6%8D%AE%E5%88%86%E6%9E%90/image-20220122200412539.png">
<meta property="og:image" content="https://thereallda.github.io/2022/01/22/R-%E7%94%A8R%E8%BF%9B%E8%A1%8CRNA-seq%E6%95%B0%E6%8D%AE%E5%88%86%E6%9E%90/image-20220122200426544.png">
<meta property="og:image" content="https://thereallda.github.io/2022/01/22/R-%E7%94%A8R%E8%BF%9B%E8%A1%8CRNA-seq%E6%95%B0%E6%8D%AE%E5%88%86%E6%9E%90/image-20220122200445711.png">
<meta property="og:image" content="https://thereallda.github.io/2022/01/22/R-%E7%94%A8R%E8%BF%9B%E8%A1%8CRNA-seq%E6%95%B0%E6%8D%AE%E5%88%86%E6%9E%90/image-20220122200502891.png">
<meta property="article:published_time" content="2022-01-23T00:38:12.000Z">
<meta property="article:modified_time" content="2022-01-22T14:03:20.951Z">
<meta property="article:author" content="Dean Li">
<meta property="article:tag" content="Data_analysis">
<meta property="article:tag" content="R">
<meta property="article:tag" content="RNA-seq">
<meta name="twitter:card" content="summary">
<meta name="twitter:image" content="https://thereallda.github.io/2022/01/22/R-%E7%94%A8R%E8%BF%9B%E8%A1%8CRNA-seq%E6%95%B0%E6%8D%AE%E5%88%86%E6%9E%90/image-20220122200228135.png">

<link rel="canonical" href="https://thereallda.github.io/2022/01/22/R-%E7%94%A8R%E8%BF%9B%E8%A1%8CRNA-seq%E6%95%B0%E6%8D%AE%E5%88%86%E6%9E%90/">


<script id="page-configurations">
  // https://hexo.io/docs/variables.html
  CONFIG.page = {
    sidebar: "",
    isHome : false,
    isPost : true,
    lang   : 'en'
  };
</script>

  <title>R-用R进行RNA-seq数据分析 | Dean's blog</title>
  






  <noscript>
  <style>
  .use-motion .brand,
  .use-motion .menu-item,
  .sidebar-inner,
  .use-motion .post-block,
  .use-motion .pagination,
  .use-motion .comments,
  .use-motion .post-header,
  .use-motion .post-body,
  .use-motion .collection-header { opacity: initial; }

  .use-motion .site-title,
  .use-motion .site-subtitle {
    opacity: initial;
    top: initial;
  }

  .use-motion .logo-line-before i { left: initial; }
  .use-motion .logo-line-after i { right: initial; }
  </style>
</noscript>

<style>.darkmode--activated{--body-bg-color:#282828;--content-bg-color:#333;--card-bg-color:#555;--text-color:#ccc;--blockquote-color:#bbb;--link-color:#ccc;--link-hover-color:#eee;--brand-color:#ddd;--brand-hover-color:#ddd;--table-row-odd-bg-color:#282828;--table-row-hover-bg-color:#363636;--menu-item-bg-color:#555;--btn-default-bg:#222;--btn-default-color:#ccc;--btn-default-border-color:#555;--btn-default-hover-bg:#666;--btn-default-hover-color:#ccc;--btn-default-hover-border-color:#666;--highlight-background:#282b2e;--highlight-foreground:#a9b7c6;--highlight-gutter-background:#34393d;--highlight-gutter-foreground:#9ca9b6}.darkmode--activated img{opacity:.75}.darkmode--activated img:hover{opacity:.9}.darkmode--activated code{color:#69dbdc;background:0 0}button.darkmode-toggle{z-index:9999}.darkmode-ignore,img{display:flex!important}</style></head>

<body itemscope itemtype="http://schema.org/WebPage">
  <div class="container use-motion">
    <div class="headband"></div>

    <header class="header" itemscope itemtype="http://schema.org/WPHeader">
      <div class="header-inner"><div class="site-brand-container">
  <div class="site-nav-toggle">
    <div class="toggle" aria-label="Toggle navigation bar">
      <span class="toggle-line toggle-line-first"></span>
      <span class="toggle-line toggle-line-middle"></span>
      <span class="toggle-line toggle-line-last"></span>
    </div>
  </div>

  <div class="site-meta">

    <a href="/" class="brand" rel="start">
      <span class="logo-line-before"><i></i></span>
      <h1 class="site-title">Dean's blog</h1>
      <span class="logo-line-after"><i></i></span>
    </a>
  </div>

  <div class="site-nav-right">
    <div class="toggle popup-trigger">
    </div>
  </div>
</div>




<nav class="site-nav">
  <ul id="menu" class="main-menu menu">
        <li class="menu-item menu-item-home">

    <a href="/" rel="section"><i class="fa fa-home fa-fw"></i>Home</a>

  </li>
        <li class="menu-item menu-item-about">

    <a href="/about/" rel="section"><i class="fa fa-user fa-fw"></i>About</a>

  </li>
        <li class="menu-item menu-item-tags">

    <a href="/tags/" rel="section"><i class="fa fa-tags fa-fw"></i>Tags</a>

  </li>
        <li class="menu-item menu-item-archives">

    <a href="/archives/" rel="section"><i class="fa fa-archive fa-fw"></i>Archives</a>

  </li>
  </ul>
</nav>




</div>
    </header>

    
  <div class="back-to-top">
    <i class="fa fa-arrow-up"></i>
    <span>0%</span>
  </div>
  <div class="reading-progress-bar"></div>


    <main class="main">
      <div class="main-inner">
        <div class="content-wrap">
          

          <div class="content post posts-expand">
            

    
  
  
  <article itemscope itemtype="http://schema.org/Article" class="post-block" lang="en">
    <link itemprop="mainEntityOfPage" href="https://thereallda.github.io/2022/01/22/R-%E7%94%A8R%E8%BF%9B%E8%A1%8CRNA-seq%E6%95%B0%E6%8D%AE%E5%88%86%E6%9E%90/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="image" content="https://avatars.githubusercontent.com/u/55836943?v=4">
      <meta itemprop="name" content="Dean Li">
      <meta itemprop="description" content="">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="Dean's blog">
    </span>
      <header class="post-header">
        <h1 class="post-title" itemprop="name headline">
          R-用R进行RNA-seq数据分析
        </h1>

        <div class="post-meta">
          
          
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-calendar"></i>
              </span>
              <span class="post-meta-item-text">Posted on</span>
              

              <time title="Created: 2022-01-22 19:38:12 / Modified: 09:03:20" itemprop="dateCreated datePublished" datetime="2022-01-22T19:38:12-05:00">2022-01-22</time>
            </span>

          
            <span class="post-meta-item" title="Views" id="busuanzi_container_page_pv" style="display: none;">
              <span class="post-meta-item-icon">
                <i class="fa fa-eye"></i>
              </span>
              <span class="post-meta-item-text">Views: </span>
              <span id="busuanzi_value_page_pv"></span>
            </span><br>
            <span class="post-meta-item" title="Symbols count in article">
              <span class="post-meta-item-icon">
                <i class="far fa-file-word"></i>
              </span>
                <span class="post-meta-item-text">Symbols count in article: </span>
              <span>20k</span>
            </span>
            <span class="post-meta-item" title="Reading time">
              <span class="post-meta-item-icon">
                <i class="far fa-clock"></i>
              </span>
                <span class="post-meta-item-text">Reading time &asymp;</span>
              <span>18 mins.</span>
            </span>

        </div>
      </header>

    
    
    
    <div class="post-body" itemprop="articleBody">

      
        <blockquote>
<p>学习最好的方式就是分享。</p>
</blockquote>
<p>最近看到一个在R上进行的RNA-seq 分析流程，恰好自己也有过RNA-seq分析的经验，所以就想结合以前的经验分享这个流程出来。<br><strong>P.S. RNA-seq  分析有多种流程，本文仅是举出其中一个例子，抛砖引玉。</strong></p>
<p><img src="/2022/01/22/R-%E7%94%A8R%E8%BF%9B%E8%A1%8CRNA-seq%E6%95%B0%E6%8D%AE%E5%88%86%E6%9E%90/image-20220122200228135.png"></p>
<p>本文将要介绍的是由<a target="_blank" rel="noopener" href="https://bioinformatics-core-shared-training.github.io/RNAseq-R/">Combine Australia</a>所提供的一个针对有参基因组的基因差异表达分析流程。</p>
<span id="more"></span>

<h1 id="分析流程的基本信息"><a href="#分析流程的基本信息" class="headerlink" title="分析流程的基本信息"></a>分析流程的基本信息</h1><ul>
<li>  目标物种 : <strong>mouse</strong>（小鼠）</li>
<li>  组织 : <strong>mammary gland</strong>（乳腺）</li>
<li>生理状态:<ul>
<li>  <strong>virgin</strong>（未怀孕）</li>
<li>  <strong>pregnant</strong>（怀孕）</li>
<li> <strong>lactating</strong> （产后哺乳期）</li>
</ul>
</li>
<li>细胞类型:<ul>
<li>  <strong>basal stem-cell enriched cells</strong> (B) </li>
<li>  <strong>committed luminal cells</strong> (L) </li>
</ul>
</li>
<li>  Six groups (3 conditions x 2 cell types) with 2 biological replicates per group</li>
<li>该文章已经发表在<strong>Nature cell biology</strong>:<ul>
<li>  <a target="_blank" rel="noopener" href="https://www.ncbi.nlm.nih.gov/pubmed/25730472">‘EGF-mediated induction of Mcl-1 at the switch to lactation is essential for alveolar cell survival’ (Fu et al. 2015)</a>published in Nature Cell Biology, with both sequence and counts available from Gene Expression Omnibus database (GEO) under accession number <a target="_blank" rel="noopener" href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE60450">GSE60450</a></li>
</ul>
</li>
</ul>
<h1 id="流程"><a href="#流程" class="headerlink" title="流程"></a>流程</h1><ul>
<li><p> <strong>Step 1:</strong> Alignment and Counting (该部分需要在linux环境或mac os系统中完成)</p>
</li>
<li><p> <strong>Step 2:</strong> <a target="_blank" rel="noopener" href="https://www.jianshu.com/p/e18aebd145d0">RNA-seq Pre-processing</a></p>
</li>
<li><p><strong>Step 3:</strong> <a target="_blank" rel="noopener" href="https://www.jianshu.com/p/1669a5e3a23a">Differential Expression for RNA-seq</a></p>
</li>
<li><p><strong>Step 4:</strong>  <a target="_blank" rel="noopener" href="https://www.jianshu.com/p/9ef0f62a6598">Annotation and Visualisation of RNA-seq results</a></p>
</li>
<li><p><strong>Step 5:</strong>  <a target="_blank" rel="noopener" href="https://www.jianshu.com/p/c2f12500c5c1">Gene-set testing</a></p>
</li>
</ul>
<h1 id="数据"><a href="#数据" class="headerlink" title="数据"></a>数据</h1><p>感兴趣的朋友也可以下载本次分析所需要的数据集，跟着动手尝试：<a target="_blank" rel="noopener" href="https://figshare.com/s/1d788fd384d33e913a2a">https://figshare.com/s/1d788fd384d33e913a2a</a></p>
<h1 id="配置"><a href="#配置" class="headerlink" title="配置"></a>配置</h1><ul>
<li>安装了R语言（3.5.0以上）及R studio（可选）</li>
<li> 需要的R包： <code>limma</code>, <code>edgeR</code>, <code>gplots</code>, <code>org.Mm.eg.db</code>, <code>EDASeq</code>, <code>RColorBrewer</code>, <code>GO.db</code>, <code>BiasedUrn</code>, <code>DESeq2</code>, <code>Glimma</code>, <code>Rsubread</code>.</li>
</ul>
<p>这里我们推荐使用<code>Bioconductor</code>安装以上这些包</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">BiocManager::install(<span class="built_in">c</span>(<span class="string">&#x27;limma&#x27;</span>, <span class="string">&#x27;edgeR&#x27;</span>, <span class="string">&#x27;gplots&#x27;</span>, <span class="string">&#x27;org.Mm.eg.db&#x27;</span>, <span class="string">&#x27;EDASeq&#x27;</span>, <span class="string">&#x27;RColorBrewer&#x27;</span>, <span class="string">&#x27;GO.db&#x27;</span>, <span class="string">&#x27;BiasedUrn&#x27;</span>, <span class="string">&#x27;DESeq2&#x27;</span>, <span class="string">&#x27;Rsubread&#x27;</span>))</span><br></pre></td></tr></table></figure>



<hr>
<p><strong>这里要注意<code>Glimma</code>包需要通过在<a target="_blank" rel="noopener" href="https://www.bioconductor.org/packages/release/bioc/html/Glimma.html">Bioconductor</a>中下载其源文件来安装</strong><br><img src="/2022/01/22/R-%E7%94%A8R%E8%BF%9B%E8%A1%8CRNA-seq%E6%95%B0%E6%8D%AE%E5%88%86%E6%9E%90/image-20220122200240029.png"></p>
<h2 id="源文件的安装方法之一"><a href="#源文件的安装方法之一" class="headerlink" title="源文件的安装方法之一"></a>源文件的安装方法之一</h2><p>源文件可以在R studio中安装，在上方工具栏中选择**”Tools –&gt; Install packages”**。</p>
<p>然后，在弹出的窗口中点击**”Install from”<strong>下拉菜单中选择</strong>Package Archive File**。</p>
<p><img src="/2022/01/22/R-%E7%94%A8R%E8%BF%9B%E8%A1%8CRNA-seq%E6%95%B0%E6%8D%AE%E5%88%86%E6%9E%90/image-20220122200248845.png"></p>
<p>最后，选择下载的源文件包即可安装。</p>
<h1 id="Step-2-RNA-seq-Pre-processing"><a href="#Step-2-RNA-seq-Pre-processing" class="headerlink" title="Step 2 RNA-seq Pre-processing"></a>Step 2 RNA-seq Pre-processing</h1><blockquote>
<p>Ref: <a target="_blank" rel="noopener" href="https://bioinformatics-core-shared-training.github.io/RNAseq-R/rna-seq-preprocessing.nb.html">https://bioinformatics-core-shared-training.github.io/RNAseq-R/rna-seq-preprocessing.nb.html</a></p>
</blockquote>
<p>本节将要介绍的是在R中进行RNA-seq数据预处理的实战代码</p>
<p>主要包括以下方面：</p>
<ol>
<li>载入mapping, counting后的数据</li>
<li>过滤低表达基因</li>
<li>对表达数据进行质量控制</li>
<li>标准化处理</li>
</ol>
<p><strong>本次流程需要的包</strong></p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">library(edgeR)</span><br><span class="line">library(limma)</span><br><span class="line">library(Glimma)</span><br><span class="line">library(gplots)</span><br><span class="line">library(org.Mm.eg.db)</span><br><span class="line">library(RColorBrewer)</span><br></pre></td></tr></table></figure>

<blockquote>
<p>温馨提示：<code>org.Mm.eg.db</code>有60多M，个人推荐没有安装过的朋友先切换到国内的镜像，再下载这个包。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">options(repos = &#x27;https://mirrors.tuna.tsinghua.edu.cn/CRAN/&#x27;)</span><br><span class="line">getOption(&#x27;repos&#x27;)</span><br><span class="line">BiocManager::install(&#x27;org.Mm.eg.db&#x27;)</span><br></pre></td></tr></table></figure>
</blockquote>
<p>另外，本次所需要的数据均可以在该网站下载：<a target="_blank" rel="noopener" href="https://figshare.com/s/1d788fd384d33e913a2a">https://figshare.com/s/1d788fd384d33e913a2a</a></p>
<p>下载的数据有：</p>
<ul>
<li>Sampleinfo.txt</li>
<li>GSE60450_Lactation-GenewiseCounts.txt</li>
<li>mouse_c2_v5.rdata</li>
<li>mouse_H_v5.rdata</li>
</ul>
<h2 id="载入数据"><a href="#载入数据" class="headerlink" title="载入数据"></a>载入数据</h2><p>首先，我们读入样本信息**”SampleInfo.txt”**</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br></pre></td><td class="code"><pre><span class="line">&gt; sampleinfo &lt;- read.delim(<span class="string">&quot;data/SampleInfo.txt&quot;</span>)</span><br><span class="line">&gt; sampleinfo</span><br><span class="line">                            FileName SampleName CellType   Status</span><br><span class="line"><span class="number">1</span>  MCL1.DG_BC2CTUACXX_ACTTGA_L002_R1    MCL1.DG  luminal   virgin</span><br><span class="line"><span class="number">2</span>  MCL1.DH_BC2CTUACXX_CAGATC_L002_R1    MCL1.DH    basal   virgin</span><br><span class="line"><span class="number">3</span>  MCL1.DI_BC2CTUACXX_ACAGTG_L002_R1    MCL1.DI    basal pregnant</span><br><span class="line"><span class="number">4</span>  MCL1.DJ_BC2CTUACXX_CGATGT_L002_R1    MCL1.DJ    basal pregnant</span><br><span class="line"><span class="number">5</span>  MCL1.DK_BC2CTUACXX_TTAGGC_L002_R1    MCL1.DK    basal  lactate</span><br><span class="line"><span class="number">6</span>  MCL1.DL_BC2CTUACXX_ATCACG_L002_R1    MCL1.DL    basal  lactate</span><br><span class="line"><span class="number">7</span>  MCL1.LA_BC2CTUACXX_GATCAG_L001_R1    MCL1.LA    basal   virgin</span><br><span class="line"><span class="number">8</span>  MCL1.LB_BC2CTUACXX_TGACCA_L001_R1    MCL1.LB  luminal   virgin</span><br><span class="line"><span class="number">9</span>  MCL1.LC_BC2CTUACXX_GCCAAT_L001_R1    MCL1.LC  luminal pregnant</span><br><span class="line"><span class="number">10</span> MCL1.LD_BC2CTUACXX_GGCTAC_L001_R1    MCL1.LD  luminal pregnant</span><br><span class="line"><span class="number">11</span> MCL1.LE_BC2CTUACXX_TAGCTT_L001_R1    MCL1.LE  luminal  lactate</span><br><span class="line"><span class="number">12</span> MCL1.LF_BC2CTUACXX_CTTGTA_L001_R1    MCL1.LF  luminal  lactate</span><br></pre></td></tr></table></figure>
<p>再读入基因计数后的数据**”GSE60450_Lactation-GenewiseCounts.txt”**</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br></pre></td><td class="code"><pre><span class="line">&gt; seqdata &lt;- read.delim(<span class="string">&quot;data/GSE60450_Lactation-GenewiseCounts.txt&quot;</span>,</span><br><span class="line">+                       stringsAsFactors = <span class="literal">FALSE</span>)</span><br><span class="line">&gt; head(seqdata)</span><br><span class="line">  EntrezGeneID Length MCL1.DG_BC2CTUACXX_ACTTGA_L002_R1 MCL1.DH_BC2CTUACXX_CAGATC_L002_R1</span><br><span class="line"><span class="number">1</span>       <span class="number">497097</span>   <span class="number">3634</span>                               <span class="number">438</span>                               <span class="number">300</span></span><br><span class="line"><span class="number">2</span>    <span class="number">100503874</span>   <span class="number">3259</span>                                 <span class="number">1</span>                                 <span class="number">0</span></span><br><span class="line"><span class="number">3</span>    <span class="number">100038431</span>   <span class="number">1634</span>                                 <span class="number">0</span>                                 <span class="number">0</span></span><br><span class="line"><span class="number">4</span>        <span class="number">19888</span>   <span class="number">9747</span>                                 <span class="number">1</span>                                 <span class="number">1</span></span><br><span class="line"><span class="number">5</span>        <span class="number">20671</span>   <span class="number">3130</span>                               <span class="number">106</span>                               <span class="number">182</span></span><br><span class="line"><span class="number">6</span>        <span class="number">27395</span>   <span class="number">4203</span>                               <span class="number">309</span>                               <span class="number">234</span></span><br><span class="line">  MCL1.DI_BC2CTUACXX_ACAGTG_L002_R1 MCL1.DJ_BC2CTUACXX_CGATGT_L002_R1</span><br><span class="line"><span class="number">1</span>                                <span class="number">65</span>                               <span class="number">237</span></span><br><span class="line"><span class="number">2</span>                                 <span class="number">1</span>                                 <span class="number">1</span></span><br><span class="line"><span class="number">3</span>                                 <span class="number">0</span>                                 <span class="number">0</span></span><br><span class="line"><span class="number">4</span>                                 <span class="number">0</span>                                 <span class="number">0</span></span><br><span class="line"><span class="number">5</span>                                <span class="number">82</span>                               <span class="number">105</span></span><br><span class="line"><span class="number">6</span>                               <span class="number">337</span>                               <span class="number">300</span></span><br><span class="line">  MCL1.DK_BC2CTUACXX_TTAGGC_L002_R1 MCL1.DL_BC2CTUACXX_ATCACG_L002_R1</span><br><span class="line"><span class="number">1</span>                               <span class="number">354</span>                               <span class="number">287</span></span><br><span class="line"><span class="number">2</span>                                 <span class="number">0</span>                                 <span class="number">4</span></span><br><span class="line"><span class="number">3</span>                                 <span class="number">0</span>                                 <span class="number">0</span></span><br><span class="line"><span class="number">4</span>                                 <span class="number">0</span>                                 <span class="number">0</span></span><br><span class="line"><span class="number">5</span>                                <span class="number">43</span>                                <span class="number">82</span></span><br><span class="line"><span class="number">6</span>                               <span class="number">290</span>                               <span class="number">270</span></span><br><span class="line">  MCL1.LA_BC2CTUACXX_GATCAG_L001_R1 MCL1.LB_BC2CTUACXX_TGACCA_L001_R1</span><br><span class="line"><span class="number">1</span>                                 <span class="number">0</span>                                 <span class="number">0</span></span><br><span class="line"><span class="number">2</span>                                 <span class="number">0</span>                                 <span class="number">0</span></span><br><span class="line"><span class="number">3</span>                                 <span class="number">0</span>                                 <span class="number">0</span></span><br><span class="line"><span class="number">4</span>                                <span class="number">10</span>                                 <span class="number">3</span></span><br><span class="line"><span class="number">5</span>                                <span class="number">16</span>                                <span class="number">25</span></span><br><span class="line"><span class="number">6</span>                               <span class="number">560</span>                               <span class="number">464</span></span><br><span class="line">  MCL1.LC_BC2CTUACXX_GCCAAT_L001_R1 MCL1.LD_BC2CTUACXX_GGCTAC_L001_R1</span><br><span class="line"><span class="number">1</span>                                 <span class="number">0</span>                                 <span class="number">0</span></span><br><span class="line"><span class="number">2</span>                                 <span class="number">0</span>                                 <span class="number">0</span></span><br><span class="line"><span class="number">3</span>                                 <span class="number">0</span>                                 <span class="number">0</span></span><br><span class="line"><span class="number">4</span>                                <span class="number">10</span>                                 <span class="number">2</span></span><br><span class="line"><span class="number">5</span>                                <span class="number">18</span>                                 <span class="number">8</span></span><br><span class="line"><span class="number">6</span>                               <span class="number">489</span>                               <span class="number">328</span></span><br><span class="line">  MCL1.LE_BC2CTUACXX_TAGCTT_L001_R1 MCL1.LF_BC2CTUACXX_CTTGTA_L001_R1</span><br><span class="line"><span class="number">1</span>                                 <span class="number">0</span>                                 <span class="number">0</span></span><br><span class="line"><span class="number">2</span>                                 <span class="number">0</span>                                 <span class="number">0</span></span><br><span class="line"><span class="number">3</span>                                 <span class="number">0</span>                                 <span class="number">0</span></span><br><span class="line"><span class="number">4</span>                                 <span class="number">0</span>                                 <span class="number">0</span></span><br><span class="line"><span class="number">5</span>                                 <span class="number">3</span>                                <span class="number">10</span></span><br><span class="line"><span class="number">6</span>                               <span class="number">307</span>                               <span class="number">342</span></span><br><span class="line">&gt; <span class="built_in">dim</span>(seqdata)   <span class="comment"># 数据包含了27179行的基因信息，1列ID、1列基因长度和12列的样本数据</span></span><br><span class="line">[<span class="number">1</span>] <span class="number">27179</span>    <span class="number">14</span></span><br></pre></td></tr></table></figure>
<p>由于我们关注的是基因计数的信息，因此为了方便下游分析处理，我们移除<code>seqdata</code>中前两列，并将第一列的ID命名为行名</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br></pre></td><td class="code"><pre><span class="line">&gt; countdata &lt;- seqdata[,-(<span class="number">1</span>:<span class="number">2</span>)]</span><br><span class="line">&gt; rownames(countdata) &lt;- seqdata[,<span class="number">1</span>]</span><br><span class="line">&gt; head(countdata)</span><br><span class="line">          MCL1.DG_BC2CTUACXX_ACTTGA_L002_R1 MCL1.DH_BC2CTUACXX_CAGATC_L002_R1</span><br><span class="line"><span class="number">497097</span>                                  <span class="number">438</span>                               <span class="number">300</span></span><br><span class="line"><span class="number">100503874</span>                                 <span class="number">1</span>                                 <span class="number">0</span></span><br><span class="line"><span class="number">100038431</span>                                 <span class="number">0</span>                                 <span class="number">0</span></span><br><span class="line"><span class="number">19888</span>                                     <span class="number">1</span>                                 <span class="number">1</span></span><br><span class="line"><span class="number">20671</span>                                   <span class="number">106</span>                               <span class="number">182</span></span><br><span class="line"><span class="number">27395</span>                                   <span class="number">309</span>                               <span class="number">234</span></span><br><span class="line">          MCL1.DI_BC2CTUACXX_ACAGTG_L002_R1 MCL1.DJ_BC2CTUACXX_CGATGT_L002_R1</span><br><span class="line"><span class="number">497097</span>                                   <span class="number">65</span>                               <span class="number">237</span></span><br><span class="line"><span class="number">100503874</span>                                 <span class="number">1</span>                                 <span class="number">1</span></span><br><span class="line"><span class="number">100038431</span>                                 <span class="number">0</span>                                 <span class="number">0</span></span><br><span class="line"><span class="number">19888</span>                                     <span class="number">0</span>                                 <span class="number">0</span></span><br><span class="line"><span class="number">20671</span>                                    <span class="number">82</span>                               <span class="number">105</span></span><br><span class="line"><span class="number">27395</span>                                   <span class="number">337</span>                               <span class="number">300</span></span><br><span class="line">          MCL1.DK_BC2CTUACXX_TTAGGC_L002_R1 MCL1.DL_BC2CTUACXX_ATCACG_L002_R1</span><br><span class="line"><span class="number">497097</span>                                  <span class="number">354</span>                               <span class="number">287</span></span><br><span class="line"><span class="number">100503874</span>                                 <span class="number">0</span>                                 <span class="number">4</span></span><br><span class="line"><span class="number">100038431</span>                                 <span class="number">0</span>                                 <span class="number">0</span></span><br><span class="line"><span class="number">19888</span>                                     <span class="number">0</span>                                 <span class="number">0</span></span><br><span class="line"><span class="number">20671</span>                                    <span class="number">43</span>                                <span class="number">82</span></span><br><span class="line"><span class="number">27395</span>                                   <span class="number">290</span>                               <span class="number">270</span></span><br><span class="line">          MCL1.LA_BC2CTUACXX_GATCAG_L001_R1 MCL1.LB_BC2CTUACXX_TGACCA_L001_R1</span><br><span class="line"><span class="number">497097</span>                                    <span class="number">0</span>                                 <span class="number">0</span></span><br><span class="line"><span class="number">100503874</span>                                 <span class="number">0</span>                                 <span class="number">0</span></span><br><span class="line"><span class="number">100038431</span>                                 <span class="number">0</span>                                 <span class="number">0</span></span><br><span class="line"><span class="number">19888</span>                                    <span class="number">10</span>                                 <span class="number">3</span></span><br><span class="line"><span class="number">20671</span>                                    <span class="number">16</span>                                <span class="number">25</span></span><br><span class="line"><span class="number">27395</span>                                   <span class="number">560</span>                               <span class="number">464</span></span><br><span class="line">          MCL1.LC_BC2CTUACXX_GCCAAT_L001_R1 MCL1.LD_BC2CTUACXX_GGCTAC_L001_R1</span><br><span class="line"><span class="number">497097</span>                                    <span class="number">0</span>                                 <span class="number">0</span></span><br><span class="line"><span class="number">100503874</span>                                 <span class="number">0</span>                                 <span class="number">0</span></span><br><span class="line"><span class="number">100038431</span>                                 <span class="number">0</span>                                 <span class="number">0</span></span><br><span class="line"><span class="number">19888</span>                                    <span class="number">10</span>                                 <span class="number">2</span></span><br><span class="line"><span class="number">20671</span>                                    <span class="number">18</span>                                 <span class="number">8</span></span><br><span class="line"><span class="number">27395</span>                                   <span class="number">489</span>                               <span class="number">328</span></span><br><span class="line">          MCL1.LE_BC2CTUACXX_TAGCTT_L001_R1 MCL1.LF_BC2CTUACXX_CTTGTA_L001_R1</span><br><span class="line"><span class="number">497097</span>                                    <span class="number">0</span>                                 <span class="number">0</span></span><br><span class="line"><span class="number">100503874</span>                                 <span class="number">0</span>                                 <span class="number">0</span></span><br><span class="line"><span class="number">100038431</span>                                 <span class="number">0</span>                                 <span class="number">0</span></span><br><span class="line"><span class="number">19888</span>                                     <span class="number">0</span>                                 <span class="number">0</span></span><br><span class="line"><span class="number">20671</span>                                     <span class="number">3</span>                                <span class="number">10</span></span><br><span class="line"><span class="number">27395</span>                                   <span class="number">307</span>                               <span class="number">342</span></span><br></pre></td></tr></table></figure>

<p>另外，我们注意到<code>countdata</code>中的列名过于冗长，而样本信息中的名字也只是前7位字母而已，因此我们使用<code>substr()</code>函数截取列名</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br></pre></td><td class="code"><pre><span class="line">&gt; sampleinfo$SampleName</span><br><span class="line"> [<span class="number">1</span>] MCL1.DG MCL1.DH MCL1.DI MCL1.DJ MCL1.DK MCL1.DL MCL1.LA MCL1.LB MCL1.LC MCL1.LD MCL1.LE MCL1.LF</span><br><span class="line"><span class="number">12</span> Levels: MCL1.DG MCL1.DH MCL1.DI MCL1.DJ MCL1.DK MCL1.DL MCL1.LA MCL1.LB MCL1.LC ... MCL1.LF</span><br><span class="line"></span><br><span class="line">&gt; colnames(countdata)</span><br><span class="line"> [<span class="number">1</span>] <span class="string">&quot;MCL1.DG_BC2CTUACXX_ACTTGA_L002_R1&quot;</span> <span class="string">&quot;MCL1.DH_BC2CTUACXX_CAGATC_L002_R1&quot;</span></span><br><span class="line"> [<span class="number">3</span>] <span class="string">&quot;MCL1.DI_BC2CTUACXX_ACAGTG_L002_R1&quot;</span> <span class="string">&quot;MCL1.DJ_BC2CTUACXX_CGATGT_L002_R1&quot;</span></span><br><span class="line"> [<span class="number">5</span>] <span class="string">&quot;MCL1.DK_BC2CTUACXX_TTAGGC_L002_R1&quot;</span> <span class="string">&quot;MCL1.DL_BC2CTUACXX_ATCACG_L002_R1&quot;</span></span><br><span class="line"> [<span class="number">7</span>] <span class="string">&quot;MCL1.LA_BC2CTUACXX_GATCAG_L001_R1&quot;</span> <span class="string">&quot;MCL1.LB_BC2CTUACXX_TGACCA_L001_R1&quot;</span></span><br><span class="line"> [<span class="number">9</span>] <span class="string">&quot;MCL1.LC_BC2CTUACXX_GCCAAT_L001_R1&quot;</span> <span class="string">&quot;MCL1.LD_BC2CTUACXX_GGCTAC_L001_R1&quot;</span></span><br><span class="line">[<span class="number">11</span>] <span class="string">&quot;MCL1.LE_BC2CTUACXX_TAGCTT_L001_R1&quot;</span> <span class="string">&quot;MCL1.LF_BC2CTUACXX_CTTGTA_L001_R1&quot;</span></span><br><span class="line">&gt; colnames(countdata) &lt;- substr(colnames(countdata), <span class="number">1</span>, <span class="number">7</span>)</span><br><span class="line">&gt; colnames(countdata)</span><br><span class="line"> [<span class="number">1</span>] <span class="string">&quot;MCL1.DG&quot;</span> <span class="string">&quot;MCL1.DH&quot;</span> <span class="string">&quot;MCL1.DI&quot;</span> <span class="string">&quot;MCL1.DJ&quot;</span> <span class="string">&quot;MCL1.DK&quot;</span> <span class="string">&quot;MCL1.DL&quot;</span> <span class="string">&quot;MCL1.LA&quot;</span> <span class="string">&quot;MCL1.LB&quot;</span> <span class="string">&quot;MCL1.LC&quot;</span></span><br><span class="line">[<span class="number">10</span>] <span class="string">&quot;MCL1.LD&quot;</span> <span class="string">&quot;MCL1.LE&quot;</span> <span class="string">&quot;MCL1.LF&quot;</span></span><br><span class="line">&gt; table(colnames(countdata)==sampleinfo$SampleName)  <span class="comment">#使用table函数检验修改后的名字与原名字是否相同</span></span><br><span class="line"></span><br><span class="line"><span class="literal">TRUE</span> </span><br><span class="line">  <span class="number">12</span> </span><br></pre></td></tr></table></figure>

<h2 id="数据过滤"><a href="#数据过滤" class="headerlink" title="数据过滤"></a>数据过滤</h2><p>对于表达量过低的基因而言，其数据在差异分析中的可信度是较差的。因此，在本流程中我们先使用<code>edgeR</code>的<code>cpm()</code>函数，将counting后的数据转换为CPM（counts-per-million），并取CPM在两个重复样本中均大于0.5的基因进入后续分析。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br></pre></td><td class="code"><pre><span class="line">&gt; myCPM &lt;- cpm(countdata)</span><br><span class="line">&gt; head(myCPM)</span><br><span class="line">              MCL1.DG     MCL1.DH     MCL1.DI     MCL1.DJ   MCL1.DK    MCL1.DL    MCL1.LA</span><br><span class="line"><span class="number">497097</span>    <span class="number">18.85684388</span> <span class="number">13.77543859</span>  <span class="number">2.69700983</span> <span class="number">10.45648006</span> <span class="number">16.442685</span> <span class="number">14.3389690</span>  <span class="number">0.0000000</span></span><br><span class="line"><span class="number">100503874</span>  <span class="number">0.04305215</span>  <span class="number">0.00000000</span>  <span class="number">0.04149246</span>  <span class="number">0.04412017</span>  <span class="number">0.000000</span>  <span class="number">0.1998463</span>  <span class="number">0.0000000</span></span><br><span class="line"><span class="number">100038431</span>  <span class="number">0.00000000</span>  <span class="number">0.00000000</span>  <span class="number">0.00000000</span>  <span class="number">0.00000000</span>  <span class="number">0.000000</span>  <span class="number">0.0000000</span>  <span class="number">0.0000000</span></span><br><span class="line"><span class="number">19888</span>      <span class="number">0.04305215</span>  <span class="number">0.04591813</span>  <span class="number">0.00000000</span>  <span class="number">0.00000000</span>  <span class="number">0.000000</span>  <span class="number">0.0000000</span>  <span class="number">0.4903857</span></span><br><span class="line"><span class="number">20671</span>      <span class="number">4.56352843</span>  <span class="number">8.35709941</span>  <span class="number">3.40238163</span>  <span class="number">4.63261775</span>  <span class="number">1.997275</span>  <span class="number">4.0968483</span>  <span class="number">0.7846171</span></span><br><span class="line"><span class="number">27395</span>     <span class="number">13.30311589</span> <span class="number">10.74484210</span> <span class="number">13.98295863</span> <span class="number">13.23605071</span> <span class="number">13.469996</span> <span class="number">13.4896224</span> <span class="number">27.4615975</span></span><br><span class="line">             MCL1.LB    MCL1.LC     MCL1.LD    MCL1.LE    MCL1.LF</span><br><span class="line"><span class="number">497097</span>     <span class="number">0.0000000</span>  <span class="number">0.0000000</span>  <span class="number">0.00000000</span>  <span class="number">0.0000000</span>  <span class="number">0.0000000</span></span><br><span class="line"><span class="number">100503874</span>  <span class="number">0.0000000</span>  <span class="number">0.0000000</span>  <span class="number">0.00000000</span>  <span class="number">0.0000000</span>  <span class="number">0.0000000</span></span><br><span class="line"><span class="number">100038431</span>  <span class="number">0.0000000</span>  <span class="number">0.0000000</span>  <span class="number">0.00000000</span>  <span class="number">0.0000000</span>  <span class="number">0.0000000</span></span><br><span class="line"><span class="number">19888</span>      <span class="number">0.1381969</span>  <span class="number">0.4496078</span>  <span class="number">0.09095771</span>  <span class="number">0.0000000</span>  <span class="number">0.0000000</span></span><br><span class="line"><span class="number">20671</span>      <span class="number">1.1516411</span>  <span class="number">0.8092940</span>  <span class="number">0.36383085</span>  <span class="number">0.1213404</span>  <span class="number">0.4055595</span></span><br><span class="line"><span class="number">27395</span>     <span class="number">21.3744588</span> <span class="number">21.9858214</span> <span class="number">14.91706476</span> <span class="number">12.4171715</span> <span class="number">13.8701357</span></span><br><span class="line"></span><br><span class="line">&gt; thresh &lt;- myCPM &gt; <span class="number">0.5</span>  <span class="comment"># This produces a logical matrix with TRUEs and FALSEs</span></span><br><span class="line">&gt; head(thresh)</span><br><span class="line">          MCL1.DG MCL1.DH MCL1.DI MCL1.DJ MCL1.DK MCL1.DL MCL1.LA MCL1.LB MCL1.LC MCL1.LD MCL1.LE</span><br><span class="line"><span class="number">497097</span>       <span class="literal">TRUE</span>    <span class="literal">TRUE</span>    <span class="literal">TRUE</span>    <span class="literal">TRUE</span>    <span class="literal">TRUE</span>    <span class="literal">TRUE</span>   <span class="literal">FALSE</span>   <span class="literal">FALSE</span>   <span class="literal">FALSE</span>   <span class="literal">FALSE</span>   <span class="literal">FALSE</span></span><br><span class="line"><span class="number">100503874</span>   <span class="literal">FALSE</span>   <span class="literal">FALSE</span>   <span class="literal">FALSE</span>   <span class="literal">FALSE</span>   <span class="literal">FALSE</span>   <span class="literal">FALSE</span>   <span class="literal">FALSE</span>   <span class="literal">FALSE</span>   <span class="literal">FALSE</span>   <span class="literal">FALSE</span>   <span class="literal">FALSE</span></span><br><span class="line"><span class="number">100038431</span>   <span class="literal">FALSE</span>   <span class="literal">FALSE</span>   <span class="literal">FALSE</span>   <span class="literal">FALSE</span>   <span class="literal">FALSE</span>   <span class="literal">FALSE</span>   <span class="literal">FALSE</span>   <span class="literal">FALSE</span>   <span class="literal">FALSE</span>   <span class="literal">FALSE</span>   <span class="literal">FALSE</span></span><br><span class="line"><span class="number">19888</span>       <span class="literal">FALSE</span>   <span class="literal">FALSE</span>   <span class="literal">FALSE</span>   <span class="literal">FALSE</span>   <span class="literal">FALSE</span>   <span class="literal">FALSE</span>   <span class="literal">FALSE</span>   <span class="literal">FALSE</span>   <span class="literal">FALSE</span>   <span class="literal">FALSE</span>   <span class="literal">FALSE</span></span><br><span class="line"><span class="number">20671</span>        <span class="literal">TRUE</span>    <span class="literal">TRUE</span>    <span class="literal">TRUE</span>    <span class="literal">TRUE</span>    <span class="literal">TRUE</span>    <span class="literal">TRUE</span>    <span class="literal">TRUE</span>    <span class="literal">TRUE</span>    <span class="literal">TRUE</span>   <span class="literal">FALSE</span>   <span class="literal">FALSE</span></span><br><span class="line"><span class="number">27395</span>        <span class="literal">TRUE</span>    <span class="literal">TRUE</span>    <span class="literal">TRUE</span>    <span class="literal">TRUE</span>    <span class="literal">TRUE</span>    <span class="literal">TRUE</span>    <span class="literal">TRUE</span>    <span class="literal">TRUE</span>    <span class="literal">TRUE</span>    <span class="literal">TRUE</span>    <span class="literal">TRUE</span></span><br><span class="line">          MCL1.LF</span><br><span class="line"><span class="number">497097</span>      <span class="literal">FALSE</span></span><br><span class="line"><span class="number">100503874</span>   <span class="literal">FALSE</span></span><br><span class="line"><span class="number">100038431</span>   <span class="literal">FALSE</span></span><br><span class="line"><span class="number">19888</span>       <span class="literal">FALSE</span></span><br><span class="line"><span class="number">20671</span>       <span class="literal">FALSE</span></span><br><span class="line"><span class="number">27395</span>        <span class="literal">TRUE</span></span><br></pre></td></tr></table></figure>
<p>用<code>table()</code>函数每个基因（每行）在几个样本中是符合条件的</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">&gt; table(rowSums(thresh))</span><br><span class="line"></span><br><span class="line">    <span class="number">0</span>     <span class="number">1</span>     <span class="number">2</span>     <span class="number">3</span>     <span class="number">4</span>     <span class="number">5</span>     <span class="number">6</span>     <span class="number">7</span>     <span class="number">8</span>     <span class="number">9</span>    <span class="number">10</span>    <span class="number">11</span>    <span class="number">12</span> </span><br><span class="line"><span class="number">10857</span>   <span class="number">518</span>   <span class="number">544</span>   <span class="number">307</span>   <span class="number">346</span>   <span class="number">307</span>   <span class="number">652</span>   <span class="number">323</span>   <span class="number">547</span>   <span class="number">343</span>   <span class="number">579</span>   <span class="number">423</span> <span class="number">11433</span> </span><br><span class="line"><span class="comment">#可以发现有11433个基因在12个样本中CPM均大于0.5 </span></span><br></pre></td></tr></table></figure>

<p>随后，保留至少在两个样本符合条件的基因。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line">&gt; keep &lt;- rowSums(thresh) &gt;= <span class="number">2</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># Subset the rows of countdata to keep the more highly expressed genes</span></span><br><span class="line">&gt; counts.keep &lt;- countdata[keep,]</span><br><span class="line">&gt; summary(keep)</span><br><span class="line">   Mode   <span class="literal">FALSE</span>    <span class="literal">TRUE</span> </span><br><span class="line">logical   <span class="number">11375</span>   <span class="number">15804</span> </span><br><span class="line">&gt; <span class="built_in">dim</span>(counts.keep)</span><br><span class="line">[<span class="number">1</span>] <span class="number">15804</span>    <span class="number">12</span></span><br></pre></td></tr></table></figure>
<p>我们保留了15804个相对表达可信的基因。至于，counts threshold的选择可以参考原文：</p>
<blockquote>
<p><strong>As a general rule, a good threshold can be chosen by identifying the CPM that corresponds to a count of 10, which in this case is about 0.5. You should filter with CPMs rather than filtering on the counts directly, as the latter does not account for differences in library sizes between samples.</strong></p>
</blockquote>
<h2 id="创建表达矩阵"><a href="#创建表达矩阵" class="headerlink" title="创建表达矩阵"></a>创建表达矩阵</h2><p>接着我们使用<code>edgeR</code>的<code>DGEList</code>函数来创建表达矩阵</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br></pre></td><td class="code"><pre><span class="line"> <span class="comment"># 将counts 转换为 DGEList 对象</span></span><br><span class="line">&gt; dgeObj &lt;- DGEList(counts.keep)</span><br><span class="line">&gt; dgeObj</span><br><span class="line">An object of <span class="built_in">class</span> <span class="string">&quot;DGEList&quot;</span></span><br><span class="line">$counts</span><br><span class="line">       MCL1.DG MCL1.DH MCL1.DI MCL1.DJ MCL1.DK MCL1.DL MCL1.LA MCL1.LB MCL1.LC MCL1.LD MCL1.LE</span><br><span class="line"><span class="number">497097</span>     <span class="number">438</span>     <span class="number">300</span>      <span class="number">65</span>     <span class="number">237</span>     <span class="number">354</span>     <span class="number">287</span>       <span class="number">0</span>       <span class="number">0</span>       <span class="number">0</span>       <span class="number">0</span>       <span class="number">0</span></span><br><span class="line"><span class="number">20671</span>      <span class="number">106</span>     <span class="number">182</span>      <span class="number">82</span>     <span class="number">105</span>      <span class="number">43</span>      <span class="number">82</span>      <span class="number">16</span>      <span class="number">25</span>      <span class="number">18</span>       <span class="number">8</span>       <span class="number">3</span></span><br><span class="line"><span class="number">27395</span>      <span class="number">309</span>     <span class="number">234</span>     <span class="number">337</span>     <span class="number">300</span>     <span class="number">290</span>     <span class="number">270</span>     <span class="number">560</span>     <span class="number">464</span>     <span class="number">489</span>     <span class="number">328</span>     <span class="number">307</span></span><br><span class="line"><span class="number">18777</span>      <span class="number">652</span>     <span class="number">515</span>     <span class="number">948</span>     <span class="number">935</span>     <span class="number">928</span>     <span class="number">791</span>     <span class="number">826</span>     <span class="number">862</span>     <span class="number">668</span>     <span class="number">646</span>     <span class="number">544</span></span><br><span class="line"><span class="number">21399</span>     <span class="number">1604</span>    <span class="number">1495</span>    <span class="number">1721</span>    <span class="number">1317</span>    <span class="number">1159</span>    <span class="number">1066</span>    <span class="number">1334</span>    <span class="number">1258</span>    <span class="number">1068</span>     <span class="number">926</span>     <span class="number">508</span></span><br><span class="line">       MCL1.LF</span><br><span class="line"><span class="number">497097</span>       <span class="number">0</span></span><br><span class="line"><span class="number">20671</span>       <span class="number">10</span></span><br><span class="line"><span class="number">27395</span>      <span class="number">342</span></span><br><span class="line"><span class="number">18777</span>      <span class="number">581</span></span><br><span class="line"><span class="number">21399</span>      <span class="number">500</span></span><br><span class="line"><span class="number">15799</span> more rows ...</span><br><span class="line"></span><br><span class="line">$samples</span><br><span class="line">        group lib.size norm.factors</span><br><span class="line">MCL1.DG     <span class="number">1</span> <span class="number">23218026</span>            <span class="number">1</span></span><br><span class="line">MCL1.DH     <span class="number">1</span> <span class="number">21768136</span>            <span class="number">1</span></span><br><span class="line">MCL1.DI     <span class="number">1</span> <span class="number">24091588</span>            <span class="number">1</span></span><br><span class="line">MCL1.DJ     <span class="number">1</span> <span class="number">22656713</span>            <span class="number">1</span></span><br><span class="line">MCL1.DK     <span class="number">1</span> <span class="number">21522033</span>            <span class="number">1</span></span><br><span class="line"><span class="number">7</span> more rows ...</span><br><span class="line"></span><br><span class="line"> <span class="comment"># 看看 dgeObj 中保存了什么信息</span></span><br><span class="line">&gt; <span class="built_in">names</span>(dgeObj)</span><br><span class="line">[<span class="number">1</span>] <span class="string">&quot;counts&quot;</span>  <span class="string">&quot;samples&quot;</span></span><br></pre></td></tr></table></figure>
<p>根据<code>CellType</code>和小鼠的<code>Status</code>，我们可以创建分组信息。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">group &lt;- paste(sampleinfo$CellType,sampleinfo$Status,sep=<span class="string">&quot;.&quot;</span>)</span><br><span class="line">group &lt;- factor(group)</span><br></pre></td></tr></table></figure>

<h2 id="质量控制"><a href="#质量控制" class="headerlink" title="质量控制"></a>质量控制</h2><p>由于我们的数据并不是正态分布的，为了下游分析的进行，使用<code>cpm</code>函数对我们的数据取log2处理</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">&gt; logcounts &lt;- cpm(dgeObj,<span class="built_in">log</span>=<span class="literal">TRUE</span>)</span><br><span class="line">&gt; <span class="comment"># 使用箱线图检查数据的分布</span></span><br><span class="line">&gt; boxplot(logcounts, xlab=<span class="string">&quot;&quot;</span>, ylab=<span class="string">&quot;Log2 counts per million&quot;</span>,las=<span class="number">2</span>)</span><br><span class="line">&gt; <span class="comment"># 使用蓝色的横线标出logCPM的中值</span></span><br><span class="line">&gt; abline(h=median(logcounts),col=<span class="string">&quot;blue&quot;</span>)</span><br><span class="line">&gt; title(<span class="string">&quot;Boxplots of logCPMs (unnormalised)&quot;</span>)</span><br></pre></td></tr></table></figure>
<p><img src="/2022/01/22/R-%E7%94%A8R%E8%BF%9B%E8%A1%8CRNA-seq%E6%95%B0%E6%8D%AE%E5%88%86%E6%9E%90/image-20220122200312506.png"></p>
<p>从箱线图中可以看出样本数据分布虽然存在差异，但这种差异程度是我们还能接受的程度。</p>
<h2 id="主成分分析"><a href="#主成分分析" class="headerlink" title="主成分分析"></a>主成分分析</h2><p>主成分分析在RNA-seq分析中是一个较为重要的步骤，其指出了样本数据中造成主要差异的因子是什么。一般而言，我们当然希望数据中最大的差异是由于<strong>处理与否</strong>引起的</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">&gt; plotMDS(dgeObj)</span><br></pre></td></tr></table></figure>
<p><img src="/2022/01/22/R-%E7%94%A8R%E8%BF%9B%E8%A1%8CRNA-seq%E6%95%B0%E6%8D%AE%E5%88%86%E6%9E%90/image-20220122200322356.png" alt="High Reproducibilty"></p>
<p>可以看到两个生物学重复样本都能较好的聚在一起</p>
<p>另外，我们也可以运用各种参数使主成分分析的可视化更加符合我们的要求，例如根据细胞类型或小鼠状态标注颜色：<br><img src="/2022/01/22/R-%E7%94%A8R%E8%BF%9B%E8%A1%8CRNA-seq%E6%95%B0%E6%8D%AE%E5%88%86%E6%9E%90/image-20220122200348849.png" alt="Colored by cell-type or condition"></p>
<p>总而言之，利用主成分分析，可以观察出引起样本变化的主要因子。</p>
<h2 id="标准化处理"><a href="#标准化处理" class="headerlink" title="标准化处理"></a>标准化处理</h2><p>此处我们采用了The trimmed mean of M-values normalization method (TMM) 去校正文库之间的组成偏好。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br></pre></td><td class="code"><pre><span class="line">&gt; dgeObj &lt;- calcNormFactors(dgeObj)</span><br><span class="line">&gt; dgeObj$samples</span><br><span class="line">        group lib.size norm.factors</span><br><span class="line">MCL1.DG     <span class="number">1</span> <span class="number">23218026</span>    <span class="number">1.2368993</span></span><br><span class="line">MCL1.DH     <span class="number">1</span> <span class="number">21768136</span>    <span class="number">1.2139485</span></span><br><span class="line">MCL1.DI     <span class="number">1</span> <span class="number">24091588</span>    <span class="number">1.1255640</span></span><br><span class="line">MCL1.DJ     <span class="number">1</span> <span class="number">22656713</span>    <span class="number">1.0698261</span></span><br><span class="line">MCL1.DK     <span class="number">1</span> <span class="number">21522033</span>    <span class="number">1.0359212</span></span><br><span class="line">MCL1.DL     <span class="number">1</span> <span class="number">20008326</span>    <span class="number">1.0872153</span></span><br><span class="line">MCL1.LA     <span class="number">1</span> <span class="number">20384562</span>    <span class="number">1.3684449</span></span><br><span class="line">MCL1.LB     <span class="number">1</span> <span class="number">21698793</span>    <span class="number">1.3653200</span></span><br><span class="line">MCL1.LC     <span class="number">1</span> <span class="number">22235847</span>    <span class="number">1.0047431</span></span><br><span class="line">MCL1.LD     <span class="number">1</span> <span class="number">21982745</span>    <span class="number">0.9232822</span></span><br><span class="line">MCL1.LE     <span class="number">1</span> <span class="number">24719697</span>    <span class="number">0.5291015</span></span><br><span class="line">MCL1.LF     <span class="number">1</span> <span class="number">24652963</span>    <span class="number">0.5354877</span></span><br><span class="line"><span class="comment"># 此步在‘samples’信息中直接添加了校正因子。</span></span><br></pre></td></tr></table></figure>

<p>我们可以单独抽一个样本出来看看校正前后的数据分布情况<br><img src="/2022/01/22/R-%E7%94%A8R%E8%BF%9B%E8%A1%8CRNA-seq%E6%95%B0%E6%8D%AE%E5%88%86%E6%9E%90/image-20220122200412539.png"></p>
<p>可以看出校正后，样本表达更加集中于0。</p>
<p>至此，数据的预处理也已经完结.</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">#保存预处理分析结果</span></span><br><span class="line">save(group,dgeObj,sampleinfo,file=<span class="string">&quot;~preprocessing.Rdata&quot;</span>)</span><br></pre></td></tr></table></figure>



<h1 id="Step-3-Differential-Expression-for-RNA-seq"><a href="#Step-3-Differential-Expression-for-RNA-seq" class="headerlink" title="Step 3: Differential Expression for RNA-seq"></a><strong>Step 3:</strong> Differential Expression for RNA-seq</h1><p>本节将要介绍的是在R中进行RNA-seq数据基因表达差异分析的实战代码.</p>
<blockquote>
<p>Ref: <a target="_blank" rel="noopener" href="https://bioinformatics-core-shared-training.github.io/RNAseq-R/rna-seq-de.nb.html">https://bioinformatics-core-shared-training.github.io/RNAseq-R/rna-seq-de.nb.html</a></p>
</blockquote>
<p>在预处理生成了表达矩阵且标准化后，接下来我们要做的就是差异分析了。</p>
<hr>
<p><strong>本次流程需要的包</strong></p>
<p>同时载入上次预处理完的数据<code>preprocessing.Rdata</code></p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">library(edgeR)</span><br><span class="line">library(limma)</span><br><span class="line">library(Glimma)</span><br><span class="line">library(gplots)</span><br><span class="line">library(org.Mm.eg.db)</span><br><span class="line">load(<span class="string">&quot;preprocessing.Rdata&quot;</span>)</span><br></pre></td></tr></table></figure>

<p>在<code>edgeR</code>包的分析流程中，构建了表达矩阵后，便需要使用<code>model.matrix</code>函数创建保存有分组信息的矩阵<strong>design matrix</strong>。该矩阵以$\color{red}{0/1}$的方式存储了分组信息。本次分析中的分组信息包括了：<strong>小鼠状态与细胞类型</strong>，现在我们先在<strong>两种因素并不存在相互作用</strong>的假设下拟合线性模型。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">&gt; group &lt;- <span class="built_in">as.character</span>(group)</span><br><span class="line">&gt; group</span><br><span class="line"> [<span class="number">1</span>] <span class="string">&quot;basal.virgin&quot;</span>     <span class="string">&quot;basal.virgin&quot;</span>     <span class="string">&quot;basal.pregnant&quot;</span>   <span class="string">&quot;basal.pregnant&quot;</span>   <span class="string">&quot;basal.lactate&quot;</span>   </span><br><span class="line"> [<span class="number">6</span>] <span class="string">&quot;basal.lactate&quot;</span>    <span class="string">&quot;luminal.virgin&quot;</span>   <span class="string">&quot;luminal.virgin&quot;</span>   <span class="string">&quot;luminal.pregnant&quot;</span> <span class="string">&quot;luminal.pregnant&quot;</span></span><br><span class="line">[<span class="number">11</span>] <span class="string">&quot;luminal.lactate&quot;</span>  <span class="string">&quot;luminal.lactate&quot;</span> </span><br></pre></td></tr></table></figure>
<p>group的分组信息包括了状态和细胞类型，因此我们使用<code>strsplit</code>函数，以‘.’为分隔把细胞类型信息和状态信息分别提取出来。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">&gt; type &lt;- sapply(strsplit(group, <span class="string">&quot;.&quot;</span>, fixed=<span class="built_in">T</span>), <span class="keyword">function</span>(x) x[<span class="number">1</span>])</span><br><span class="line">&gt; status &lt;- sapply(strsplit(group, <span class="string">&quot;.&quot;</span>, fixed=<span class="built_in">T</span>), <span class="keyword">function</span>(x) x[<span class="number">2</span>])</span><br><span class="line">&gt; type</span><br><span class="line"> [<span class="number">1</span>] <span class="string">&quot;basal&quot;</span>   <span class="string">&quot;basal&quot;</span>   <span class="string">&quot;basal&quot;</span>   <span class="string">&quot;basal&quot;</span>   <span class="string">&quot;basal&quot;</span>   <span class="string">&quot;basal&quot;</span>   <span class="string">&quot;luminal&quot;</span> <span class="string">&quot;luminal&quot;</span> <span class="string">&quot;luminal&quot;</span></span><br><span class="line">[<span class="number">10</span>] <span class="string">&quot;luminal&quot;</span> <span class="string">&quot;luminal&quot;</span> <span class="string">&quot;luminal&quot;</span></span><br><span class="line">&gt; status</span><br><span class="line"> [<span class="number">1</span>] <span class="string">&quot;virgin&quot;</span>   <span class="string">&quot;virgin&quot;</span>   <span class="string">&quot;pregnant&quot;</span> <span class="string">&quot;pregnant&quot;</span> <span class="string">&quot;lactate&quot;</span>  <span class="string">&quot;lactate&quot;</span>  <span class="string">&quot;virgin&quot;</span>   <span class="string">&quot;virgin&quot;</span>  </span><br><span class="line"> [<span class="number">9</span>] <span class="string">&quot;pregnant&quot;</span> <span class="string">&quot;pregnant&quot;</span> <span class="string">&quot;lactate&quot;</span>  <span class="string">&quot;lactate&quot;</span> </span><br></pre></td></tr></table></figure>

<p>成功提取后，构建分组信息矩阵</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br></pre></td><td class="code"><pre><span class="line">&gt; design &lt;- model.matrix(~ type + status)</span><br><span class="line">&gt; design</span><br><span class="line">   (Intercept) typeluminal statuspregnant statusvirgin</span><br><span class="line"><span class="number">1</span>            <span class="number">1</span>           <span class="number">0</span>              <span class="number">0</span>            <span class="number">1</span></span><br><span class="line"><span class="number">2</span>            <span class="number">1</span>           <span class="number">0</span>              <span class="number">0</span>            <span class="number">1</span></span><br><span class="line"><span class="number">3</span>            <span class="number">1</span>           <span class="number">0</span>              <span class="number">1</span>            <span class="number">0</span></span><br><span class="line"><span class="number">4</span>            <span class="number">1</span>           <span class="number">0</span>              <span class="number">1</span>            <span class="number">0</span></span><br><span class="line"><span class="number">5</span>            <span class="number">1</span>           <span class="number">0</span>              <span class="number">0</span>            <span class="number">0</span></span><br><span class="line"><span class="number">6</span>            <span class="number">1</span>           <span class="number">0</span>              <span class="number">0</span>            <span class="number">0</span></span><br><span class="line"><span class="number">7</span>            <span class="number">1</span>           <span class="number">1</span>              <span class="number">0</span>            <span class="number">1</span></span><br><span class="line"><span class="number">8</span>            <span class="number">1</span>           <span class="number">1</span>              <span class="number">0</span>            <span class="number">1</span></span><br><span class="line"><span class="number">9</span>            <span class="number">1</span>           <span class="number">1</span>              <span class="number">1</span>            <span class="number">0</span></span><br><span class="line"><span class="number">10</span>           <span class="number">1</span>           <span class="number">1</span>              <span class="number">1</span>            <span class="number">0</span></span><br><span class="line"><span class="number">11</span>           <span class="number">1</span>           <span class="number">1</span>              <span class="number">0</span>            <span class="number">0</span></span><br><span class="line"><span class="number">12</span>           <span class="number">1</span>           <span class="number">1</span>              <span class="number">0</span>            <span class="number">0</span></span><br><span class="line"><span class="built_in">attr</span>(,<span class="string">&quot;assign&quot;</span>)</span><br><span class="line">[<span class="number">1</span>] <span class="number">0</span> <span class="number">1</span> <span class="number">2</span> <span class="number">2</span></span><br><span class="line"><span class="built_in">attr</span>(,<span class="string">&quot;contrasts&quot;</span>)</span><br><span class="line"><span class="built_in">attr</span>(,<span class="string">&quot;contrasts&quot;</span>)$type</span><br><span class="line">[<span class="number">1</span>] <span class="string">&quot;contr.treatment&quot;</span></span><br><span class="line"></span><br><span class="line"><span class="built_in">attr</span>(,<span class="string">&quot;contrasts&quot;</span>)$status</span><br><span class="line">[<span class="number">1</span>] <span class="string">&quot;contr.treatment&quot;</span></span><br></pre></td></tr></table></figure>

<h2 id="预测基因表达的离散度"><a href="#预测基因表达的离散度" class="headerlink" title="预测基因表达的离散度"></a>预测基因表达的离散度</h2><p>这里根据每个基因在不同样本之间的表达量对基因表达的离散度（dispersion）进行预测。对基因表达离散度的预测将会帮助我们对各个样本的counts进行校正。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">&gt; dgeObj &lt;- estimateCommonDisp(dgeObj)</span><br><span class="line"><span class="comment"># Then we estimate gene-wise dispersion estimates, allowing a possible trend with averge count size:</span></span><br><span class="line">&gt; dgeObj &lt;- estimateGLMTrendedDisp(dgeObj)</span><br><span class="line">&gt; dgeObj &lt;- estimateTagwiseDisp(dgeObj)</span><br><span class="line"> <span class="comment">#Plot the estimated dispersions</span></span><br><span class="line">&gt; plotBCV(dgeObj)</span><br></pre></td></tr></table></figure>
<p><img src="/2022/01/22/R-%E7%94%A8R%E8%BF%9B%E8%A1%8CRNA-seq%E6%95%B0%E6%8D%AE%E5%88%86%E6%9E%90/image-20220122200426544.png"></p>
<p>实际上以上三句代码可以用一个函数代替。在真实的分析当中也是使用<code>estimateDisp()</code>预测基因离散度，上面的拆分只是为了帮助我们理解离散度的预测是由哪几方面出发的。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">dgeObj &lt;- estimateDisp(dgeObj, design = design)</span><br></pre></td></tr></table></figure>

<h2 id="差异表达分析"><a href="#差异表达分析" class="headerlink" title="差异表达分析"></a>差异表达分析</h2><p>预测离散度后，简单来说，我们将使用<code>edgeR::glmFit()</code>进行模型的拟合；使用<code>edgeR::glmLRT()</code>检验基因表达差异是否显著。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line">&gt; fit &lt;- glmFit(dgeObj, design)</span><br><span class="line"><span class="comment"># 看一眼系数</span></span><br><span class="line">&gt; head(coef(fit))</span><br><span class="line">       (Intercept) typeluminal statuspregnant statusvirgin</span><br><span class="line"><span class="number">497097</span>  -<span class="number">11.187922</span> -<span class="number">7.58804851</span>     -<span class="number">0.7085514</span>  -<span class="number">0.09305118</span></span><br><span class="line"><span class="number">20671</span>   -<span class="number">12.715063</span> -<span class="number">1.85287334</span>      <span class="number">0.2269001</span>   <span class="number">0.49554506</span></span><br><span class="line"><span class="number">27395</span>   -<span class="number">11.221391</span>  <span class="number">0.56368066</span>     -<span class="number">0.1415910</span>  -<span class="number">0.29221577</span></span><br><span class="line"><span class="number">18777</span>   -<span class="number">10.146793</span>  <span class="number">0.08280255</span>     -<span class="number">0.1845489</span>  -<span class="number">0.48795441</span></span><br><span class="line"><span class="number">21399</span>    -<span class="number">9.909825</span> -<span class="number">0.24195503</span>      <span class="number">0.1753606</span>   <span class="number">0.13494615</span></span><br><span class="line"><span class="number">58175</span>   -<span class="number">16.310131</span>  <span class="number">3.09936215</span>      <span class="number">1.1975518</span>   <span class="number">0.84742701</span></span><br></pre></td></tr></table></figure>

<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># Conduct likelihood ratio tests for luminal vs basal and show the top genes:</span></span><br><span class="line">&gt; lrt.BvsL &lt;- glmLRT(fit, coef=<span class="number">2</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment"># 查看前10个最显著的差异表达基因</span></span><br><span class="line">&gt; topTags(lrt.BvsL)</span><br><span class="line">Coefficient:  typeluminal </span><br><span class="line">           logFC    logCPM       LR       PValue          FDR</span><br><span class="line"><span class="number">110308</span> -<span class="number">8.940579</span> <span class="number">10.264297</span> <span class="number">24.89789</span> <span class="number">6.044844e-07</span> <span class="number">0.0004377961</span></span><br><span class="line"><span class="number">50916</span>  -<span class="number">8.636503</span>  <span class="number">5.749781</span> <span class="number">24.80037</span> <span class="number">6.358512e-07</span> <span class="number">0.0004377961</span></span><br><span class="line"><span class="number">12293</span>  -<span class="number">8.362247</span>  <span class="number">6.794788</span> <span class="number">24.68526</span> <span class="number">6.749827e-07</span> <span class="number">0.0004377961</span></span><br><span class="line"><span class="number">56069</span>  -<span class="number">8.419433</span>  <span class="number">6.124377</span> <span class="number">24.41532</span> <span class="number">7.764861e-07</span> <span class="number">0.0004377961</span></span><br><span class="line"><span class="number">24117</span>  -<span class="number">9.290691</span>  <span class="number">6.757163</span> <span class="number">24.32506</span> <span class="number">8.137331e-07</span> <span class="number">0.0004377961</span></span><br><span class="line"><span class="number">12818</span>  -<span class="number">8.216790</span>  <span class="number">8.172247</span> <span class="number">24.24233</span> <span class="number">8.494462e-07</span> <span class="number">0.0004377961</span></span><br><span class="line"><span class="number">22061</span>  -<span class="number">8.034712</span>  <span class="number">7.255370</span> <span class="number">24.16987</span> <span class="number">8.820135e-07</span> <span class="number">0.0004377961</span></span><br><span class="line"><span class="number">12797</span>  -<span class="number">9.001419</span>  <span class="number">9.910795</span> <span class="number">24.12854</span> <span class="number">9.011487e-07</span> <span class="number">0.0004377961</span></span><br><span class="line"><span class="number">50706</span>  -<span class="number">7.697022</span> <span class="number">10.809629</span> <span class="number">24.06926</span> <span class="number">9.293193e-07</span> <span class="number">0.0004377961</span></span><br><span class="line"><span class="number">237979</span> -<span class="number">8.167451</span>  <span class="number">5.215921</span> <span class="number">24.03528</span> <span class="number">9.458678e-07</span> <span class="number">0.0004377961</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># 提取所有基因的差异分析结果</span></span><br><span class="line">de.res &lt;- topTags(lrt.BvsL, n = <span class="literal">Inf</span>)</span><br></pre></td></tr></table></figure>


<p>最后，保存差异分析的结果.</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">save(lrt.BvsL,dgeObj,group,file=<span class="string">&quot;DE.Rdata&quot;</span>)</span><br></pre></td></tr></table></figure>



<h1 id="Step-4-Annotation-and-Visualisation-of-RNA-seq-results"><a href="#Step-4-Annotation-and-Visualisation-of-RNA-seq-results" class="headerlink" title="Step 4:  Annotation and Visualisation of RNA-seq results"></a><strong>Step 4:</strong>  Annotation and Visualisation of RNA-seq results</h1><p>本节将介绍RNA-seq差异分析结果可视化的实战代码。</p>
<blockquote>
<p>Ref: <a target="_blank" rel="noopener" href="https://bioinformatics-core-shared-training.github.io/RNAseq-R/rna-seq-annotation-visualisation.nb.html">https://bioinformatics-core-shared-training.github.io/RNAseq-R/rna-seq-annotation-visualisation.nb.html</a></p>
</blockquote>
<p>在差异分析后我们经常要可视化我们的分析结果，例如我们可以使用火山图将差异表达的基因画出来。</p>
<h2 id="基因注释"><a href="#基因注释" class="headerlink" title="基因注释"></a>基因注释</h2><p>在本次分析中，我们都是采用基因的ID来进行的，但如果想便于我们自己查阅的话，单纯的ID是比较麻烦的，我们还得一个个ID找它对应的基因名。因此，我们可以在差异分析结果的表格中添加上基因的symbol，genename等信息。</p>
<p>由于这次分析的物种是小鼠，我们采用以下的注释包：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">&gt; library(org.Mm.eg.db)</span><br><span class="line"></span><br><span class="line"><span class="comment"># 同时载入之前分析的数据</span></span><br><span class="line">&gt; load(<span class="string">&quot;DE.Rdata&quot;</span>)</span><br></pre></td></tr></table></figure>

<p>让我们查看一下该注释包存有的数据</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">&gt; keytypes(org.Mm.eg.db)</span><br><span class="line"> [<span class="number">1</span>] <span class="string">&quot;ACCNUM&quot;</span>       <span class="string">&quot;ALIAS&quot;</span>        <span class="string">&quot;ENSEMBL&quot;</span>      <span class="string">&quot;ENSEMBLPROT&quot;</span>  <span class="string">&quot;ENSEMBLTRANS&quot;</span> <span class="string">&quot;ENTREZID&quot;</span>    </span><br><span class="line"> [<span class="number">7</span>] <span class="string">&quot;ENZYME&quot;</span>       <span class="string">&quot;EVIDENCE&quot;</span>     <span class="string">&quot;EVIDENCEALL&quot;</span>  <span class="string">&quot;GENENAME&quot;</span>     <span class="string">&quot;GO&quot;</span>           <span class="string">&quot;GOALL&quot;</span>       </span><br><span class="line">[<span class="number">13</span>] <span class="string">&quot;IPI&quot;</span>          <span class="string">&quot;MGI&quot;</span>          <span class="string">&quot;ONTOLOGY&quot;</span>     <span class="string">&quot;ONTOLOGYALL&quot;</span>  <span class="string">&quot;PATH&quot;</span>         <span class="string">&quot;PFAM&quot;</span>        </span><br><span class="line">[<span class="number">19</span>] <span class="string">&quot;PMID&quot;</span>         <span class="string">&quot;PROSITE&quot;</span>      <span class="string">&quot;REFSEQ&quot;</span>       <span class="string">&quot;SYMBOL&quot;</span>       <span class="string">&quot;UNIGENE&quot;</span>      <span class="string">&quot;UNIPROT&quot;</span></span><br></pre></td></tr></table></figure>
<p>在以上的数据中，我们需要的**”ENTREZID”**  、**”GENENAME”<strong>、</strong>“SYMBOL”** 都在其中。</p>
<p>接下来，我们只要对应已有结果中的<strong>ENTREZID</strong>提取相应的<strong>GENENAME</strong>和<strong>SYMBOL</strong>即可。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># 将差异分析结果存储到results变量当中</span></span><br><span class="line">&gt; results &lt;- as.data.frame(topTags(lrt.BvsL,n = <span class="literal">Inf</span>))</span><br><span class="line"></span><br><span class="line">&gt; ann &lt;- select(org.Mm.eg.db,keys=rownames(results),</span><br><span class="line">+               columns=<span class="built_in">c</span>(<span class="string">&quot;ENTREZID&quot;</span>,<span class="string">&quot;SYMBOL&quot;</span>,<span class="string">&quot;GENENAME&quot;</span>))</span><br><span class="line"><span class="string">&#x27;select()&#x27;</span> returned <span class="number">1</span>:<span class="number">1</span> mapping between keys and columns</span><br><span class="line">&gt; head(ann)</span><br><span class="line">  ENTREZID   SYMBOL                                                   GENENAME</span><br><span class="line"><span class="number">1</span>   <span class="number">110308</span>     Krt5                                                  keratin <span class="number">5</span></span><br><span class="line"><span class="number">2</span>    <span class="number">50916</span>     Irx4                                        Iroquois homeobox <span class="number">4</span></span><br><span class="line"><span class="number">3</span>    <span class="number">12293</span> Cacna2d1 calcium channel, voltage-dependent, alpha2/delta subunit <span class="number">1</span></span><br><span class="line"><span class="number">4</span>    <span class="number">56069</span>    Il17b                                            interleukin <span class="number">17</span>B</span><br><span class="line"><span class="number">5</span>    <span class="number">24117</span>     Wif1                                    Wnt inhibitory factor <span class="number">1</span></span><br><span class="line"><span class="number">6</span>    <span class="number">12818</span>  Col14a1                                collagen, type XIV, alpha <span class="number">1</span></span><br></pre></td></tr></table></figure>
<p>利用<code>cbind</code>函数将新的注释信息与原表格合并之后即可</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br></pre></td><td class="code"><pre><span class="line">&gt; results.annotated &lt;- cbind(results, ann)</span><br><span class="line"></span><br><span class="line">&gt; head(results.annotated)</span><br><span class="line">           logFC    logCPM       LR       PValue          FDR ENTREZID   SYMBOL</span><br><span class="line"><span class="number">110308</span> -<span class="number">8.940579</span> <span class="number">10.264297</span> <span class="number">24.89789</span> <span class="number">6.044844e-07</span> <span class="number">0.0004377961</span>   <span class="number">110308</span>     Krt5</span><br><span class="line"><span class="number">50916</span>  -<span class="number">8.636503</span>  <span class="number">5.749781</span> <span class="number">24.80037</span> <span class="number">6.358512e-07</span> <span class="number">0.0004377961</span>    <span class="number">50916</span>     Irx4</span><br><span class="line"><span class="number">12293</span>  -<span class="number">8.362247</span>  <span class="number">6.794788</span> <span class="number">24.68526</span> <span class="number">6.749827e-07</span> <span class="number">0.0004377961</span>    <span class="number">12293</span> Cacna2d1</span><br><span class="line"><span class="number">56069</span>  -<span class="number">8.419433</span>  <span class="number">6.124377</span> <span class="number">24.41532</span> <span class="number">7.764861e-07</span> <span class="number">0.0004377961</span>    <span class="number">56069</span>    Il17b</span><br><span class="line"><span class="number">24117</span>  -<span class="number">9.290691</span>  <span class="number">6.757163</span> <span class="number">24.32506</span> <span class="number">8.137331e-07</span> <span class="number">0.0004377961</span>    <span class="number">24117</span>     Wif1</span><br><span class="line"><span class="number">12818</span>  -<span class="number">8.216790</span>  <span class="number">8.172247</span> <span class="number">24.24233</span> <span class="number">8.494462e-07</span> <span class="number">0.0004377961</span>    <span class="number">12818</span>  Col14a1</span><br><span class="line">                                                         GENENAME</span><br><span class="line"><span class="number">110308</span>                                                  keratin <span class="number">5</span></span><br><span class="line"><span class="number">50916</span>                                         Iroquois homeobox <span class="number">4</span></span><br><span class="line"><span class="number">12293</span>  calcium channel, voltage-dependent, alpha2/delta subunit <span class="number">1</span></span><br><span class="line"><span class="number">56069</span>                                             interleukin <span class="number">17</span>B</span><br><span class="line"><span class="number">24117</span>                                     Wnt inhibitory factor <span class="number">1</span></span><br><span class="line"><span class="number">12818</span>                                 collagen, type XIV, alpha <span class="number">1</span></span><br></pre></td></tr></table></figure>
<p>此外，除了基因名的信息之外，还可以利用有基因组位点信息的包，如<code>TxDb.Mmusculus.UCSC.mm10.knownGene</code>(小鼠的)和<code>GenomicRanges</code>来注释基因座信息。</p>
<h2 id="可视化"><a href="#可视化" class="headerlink" title="可视化"></a>可视化</h2><p>火山图是一种常用的差异表达分析结果可视化形式。该图横轴展示基因表达差异倍数，纵轴表示基因表达差异的显著性。</p>
<p>先看看最简单的图</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">&gt; signif &lt;- -log10(results.annotated$FDR)</span><br><span class="line">&gt; plot(results.annotated$logFC,<span class="built_in">signif</span>,pch=<span class="number">16</span>)</span><br></pre></td></tr></table></figure>
<p><img src="/2022/01/22/R-%E7%94%A8R%E8%BF%9B%E8%A1%8CRNA-seq%E6%95%B0%E6%8D%AE%E5%88%86%E6%9E%90/image-20220122200445711.png"></p>
<p>当然，各位R学习者肯定不会止步于画了这个图就算了。让我们用<code>ggplot</code>画出更漂亮的图吧！</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br></pre></td><td class="code"><pre><span class="line">&gt; library(ggplot2)</span><br><span class="line">&gt; library(ggtheme)</span><br><span class="line"></span><br><span class="line"><span class="comment"># 首先添加上threshold</span></span><br><span class="line">&gt; results.annotated$threshold &lt;- as.factor(</span><br><span class="line">+   ifelse(results.annotated$FDR &lt; <span class="number">0.05</span> &amp; <span class="built_in">abs</span>(results.annotated$logFC) &gt;=log2(<span class="number">2</span>),</span><br><span class="line">+   ifelse(results.annotated$logFC &gt; log2(<span class="number">2</span>) ,<span class="string">&#x27;Up&#x27;</span>,<span class="string">&#x27;Down&#x27;</span>),<span class="string">&#x27;Not&#x27;</span>))</span><br><span class="line"></span><br><span class="line">&gt; plt &lt;- ggplot(data=results.annotated, aes(x=logFC, y =-log10(FDR), colour=threshold,fill=threshold)) +</span><br><span class="line">+   scale_color_manual(values=<span class="built_in">c</span>(<span class="string">&quot;blue&quot;</span>, <span class="string">&quot;grey&quot;</span>,<span class="string">&quot;red&quot;</span>))+</span><br><span class="line">+   geom_point(alpha=<span class="number">0.4</span>, size=<span class="number">1.2</span>) +</span><br><span class="line">+   theme_bw(base_size = <span class="number">12</span>, base_family = <span class="string">&quot;Times&quot;</span>) +</span><br><span class="line">+   geom_vline(xintercept=<span class="built_in">c</span>(-<span class="number">0.5</span>,<span class="number">0.5</span>),lty=<span class="number">4</span>,col=<span class="string">&quot;grey&quot;</span>,lwd=<span class="number">0.6</span>)+</span><br><span class="line">+   geom_hline(yintercept = -log10(<span class="number">0.05</span>),lty=<span class="number">4</span>,col=<span class="string">&quot;grey&quot;</span>,lwd=<span class="number">0.6</span>)+</span><br><span class="line">+   theme(legend.position=<span class="string">&quot;right&quot;</span>,</span><br><span class="line">+         panel.grid=element_blank(),</span><br><span class="line">+         legend.title = element_blank(),</span><br><span class="line">+         legend.text= element_text(face=<span class="string">&quot;bold&quot;</span>, color=<span class="string">&quot;black&quot;</span>,family = <span class="string">&quot;Times&quot;</span>, size=<span class="number">8</span>),</span><br><span class="line">+         plot.title = element_text(hjust = <span class="number">0.5</span>),</span><br><span class="line">+         axis.text.x = element_text(face=<span class="string">&quot;bold&quot;</span>, color=<span class="string">&quot;black&quot;</span>, size=<span class="number">12</span>),</span><br><span class="line">+         axis.text.y = element_text(face=<span class="string">&quot;bold&quot;</span>,  color=<span class="string">&quot;black&quot;</span>, size=<span class="number">12</span>),</span><br><span class="line">+         axis.title.x = element_text(face=<span class="string">&quot;bold&quot;</span>, color=<span class="string">&quot;black&quot;</span>, size=<span class="number">12</span>),</span><br><span class="line">+         axis.title.y = element_text(face=<span class="string">&quot;bold&quot;</span>,color=<span class="string">&quot;black&quot;</span>, size=<span class="number">12</span>)) +</span><br><span class="line">+   labs( x=<span class="string">&quot;log2 (Fold Change)&quot;</span>,y=<span class="string">&quot;-log10 (p-value)&quot;</span>)</span><br><span class="line"></span><br><span class="line">&gt; plt</span><br></pre></td></tr></table></figure>

<p><img src="/2022/01/22/R-%E7%94%A8R%E8%BF%9B%E8%A1%8CRNA-seq%E6%95%B0%E6%8D%AE%E5%88%86%E6%9E%90/image-20220122200502891.png"></p>
<p>除了火山图，还有各种对差异分析结果可视化的图就不在此展开了，有兴趣的朋友可以到原网站上继续了解。</p>
<h1 id="Step-5-Gene-set-testing"><a href="#Step-5-Gene-set-testing" class="headerlink" title="Step 5:  Gene-set testing"></a><strong>Step 5:</strong>  Gene-set testing</h1><p>本节将介绍基因富集分析的实战代码。</p>
<blockquote>
<p>Ref: <a target="_blank" rel="noopener" href="https://bioinformatics-core-shared-training.github.io/RNAseq-R/rna-seq-gene-set-testing.nb.html">https://bioinformatics-core-shared-training.github.io/RNAseq-R/rna-seq-gene-set-testing.nb.html</a></p>
</blockquote>
<p>在基因差异分析后，我们会得到一个包含了差异表达基因信息的表格。但有时过于<strong>冗长的</strong>基因列表或是<strong>繁杂的</strong>信息会阻碍我们从中提取有效的信息。因此，科学家们时常会采用基因富集分析的办法，去掌握<strong>差异表达基因</strong>真正影响到的<strong>信号通路</strong>或是<strong>代谢途径</strong>等。</p>
<p>例如，<strong>Gene Ontology（GO）</strong>或是<strong>KEGG</strong>就是较为常用的基因富集分析，不仅可以帮助我们从冗长的数据中提取出<strong>有生物学意义的信息</strong>，还能有助于我们从<strong>系统的水平</strong>观察细胞/机体的变化。</p>
<p>接下来，我们就以<code>GOseq</code>这个包向大家展示在R中处理的GO富集分析流程。当然我们只是简要地介绍一下，如果想要深入了解的朋友，可以查看Bioconductor上的帮助文档：<a target="_blank" rel="noopener" href="https://www.bioconductor.org/packages/release/bioc/vignettes/goseq/inst/doc/goseq.pdf">GOseq</a></p>
<p><strong>本次流程所需要的包和数据集</strong></p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">library(edgeR)</span><br><span class="line">library(goseq)</span><br><span class="line"></span><br><span class="line"><span class="comment"># 此前差异分析的结果</span></span><br><span class="line">load(<span class="string">&quot;DE.Rdata&quot;</span>)</span><br></pre></td></tr></table></figure>

<h1 id="GOseq-analysis-pipeline"><a href="#GOseq-analysis-pipeline" class="headerlink" title="GOseq analysis pipeline"></a>GOseq analysis pipeline</h1><figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br></pre></td><td class="code"><pre><span class="line">results &lt;- as.data.frame(topTags(lrt.BvsL, n = <span class="literal">Inf</span>))</span><br><span class="line"><span class="comment"># 筛选显著的差异表达基因进行分析</span></span><br><span class="line">genes &lt;- results$FDR &lt; <span class="number">0.01</span></span><br><span class="line"><span class="built_in">names</span>(genes) &lt;- rownames(results)</span><br><span class="line"></span><br><span class="line"><span class="comment"># Fit the Probability Weighting Function (PWF) ，以生成零分布</span></span><br><span class="line">pwf &lt;- nullp(genes, <span class="string">&quot;mm10&quot;</span>,<span class="string">&quot;knownGene&quot;</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment"># 富集分析</span></span><br><span class="line">go.results &lt;- goseq(pwf, <span class="string">&quot;mm10&quot;</span>,<span class="string">&quot;knownGene&quot;</span>)</span><br><span class="line">head(go.results,<span class="number">10</span>)</span><br><span class="line">        category over_represented_pvalue under_represented_pvalue numDEInCat numInCat</span><br><span class="line"><span class="number">16121</span> GO:<span class="number">0071944</span>            <span class="number">4.062458e-63</span>                        <span class="number">1</span>        <span class="number">933</span>     <span class="number">3543</span></span><br><span class="line"><span class="number">2723</span>  GO:<span class="number">0005886</span>            <span class="number">2.686421e-62</span>                        <span class="number">1</span>        <span class="number">908</span>     <span class="number">3420</span></span><br><span class="line"><span class="number">2540</span>  GO:<span class="number">0005576</span>            <span class="number">4.492776e-57</span>                        <span class="number">1</span>        <span class="number">469</span>     <span class="number">1457</span></span><br><span class="line"><span class="number">11253</span> GO:<span class="number">0044421</span>            <span class="number">1.026496e-50</span>                        <span class="number">1</span>        <span class="number">391</span>     <span class="number">1175</span></span><br><span class="line"><span class="number">11283</span> GO:<span class="number">0044459</span>            <span class="number">2.158946e-49</span>                        <span class="number">1</span>        <span class="number">563</span>     <span class="number">1893</span></span><br><span class="line"><span class="number">2562</span>  GO:<span class="number">0005615</span>            <span class="number">2.018827e-47</span>                        <span class="number">1</span>        <span class="number">336</span>      <span class="number">975</span></span><br><span class="line"><span class="number">7483</span>  GO:<span class="number">0031224</span>            <span class="number">2.789374e-41</span>                        <span class="number">1</span>        <span class="number">833</span>     <span class="number">3414</span></span><br><span class="line"><span class="number">7485</span>  GO:<span class="number">0031226</span>            <span class="number">3.115084e-40</span>                        <span class="number">1</span>        <span class="number">323</span>      <span class="number">923</span></span><br><span class="line"><span class="number">2724</span>  GO:<span class="number">0005887</span>            <span class="number">7.600430e-40</span>                        <span class="number">1</span>        <span class="number">309</span>      <span class="number">867</span></span><br><span class="line"><span class="number">4426</span>  GO:<span class="number">0009653</span>            <span class="number">1.471331e-39</span>                        <span class="number">1</span>        <span class="number">592</span>     <span class="number">2186</span></span><br><span class="line">                                        term ontology</span><br><span class="line"><span class="number">16121</span>                         cell periphery       CC</span><br><span class="line"><span class="number">2723</span>                         plasma membrane       CC</span><br><span class="line"><span class="number">2540</span>                    extracellular region       CC</span><br><span class="line"><span class="number">11253</span>              extracellular region part       CC</span><br><span class="line"><span class="number">11283</span>                   plasma membrane part       CC</span><br><span class="line"><span class="number">2562</span>                     extracellular space       CC</span><br><span class="line"><span class="number">7483</span>         intrinsic component of membrane       CC</span><br><span class="line"><span class="number">7485</span>  intrinsic component of plasma membrane       CC</span><br><span class="line"><span class="number">2724</span>   integral component of plasma membrane       CC</span><br><span class="line"><span class="number">4426</span>      anatomical structure morphogenesis       BP</span><br></pre></td></tr></table></figure>

<p>GO的分析也就到此为止了，至于后续的可视化的话就不先在此展开了。实际上还有许多R包也能在R中进行基因富集分析，例如Y叔的<code>clusterProfiler</code>，<code>fgsea</code>，<code>camera</code>等等。</p>
<p>关于RNA-seq分析流程到这就先告一段落了，文章中有任何错误都请各位小伙伴指出，共同学习！</p>
<p>完。</p>

    </div>

    
    
    
        

<div>
<ul class="post-copyright">
  <li class="post-copyright-author">
    <strong>Post author:  </strong>Dean Li
  </li>
  <li class="post-copyright-link">
    <strong>Post link: </strong>
    <a href="https://thereallda.github.io/2022/01/22/R-%E7%94%A8R%E8%BF%9B%E8%A1%8CRNA-seq%E6%95%B0%E6%8D%AE%E5%88%86%E6%9E%90/" title="R-用R进行RNA-seq数据分析">https://thereallda.github.io/2022/01/22/R-用R进行RNA-seq数据分析/</a>
  </li>
  <li class="post-copyright-license">
    <strong>Copyright Notice:  </strong>All articles in this blog are licensed under <a href="https://creativecommons.org/licenses/by-nc-sa/4.0/" rel="noopener" target="_blank"><i class="fab fa-fw fa-creative-commons"></i>BY-NC-SA</a> unless stating additionally.
  </li>
</ul>
</div>


      <footer class="post-footer">
          <div class="post-tags">
              <a href="/tags/Data-analysis/" rel="tag"># Data_analysis</a>
              <a href="/tags/R/" rel="tag"># R</a>
              <a href="/tags/RNA-seq/" rel="tag"># RNA-seq</a>
          </div>

        


        
    <div class="post-nav">
      <div class="post-nav-item">
    <a href="/2022/01/19/R-ggplot2%E7%AE%80%E4%BB%8B/" rel="prev" title="R-ggplot2简介">
      <i class="fa fa-chevron-left"></i> R-ggplot2简介
    </a></div>
      <div class="post-nav-item">
    <a href="/2022/01/23/R-TCGA%E6%95%B0%E6%8D%AE%E6%8C%96%E6%8E%98%E5%A4%8D%E7%8E%B0%E4%B9%8BBRCA/" rel="next" title="R-TCGA数据挖掘复现之BRCA">
      R-TCGA数据挖掘复现之BRCA <i class="fa fa-chevron-right"></i>
    </a></div>
    </div>
      </footer>
    
  </article>
  
  
  



          </div>
          
    <div class="comments" id="gitalk-container"></div>

<script>
  window.addEventListener('tabs:register', () => {
    let { activeClass } = CONFIG.comments;
    if (CONFIG.comments.storage) {
      activeClass = localStorage.getItem('comments_active') || activeClass;
    }
    if (activeClass) {
      let activeTab = document.querySelector(`a[href="#comment-${activeClass}"]`);
      if (activeTab) {
        activeTab.click();
      }
    }
  });
  if (CONFIG.comments.storage) {
    window.addEventListener('tabs:click', event => {
      if (!event.target.matches('.tabs-comment .tab-content .tab-pane')) return;
      let commentClass = event.target.classList[1];
      localStorage.setItem('comments_active', commentClass);
    });
  }
</script>

        </div>
          
  
  <div class="toggle sidebar-toggle">
    <span class="toggle-line toggle-line-first"></span>
    <span class="toggle-line toggle-line-middle"></span>
    <span class="toggle-line toggle-line-last"></span>
  </div>

  <aside class="sidebar">
    <div class="sidebar-inner">

      <ul class="sidebar-nav motion-element">
        <li class="sidebar-nav-toc">
          Table of Contents
        </li>
        <li class="sidebar-nav-overview">
          Overview
        </li>
      </ul>

      <!--noindex-->
      <div class="post-toc-wrap sidebar-panel">
          <div class="post-toc motion-element"><ol class="nav"><li class="nav-item nav-level-1"><a class="nav-link" href="#%E5%88%86%E6%9E%90%E6%B5%81%E7%A8%8B%E7%9A%84%E5%9F%BA%E6%9C%AC%E4%BF%A1%E6%81%AF"><span class="nav-number">1.</span> <span class="nav-text">分析流程的基本信息</span></a></li><li class="nav-item nav-level-1"><a class="nav-link" href="#%E6%B5%81%E7%A8%8B"><span class="nav-number">2.</span> <span class="nav-text">流程</span></a></li><li class="nav-item nav-level-1"><a class="nav-link" href="#%E6%95%B0%E6%8D%AE"><span class="nav-number">3.</span> <span class="nav-text">数据</span></a></li><li class="nav-item nav-level-1"><a class="nav-link" href="#%E9%85%8D%E7%BD%AE"><span class="nav-number">4.</span> <span class="nav-text">配置</span></a><ol class="nav-child"><li class="nav-item nav-level-2"><a class="nav-link" href="#%E6%BA%90%E6%96%87%E4%BB%B6%E7%9A%84%E5%AE%89%E8%A3%85%E6%96%B9%E6%B3%95%E4%B9%8B%E4%B8%80"><span class="nav-number">4.1.</span> <span class="nav-text">源文件的安装方法之一</span></a></li></ol></li><li class="nav-item nav-level-1"><a class="nav-link" href="#Step-2-RNA-seq-Pre-processing"><span class="nav-number">5.</span> <span class="nav-text">Step 2 RNA-seq Pre-processing</span></a><ol class="nav-child"><li class="nav-item nav-level-2"><a class="nav-link" href="#%E8%BD%BD%E5%85%A5%E6%95%B0%E6%8D%AE"><span class="nav-number">5.1.</span> <span class="nav-text">载入数据</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#%E6%95%B0%E6%8D%AE%E8%BF%87%E6%BB%A4"><span class="nav-number">5.2.</span> <span class="nav-text">数据过滤</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#%E5%88%9B%E5%BB%BA%E8%A1%A8%E8%BE%BE%E7%9F%A9%E9%98%B5"><span class="nav-number">5.3.</span> <span class="nav-text">创建表达矩阵</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#%E8%B4%A8%E9%87%8F%E6%8E%A7%E5%88%B6"><span class="nav-number">5.4.</span> <span class="nav-text">质量控制</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#%E4%B8%BB%E6%88%90%E5%88%86%E5%88%86%E6%9E%90"><span class="nav-number">5.5.</span> <span class="nav-text">主成分分析</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#%E6%A0%87%E5%87%86%E5%8C%96%E5%A4%84%E7%90%86"><span class="nav-number">5.6.</span> <span class="nav-text">标准化处理</span></a></li></ol></li><li class="nav-item nav-level-1"><a class="nav-link" href="#Step-3-Differential-Expression-for-RNA-seq"><span class="nav-number">6.</span> <span class="nav-text">Step 3: Differential Expression for RNA-seq</span></a><ol class="nav-child"><li class="nav-item nav-level-2"><a class="nav-link" href="#%E9%A2%84%E6%B5%8B%E5%9F%BA%E5%9B%A0%E8%A1%A8%E8%BE%BE%E7%9A%84%E7%A6%BB%E6%95%A3%E5%BA%A6"><span class="nav-number">6.1.</span> <span class="nav-text">预测基因表达的离散度</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#%E5%B7%AE%E5%BC%82%E8%A1%A8%E8%BE%BE%E5%88%86%E6%9E%90"><span class="nav-number">6.2.</span> <span class="nav-text">差异表达分析</span></a></li></ol></li><li class="nav-item nav-level-1"><a class="nav-link" href="#Step-4-Annotation-and-Visualisation-of-RNA-seq-results"><span class="nav-number">7.</span> <span class="nav-text">Step 4:  Annotation and Visualisation of RNA-seq results</span></a><ol class="nav-child"><li class="nav-item nav-level-2"><a class="nav-link" href="#%E5%9F%BA%E5%9B%A0%E6%B3%A8%E9%87%8A"><span class="nav-number">7.1.</span> <span class="nav-text">基因注释</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#%E5%8F%AF%E8%A7%86%E5%8C%96"><span class="nav-number">7.2.</span> <span class="nav-text">可视化</span></a></li></ol></li><li class="nav-item nav-level-1"><a class="nav-link" href="#Step-5-Gene-set-testing"><span class="nav-number">8.</span> <span class="nav-text">Step 5:  Gene-set testing</span></a></li><li class="nav-item nav-level-1"><a class="nav-link" href="#GOseq-analysis-pipeline"><span class="nav-number">9.</span> <span class="nav-text">GOseq analysis pipeline</span></a></li></ol></div>
      </div>
      <!--/noindex-->

      <div class="site-overview-wrap sidebar-panel">
        <div class="site-author motion-element" itemprop="author" itemscope itemtype="http://schema.org/Person">
    <img class="site-author-image" itemprop="image" alt="Dean Li"
      src="https://avatars.githubusercontent.com/u/55836943?v=4">
  <p class="site-author-name" itemprop="name">Dean Li</p>
  <div class="site-description" itemprop="description"></div>
</div>
<div class="site-state-wrap motion-element">
  <nav class="site-state">
      <div class="site-state-item site-state-posts">
          <a href="/archives/">
        
          <span class="site-state-item-count">66</span>
          <span class="site-state-item-name">posts</span>
        </a>
      </div>
      <div class="site-state-item site-state-tags">
            <a href="/tags/">
          
        <span class="site-state-item-count">21</span>
        <span class="site-state-item-name">tags</span></a>
      </div>
  </nav>
</div>
  <div class="links-of-author motion-element">
      <span class="links-of-author-item">
        <a href="https://github.com/thereallda" title="GitHub → https:&#x2F;&#x2F;github.com&#x2F;thereallda" rel="noopener" target="_blank"><i class="fab fa-github fa-fw"></i>GitHub</a>
      </span>
      <span class="links-of-author-item">
        <a href="mailto:lideanlda@foxmail.com" title="E-Mail → mailto:lideanlda@foxmail.com" rel="noopener" target="_blank"><i class="fa fa-envelope fa-fw"></i>E-Mail</a>
      </span>
      <span class="links-of-author-item">
        <a href="https://www.jianshu.com/u/ab8a1c7d0510" title="简书 → https:&#x2F;&#x2F;www.jianshu.com&#x2F;u&#x2F;ab8a1c7d0510" rel="noopener" target="_blank"><i class="fa fa-book fa-fw"></i>简书</a>
      </span>
  </div>



      </div>

    </div>
  </aside>
  <div id="sidebar-dimmer"></div>


      </div>
    </main>

    <footer class="footer">
      <div class="footer-inner">
        

        

<div class="copyright">
  
  &copy; 
  <span itemprop="copyrightYear">2025</span>
  <span class="with-love">
    <i class="fa fa-heart"></i>
  </span>
  <span class="author" itemprop="copyrightHolder">Dean Li</span>
</div>
  <div class="powered-by">Powered by <a href="https://hexo.io/" class="theme-link" rel="noopener" target="_blank">Hexo</a> & <a href="https://theme-next.org/" class="theme-link" rel="noopener" target="_blank">NexT.Gemini</a>
  </div>



        
<div class="busuanzi-count">
  <script async src="https://busuanzi.ibruce.info/busuanzi/2.3/busuanzi.pure.mini.js"></script>
    <span class="post-meta-item" id="busuanzi_container_site_uv" style="display: none;">
      <span class="post-meta-item-icon">
        <i class="fa fa-user"></i>
      </span>
      <span class="site-uv" title="Total Visitors">
        <span id="busuanzi_value_site_uv"></span>
      </span>
    </span>
    <span class="post-meta-divider">|</span>
    <span class="post-meta-item" id="busuanzi_container_site_pv" style="display: none;">
      <span class="post-meta-item-icon">
        <i class="fa fa-eye"></i>
      </span>
      <span class="site-pv" title="Total Views">
        <span id="busuanzi_value_site_pv"></span>
      </span>
    </span>
</div>








      </div>
    </footer>
  </div>

  
  <script src="/lib/anime.min.js"></script>
  <script src="/lib/velocity/velocity.min.js"></script>
  <script src="/lib/velocity/velocity.ui.min.js"></script>

<script src="/js/utils.js"></script>

<script src="/js/motion.js"></script>


<script src="/js/schemes/pisces.js"></script>


<script src="/js/next-boot.js"></script>




  











<script>
document.querySelectorAll('.pdfobject-container').forEach(element => {
  let url = element.dataset.target;
  let pdfOpenParams = {
    navpanes : 0,
    toolbar  : 0,
    statusbar: 0,
    pagemode : 'thumbs',
    view     : 'FitH'
  };
  let pdfOpenFragment = '#' + Object.entries(pdfOpenParams).map(([key, value]) => `${key}=${encodeURIComponent(value)}`).join('&');
  let fullURL = `/lib/pdf/web/viewer.html?file=${encodeURIComponent(url)}${pdfOpenFragment}`;

  if (NexT.utils.supportsPDFs()) {
    element.innerHTML = `<embed class="pdfobject" src="${url + pdfOpenFragment}" type="application/pdf" style="height: ${element.dataset.height};">`;
  } else {
    element.innerHTML = `<iframe src="${fullURL}" style="height: ${element.dataset.height};" frameborder="0"></iframe>`;
  }
});
</script>




  

  

  
<script src="https://cdn.jsdelivr.net/npm/darkmode-js@1.5.7/lib/darkmode-js.min.js"></script>

<script>
var options = {
  bottom: '64px',
  right: '32px',
  left: 'unset',
  time: '0.5s',
  mixColor: 'transparent',
  backgroundColor: 'transparent',
  buttonColorDark: '#100f2c',
  buttonColorLight: '#fff',
  saveInCookies: true,
  label: '🌓',
  autoMatchOsTheme: true
}
const darkmode = new Darkmode(options);
window.darkmode = darkmode;
darkmode.showWidget();
</script>

<link rel="stylesheet" href="//cdn.jsdelivr.net/npm/gitalk@1/dist/gitalk.min.css">

<script>
NexT.utils.loadComments(document.querySelector('#gitalk-container'), () => {
  NexT.utils.getScript('//cdn.jsdelivr.net/npm/gitalk@1/dist/gitalk.min.js', () => {
    var gitalk = new Gitalk({
      clientID    : 'aea502b6092de143578a',
      clientSecret: '7bb9f5afb99d67ab37d8b296f39782af05dc8614',
      repo        : 'thereallda.GitHub.io',
      owner       : 'thereallda',
      admin       : ['thereallda'],
      id          : '1e7d92f33663627422e5eb58f9ca98a9',
        language: '',
      distractionFreeMode: true
    });
    gitalk.render('gitalk-container');
  }, window.Gitalk);
});
</script>

</body>
</html>
