<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
<meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=2">
<meta name="theme-color" content="#222">
<meta name="generator" content="Hexo 5.4.0">
  <link rel="apple-touch-icon" sizes="180x180" href="/images/apple-touch-icon.png">
  <link rel="icon" type="image/png" sizes="32x32" href="/images/favicon-32x32.png">
  <link rel="icon" type="image/png" sizes="16x16" href="/images/favicon-16x16.png">

<link rel="stylesheet" href="/css/main.css">


<link rel="stylesheet" href="/lib/font-awesome/css/all.min.css">

<script id="hexo-configurations">
    var NexT = window.NexT || {};
    var CONFIG = {"hostname":"thereallda.github.io","root":"/","scheme":"Gemini","version":"7.8.0","exturl":false,"sidebar":{"position":"left","display":"post","padding":18,"offset":12,"onmobile":false},"copycode":{"enable":true,"show_result":false,"style":"mac"},"back2top":{"enable":true,"sidebar":false,"scrollpercent":false},"bookmark":{"enable":false,"color":"#222","save":"auto"},"fancybox":false,"mediumzoom":false,"lazyload":false,"pangu":false,"comments":{"style":"tabs","active":"gitalk","storage":true,"lazyload":false,"nav":null,"activeClass":"gitalk"},"algolia":{"hits":{"per_page":10},"labels":{"input_placeholder":"Search for Posts","hits_empty":"We didn't find any results for the search: ${query}","hits_stats":"${hits} results found in ${time} ms"}},"localsearch":{"enable":false,"trigger":"auto","top_n_per_article":1,"unescape":false,"preload":false},"motion":{"enable":true,"async":false,"transition":{"post_block":"fadeIn","post_header":"slideDownIn","post_body":"slideDownIn","coll_header":"slideLeftIn","sidebar":"slideUpIn"}}};
  </script>

  <meta name="description" content="CytoTRACE2CytoTRACE2是 Newman Lab 开发的用于预测scRNA-seq数据中细胞干性和发育潜能的计算方法。 CytoTRACE2本质是一种深度学习框架，该模型在涵盖 28 种组织类型和整个发育范围的人类和小鼠 scRNA-seq 数据集的训练和验证（总共31个数据）。利用该模型，CytoTRACE2可以将细胞分类为终末分化（differentiated: 0）和全能细胞">
<meta property="og:type" content="article">
<meta property="og:title" content="CytoTRACE2-basic">
<meta property="og:url" content="https://thereallda.github.io/2024/12/10/cytotrace2-basic/index.html">
<meta property="og:site_name" content="Dean&#39;s blog">
<meta property="og:description" content="CytoTRACE2CytoTRACE2是 Newman Lab 开发的用于预测scRNA-seq数据中细胞干性和发育潜能的计算方法。 CytoTRACE2本质是一种深度学习框架，该模型在涵盖 28 种组织类型和整个发育范围的人类和小鼠 scRNA-seq 数据集的训练和验证（总共31个数据）。利用该模型，CytoTRACE2可以将细胞分类为终末分化（differentiated: 0）和全能细胞">
<meta property="og:locale" content="en_US">
<meta property="og:image" content="https://thereallda.github.io/2024/12/10/cytotrace2-basic/clipboard-768829770.png">
<meta property="og:image" content="https://thereallda.github.io/2024/12/10/cytotrace2-basic/unnamed-chunk-5-1.png">
<meta property="og:image" content="https://thereallda.github.io/2024/12/10/cytotrace2-basic/unnamed-chunk-6-1.png">
<meta property="og:image" content="https://thereallda.github.io/2024/12/10/cytotrace2-basic/unnamed-chunk-7-1.png">
<meta property="og:image" content="https://thereallda.github.io/2024/12/10/cytotrace2-basic/unnamed-chunk-8-1.png">
<meta property="og:image" content="https://thereallda.github.io/2024/12/10/cytotrace2-basic/unnamed-chunk-9-1.png">
<meta property="og:image" content="https://thereallda.github.io/2024/12/10/cytotrace2-basic/unnamed-chunk-12-1.png">
<meta property="og:image" content="https://thereallda.github.io/2024/12/10/cytotrace2-basic/unnamed-chunk-13-1.png">
<meta property="og:image" content="https://thereallda.github.io/2024/12/10/cytotrace2-basic/unnamed-chunk-14-1.png">
<meta property="article:published_time" content="2024-12-10T20:21:00.000Z">
<meta property="article:modified_time" content="2024-12-10T20:22:51.856Z">
<meta property="article:author" content="Dean Li">
<meta property="article:tag" content="R">
<meta property="article:tag" content="single-cell">
<meta name="twitter:card" content="summary">
<meta name="twitter:image" content="https://thereallda.github.io/2024/12/10/cytotrace2-basic/clipboard-768829770.png">

<link rel="canonical" href="https://thereallda.github.io/2024/12/10/cytotrace2-basic/">


<script id="page-configurations">
  // https://hexo.io/docs/variables.html
  CONFIG.page = {
    sidebar: "",
    isHome : false,
    isPost : true,
    lang   : 'en'
  };
</script>

  <title>CytoTRACE2-basic | Dean's blog</title>
  






  <noscript>
  <style>
  .use-motion .brand,
  .use-motion .menu-item,
  .sidebar-inner,
  .use-motion .post-block,
  .use-motion .pagination,
  .use-motion .comments,
  .use-motion .post-header,
  .use-motion .post-body,
  .use-motion .collection-header { opacity: initial; }

  .use-motion .site-title,
  .use-motion .site-subtitle {
    opacity: initial;
    top: initial;
  }

  .use-motion .logo-line-before i { left: initial; }
  .use-motion .logo-line-after i { right: initial; }
  </style>
</noscript>

<style>.darkmode--activated{--body-bg-color:#282828;--content-bg-color:#333;--card-bg-color:#555;--text-color:#ccc;--blockquote-color:#bbb;--link-color:#ccc;--link-hover-color:#eee;--brand-color:#ddd;--brand-hover-color:#ddd;--table-row-odd-bg-color:#282828;--table-row-hover-bg-color:#363636;--menu-item-bg-color:#555;--btn-default-bg:#222;--btn-default-color:#ccc;--btn-default-border-color:#555;--btn-default-hover-bg:#666;--btn-default-hover-color:#ccc;--btn-default-hover-border-color:#666;--highlight-background:#282b2e;--highlight-foreground:#a9b7c6;--highlight-gutter-background:#34393d;--highlight-gutter-foreground:#9ca9b6}.darkmode--activated img{opacity:.75}.darkmode--activated img:hover{opacity:.9}.darkmode--activated code{color:#69dbdc;background:0 0}button.darkmode-toggle{z-index:9999}.darkmode-ignore,img{display:flex!important}</style></head>

<body itemscope itemtype="http://schema.org/WebPage">
  <div class="container use-motion">
    <div class="headband"></div>

    <header class="header" itemscope itemtype="http://schema.org/WPHeader">
      <div class="header-inner"><div class="site-brand-container">
  <div class="site-nav-toggle">
    <div class="toggle" aria-label="Toggle navigation bar">
      <span class="toggle-line toggle-line-first"></span>
      <span class="toggle-line toggle-line-middle"></span>
      <span class="toggle-line toggle-line-last"></span>
    </div>
  </div>

  <div class="site-meta">

    <a href="/" class="brand" rel="start">
      <span class="logo-line-before"><i></i></span>
      <h1 class="site-title">Dean's blog</h1>
      <span class="logo-line-after"><i></i></span>
    </a>
  </div>

  <div class="site-nav-right">
    <div class="toggle popup-trigger">
    </div>
  </div>
</div>




<nav class="site-nav">
  <ul id="menu" class="main-menu menu">
        <li class="menu-item menu-item-home">

    <a href="/" rel="section"><i class="fa fa-home fa-fw"></i>Home</a>

  </li>
        <li class="menu-item menu-item-about">

    <a href="/about/" rel="section"><i class="fa fa-user fa-fw"></i>About</a>

  </li>
        <li class="menu-item menu-item-tags">

    <a href="/tags/" rel="section"><i class="fa fa-tags fa-fw"></i>Tags</a>

  </li>
        <li class="menu-item menu-item-archives">

    <a href="/archives/" rel="section"><i class="fa fa-archive fa-fw"></i>Archives</a>

  </li>
  </ul>
</nav>




</div>
    </header>

    
  <div class="back-to-top">
    <i class="fa fa-arrow-up"></i>
    <span>0%</span>
  </div>
  <div class="reading-progress-bar"></div>


    <main class="main">
      <div class="main-inner">
        <div class="content-wrap">
          

          <div class="content post posts-expand">
            

    
  
  
  <article itemscope itemtype="http://schema.org/Article" class="post-block" lang="en">
    <link itemprop="mainEntityOfPage" href="https://thereallda.github.io/2024/12/10/cytotrace2-basic/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="image" content="https://avatars.githubusercontent.com/u/55836943?v=4">
      <meta itemprop="name" content="Dean Li">
      <meta itemprop="description" content="">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="Dean's blog">
    </span>
      <header class="post-header">
        <h1 class="post-title" itemprop="name headline">
          CytoTRACE2-basic
        </h1>

        <div class="post-meta">
          
          
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-calendar"></i>
              </span>
              <span class="post-meta-item-text">Posted on</span>
              

              <time title="Created: 2024-12-10 15:21:00 / Modified: 15:22:51" itemprop="dateCreated datePublished" datetime="2024-12-10T15:21:00-05:00">2024-12-10</time>
            </span>

          
            <span class="post-meta-item" title="Views" id="busuanzi_container_page_pv" style="display: none;">
              <span class="post-meta-item-icon">
                <i class="fa fa-eye"></i>
              </span>
              <span class="post-meta-item-text">Views: </span>
              <span id="busuanzi_value_page_pv"></span>
            </span><br>
            <span class="post-meta-item" title="Symbols count in article">
              <span class="post-meta-item-icon">
                <i class="far fa-file-word"></i>
              </span>
                <span class="post-meta-item-text">Symbols count in article: </span>
              <span>13k</span>
            </span>
            <span class="post-meta-item" title="Reading time">
              <span class="post-meta-item-icon">
                <i class="far fa-clock"></i>
              </span>
                <span class="post-meta-item-text">Reading time &asymp;</span>
              <span>12 mins.</span>
            </span>

        </div>
      </header>

    
    
    
    <div class="post-body" itemprop="articleBody">

      
        <h2 id="CytoTRACE2"><a href="#CytoTRACE2" class="headerlink" title="CytoTRACE2"></a>CytoTRACE2</h2><p>CytoTRACE2是 <a target="_blank" rel="noopener" href="https://anlab.stanford.edu/">Newman Lab</a> 开发的用于预测scRNA-seq数据中细胞干性和发育潜能的计算方法。</p>
<p>CytoTRACE2本质是一种深度学习框架，该模型在涵盖 28 种组织类型和整个发育范围的人类和小鼠 scRNA-seq 数据集的训练和验证（总共31个数据）。利用该模型，CytoTRACE2可以将细胞分类为终末分化（differentiated: 0）和全能细胞（totipotent: 1）.</p>
<p><img src="/2024/12/10/cytotrace2-basic/clipboard-768829770.png"></p>
<p>本文简单介绍CytoTRACE2 R包的基本用法，具体原理可参考他们目前的预印本 <a target="_blank" rel="noopener" href="https://doi.org/10.1101/2024.03.19.585637">https://doi.org/10.1101/2024.03.19.585637</a></p>
<span id="more"></span>

<h2 id="安装"><a href="#安装" class="headerlink" title="安装"></a>安装</h2><figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">devtools::install_github(<span class="string">&quot;digitalcytometry/cytotrace2&quot;</span>, subdir = <span class="string">&quot;cytotrace2_r&quot;</span>) <span class="comment">#installing</span></span><br></pre></td></tr></table></figure>

<h2 id="数据准备"><a href="#数据准备" class="headerlink" title="数据准备"></a>数据准备</h2><p>CytoTRACE2需要的输入数据包括单细胞表达数据，这里利用一个已发表的胰腺的10X scRNA-seq作为示例 （<a target="_blank" rel="noopener" href="https://doi.org/10.1242/dev.173849">Bastidas-Ponce et al., 2019</a>）</p>
<p>这里下载CytoTRACE2提供的示例数据集，包括了：</p>
<ul>
<li><p>  表达数据（<code>expression_data</code>）：小鼠胰腺上皮的2280细胞的10X scRNA-seq数据。该数据已经进行归一化处理</p>
</li>
<li><p>  注释数据（<code>annotation</code>）：表型注释</p>
</li>
</ul>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># download the .rds file (this will download the file to your working directory)</span></span><br><span class="line">download.file(<span class="string">&quot;https://drive.google.com/uc?export=download&amp;id=1ivi9TBlmzVTDGzNWQrXXeyL68Wug989K&quot;</span>, <span class="string">&quot;Pancreas_10x_downsampled.rds&quot;</span>)</span><br></pre></td></tr></table></figure>

<h2 id="运行CytoTRACE2"><a href="#运行CytoTRACE2" class="headerlink" title="运行CytoTRACE2"></a>运行CytoTRACE2</h2><p>运行 <code>cytotrace2()</code> 进行细胞干性推断，函数运行时执行以下操作：</p>
<ol>
<li><p> 预处理数据：在每个细胞中对表达数据进行降序排序，使用排序后的秩（rank）进行预测</p>
</li>
<li><p>预测细胞发育状态：读入模型预训练参数进行预测</p>
<blockquote>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line"># full model</span><br><span class="line">parameter_dict &lt;- readRDS(system.file(&quot;extdata&quot;, &quot;parameter_dict_17.rds&quot;, package = &quot;CytoTRACE2&quot;))</span><br><span class="line"># else </span><br><span class="line">parameter_dict &lt;- readRDS(system.file(&quot;extdata&quot;, &quot;parameter_dict_5_best.rds&quot;, package = &quot;CytoTRACE2&quot;))</span><br></pre></td></tr></table></figure></blockquote>
</li>
<li><p> 后处理预测结果：利用diffusion方法平滑预测值，并利用kNN和binning的方法重新正则化预测值</p>
</li>
</ol>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br><span class="line">69</span><br><span class="line">70</span><br><span class="line">71</span><br><span class="line">72</span><br><span class="line">73</span><br><span class="line">74</span><br><span class="line">75</span><br><span class="line">76</span><br><span class="line">77</span><br><span class="line">78</span><br><span class="line">79</span><br><span class="line">80</span><br><span class="line">81</span><br><span class="line">82</span><br><span class="line">83</span><br><span class="line">84</span><br><span class="line">85</span><br><span class="line">86</span><br><span class="line">87</span><br><span class="line">88</span><br><span class="line">89</span><br><span class="line">90</span><br><span class="line">91</span><br><span class="line">92</span><br><span class="line">93</span><br><span class="line">94</span><br><span class="line">95</span><br><span class="line">96</span><br><span class="line">97</span><br><span class="line">98</span><br><span class="line">99</span><br><span class="line">100</span><br><span class="line">101</span><br><span class="line">102</span><br><span class="line">103</span><br><span class="line">104</span><br><span class="line">105</span><br><span class="line">106</span><br><span class="line">107</span><br><span class="line">108</span><br><span class="line">109</span><br><span class="line">110</span><br><span class="line">111</span><br><span class="line">112</span><br><span class="line">113</span><br><span class="line">114</span><br><span class="line">115</span><br><span class="line">116</span><br><span class="line">117</span><br><span class="line">118</span><br><span class="line">119</span><br><span class="line">120</span><br><span class="line">121</span><br><span class="line">122</span><br><span class="line">123</span><br><span class="line">124</span><br><span class="line">125</span><br><span class="line">126</span><br><span class="line">127</span><br><span class="line">128</span><br><span class="line">129</span><br><span class="line">130</span><br><span class="line">131</span><br><span class="line">132</span><br><span class="line">133</span><br><span class="line">134</span><br><span class="line">135</span><br><span class="line">136</span><br><span class="line">137</span><br><span class="line">138</span><br><span class="line">139</span><br><span class="line">140</span><br><span class="line">141</span><br><span class="line">142</span><br><span class="line">143</span><br><span class="line">144</span><br><span class="line">145</span><br><span class="line">146</span><br><span class="line">147</span><br><span class="line">148</span><br><span class="line">149</span><br><span class="line">150</span><br><span class="line">151</span><br><span class="line">152</span><br><span class="line">153</span><br><span class="line">154</span><br><span class="line">155</span><br><span class="line">156</span><br><span class="line">157</span><br><span class="line">158</span><br><span class="line">159</span><br><span class="line">160</span><br><span class="line">161</span><br><span class="line">162</span><br><span class="line">163</span><br><span class="line">164</span><br><span class="line">165</span><br><span class="line">166</span><br><span class="line">167</span><br><span class="line">168</span><br><span class="line">169</span><br><span class="line">170</span><br><span class="line">171</span><br><span class="line">172</span><br><span class="line">173</span><br><span class="line">174</span><br><span class="line">175</span><br><span class="line">176</span><br><span class="line">177</span><br><span class="line">178</span><br><span class="line">179</span><br><span class="line">180</span><br><span class="line">181</span><br><span class="line">182</span><br><span class="line">183</span><br><span class="line">184</span><br><span class="line">185</span><br><span class="line">186</span><br><span class="line">187</span><br><span class="line">188</span><br><span class="line">189</span><br><span class="line">190</span><br><span class="line">191</span><br><span class="line">192</span><br><span class="line">193</span><br><span class="line">194</span><br><span class="line">195</span><br><span class="line">196</span><br><span class="line">197</span><br><span class="line">198</span><br><span class="line">199</span><br><span class="line">200</span><br><span class="line">201</span><br><span class="line">202</span><br><span class="line">203</span><br><span class="line">204</span><br><span class="line">205</span><br><span class="line">206</span><br><span class="line">207</span><br><span class="line">208</span><br><span class="line">209</span><br><span class="line">210</span><br><span class="line">211</span><br><span class="line">212</span><br><span class="line">213</span><br><span class="line">214</span><br><span class="line">215</span><br><span class="line">216</span><br><span class="line">217</span><br><span class="line">218</span><br><span class="line">219</span><br><span class="line">220</span><br><span class="line">221</span><br></pre></td><td class="code"><pre><span class="line">library(CytoTRACE2) <span class="comment">#loading</span></span><br><span class="line"></span><br><span class="line"><span class="comment">## Loading required package: data.table</span></span><br><span class="line"></span><br><span class="line"><span class="comment">## Loading required package: doParallel</span></span><br><span class="line"></span><br><span class="line"><span class="comment">## Loading required package: foreach</span></span><br><span class="line"></span><br><span class="line"><span class="comment">## Loading required package: iterators</span></span><br><span class="line"></span><br><span class="line"><span class="comment">## Loading required package: parallel</span></span><br><span class="line"></span><br><span class="line"><span class="comment">## Loading required package: dplyr</span></span><br><span class="line"></span><br><span class="line"><span class="comment">## </span></span><br><span class="line"><span class="comment">## Attaching package: &#x27;dplyr&#x27;</span></span><br><span class="line"></span><br><span class="line"><span class="comment">## The following objects are masked from &#x27;package:data.table&#x27;:</span></span><br><span class="line"><span class="comment">## </span></span><br><span class="line"><span class="comment">##     between, first, last</span></span><br><span class="line"></span><br><span class="line"><span class="comment">## The following objects are masked from &#x27;package:stats&#x27;:</span></span><br><span class="line"><span class="comment">## </span></span><br><span class="line"><span class="comment">##     filter, lag</span></span><br><span class="line"></span><br><span class="line"><span class="comment">## The following objects are masked from &#x27;package:base&#x27;:</span></span><br><span class="line"><span class="comment">## </span></span><br><span class="line"><span class="comment">##     intersect, setdiff, setequal, union</span></span><br><span class="line"></span><br><span class="line"><span class="comment">## Loading required package: ggplot2</span></span><br><span class="line"></span><br><span class="line"><span class="comment">## Loading required package: HiClimR</span></span><br><span class="line"></span><br><span class="line"><span class="comment">## Loading required package: magrittr</span></span><br><span class="line"></span><br><span class="line"><span class="comment">## Loading required package: Matrix</span></span><br><span class="line"></span><br><span class="line"><span class="comment">## Loading required package: plyr</span></span><br><span class="line"></span><br><span class="line"><span class="comment">## ------------------------------------------------------------------------------</span></span><br><span class="line"></span><br><span class="line"><span class="comment">## You have loaded plyr after dplyr - this is likely to cause problems.</span></span><br><span class="line"><span class="comment">## If you need functions from both plyr and dplyr, please load plyr first, then dplyr:</span></span><br><span class="line"><span class="comment">## library(plyr); library(dplyr)</span></span><br><span class="line"></span><br><span class="line"><span class="comment">## ------------------------------------------------------------------------------</span></span><br><span class="line"></span><br><span class="line"><span class="comment">## </span></span><br><span class="line"><span class="comment">## Attaching package: &#x27;plyr&#x27;</span></span><br><span class="line"></span><br><span class="line"><span class="comment">## The following objects are masked from &#x27;package:dplyr&#x27;:</span></span><br><span class="line"><span class="comment">## </span></span><br><span class="line"><span class="comment">##     arrange, count, desc, failwith, id, mutate, rename, summarise,</span></span><br><span class="line"><span class="comment">##     summarize</span></span><br><span class="line"></span><br><span class="line"><span class="comment">## Loading required package: RANN</span></span><br><span class="line"></span><br><span class="line"><span class="comment">## Loading required package: Rfast</span></span><br><span class="line"></span><br><span class="line"><span class="comment">## Loading required package: Rcpp</span></span><br><span class="line"></span><br><span class="line"><span class="comment">## Loading required package: RcppZiggurat</span></span><br><span class="line"></span><br><span class="line"><span class="comment">## Loading required package: RcppParallel</span></span><br><span class="line"></span><br><span class="line"><span class="comment">## </span></span><br><span class="line"><span class="comment">## Attaching package: &#x27;RcppParallel&#x27;</span></span><br><span class="line"></span><br><span class="line"><span class="comment">## The following object is masked from &#x27;package:Rcpp&#x27;:</span></span><br><span class="line"><span class="comment">## </span></span><br><span class="line"><span class="comment">##     LdFlags</span></span><br><span class="line"></span><br><span class="line"><span class="comment">## </span></span><br><span class="line"><span class="comment">## Rfast: 2.1.0</span></span><br><span class="line"></span><br><span class="line"><span class="comment">##  ___ __ __ __ __    __ __ __ __ __ _             _               __ __ __ __ __     __ __ __ __ __ __   </span></span><br><span class="line"><span class="comment">## |  __ __ __ __  |  |  __ __ __ __ _/            / \             |  __ __ __ __ /   /__ __ _   _ __ __\  </span></span><br><span class="line"><span class="comment">## | |           | |  | |                         / _ \            | |                        / /          </span></span><br><span class="line"><span class="comment">## | |           | |  | |                        / / \ \           | |                       / /          </span></span><br><span class="line"><span class="comment">## | |           | |  | |                       / /   \ \          | |                      / /          </span></span><br><span class="line"><span class="comment">## | |__ __ __ __| |  | |__ __ __ __           / /     \ \         | |__ __ __ __ _        / /__/\          </span></span><br><span class="line"><span class="comment">## |    __ __ __ __|  |  __ __ __ __|         / /__ _ __\ \        |_ __ __ __ _   |      / ___  /           </span></span><br><span class="line"><span class="comment">## |   \              | |                    / _ _ _ _ _ _ \                     | |      \/  / /       </span></span><br><span class="line"><span class="comment">## | |\ \             | |                   / /           \ \                    | |         / /          </span></span><br><span class="line"><span class="comment">## | | \ \            | |                  / /             \ \                   | |        / /          </span></span><br><span class="line"><span class="comment">## | |  \ \           | |                 / /               \ \                  | |       / /          </span></span><br><span class="line"><span class="comment">## | |   \ \__ __ _   | |                / /                 \ \     _ __ __ __ _| |      / /          </span></span><br><span class="line"><span class="comment">## |_|    \__ __ __\  |_|               /_/                   \_\   /_ __ __ __ ___|      \/             team</span></span><br><span class="line"></span><br><span class="line"><span class="comment">## </span></span><br><span class="line"><span class="comment">## Attaching package: &#x27;Rfast&#x27;</span></span><br><span class="line"></span><br><span class="line"><span class="comment">## The following object is masked from &#x27;package:dplyr&#x27;:</span></span><br><span class="line"><span class="comment">## </span></span><br><span class="line"><span class="comment">##     nth</span></span><br><span class="line"></span><br><span class="line"><span class="comment">## The following object is masked from &#x27;package:data.table&#x27;:</span></span><br><span class="line"><span class="comment">## </span></span><br><span class="line"><span class="comment">##     transpose</span></span><br><span class="line"></span><br><span class="line"><span class="comment">## Loading required package: Seurat</span></span><br><span class="line"></span><br><span class="line"><span class="comment">## Loading required package: SeuratObject</span></span><br><span class="line"></span><br><span class="line"><span class="comment">## Loading required package: sp</span></span><br><span class="line"></span><br><span class="line"><span class="comment">## &#x27;SeuratObject&#x27; was built under R 4.4.0 but the current version is</span></span><br><span class="line"><span class="comment">## 4.4.1; it is recomended that you reinstall &#x27;SeuratObject&#x27; as the ABI</span></span><br><span class="line"><span class="comment">## for R may have changed</span></span><br><span class="line"></span><br><span class="line"><span class="comment">## </span></span><br><span class="line"><span class="comment">## Attaching package: &#x27;SeuratObject&#x27;</span></span><br><span class="line"></span><br><span class="line"><span class="comment">## The following objects are masked from &#x27;package:base&#x27;:</span></span><br><span class="line"><span class="comment">## </span></span><br><span class="line"><span class="comment">##     intersect, t</span></span><br><span class="line"></span><br><span class="line"><span class="comment">## Loading required package: stringr</span></span><br><span class="line"></span><br><span class="line"><span class="comment">## Warning: replacing previous import &#x27;data.table::first&#x27; by &#x27;dplyr::first&#x27; when</span></span><br><span class="line"><span class="comment">## loading &#x27;CytoTRACE2&#x27;</span></span><br><span class="line"></span><br><span class="line"><span class="comment">## Warning: replacing previous import &#x27;data.table::last&#x27; by &#x27;dplyr::last&#x27; when</span></span><br><span class="line"><span class="comment">## loading &#x27;CytoTRACE2&#x27;</span></span><br><span class="line"></span><br><span class="line"><span class="comment">## Warning: replacing previous import &#x27;data.table::between&#x27; by &#x27;dplyr::between&#x27;</span></span><br><span class="line"><span class="comment">## when loading &#x27;CytoTRACE2&#x27;</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># load rds</span></span><br><span class="line">scdata &lt;- readRDS(<span class="string">&quot;Pancreas_10x_downsampled.rds&quot;</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment"># extract expression data</span></span><br><span class="line">expression_data &lt;- scdata$expression_data</span><br><span class="line"><span class="built_in">dim</span>(expression_data)</span><br><span class="line"></span><br><span class="line"><span class="comment">## [1] 17326  2280</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># running CytoTRACE 2 main function - cytotrace2 - with default parameters</span></span><br><span class="line">cytotrace2_result &lt;- cytotrace2(expression_data, </span><br><span class="line">                                species = <span class="string">&#x27;mouse&#x27;</span>,</span><br><span class="line">                                is_seurat = <span class="built_in">F</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment">## cytotrace2: Started loading data</span></span><br><span class="line"></span><br><span class="line"><span class="comment">## Dataset contains 17326 genes and 2280 cells.</span></span><br><span class="line"></span><br><span class="line"><span class="comment">## The passed subsample size is greater than the number of cells in dataset.</span></span><br><span class="line"><span class="comment">## Now setting subsample size to 2280</span></span><br><span class="line"></span><br><span class="line"><span class="comment">## cytotrace2: Running on 1 subsample(s) approximately of length 2280</span></span><br><span class="line"></span><br><span class="line"><span class="comment">## cytotrace2: Started running on subsample(s). This will take a few minutes.</span></span><br><span class="line"></span><br><span class="line"><span class="comment">## cytotrace2: Started preprocessing.</span></span><br><span class="line"></span><br><span class="line"><span class="comment">## 13191 input genes mapped to model genes.</span></span><br><span class="line"></span><br><span class="line"><span class="comment">## cytotrace2: Started prediction.</span></span><br><span class="line"></span><br><span class="line"><span class="comment">## This section will run using  5 / 11 core(s).</span></span><br><span class="line"></span><br><span class="line"><span class="comment">## cytotrace2: Started postprocessing.</span></span><br><span class="line"></span><br><span class="line"><span class="comment">## cytotrace2: Running with fast mode (subsamples are processed in parallel)</span></span><br><span class="line"></span><br><span class="line"><span class="comment">## This section will run on 2 sub-sample(s) of approximately 1140 cells each using 2 / 11 core(s).</span></span><br><span class="line"></span><br><span class="line"><span class="comment">## cytotrace2: Finished</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># extract annotation data</span></span><br><span class="line">annotation &lt;- scdata$annotation</span><br><span class="line">head(annotation)</span><br><span class="line"></span><br><span class="line"><span class="comment">##                        phenotype1 phenotype2        potency</span></span><br><span class="line"><span class="comment">## CACCAGGCAGGTGCCT-1-0   Alpha cell      Alpha Differentiated</span></span><br><span class="line"><span class="comment">## CCCAGTTAGGGAACGG-1-2   Alpha cell      Alpha Differentiated</span></span><br><span class="line"><span class="comment">## CGGACTGCATCTACGA-1-2 Epsilon cell    Epsilon Differentiated</span></span><br><span class="line"><span class="comment">## TACCTTAGTTCCAACA-1-2   Alpha cell      Alpha Differentiated</span></span><br><span class="line"><span class="comment">## TACTTACCAGACAGGT-1-2   Alpha cell      Alpha Differentiated</span></span><br><span class="line"><span class="comment">## ATGAGGGGTGGTACAG-1-3    Beta cell       Beta Differentiated</span></span><br><span class="line"><span class="comment">##                      absolute_granular_ordering relative_ordering</span></span><br><span class="line"><span class="comment">## CACCAGGCAGGTGCCT-1-0                         23                 6</span></span><br><span class="line"><span class="comment">## CCCAGTTAGGGAACGG-1-2                         23                 6</span></span><br><span class="line"><span class="comment">## CGGACTGCATCTACGA-1-2                         23                 6</span></span><br><span class="line"><span class="comment">## TACCTTAGTTCCAACA-1-2                         23                 6</span></span><br><span class="line"><span class="comment">## TACTTACCAGACAGGT-1-2                         23                 6</span></span><br><span class="line"><span class="comment">## ATGAGGGGTGGTACAG-1-3                         23                 6</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># generate prediction and phenotype association plots with plotData function</span></span><br><span class="line">plots &lt;- plotData(cytotrace2_result = cytotrace2_result, </span><br><span class="line">                  annotation = annotation,</span><br><span class="line">                  expression_data = expression_data</span><br><span class="line">                  )</span><br><span class="line"></span><br><span class="line"><span class="comment">## Preparing input for visualization.</span></span><br><span class="line"></span><br><span class="line"><span class="comment">## Creating plots.</span></span><br><span class="line"></span><br><span class="line"><span class="comment">## Done. You can access any plot directly from the returned list by &#x27;$&#x27; operator (i.e. plots$CytoTRACE2_Potency_UMAP).</span></span><br><span class="line"></span><br><span class="line">table(annotation$phenotype1)</span><br><span class="line"></span><br><span class="line"><span class="comment">## </span></span><br><span class="line"><span class="comment">##                        Alpha cell                         Beta cell </span></span><br><span class="line"><span class="comment">##                               276                               258 </span></span><br><span class="line"><span class="comment">##                        Delta cell          Endocrine precursor cell </span></span><br><span class="line"><span class="comment">##                                20                               509 </span></span><br><span class="line"><span class="comment">##              Endocrine progenitor                      Epsilon cell </span></span><br><span class="line"><span class="comment">##                               499                                46 </span></span><br><span class="line"><span class="comment">##           Immature endocrine cell Multipotent pancreatic progenitor </span></span><br><span class="line"><span class="comment">##                               427                               245</span></span><br><span class="line"></span><br><span class="line">table(cytotrace2_result$CytoTRACE2_Potency)</span><br><span class="line"></span><br><span class="line"><span class="comment">## </span></span><br><span class="line"><span class="comment">## Differentiated      Unipotent    Oligopotent    Multipotent    Pluripotent </span></span><br><span class="line"><span class="comment">##            952            598            501            229              0 </span></span><br><span class="line"><span class="comment">##     Totipotent </span></span><br><span class="line"><span class="comment">##              0</span></span><br><span class="line"></span><br><span class="line">plots$CytoTRACE2_UMAP</span><br></pre></td></tr></table></figure>

<p><img src="/2024/12/10/cytotrace2-basic/unnamed-chunk-5-1.png"></p>
<h3 id="Potency-score-distribution-by-phenotype"><a href="#Potency-score-distribution-by-phenotype" class="headerlink" title="Potency score distribution by phenotype"></a><em><strong>Potency score distribution by phenotype</strong></em></h3><figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">plots$CytoTRACE2_Boxplot_byPheno</span><br></pre></td></tr></table></figure>

<p><img src="/2024/12/10/cytotrace2-basic/unnamed-chunk-6-1.png"></p>
<h3 id="Potency-category"><a href="#Potency-category" class="headerlink" title="Potency category"></a><em><strong>Potency category</strong></em></h3><figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">plots$CytoTRACE2_Potency_UMAP</span><br></pre></td></tr></table></figure>

<p><img src="/2024/12/10/cytotrace2-basic/unnamed-chunk-7-1.png"></p>
<h3 id="Relative-order"><a href="#Relative-order" class="headerlink" title="Relative order"></a><em><strong>Relative order</strong></em></h3><figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">plots$CytoTRACE2_Relative_UMAP</span><br></pre></td></tr></table></figure>

<p><img src="/2024/12/10/cytotrace2-basic/unnamed-chunk-8-1.png"></p>
<h3 id="Phenotype"><a href="#Phenotype" class="headerlink" title="Phenotype"></a>Phenotype</h3><figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">plots$Phenotype_UMAP</span><br></pre></td></tr></table></figure>

<p><img src="/2024/12/10/cytotrace2-basic/unnamed-chunk-9-1.png"></p>
<h3 id="Seurat-object-input"><a href="#Seurat-object-input" class="headerlink" title="Seurat object input"></a>Seurat object input</h3><figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><span class="line">library(Seurat)</span><br><span class="line">seu &lt;- CreateSeuratObject(expression_data, meta.data = annotation)</span><br><span class="line"></span><br><span class="line"><span class="comment">## Warning: Data is of class data.frame. Coercing to dgCMatrix.</span></span><br><span class="line"></span><br><span class="line">seu</span><br><span class="line"></span><br><span class="line"><span class="comment">## An object of class Seurat </span></span><br><span class="line"><span class="comment">## 17326 features across 2280 samples within 1 assay </span></span><br><span class="line"><span class="comment">## Active assay: RNA (17326 features, 0 variable features)</span></span><br><span class="line"><span class="comment">##  1 layer present: counts</span></span><br></pre></td></tr></table></figure>

<p>run cytotrace2</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br></pre></td><td class="code"><pre><span class="line">cytotrace2_result &lt;- cytotrace2(seu, is_seurat = <span class="literal">TRUE</span>, slot_type = <span class="string">&quot;counts&quot;</span>, species = <span class="string">&#x27;mouse&#x27;</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment">## cytotrace2: Started loading data</span></span><br><span class="line"></span><br><span class="line"><span class="comment">## Dataset contains 17326 genes and 2280 cells.</span></span><br><span class="line"></span><br><span class="line"><span class="comment">## The passed subsample size is greater than the number of cells in dataset.</span></span><br><span class="line"><span class="comment">## Now setting subsample size to 2280</span></span><br><span class="line"></span><br><span class="line"><span class="comment">## cytotrace2: Running on 1 subsample(s) approximately of length 2280</span></span><br><span class="line"></span><br><span class="line"><span class="comment">## cytotrace2: Started running on subsample(s). This will take a few minutes.</span></span><br><span class="line"></span><br><span class="line"><span class="comment">## cytotrace2: Started preprocessing.</span></span><br><span class="line"></span><br><span class="line"><span class="comment">## 13191 input genes mapped to model genes.</span></span><br><span class="line"></span><br><span class="line"><span class="comment">## cytotrace2: Started prediction.</span></span><br><span class="line"></span><br><span class="line"><span class="comment">## This section will run using  5 / 11 core(s).</span></span><br><span class="line"></span><br><span class="line"><span class="comment">## cytotrace2: Started postprocessing.</span></span><br><span class="line"></span><br><span class="line"><span class="comment">## cytotrace2: Running with fast mode (subsamples are processed in parallel)</span></span><br><span class="line"></span><br><span class="line"><span class="comment">## This section will run on 2 sub-sample(s) of approximately 1140 cells each using 2 / 11 core(s).</span></span><br><span class="line"></span><br><span class="line"><span class="comment">## cytotrace2: Finished</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># plotting</span></span><br><span class="line">plots &lt;- plotData(cytotrace2_result = cytotrace2_result, </span><br><span class="line">                  annotation = annotation, </span><br><span class="line">                  is_seurat = <span class="literal">TRUE</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment">## Preparing input for visualization.</span></span><br><span class="line"></span><br><span class="line"><span class="comment">## Creating plots.</span></span><br><span class="line"></span><br><span class="line"><span class="comment">## Done. You can access any plot directly from the returned list by &#x27;$&#x27; operator (i.e. plots$CytoTRACE2_Potency_UMAP).</span></span><br></pre></td></tr></table></figure>

<p>可视化</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">plots$CytoTRACE2_Potency_UMAP</span><br></pre></td></tr></table></figure>

<p><img src="/2024/12/10/cytotrace2-basic/unnamed-chunk-12-1.png"></p>
<p>也可以用seurat流程跑完后的对象进行UMAP可视化</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br><span class="line">69</span><br><span class="line">70</span><br><span class="line">71</span><br><span class="line">72</span><br><span class="line">73</span><br><span class="line">74</span><br><span class="line">75</span><br><span class="line">76</span><br><span class="line">77</span><br><span class="line">78</span><br><span class="line">79</span><br><span class="line">80</span><br><span class="line">81</span><br><span class="line">82</span><br><span class="line">83</span><br><span class="line">84</span><br><span class="line">85</span><br><span class="line">86</span><br><span class="line">87</span><br><span class="line">88</span><br><span class="line">89</span><br><span class="line">90</span><br><span class="line">91</span><br><span class="line">92</span><br><span class="line">93</span><br><span class="line">94</span><br><span class="line">95</span><br><span class="line">96</span><br><span class="line">97</span><br><span class="line">98</span><br><span class="line">99</span><br><span class="line">100</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># Normalize the data</span></span><br><span class="line">cytotrace2_result &lt;- NormalizeData(cytotrace2_result)</span><br><span class="line"></span><br><span class="line"><span class="comment">## Normalizing layer: counts</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># Find variable features</span></span><br><span class="line">cytotrace2_result &lt;- FindVariableFeatures(cytotrace2_result, selection.method = <span class="string">&quot;vst&quot;</span>, nfeatures = <span class="number">2000</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment">## Finding variable features for layer counts</span></span><br><span class="line"></span><br><span class="line">all.genes &lt;- rownames(cytotrace2_result)</span><br><span class="line"><span class="comment"># Scale the data</span></span><br><span class="line">cytotrace2_result &lt;- ScaleData(cytotrace2_result, features=all.genes)</span><br><span class="line"></span><br><span class="line"><span class="comment">## Centering and scaling data matrix</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># Perform PCA</span></span><br><span class="line">cytotrace2_result &lt;- RunPCA(cytotrace2_result, npcs = <span class="number">30</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment">## PC_ 1 </span></span><br><span class="line"><span class="comment">## Positive:  Cpe, Pcsk1n, Pax6, Chga, Fam183b, Hmgn3, Chgb, Scg3, Map1b, Cryba2 </span></span><br><span class="line"><span class="comment">##     Isl1, Slc38a5, Neurod1, Prnp, Gnao1, Scgn, Gnas, Gch1, Aplp1, Fev </span></span><br><span class="line"><span class="comment">##     Dbpht2, Pcsk2, Vwa5b2, Pyy, Pam, Pim2, Cnih2, Mafb, Abcc8, Tspan7 </span></span><br><span class="line"><span class="comment">## Negative:  H19, Wfdc2, Tead2, Sparc, Smco4, Phgdh, Rbp1, Rpl36a, Cd24a, Mgst1 </span></span><br><span class="line"><span class="comment">##     Adamts1, 1700011H14Rik, Gsta3, Stmn1, Ldha, Rpl39, Hmga2, Spp1, Ptn, Rpl12 </span></span><br><span class="line"><span class="comment">##     Galk1, Wfdc15b, Cdkn1c, Sox9, Serpinh1, Sox4, Habp2, Hes1, Vim, Rps12 </span></span><br><span class="line"><span class="comment">## PC_ 2 </span></span><br><span class="line"><span class="comment">## Positive:  Btbd17, Ppp1r14a, Gadd45a, Neurod2, Igfbpl1, Sult2b1, Mfng, Neurog3, Pax4, Tubb3 </span></span><br><span class="line"><span class="comment">##     Btg2, Mfap4, Smarcd2, Tmsb4x, Plk3, Cbfa2t3, Cck, Notum, Actg1, Hes6 </span></span><br><span class="line"><span class="comment">##     Sulf2, Lynx1, Gfra3, Epb42, Lingo1, Gdpd1, Foxa3, Cdc14b, Hpca, Rcor2 </span></span><br><span class="line"><span class="comment">## Negative:  Atp1b1, Rbp4, Tagln2, Tuba1b, Akr1c19, Id2, Gcsh, Fxyd6, Rnase4, Bex4 </span></span><br><span class="line"><span class="comment">##     1700011H14Rik, Tmem27, Siva1, Phgdh, Tubb2a, Meis2, Cmtm8, Mgst3, Ociad2, Crip1 </span></span><br><span class="line"><span class="comment">##     Adamts1, Parm1, Fkbp2, Fabp3, Id3, Eif4ebp1, Spc25, Sdc4, Cpn1, Cyr61 </span></span><br><span class="line"><span class="comment">## PC_ 3 </span></span><br><span class="line"><span class="comment">## Positive:  Rpl13a, Rpl14, Rps14, mt-Atp6, Eef1a1, Rpl10, Rpl41, Ftl1, mt-Co3, Rps16 </span></span><br><span class="line"><span class="comment">##     Rps18, Rps27a, Rps3a1, Rpl26, mt-Cytb, Rps6, mt-Nd1, Rpl13, Rpl11, Rpl6 </span></span><br><span class="line"><span class="comment">##     Rpl9, mt-Co1, Rps5, Rps7, Ppia, Rps2, Rpl17, Rplp0, Rps9, Rpl32 </span></span><br><span class="line"><span class="comment">## Negative:  Neurog3, Anxa2, Sparc, Col9a3, Col9a1, Fxyd2, Amotl2, Mpc1, Aurka, Tpx2 </span></span><br><span class="line"><span class="comment">##     Cotl1, Ckb, Cenpf, Gpx2, Mia, Spp1, Cd82, Cdk2ap1, Cdc20, Lurap1l </span></span><br><span class="line"><span class="comment">##     Pbk, Gsta3, Racgap1, Hmmr, Cxcl12, Aurkb, Ccnb1, Aldh1b1, Tpm1, Tpm4 </span></span><br><span class="line"><span class="comment">## PC_ 4 </span></span><br><span class="line"><span class="comment">## Positive:  Cldn4, Runx1t1, Fev, Irf2bpl, Cacna2d1, Hnrnpa1, Pdk3, Slc35d3, Grb10, Gm43861 </span></span><br><span class="line"><span class="comment">##     Akr1c19, Jun, 1110012L19Rik, BC023829, Glud1, Vdr, Tm4sf4, Emb, St18, Celf3 </span></span><br><span class="line"><span class="comment">##     Pgf, Ids, Pcyt1b, Cited2, Maff, Elavl4, Sulf1, Neurod1, Serpini1, Ddc </span></span><br><span class="line"><span class="comment">## Negative:  Ppp1r1a, Iapp, Pdia6, Scg2, Calr, Ttr, Higd1a, G6pc2, Tmed3, Tmsb15l </span></span><br><span class="line"><span class="comment">##     Papss2, Sdf2l1, Hspa5, Clu, Ins1, Tmem27, Ly6e, Cd82, Slc30a8, Dap </span></span><br><span class="line"><span class="comment">##     Igfbp7, Ins2, Ociad2, Gcg, Ttc28, Qsox1, Bcl2, Gapdh, Irx1, Npy </span></span><br><span class="line"><span class="comment">## PC_ 5 </span></span><br><span class="line"><span class="comment">## Positive:  Cxcl12, Pdzk1ip1, Anxa2, Grin3a, Dcdc2a, Krt18, 8430408G22Rik, Pamr1, Anxa5, Bicc1 </span></span><br><span class="line"><span class="comment">##     Capg, Muc1, Krt8, Krt7, S100a10, Cldn3, Cystm1, Spp1, Gsta3, Trim47 </span></span><br><span class="line"><span class="comment">##     Smoc2, Ppp1r1b, Col9a3, Vim, Ndrg2, Tsc22d1, Cat, Cldn4, Txnip, Nfib </span></span><br><span class="line"><span class="comment">## Negative:  Vtn, 2810417H13Rik, Hist1h2ap, Chst2, Tuba1a, Top2a, P2rx1, Aurkb, Rrm2, Fbxo5 </span></span><br><span class="line"><span class="comment">##     Ccna2, Cpn1, Spc25, H2afz, Cenpf, Birc5, Ccnb1, Pbk, Gcat, Smc2 </span></span><br><span class="line"><span class="comment">##     Ube2c, Incenp, Mki67, Car14, Rad51ap1, Gc, Cdk1, Plk1, Prc1, Smc4</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># Find neighbors</span></span><br><span class="line">cytotrace2_result &lt;- FindNeighbors(cytotrace2_result, dims = <span class="number">1</span>:<span class="number">20</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment">## Computing nearest neighbor graph</span></span><br><span class="line"></span><br><span class="line"><span class="comment">## Computing SNN</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># Find clusters</span></span><br><span class="line">cytotrace2_result &lt;- FindClusters(cytotrace2_result, resolution = <span class="number">0.1</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment">## Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck</span></span><br><span class="line"><span class="comment">## </span></span><br><span class="line"><span class="comment">## Number of nodes: 2280</span></span><br><span class="line"><span class="comment">## Number of edges: 73179</span></span><br><span class="line"><span class="comment">## </span></span><br><span class="line"><span class="comment">## Running Louvain algorithm...</span></span><br><span class="line"><span class="comment">## Maximum modularity in 10 random starts: 0.9505</span></span><br><span class="line"><span class="comment">## Number of communities: 4</span></span><br><span class="line"><span class="comment">## Elapsed time: 0 seconds</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># Run UMAP</span></span><br><span class="line">cytotrace2_result &lt;- RunUMAP(cytotrace2_result, dims = <span class="number">1</span>:<span class="number">20</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment">## 14:12:35 UMAP embedding parameters a = 0.9922 b = 1.112</span></span><br><span class="line"></span><br><span class="line"><span class="comment">## 14:12:35 Read 2280 rows and found 20 numeric columns</span></span><br><span class="line"></span><br><span class="line"><span class="comment">## 14:12:35 Using Annoy for neighbor search, n_neighbors = 30</span></span><br><span class="line"></span><br><span class="line"><span class="comment">## 14:12:35 Building Annoy index with metric = cosine, n_trees = 50</span></span><br><span class="line"></span><br><span class="line"><span class="comment">## 0%   10   20   30   40   50   60   70   80   90   100%</span></span><br><span class="line"></span><br><span class="line"><span class="comment">## [----|----|----|----|----|----|----|----|----|----|</span></span><br><span class="line"></span><br><span class="line"><span class="comment">## **************************************************|</span></span><br><span class="line"><span class="comment">## 14:12:35 Writing NN index file to temp file /var/folders/yk/896pglhn555drshbgn27f71h0000gn/T//RtmpD9l2F4/file95af46be469e</span></span><br><span class="line"><span class="comment">## 14:12:35 Searching Annoy index using 1 thread, search_k = 3000</span></span><br><span class="line"><span class="comment">## 14:12:35 Annoy recall = 100%</span></span><br><span class="line"><span class="comment">## 14:12:35 Commencing smooth kNN distance calibration using 1 thread with target n_neighbors = 30</span></span><br><span class="line"><span class="comment">## 14:12:35 Initializing from normalized Laplacian + noise (using RSpectra)</span></span><br><span class="line"><span class="comment">## 14:12:35 Commencing optimization for 500 epochs, with 87006 positive edges</span></span><br><span class="line"><span class="comment">## 14:12:37 Optimization finished</span></span><br><span class="line"></span><br><span class="line">DimPlot(cytotrace2_result, group.by = <span class="string">&#x27;CytoTRACE2_Potency&#x27;</span>)</span><br></pre></td></tr></table></figure>

<p><img src="/2024/12/10/cytotrace2-basic/unnamed-chunk-13-1.png"></p>
<p>或者看score的分布</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">FeaturePlot(cytotrace2_result, features = <span class="string">&#x27;CytoTRACE2_Score&#x27;</span>)</span><br></pre></td></tr></table></figure>

<p><img src="/2024/12/10/cytotrace2-basic/unnamed-chunk-14-1.png"></p>
<blockquote>
<p>Ref:</p>
<p><a target="_blank" rel="noopener" href="https://github.com/digitalcytometry/cytotrace2/tree/main">https://github.com/digitalcytometry/cytotrace2/tree/main</a></p>
<p><a target="_blank" rel="noopener" href="https://www.biorxiv.org/content/10.1101/2024.03.19.585637v1.full">https://www.biorxiv.org/content/10.1101/2024.03.19.585637v1.full</a></p>
</blockquote>

    </div>

    
    
    
        

<div>
<ul class="post-copyright">
  <li class="post-copyright-author">
    <strong>Post author:  </strong>Dean Li
  </li>
  <li class="post-copyright-link">
    <strong>Post link: </strong>
    <a href="https://thereallda.github.io/2024/12/10/cytotrace2-basic/" title="CytoTRACE2-basic">https://thereallda.github.io/2024/12/10/cytotrace2-basic/</a>
  </li>
  <li class="post-copyright-license">
    <strong>Copyright Notice:  </strong>All articles in this blog are licensed under <a href="https://creativecommons.org/licenses/by-nc-sa/4.0/" rel="noopener" target="_blank"><i class="fab fa-fw fa-creative-commons"></i>BY-NC-SA</a> unless stating additionally.
  </li>
</ul>
</div>


      <footer class="post-footer">
          <div class="post-tags">
              <a href="/tags/R/" rel="tag"># R</a>
              <a href="/tags/single-cell/" rel="tag"># single-cell</a>
          </div>

        


        
    <div class="post-nav">
      <div class="post-nav-item">
    <a href="/2024/09/25/R-SeuratV5-to-H5AD/" rel="prev" title="R-SeuratV5-to-H5AD">
      <i class="fa fa-chevron-left"></i> R-SeuratV5-to-H5AD
    </a></div>
      <div class="post-nav-item">
    <a href="/2025/01/12/Seurat_pvalue/" rel="next" title="Seurat-VlnPlot-pvalues">
      Seurat-VlnPlot-pvalues <i class="fa fa-chevron-right"></i>
    </a></div>
    </div>
      </footer>
    
  </article>
  
  
  



          </div>
          
    <div class="comments" id="gitalk-container"></div>

<script>
  window.addEventListener('tabs:register', () => {
    let { activeClass } = CONFIG.comments;
    if (CONFIG.comments.storage) {
      activeClass = localStorage.getItem('comments_active') || activeClass;
    }
    if (activeClass) {
      let activeTab = document.querySelector(`a[href="#comment-${activeClass}"]`);
      if (activeTab) {
        activeTab.click();
      }
    }
  });
  if (CONFIG.comments.storage) {
    window.addEventListener('tabs:click', event => {
      if (!event.target.matches('.tabs-comment .tab-content .tab-pane')) return;
      let commentClass = event.target.classList[1];
      localStorage.setItem('comments_active', commentClass);
    });
  }
</script>

        </div>
          
  
  <div class="toggle sidebar-toggle">
    <span class="toggle-line toggle-line-first"></span>
    <span class="toggle-line toggle-line-middle"></span>
    <span class="toggle-line toggle-line-last"></span>
  </div>

  <aside class="sidebar">
    <div class="sidebar-inner">

      <ul class="sidebar-nav motion-element">
        <li class="sidebar-nav-toc">
          Table of Contents
        </li>
        <li class="sidebar-nav-overview">
          Overview
        </li>
      </ul>

      <!--noindex-->
      <div class="post-toc-wrap sidebar-panel">
          <div class="post-toc motion-element"><ol class="nav"><li class="nav-item nav-level-2"><a class="nav-link" href="#CytoTRACE2"><span class="nav-number">1.</span> <span class="nav-text">CytoTRACE2</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#%E5%AE%89%E8%A3%85"><span class="nav-number">2.</span> <span class="nav-text">安装</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#%E6%95%B0%E6%8D%AE%E5%87%86%E5%A4%87"><span class="nav-number">3.</span> <span class="nav-text">数据准备</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#%E8%BF%90%E8%A1%8CCytoTRACE2"><span class="nav-number">4.</span> <span class="nav-text">运行CytoTRACE2</span></a><ol class="nav-child"><li class="nav-item nav-level-3"><a class="nav-link" href="#Potency-score-distribution-by-phenotype"><span class="nav-number">4.1.</span> <span class="nav-text">Potency score distribution by phenotype</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#Potency-category"><span class="nav-number">4.2.</span> <span class="nav-text">Potency category</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#Relative-order"><span class="nav-number">4.3.</span> <span class="nav-text">Relative order</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#Phenotype"><span class="nav-number">4.4.</span> <span class="nav-text">Phenotype</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#Seurat-object-input"><span class="nav-number">4.5.</span> <span class="nav-text">Seurat object input</span></a></li></ol></li></ol></div>
      </div>
      <!--/noindex-->

      <div class="site-overview-wrap sidebar-panel">
        <div class="site-author motion-element" itemprop="author" itemscope itemtype="http://schema.org/Person">
    <img class="site-author-image" itemprop="image" alt="Dean Li"
      src="https://avatars.githubusercontent.com/u/55836943?v=4">
  <p class="site-author-name" itemprop="name">Dean Li</p>
  <div class="site-description" itemprop="description"></div>
</div>
<div class="site-state-wrap motion-element">
  <nav class="site-state">
      <div class="site-state-item site-state-posts">
          <a href="/archives/">
        
          <span class="site-state-item-count">66</span>
          <span class="site-state-item-name">posts</span>
        </a>
      </div>
      <div class="site-state-item site-state-tags">
            <a href="/tags/">
          
        <span class="site-state-item-count">21</span>
        <span class="site-state-item-name">tags</span></a>
      </div>
  </nav>
</div>
  <div class="links-of-author motion-element">
      <span class="links-of-author-item">
        <a href="https://github.com/thereallda" title="GitHub → https:&#x2F;&#x2F;github.com&#x2F;thereallda" rel="noopener" target="_blank"><i class="fab fa-github fa-fw"></i>GitHub</a>
      </span>
      <span class="links-of-author-item">
        <a href="mailto:lideanlda@foxmail.com" title="E-Mail → mailto:lideanlda@foxmail.com" rel="noopener" target="_blank"><i class="fa fa-envelope fa-fw"></i>E-Mail</a>
      </span>
      <span class="links-of-author-item">
        <a href="https://scholar.google.com/citations?user=zMno6_8AAAAJ" title="Google → https:&#x2F;&#x2F;scholar.google.com&#x2F;citations?user&#x3D;zMno6_8AAAAJ" rel="noopener" target="_blank"><i class="fab fa-google fa-fw"></i>Google</a>
      </span>
      <span class="links-of-author-item">
        <a href="https://www.jianshu.com/u/ab8a1c7d0510" title="简书 → https:&#x2F;&#x2F;www.jianshu.com&#x2F;u&#x2F;ab8a1c7d0510" rel="noopener" target="_blank"><i class="fa fa-book fa-fw"></i>简书</a>
      </span>
  </div>



      </div>

    </div>
  </aside>
  <div id="sidebar-dimmer"></div>


      </div>
    </main>

    <footer class="footer">
      <div class="footer-inner">
        

        

<div class="copyright">
  
  &copy; 
  <span itemprop="copyrightYear">2025</span>
  <span class="with-love">
    <i class="fa fa-heart"></i>
  </span>
  <span class="author" itemprop="copyrightHolder">Dean Li</span>
</div>
  <div class="powered-by">Powered by <a href="https://hexo.io/" class="theme-link" rel="noopener" target="_blank">Hexo</a> & <a href="https://theme-next.org/" class="theme-link" rel="noopener" target="_blank">NexT.Gemini</a>
  </div>



        
<div class="busuanzi-count">
  <script async src="https://busuanzi.ibruce.info/busuanzi/2.3/busuanzi.pure.mini.js"></script>
    <span class="post-meta-item" id="busuanzi_container_site_uv" style="display: none;">
      <span class="post-meta-item-icon">
        <i class="fa fa-user"></i>
      </span>
      <span class="site-uv" title="Total Visitors">
        <span id="busuanzi_value_site_uv"></span>
      </span>
    </span>
    <span class="post-meta-divider">|</span>
    <span class="post-meta-item" id="busuanzi_container_site_pv" style="display: none;">
      <span class="post-meta-item-icon">
        <i class="fa fa-eye"></i>
      </span>
      <span class="site-pv" title="Total Views">
        <span id="busuanzi_value_site_pv"></span>
      </span>
    </span>
</div>








      </div>
    </footer>
  </div>

  
  <script src="/lib/anime.min.js"></script>
  <script src="/lib/velocity/velocity.min.js"></script>
  <script src="/lib/velocity/velocity.ui.min.js"></script>

<script src="/js/utils.js"></script>

<script src="/js/motion.js"></script>


<script src="/js/schemes/pisces.js"></script>


<script src="/js/next-boot.js"></script>




  











<script>
document.querySelectorAll('.pdfobject-container').forEach(element => {
  let url = element.dataset.target;
  let pdfOpenParams = {
    navpanes : 0,
    toolbar  : 0,
    statusbar: 0,
    pagemode : 'thumbs',
    view     : 'FitH'
  };
  let pdfOpenFragment = '#' + Object.entries(pdfOpenParams).map(([key, value]) => `${key}=${encodeURIComponent(value)}`).join('&');
  let fullURL = `/lib/pdf/web/viewer.html?file=${encodeURIComponent(url)}${pdfOpenFragment}`;

  if (NexT.utils.supportsPDFs()) {
    element.innerHTML = `<embed class="pdfobject" src="${url + pdfOpenFragment}" type="application/pdf" style="height: ${element.dataset.height};">`;
  } else {
    element.innerHTML = `<iframe src="${fullURL}" style="height: ${element.dataset.height};" frameborder="0"></iframe>`;
  }
});
</script>




  

  

  
<script src="https://cdn.jsdelivr.net/npm/darkmode-js@1.5.7/lib/darkmode-js.min.js"></script>

<script>
var options = {
  bottom: '64px',
  right: '32px',
  left: 'unset',
  time: '0.5s',
  mixColor: 'transparent',
  backgroundColor: 'transparent',
  buttonColorDark: '#100f2c',
  buttonColorLight: '#fff',
  saveInCookies: true,
  label: '🌓',
  autoMatchOsTheme: true
}
const darkmode = new Darkmode(options);
window.darkmode = darkmode;
darkmode.showWidget();
</script>

<link rel="stylesheet" href="//cdn.jsdelivr.net/npm/gitalk@1/dist/gitalk.min.css">

<script>
NexT.utils.loadComments(document.querySelector('#gitalk-container'), () => {
  NexT.utils.getScript('//cdn.jsdelivr.net/npm/gitalk@1/dist/gitalk.min.js', () => {
    var gitalk = new Gitalk({
      clientID    : 'aea502b6092de143578a',
      clientSecret: '7bb9f5afb99d67ab37d8b296f39782af05dc8614',
      repo        : 'thereallda.GitHub.io',
      owner       : 'thereallda',
      admin       : ['thereallda'],
      id          : 'e28911c1fabca29b49323ec027921c82',
        language: '',
      distractionFreeMode: true
    });
    gitalk.render('gitalk-container');
  }, window.Gitalk);
});
</script>

</body>
</html>
